<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.38">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>1&nbsp; Introduction into R – R  cookbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../code/data_types.html" rel="next">
<link href="../index.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-b663525e4b92bb903b2199e2d401bf36.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-699dbeaba9dd29303344692a4c1b088c.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">

<script src="../site_libs/pagedtable-1.1/js/pagedtable.js"></script>



</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../code/1_intro.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction into R</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">R cookbook</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/1_intro.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction into R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/data_types.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/data_transformations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data transformations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/control_structures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Control Structures</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Statistics in R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/plotting_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Basic plotting</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/parsing_output_from_annotations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Working with annotation data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/2_misc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Miscellaneous</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#good-practices-for-coding" id="toc-good-practices-for-coding" class="nav-link active" data-scroll-target="#good-practices-for-coding"><span class="header-section-number">1.1</span> Good practices for coding</a></li>
  <li><a href="#the-example-data" id="toc-the-example-data" class="nav-link" data-scroll-target="#the-example-data"><span class="header-section-number">1.2</span> The example data</a></li>
  <li><a href="#working-in-r" id="toc-working-in-r" class="nav-link" data-scroll-target="#working-in-r"><span class="header-section-number">1.3</span> Working in R</a>
  <ul class="collapse">
  <li><a href="#opening-r-via-the-terminal" id="toc-opening-r-via-the-terminal" class="nav-link" data-scroll-target="#opening-r-via-the-terminal"><span class="header-section-number">1.3.1</span> Opening R via the terminal</a></li>
  <li><a href="#rstudio-everything-in-one-place" id="toc-rstudio-everything-in-one-place" class="nav-link" data-scroll-target="#rstudio-everything-in-one-place"><span class="header-section-number">1.3.2</span> RStudio (everything in one place):</a></li>
  </ul></li>
  <li><a href="#documenting-code" id="toc-documenting-code" class="nav-link" data-scroll-target="#documenting-code"><span class="header-section-number">1.4</span> Documenting code</a>
  <ul class="collapse">
  <li><a href="#markdown" id="toc-markdown" class="nav-link" data-scroll-target="#markdown"><span class="header-section-number">1.4.1</span> Markdown</a></li>
  <li><a href="#r-code-in-markdown" id="toc-r-code-in-markdown" class="nav-link" data-scroll-target="#r-code-in-markdown"><span class="header-section-number">1.4.2</span> R code in Markdown</a></li>
  <li><a href="#software" id="toc-software" class="nav-link" data-scroll-target="#software"><span class="header-section-number">1.4.3</span> Software</a></li>
  <li><a href="#execution-options" id="toc-execution-options" class="nav-link" data-scroll-target="#execution-options"><span class="header-section-number">1.4.4</span> Execution options</a></li>
  <li><a href="#using-languages-other-than-r" id="toc-using-languages-other-than-r" class="nav-link" data-scroll-target="#using-languages-other-than-r"><span class="header-section-number">1.4.5</span> Using languages other than R</a></li>
  </ul></li>
  <li><a href="#getting-help" id="toc-getting-help" class="nav-link" data-scroll-target="#getting-help"><span class="header-section-number">1.5</span> Getting help</a></li>
  <li><a href="#what-is-a-workspace" id="toc-what-is-a-workspace" class="nav-link" data-scroll-target="#what-is-a-workspace"><span class="header-section-number">1.6</span> What is a workspace?</a></li>
  <li><a href="#the-working-directory" id="toc-the-working-directory" class="nav-link" data-scroll-target="#the-working-directory"><span class="header-section-number">1.7</span> The working directory</a></li>
  <li><a href="#packages" id="toc-packages" class="nav-link" data-scroll-target="#packages"><span class="header-section-number">1.8</span> Packages</a>
  <ul class="collapse">
  <li><a href="#installing-packages" id="toc-installing-packages" class="nav-link" data-scroll-target="#installing-packages"><span class="header-section-number">1.8.1</span> Installing packages</a></li>
  <li><a href="#updating-packages" id="toc-updating-packages" class="nav-link" data-scroll-target="#updating-packages"><span class="header-section-number">1.8.2</span> Updating packages</a></li>
  <li><a href="#loading-packages-into-your-current-r-session" id="toc-loading-packages-into-your-current-r-session" class="nav-link" data-scroll-target="#loading-packages-into-your-current-r-session"><span class="header-section-number">1.8.3</span> Loading packages into your current R session</a></li>
  </ul></li>
  <li><a href="#the-assignment-operator--" id="toc-the-assignment-operator--" class="nav-link" data-scroll-target="#the-assignment-operator--"><span class="header-section-number">1.9</span> The assignment operator <code>&lt;-</code></a></li>
  <li><a href="#use-build-in-functions" id="toc-use-build-in-functions" class="nav-link" data-scroll-target="#use-build-in-functions"><span class="header-section-number">1.10</span> Use build in functions</a>
  <ul class="collapse">
  <li><a href="#call-the-default-values-of-a-function" id="toc-call-the-default-values-of-a-function" class="nav-link" data-scroll-target="#call-the-default-values-of-a-function"><span class="header-section-number">1.10.1</span> Call the default values of a function</a></li>
  </ul></li>
  <li><a href="#read-data-into-r" id="toc-read-data-into-r" class="nav-link" data-scroll-target="#read-data-into-r"><span class="header-section-number">1.11</span> Read data into R</a></li>
  <li><a href="#write-data-into-a-text-file" id="toc-write-data-into-a-text-file" class="nav-link" data-scroll-target="#write-data-into-a-text-file"><span class="header-section-number">1.12</span> Write data into a text file</a></li>
  <li><a href="#useful-functions-in-r" id="toc-useful-functions-in-r" class="nav-link" data-scroll-target="#useful-functions-in-r"><span class="header-section-number">1.13</span> Useful functions in R</a>
  <ul class="collapse">
  <li><a href="#base-functions" id="toc-base-functions" class="nav-link" data-scroll-target="#base-functions"><span class="header-section-number">1.13.1</span> Base functions</a></li>
  <li><a href="#the-unique-command" id="toc-the-unique-command" class="nav-link" data-scroll-target="#the-unique-command"><span class="header-section-number">1.13.2</span> The unique command</a></li>
  <li><a href="#the-merge-command" id="toc-the-merge-command" class="nav-link" data-scroll-target="#the-merge-command"><span class="header-section-number">1.13.3</span> The merge command</a></li>
  </ul></li>
  <li><a href="#combine-commands-into-one-line" id="toc-combine-commands-into-one-line" class="nav-link" data-scroll-target="#combine-commands-into-one-line"><span class="header-section-number">1.14</span> Combine commands into one line</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction into R</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>R is a statistical programming language and environment and free, open source and in active development. This tutorial will introduce into the basic concepts of R.</p>
<p>This tutorial will work with example data for two datasets:</p>
<p><strong>1. Growth data</strong></p>
<p>We have two data files that work with a similar experimental setup:</p>
<p>1.1. <strong>Growth_Data.txt</strong></p>
<p>During this experiments, we are doing plant growth experiments and treated our plants with different microbes wondering if any microbe affects plant growth in a positive way.</p>
<p>This file contains measurements of root length and shoot fresh weight for plants grown under control treatments (=MgCL) or when treating with 4 different bacteria. For simplicity only 1 biological experiment with 7-10 individual measurements per treatment were included.</p>
<p>1.2. <strong>Timecourse.txt</strong></p>
<p>We found positive effects for some of the strains tested above and now we want to know how long it takes for this effect to appear. To answer this, we measured root length of our plants when adding our microbe and compared it to control treatments at 5 different time points. Something unique with this dataframe is that we have empty cells (=NAs) and we need to deal with them as some R functions don’t like empty cells.</p>
<section id="good-practices-for-coding" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="good-practices-for-coding"><span class="header-section-number">1.1</span> Good practices for coding</h2>
<p>These practices are useful regardless of the computational language you use.</p>
<ul>
<li>Record what program versions you used</li>
<li>For each project, document who wrote the code, when you did it and why</li>
<li>Put dependencies in the beginning (ie packages)</li>
<li>Record the working directory (wdir)</li>
<li>Document ALL your code and comment it (using the # symbol)</li>
<li>Comment code in detail, so that you can still understand it after 5 years</li>
<li>Break code into smaller pieces for better readability</li>
<li>Test each line of code and build in control steps</li>
<li>If you work with random numbers, report the seed</li>
<li>Use sessionInfo() at the end of the script, which documents all the packages used within R for the current project</li>
<li>For larger files: save objects not workspaces (for space reasons)</li>
<li>Have descriptive names for objects, short and simple but easy enough to understand what they mean</li>
</ul>
</section>
<section id="the-example-data" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="the-example-data"><span class="header-section-number">1.2</span> The example data</h2>
<p>Let’s have a look at the data structure for our first dataframe:</p>
<div class="cell">
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["fct"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["fct"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["fct"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>We can see that we test our bacteria under different nutrient conditions (noP and P, which equals to normal phosphorous and low phosphorus concentrations added) and different treatments (=conditions), which are control treatments (MgCl) and different strains of microbes (Strain 101, Strain28, etc.). For each of these treatment we measured the shoot fresh weight in mg and the root length (in cm).</p>
<p>The timecourse data looks similar, we just have an extra column for the different timepoints and we only have measurements for the root length.</p>
<p><strong>2. Annotation data</strong></p>
<p>This file is specific to the output of the <a href="https://github.com/ndombrowski/Annotation_workflow">Spang_team annotation pipeline</a> but this workflow can be used for any type of categorical data one wants to summarize.</p>
<p>Using this workflow, we generated the file <strong>UAP2_Annotation_table_u.txt</strong>, which includes annotations for a set of 46 DPANN genomes. This includes annotations across several databases (arCOG, KO, PFAM, …) for each individual protein found across all these 46 genomes.</p>
<p>Specifically, we want to learn how to:</p>
<ul>
<li>Make a count table for each genome</li>
<li>Make a count table for clusters of interest</li>
<li>Make a heatmap for genes of interest</li>
<li>Merge our results with some pre-sorted tables</li>
</ul>
<p>For this to work we have some additional files to make our job easier:</p>
<ul>
<li><strong>mapping.txt</strong> = a list that defines to what cluster (i.e.&nbsp;grouping based on a phylogenetic tree) or bins belong to</li>
<li><strong>Genes_of_interest</strong> = a list of genes we are interested in and that we want to plot in a heatmap</li>
<li><strong>ar14_arCOGdef19.txt</strong> = metadata for the arCOG annotations</li>
<li><strong>Metabolism_Table_KO_Apr2020.txt</strong> = metadata for KOs and sorted by pathways</li>
</ul>
<p>The annotation table looks like this:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 4%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 4%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 5%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 3%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 3%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 6%">
<col style="width: 15%">
<col style="width: 1%">
<col style="width: 3%">
<col style="width: 7%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">accession</th>
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">TaxString</th>
<th style="text-align: left;">NewContigID</th>
<th style="text-align: left;">OldContigId</th>
<th style="text-align: left;">ContigIdMerge</th>
<th style="text-align: right;">ContigNewLength</th>
<th style="text-align: right;">GC</th>
<th style="text-align: left;">ProteinID</th>
<th style="text-align: right;">ProteinGC</th>
<th style="text-align: right;">ProteinLength</th>
<th style="text-align: left;">Prokka</th>
<th style="text-align: left;">arcogs</th>
<th style="text-align: left;">arcogs_geneID</th>
<th style="text-align: left;">arcogs_Description</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">arcogs_evalue</th>
<th style="text-align: left;">KO_hmm</th>
<th style="text-align: left;">e_value</th>
<th style="text-align: left;">bit_score</th>
<th style="text-align: left;">bit_score_cutoff</th>
<th style="text-align: left;">Definition</th>
<th style="text-align: left;">confidence</th>
<th style="text-align: left;">PFAM_hmm</th>
<th style="text-align: left;">PFAM_description</th>
<th style="text-align: left;">Pfam_Evalue</th>
<th style="text-align: left;">TIRGR</th>
<th style="text-align: left;">TIGR_description</th>
<th style="text-align: left;">EC</th>
<th style="text-align: left;">TIGR_Evalue</th>
<th style="text-align: left;">CAZy</th>
<th style="text-align: left;">CAZy_evalue</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">TDBD_ID</th>
<th style="text-align: left;">TPDB_evalue</th>
<th style="text-align: left;">HydDB</th>
<th style="text-align: left;">Description.1</th>
<th style="text-align: left;">HydDB_evalue</th>
<th style="text-align: left;">PFAM</th>
<th style="text-align: left;">PFAMdescription</th>
<th style="text-align: left;">IPR</th>
<th style="text-align: left;">IPRdescription</th>
<th style="text-align: left;">TopHit</th>
<th style="text-align: left;">E_value</th>
<th style="text-align: left;">PecID</th>
<th style="text-align: left;">TaxID</th>
<th style="text-align: left;">TaxString.1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00010</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00010</td>
<td style="text-align: right;">0.094</td>
<td style="text-align: right;">350</td>
<td style="text-align: left;">Digeranylgeranylglycerophospholipid reductase</td>
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Geranylgeranyl_reductase,_flavoprotein”</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">2.80E-75</td>
<td style="text-align: left;">K17830</td>
<td style="text-align: left;">2.10E-69</td>
<td style="text-align: left;">243.3</td>
<td style="text-align: left;">246.87</td>
<td style="text-align: left;">digeranylgeranylglycerophospholipid_reductase_[EC:1.3.1.101_1.3.7.11]</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">FAD_binding_domain</td>
<td style="text-align: left;">5.90E-09</td>
<td style="text-align: left;">TIGR02032</td>
<td style="text-align: left;">geranylgeranyl_reductase_family</td>
<td style="text-align: left;">1.3.1.-</td>
<td style="text-align: left;">1.40E-49</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR036188;_IPR002938</td>
<td style="text-align: left;">FAD/NAD(P)-binding_domain_superfamily;_FAD-binding_domain</td>
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">FAD_binding_domain</td>
<td style="text-align: left;">KYK22416.1_hypothetical_protein_AYK24_02275_[Thermoplasmatales_archaeon_SG8-52-4]</td>
<td style="text-align: left;">2.80E-58</td>
<td style="text-align: left;">36.5</td>
<td style="text-align: left;">1803819</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Thermoplasmata,Thermoplasmatales,none,none,Thermoplasmatales_archaeon_SG8-52-4”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00020</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00020</td>
<td style="text-align: right;">0.057</td>
<td style="text-align: right;">105</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00030</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00030</td>
<td style="text-align: right;">0.092</td>
<td style="text-align: right;">304</td>
<td style="text-align: left;">tRNA-2-methylthio-N(6)-dimethylallyladenosine synthase</td>
<td style="text-align: left;">arCOG01358</td>
<td style="text-align: left;">MiaB</td>
<td style="text-align: left;">2-methylthioadenine_synthetase</td>
<td style="text-align: left;">J</td>
<td style="text-align: left;">5.80E-90</td>
<td style="text-align: left;">K15865</td>
<td style="text-align: left;">1.20E-109</td>
<td style="text-align: left;">375.8</td>
<td style="text-align: left;">343.7</td>
<td style="text-align: left;">threonylcarbamoyladenosine_tRNA_methylthiotransferase_CDKAL1_[EC:2.8.4.5]</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF04055</td>
<td style="text-align: left;">Radical_SAM_superfamily</td>
<td style="text-align: left;">2.60E-21</td>
<td style="text-align: left;">TIGR01578</td>
<td style="text-align: left;">“MiaB-like_tRNA_modifying_enzyme,_archaeal-type”</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">1.70E-104</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR005839;_IPR002792;_IPR006466;_IPR006638;_IPR007197;_IPR020612;_IPR023404</td>
<td style="text-align: left;">“Methylthiotransferase;_TRAM_domain;_MiaB-like_tRNA_modifying_enzyme,_archaeal-type;_Elp3/MiaB/NifB;_Radical_SAM;_Methylthiotransferase,_conserved_site;_Radical_SAM,_alpha/beta_horseshoe”</td>
<td style="text-align: left;">PF01938;_PF04055</td>
<td style="text-align: left;">TRAM_domain;_Radical_SAM_superfamily</td>
<td style="text-align: left;">OIO63284.1_hypothetical_protein_AUJ83_01460_[Candidatus_Woesearchaeota_archaeon_CG1_02_33_12]</td>
<td style="text-align: left;">8.20E-91</td>
<td style="text-align: left;">55.6</td>
<td style="text-align: left;">1805422</td>
<td style="text-align: left;">“Archaea,Candidatus_Woesearchaeota,none,none,none,none,Candidatus_Woesearchaeota_archaeon_CG1_02_33_12”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00040</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00040</td>
<td style="text-align: right;">0.065</td>
<td style="text-align: right;">92</td>
<td style="text-align: left;">Enolase</td>
<td style="text-align: left;">arCOG01169</td>
<td style="text-align: left;">Eno</td>
<td style="text-align: left;">Enolase</td>
<td style="text-align: left;">G</td>
<td style="text-align: left;">4.00E-26</td>
<td style="text-align: left;">K01689</td>
<td style="text-align: left;">5.10E-20</td>
<td style="text-align: left;">81</td>
<td style="text-align: left;">48.73</td>
<td style="text-align: left;">enolase_[EC:4.2.1.11]</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain”</td>
<td style="text-align: left;">1.20E-20</td>
<td style="text-align: left;">TIGR01060</td>
<td style="text-align: left;">phosphopyruvate_hydratase</td>
<td style="text-align: left;">4.2.1.11</td>
<td style="text-align: left;">1.10E-25</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR020810;_IPR036849;_IPR000941;_IPR020809</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain;_Enolase-like,_C-terminal_domain_superfamily;_Enolase;_Enolase,_conserved_site”</td>
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain”</td>
<td style="text-align: left;">BAW30993.1_2-phosphoglycerate_dehydratase_[Methanothermobacter_sp._MT-2]</td>
<td style="text-align: left;">3.80E-23</td>
<td style="text-align: left;">70.7</td>
<td style="text-align: left;">1898379</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Methanobacteria,Methanobacteriales,Methanobacteriaceae,Methanothermobacter,Methanothermobacter_sp._MT-2”</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00050</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00050</td>
<td style="text-align: right;">0.059</td>
<td style="text-align: right;">218</td>
<td style="text-align: left;">30S ribosomal protein S2</td>
<td style="text-align: left;">arCOG04245</td>
<td style="text-align: left;">RpsB</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">J</td>
<td style="text-align: left;">8.90E-64</td>
<td style="text-align: left;">K02998</td>
<td style="text-align: left;">1.60E-46</td>
<td style="text-align: left;">167.8</td>
<td style="text-align: left;">210.97</td>
<td style="text-align: left;">small_subunit_ribosomal_protein_SAe</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">7.00E-24</td>
<td style="text-align: left;">TIGR01012</td>
<td style="text-align: left;">ribosomal_protein_uS2</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">1.00E-72</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR005707;_IPR023454;_IPR023591;_IPR018130;_IPR001865</td>
<td style="text-align: left;">“Ribosomal_protein_S2,_eukaryotic/archaeal;_Ribosomal_protein_S2,_archaeal;_Ribosomal_protein_S2,_flavodoxin-like_domain_superfamily;_Ribosomal_protein_S2,_conserved_site;_Ribosomal_protein_S2”</td>
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">A0B6E5.1_RecName:_Full=30S_ribosomal_protein_S2</td>
<td style="text-align: left;">1.20E-56</td>
<td style="text-align: left;">52.8</td>
<td style="text-align: left;">349307</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Methanomicrobia,Methanosarcinales,Methanotrichaceae,Methanothrix,Methanothrix_thermoacetophila”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00060</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00060</td>
<td style="text-align: right;">0.100</td>
<td style="text-align: right;">280</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">arCOG01728</td>
<td style="text-align: left;">Mho1</td>
<td style="text-align: left;">“Predicted_class_III_extradiol_dioxygenase,_MEMO1_family”</td>
<td style="text-align: left;">R</td>
<td style="text-align: left;">8.80E-74</td>
<td style="text-align: left;">K06990</td>
<td style="text-align: left;">1.40E-72</td>
<td style="text-align: left;">253.3</td>
<td style="text-align: left;">47</td>
<td style="text-align: left;">MEMO1_family_protein</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">Memo-like_protein</td>
<td style="text-align: left;">4.50E-77</td>
<td style="text-align: left;">TIGR04336</td>
<td style="text-align: left;">AmmeMemoRadiSam_system_protein_B</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">5.30E-91</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR002737</td>
<td style="text-align: left;">MEMO1_family</td>
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">Memo-like_protein</td>
<td style="text-align: left;">OYT50994.1_AmmeMemoRadiSam_system_protein_B_[Candidatus_Bathyarchaeota_archaeon_ex4484_135]</td>
<td style="text-align: left;">1.30E-66</td>
<td style="text-align: left;">47</td>
<td style="text-align: left;">2012509</td>
<td style="text-align: left;">“Archaea,Candidatus_Bathyarchaeota,none,none,none,none,Candidatus_Bathyarchaeota_archaeon_ex4484_135”</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="working-in-r" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="working-in-r"><span class="header-section-number">1.3</span> Working in R</h2>
<section id="opening-r-via-the-terminal" class="level3" data-number="1.3.1">
<h3 data-number="1.3.1" class="anchored" data-anchor-id="opening-r-via-the-terminal"><span class="header-section-number">1.3.1</span> Opening R via the terminal</h3>
<p>If you work with linux or want to start R from the terminal then open your terminal, change your directory to the R_exercises folder and then just type <code>R</code>.</p>
<p>Then, you should see something like this:</p>
<p align="left">
<img height="300" src="../images/R.png">
</p>
<p>To check your R version, start and quit R you can type the following in your terminal:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Ask what R version we have</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>R.version</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#start R</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>R</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co">#exit R</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">q</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="rstudio-everything-in-one-place" class="level3" data-number="1.3.2">
<h3 data-number="1.3.2" class="anchored" data-anchor-id="rstudio-everything-in-one-place"><span class="header-section-number">1.3.2</span> RStudio (everything in one place):</h3>
<p>R is command-line only while RStudio is a GUI ((graphical user interface)) version of R. Therefore, working in RStudio makes everything a bit more interactive.</p>
<p>RStudio includes the following:</p>
<ul>
<li>Script separate from command-line (left-hand screen)</li>
<li>Lists your variables (upper, right-hand corner)</li>
<li>Manual and an extensive help function</li>
<li>Easy install of new packages</li>
<li>Plots are shown within RStudio</li>
</ul>
<p align="left">
<img height="400" src="../images/RStudio.png">
</p>
</section>
</section>
<section id="documenting-code" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="documenting-code"><span class="header-section-number">1.4</span> Documenting code</h2>
<section id="markdown" class="level3" data-number="1.4.1">
<h3 data-number="1.4.1" class="anchored" data-anchor-id="markdown"><span class="header-section-number">1.4.1</span> Markdown</h3>
<p>Markdown is a lightweight markup language that you can use to add formatting elements to plain text text documents.</p>
<p>Some examples:</p>
<ul>
<li>Headings are defined with ‘#’, or ‘##’, or ‘###’ for first, second and third level.</li>
<li>Lists are created by using ’*’ for (bullets) and ‘1’, ‘2’, … for numbered lists.</li>
</ul>
<p>But why should we bother to write with Markdown when you can press buttons in an interface to format your text?</p>
<ul>
<li>Its used for a lot of things, including code documentation or building websites</li>
<li>Files containing Markdown-formatted text can be opened a lot application making it extremely portable</li>
<li>You can work with Markdown on different operating systems</li>
<li>Its used by a lot of tools, such as github, jupyter or RStudio.</li>
</ul>
<p>Its not the goal of this tutorial to introduce into markdown, but there is some good material online:</p>
<ul>
<li><a href="https://www.markdowntutorial.com">A step-by-step tutorial</a></li>
<li><a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet">A cheatsheet</a></li>
</ul>
</section>
<section id="r-code-in-markdown" class="level3" data-number="1.4.2">
<h3 data-number="1.4.2" class="anchored" data-anchor-id="r-code-in-markdown"><span class="header-section-number">1.4.2</span> R code in Markdown</h3>
<ul>
<li>The R-code is embedded in between the<br>
<code>```{r}</code> and <code>```</code> symbols.</li>
</ul>
<p>In Rstudio, on the top-right position of such a section you will find three symbols. Pressing the middle one will run all code chunks above, while the right symbol will run the current R-chunk.</p>
<p align="left">
<img width="800" src="../images/R2.png">
</p>
<p>An important menu-button is the <strong>Knit</strong> or <strong>Render</strong> button at the top, left-hand corner of RStudio. Pressing this button will create the final, rendered document (i.e.&nbsp;a HTML or PDF)</p>
</section>
<section id="software" class="level3" data-number="1.4.3">
<h3 data-number="1.4.3" class="anchored" data-anchor-id="software"><span class="header-section-number">1.4.3</span> Software</h3>
<section id="rmarkdown" class="level4" data-number="1.4.3.1">
<h4 data-number="1.4.3.1" class="anchored" data-anchor-id="rmarkdown"><span class="header-section-number">1.4.3.1</span> Rmarkdown</h4>
<p>One nice way to commend code is combine R-code with informative text in R markdown format (rmd).</p>
<p>R Markdown supports dozens of static and dynamic output formats including: HTML, PDF, MS Word, Beamer, HTML5 slides, Tufte-style handouts, books, dashboards, shiny applications, scientific articles, websites, and more.</p>
<p>The R markdown file specifies code chunks which will be executed in R (or python or bash) and plain text which will be written to the report as is. A report is created by rendering the file in R, then the R-code is executed and the results are merged in a pdf or html output.</p>
<p>How you create a Rmarkdown document in R:</p>
<ul>
<li>Open RStudio</li>
<li>press <code>File/new File/R markdown</code></li>
</ul>
<p>This will create an R markdown file that already contains some R code and text. You can also open this document (the rmd file) in RStudio and see how the code looks.</p>
<p>You can also open a new file in any text editor and save it with the .rmd extension.</p>
</section>
<section id="quarto" class="level4" data-number="1.4.3.2">
<h4 data-number="1.4.3.2" class="anchored" data-anchor-id="quarto"><span class="header-section-number">1.4.3.2</span> Quarto</h4>
<p><a href="https://quarto.org">Quarto<c2><ae></ae></c2></a> is the predecessor of RMarkdown and is an open-source scientific and technical publishing system built on Pandoc and allows to:</p>
<ul>
<li>Create dynamic content with Python, R, Julia, and Observable.</li>
<li>Author documents as plain text markdown or Jupyter notebooks.</li>
<li>Publish high-quality articles, reports, presentations, websites, blogs, and books in HTML, PDF, MS Word, ePub, and more.</li>
<li>Author with scientific markdown, including equations, citations, crossrefs, figure panels, callouts, advanced layout, and more.</li>
</ul>
<p>If you installed the newest version of RStudio, Quarto is already installed and we can create a quarto document with</p>
<ul>
<li>Open RStudio</li>
<li>press <code>File/New File/Quarto document</code></li>
</ul>
<p>Same as with RMarkdown, we document with Markdown (and HTML if we want), so knowing some basics is very useful.</p>
</section>
</section>
<section id="execution-options" class="level3" data-number="1.4.4">
<h3 data-number="1.4.4" class="anchored" data-anchor-id="execution-options"><span class="header-section-number">1.4.4</span> Execution options</h3>
<p>There are a wide variety of options available for customizing output from executed code</p>
<ul>
<li><code>include = FALSE</code> — prevents code and results from appearing in the finished file. R Markdown still runs the code in the chunk, and the results can be used by other chunks.</li>
<li><code>echo = FALSE</code> — prevents code, but not the results from appearing in the finished file. This is a useful way to embed figures.</li>
<li><code>message = FALSE</code> — prevents messages that are generated by code from appearing in the finished file.</li>
<li><code>warning = FALSE</code> — prevents warnings that are generated by code from appearing in the finished.</li>
<li><code>fig.cap = "..."</code> — adds a caption to graphical results.</li>
</ul>
<p>For setting these options inside a quarto document, see more <a href="https://quarto.org/docs/computations/execution-options.html">here</a>.</p>
</section>
<section id="using-languages-other-than-r" class="level3" data-number="1.4.5">
<h3 data-number="1.4.5" class="anchored" data-anchor-id="using-languages-other-than-r"><span class="header-section-number">1.4.5</span> Using languages other than R</h3>
<p>R Markdown/Quarto support several languages, such as bash and python, and you can call them in the same way as R code.</p>
<p>This is useful if you for example modify a dataframe in bash but then want continue to work on the data R. With proper documenting you can document the code in the same file.</p>
<p>Below is just an example, we see that we only need to “tell” R to use bash instead of R inside the top of the code chunk.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#run echo to print something to the screen</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'hello world'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>hello world</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#run echo and follow with a sed to modify text</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'a b c'</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/ /\|/g'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>a|b|c</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#list qmd files we have in our directory</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>  <span class="pp">*</span>qmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1_intro.qmd
2_misc.qmd
basic_operations.qmd
control_structures.qmd
data_transformations.qmd
data_types.qmd
parsing_output_from_annotations.qmd
plotting_basics.qmd
stats.qmd
test.qmd</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#show whats in our data</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> ../data/Growth_Data.txt </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SampleID    Nutrient    Condition   FW_shoot_mg     Rootlength
noP noP MgCl    10.26   5.931015152
noP noP MgCl    6.52    5.74344697
noP noP MgCl    12.17   6.834719697
noP noP MgCl    11.37   6.742734848
noP noP MgCl    9.8 6.736886364
noP noP MgCl    3.75    4.236348485
noP noP MgCl    5.38    4.753484848
noP noP MgCl    4.53    5.532333333
noP noP MgCl    7.75    5.484363636</code></pre>
</div>
</div>
<p>A general introduction into bash and awk is provided in separate tutorials that are also available on <a href="https://github.com/ndombrowski/AWK_tutorial">github.</a></p>
</section>
</section>
<section id="getting-help" class="level2" data-number="1.5">
<h2 data-number="1.5" class="anchored" data-anchor-id="getting-help"><span class="header-section-number">1.5</span> Getting help</h2>
<ol type="1">
<li>Some good places to check for things online are:</li>
</ol>
<ul>
<li>www.r-project.org</li>
<li>stack overflow</li>
<li>many more</li>
</ul>
<ol start="2" type="1">
<li>Inside of R, we can get help on functions and other things by typing either of the following:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help</span>(mean)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>?mean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="what-is-a-workspace" class="level2" data-number="1.6">
<h2 data-number="1.6" class="anchored" data-anchor-id="what-is-a-workspace"><span class="header-section-number">1.6</span> What is a workspace?</h2>
<p>The workspace is your current R working environment and includes any user-defined objects (i.e.&nbsp;vectors, matrices, data frames, lists, functions).</p>
<p>At the end of an R session, the user can save an image of the current workspace that is automatically reloaded the next time R is started.</p>
<p>We can check our workspace as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print the current working directory</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Users/ninadombrowski/Desktop/WorkingDir/Notebooks/Code_snippets/R/code"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#list the objects in the current workspace</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>()   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "annotation_data" "growth_data"    </code></pre>
</div>
</div>
</section>
<section id="the-working-directory" class="level2" data-number="1.7">
<h2 data-number="1.7" class="anchored" data-anchor-id="the-working-directory"><span class="header-section-number">1.7</span> The working directory</h2>
<p>The directory from which you work is usually first set from where you start R or where the script resides (the latter is the case for this example). But it can be re-set to find your data more easily. Ideally, you make one wdir per project and define the path in the script (see later below). It is recommended to have a similar format for these project folders, i.e.&nbsp;consider to create subfolders for input and output files. From the wdir you set you can load files using absolute and relative paths.</p>
<p>An example would be something with a structure like this:</p>
<p align="left">
<img height="200" src="../images/FolderStructure.png">
</p>
<p>In this example you see that we have 4 projects, and in each folder we have the R_script and folders for the required input and required output files. Also useful to have is a text file with the session info and if you plan to share this with others it is also good to have a README file that provides some background on the analysis done.</p>
<p>Options to see the working dir and set the working directory in R are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print your wdir</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">#setting your wdir</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="fu">getwd</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="packages" class="level2" data-number="1.8">
<h2 data-number="1.8" class="anchored" data-anchor-id="packages"><span class="header-section-number">1.8</span> Packages</h2>
<p>Packages are a collection of functions and tools that can be added to R that are often contributed by the community.</p>
<ul>
<li>There might be incompatibilities and packages are updated frequently but updating can break dependencies.</li>
<li>You need to install packages and load them EVERY TIME you want to use them. Therefore, ideally add them at the beginning of your scripts.</li>
</ul>
<section id="installing-packages" class="level3" data-number="1.8.1">
<h3 data-number="1.8.1" class="anchored" data-anchor-id="installing-packages"><span class="header-section-number">1.8.1</span> Installing packages</h3>
<p>We have two ways to install packages:</p>
<ol type="1">
<li>Via the console by typing:</li>
</ol>
<p><code>install.packages("package-name")</code></p>
<p>This will download a package from one of the CRAN mirrors assuming that a binary is available for your operating system. If you have not set a preferred CRAN mirror in your options(), then a menu will pop up asking you to choose a location.</p>
<ol start="2" type="1">
<li>Using R studio:</li>
</ol>
<p>Go to the lower right hand-side window and click on packages and then install. Find the packages your are interested in.</p>
<p>Notice: If libraries come with their own data (i.e.&nbsp;example tables), then the data needs to be loaded separately. I.e. via <code>data(cars)</code> to load the cars data from the cars package.</p>
</section>
<section id="updating-packages" class="level3" data-number="1.8.2">
<h3 data-number="1.8.2" class="anchored" data-anchor-id="updating-packages"><span class="header-section-number">1.8.2</span> Updating packages</h3>
<ul>
<li>Use <code>old.packages()</code> to list all your locally installed packages that are now out of date.</li>
<li><code>update.packages()</code> will update all packages in the known libraries interactively. This can take a while if you haven not done it recently. To update everything without any user intervention, use the ask = FALSE argument.</li>
</ul>
</section>
<section id="loading-packages-into-your-current-r-session" class="level3" data-number="1.8.3">
<h3 data-number="1.8.3" class="anchored" data-anchor-id="loading-packages-into-your-current-r-session"><span class="header-section-number">1.8.3</span> Loading packages into your current R session</h3>
<p>R will not remember what libraries you have loaded after you closed R. Therefore, you need to load libraries every time you re-open R. Here, we will load some libraries that are usually quite helpful and it is recommended to make the libraries you load part of each of your scripts. For example like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#some example packages needed for Rmarkdown</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(knitr)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kableExtra)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="the-assignment-operator--" class="level2" data-number="1.9">
<h2 data-number="1.9" class="anchored" data-anchor-id="the-assignment-operator--"><span class="header-section-number">1.9</span> The assignment operator <code>&lt;-</code></h2>
<p>The <code>&lt;-</code> symbol assigns a value to a variable,</p>
<p>General rules for the syntax R uses:</p>
<ul>
<li>R is case sensitive</li>
<li>If a variable exists, it will overwrite it with a new variable without asking</li>
<li>If you work with characters, i.e.&nbsp;words like ‘hello’, then this needs to be written with quotes around it: “hello” (this will become clearer below)</li>
<li><code>ls()</code> shows all the variables that are known by the system at the moment</li>
<li>you can remove individual objects with <code>rm()</code> and remove all objects <code>rm(list=ls())</code></li>
</ul>
<p>We can store more or less everything in a variable and use it later. For example, we can store numbers and do some math with them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#store some numbers</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span><span class="dv">4</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co">#do some simple math with the numbers we have stored</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>x<span class="sc">+</span>y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5</code></pre>
</div>
</div>
</section>
<section id="use-build-in-functions" class="level2" data-number="1.10">
<h2 data-number="1.10" class="anchored" data-anchor-id="use-build-in-functions"><span class="header-section-number">1.10</span> Use build in functions</h2>
<p>Functions are build in code that we can use to make our life easier, i.e.&nbsp;we can calculate lengths of vectors, do math or do statistical analyses.</p>
<p>Base R already knows many useful functions but loading new packages greatly increases our repertoire.</p>
<p>A list of most used functions can be found <a href="**http://www.statmethods.net/management/functions.html">here</a></p>
<p>A function consists of:</p>
<ol type="1">
<li>Function name</li>
<li>Arguments (optional, some might be set with a default) = control how exactly the function behaves</li>
<li>Body of the function = defines what the function does</li>
</ol>
<p>As an example lets test some simple functions: print and log:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#use the print function</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="dv">3</span><span class="sc">+</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#use the log function</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">log</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.302585</code></pre>
</div>
</div>
<section id="call-the-default-values-of-a-function" class="level3" data-number="1.10.1">
<h3 data-number="1.10.1" class="anchored" data-anchor-id="call-the-default-values-of-a-function"><span class="header-section-number">1.10.1</span> Call the default values of a function</h3>
<p>Every function comes with a set of arguments that you can set but that usually also have some default values. In R Studio you can easily access all those details with the help function.</p>
<ul>
<li><code>?</code> allows us to first of all check exactly what a function is doing. If you scroll down to the bottom of the help page you also get some examples on how to use a function.</li>
<li>More specifically the help function also allows us to get details on the arguments of a function.</li>
<li>For example, if we check the help page of <strong>read.table</strong> we see that by default this function does not read in a header and if we want to provide a header we have to change that argument.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#let's check what **log** is doing</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>?log</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co">#lets check the default arguments of a function</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>?read.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other useful functions:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Name</th>
<th style="text-align: left;">Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ls()</td>
<td style="text-align: left;">List objects in your current workspace</td>
</tr>
<tr class="even">
<td style="text-align: left;">rm(object)</td>
<td style="text-align: left;">Remove object from your current workspace</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rm(list = ls())</td>
<td style="text-align: left;">Remove all objects from your current workspace</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="read-data-into-r" class="level2" data-number="1.11">
<h2 data-number="1.11" class="anchored" data-anchor-id="read-data-into-r"><span class="header-section-number">1.11</span> Read data into R</h2>
<p>To work with any kind of data we need to first read the data into R to be able to work with it.</p>
<p>For tables, there are some things to be aware of:</p>
<ul>
<li>It matters what separator our table uses to specify individual columns. I.e. some programs store data using commas while others use tab as default delimiter. By default R assume we use a tab, but we can change this behavior when we read in the table.</li>
<li>Do not have any hash symbols (#) in your table. R will read this as a commented cell and not read the table from that point onward</li>
<li>Avoid empty cells, as these sometimes can mess up your data.</li>
</ul>
<p>For now, let’s read in the table with our growth data and store it under the variable name <strong>growth_data</strong>. To read in this file we need to direct it to the correct path as we do not have the file in the working directory but in a subdirectory namd <code>data</code>.</p>
<p>Options that are good to keep in mind when reading in a table:</p>
<ul>
<li><code>sep</code> = define that our field separator is a tab. A tab is written like this <code>/t</code>. If your data is using a space or comma, you can change that here.</li>
<li><code>header</code> = tell R that our data comes with a custom header (the first row in your dataframe)</li>
<li><code>quote</code> = deals with some annoying issue with data formatting in excel files</li>
</ul>
<p>General notice:</p>
<ul>
<li>To view data, the <code>head()</code> command is extremely practical, use it always when you modify data to check if everything went alright</li>
<li><code>dim()</code> is another useful function that displays the dimensions of a table, i.e.&nbsp;how many rows and columns we have. Again, this is useful to verify our data after we have transformed it to check if everything went alright.</li>
<li><code>colnames()</code> allows to only we the column names</li>
<li><code>rownames()</code> allows to only we the row names. Usually these are numbers, but we can also add anything else into the rows.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#read in data</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>timecourse <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/Timecourse.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T,  <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>growth_data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/Growth_Data.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T,  <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#check the first few lines of our data</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the dimensions of our data</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 105   5</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the column names</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SampleID"    "Nutrient"    "Condition"   "FW_shoot_mg" "Rootlength" </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the row names</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12" 
 [13] "13"  "14"  "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24" 
 [25] "25"  "26"  "27"  "28"  "29"  "30"  "31"  "32"  "33"  "34"  "35"  "36" 
 [37] "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44"  "45"  "46"  "47"  "48" 
 [49] "49"  "50"  "51"  "52"  "53"  "54"  "55"  "56"  "57"  "58"  "59"  "60" 
 [61] "61"  "62"  "63"  "64"  "65"  "66"  "67"  "68"  "69"  "70"  "71"  "72" 
 [73] "73"  "74"  "75"  "76"  "77"  "78"  "79"  "80"  "81"  "82"  "83"  "84" 
 [85] "85"  "86"  "87"  "88"  "89"  "90"  "91"  "92"  "93"  "94"  "95"  "96" 
 [97] "97"  "98"  "99"  "100" "101" "102" "103" "104" "105"</code></pre>
</div>
</div>
<p>Useful comments:</p>
<p>Sometimes we have to deal with really large data that take long to load with read.table. The function <code>fread()</code> from the data.table package is a very nice alternative.</p>
<p><strong>This script sometimes uses kable for making tables visually attractive in html. Whenever you see a function using kable simply replace it with the <code>head()</code> function, i.e.&nbsp;write <code>head(growth_data)</code></strong></p>
</section>
<section id="write-data-into-a-text-file" class="level2" data-number="1.12">
<h2 data-number="1.12" class="anchored" data-anchor-id="write-data-into-a-text-file"><span class="header-section-number">1.12</span> Write data into a text file</h2>
<p>Now, if we would have modified the table we might want to store it on your computer. We can do this using <code>write.table()</code> and below we use a different output directory. Notice, we always start from the location we set as working directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(growth_data, <span class="st">"../output_examples/growth_data_changed.txt"</span>,  <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> T, <span class="at">quote =</span>F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Arguments:</p>
<ul>
<li><code>sep</code> –&gt; we define what delimiter we want to use</li>
<li><code>row.names</code> = T –&gt; we want to include whatever is in the rownames</li>
<li><code>quote</code> = F –&gt; we do not want R to add any quotes around our columns.</li>
</ul>
</section>
<section id="useful-functions-in-r" class="level2" data-number="1.13">
<h2 data-number="1.13" class="anchored" data-anchor-id="useful-functions-in-r"><span class="header-section-number">1.13</span> Useful functions in R</h2>
<section id="base-functions" class="level3" data-number="1.13.1">
<h3 data-number="1.13.1" class="anchored" data-anchor-id="base-functions"><span class="header-section-number">1.13.1</span> Base functions</h3>
<p>R comes with the <code>base</code> package. This package contains the basic functions which let R function as a language: arithmetic, input/output, basic programming support, etc. Its contents are available through inheritance from any environment.</p>
<p>For a complete list of functions, use library(help = “base”).</p>
<p>Apart from the elementary operations, common arithmetic functions are available: log, exp, sin, cos, tan, sqrt, etc. Other useful functions one can use on vectors are:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Name</th>
<th style="text-align: left;">Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">max</td>
<td style="text-align: left;">select smallest element</td>
</tr>
<tr class="even">
<td style="text-align: left;">min</td>
<td style="text-align: left;">select largest element</td>
</tr>
<tr class="odd">
<td style="text-align: left;">length</td>
<td style="text-align: left;">gives the number of elements</td>
</tr>
<tr class="even">
<td style="text-align: left;">sum</td>
<td style="text-align: left;">sums all elements</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mean</td>
<td style="text-align: left;">obtains the mean value</td>
</tr>
<tr class="even">
<td style="text-align: left;">var</td>
<td style="text-align: left;">unbiased sample variance</td>
</tr>
<tr class="odd">
<td style="text-align: left;">sort</td>
<td style="text-align: left;">see exercise 2c</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="the-unique-command" class="level3" data-number="1.13.2">
<h3 data-number="1.13.2" class="anchored" data-anchor-id="the-unique-command"><span class="header-section-number">1.13.2</span> The unique command</h3>
<p><code>unique()</code> allows to determine duplicate rows and allows us to subset our data for certain categories. For example, for very large dataframes we often can simplify things.</p>
<p>Here, if we have a lot of treatments and did the experiment a long time ago, we might want to ask for a table that lists the treatments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make unique contig list that still contains info of our bin ID</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>mapping_file <span class="ot">&lt;-</span> <span class="fu">unique</span>(growth_data[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Nutrient"</span>, <span class="st">"Condition"</span>)])</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mapping_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","_rn_":"1"},{"1":"noP_101","2":"noP","3":"Strain101","_rn_":"14"},{"1":"noP_230","2":"noP","3":"Strain230","_rn_":"27"},{"1":"noP_28","2":"noP","3":"Strain28","_rn_":"38"},{"1":"P","2":"P","3":"MgCl","_rn_":"53"},{"1":"P_101","2":"P","3":"Strain101","_rn_":"68"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
<section id="the-merge-command" class="level3" data-number="1.13.3">
<h3 data-number="1.13.3" class="anchored" data-anchor-id="the-merge-command"><span class="header-section-number">1.13.3</span> The merge command</h3>
<p>We can also add additional metadata to our growth data.</p>
<p>One way to do this is the <strong>cbind()</strong> or <strong>rbind()</strong> functions. However, these functions require the two dataframes to have the exact number of columns or rows, which we do not have.</p>
<p>Here, the <code>merge()</code> function of the data.table package is very useful to merge data with different dimensions as long as they have a common pattern (i.e.&nbsp;the SampleID).</p>
<p>First lets build an artificial mapping file that incldues the number of days we grew our plants:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make mapping that contains our basic sample info</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>mapping_file <span class="ot">&lt;-</span> <span class="fu">unique</span>(growth_data[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Nutrient"</span>, <span class="st">"Condition"</span>)])</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co">#add a new column, where we list our experiment ID</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>mapping_file<span class="sc">$</span>Comment <span class="ot">&lt;-</span> <span class="st">"FirstExperiment"</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mapping_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Comment"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"FirstExperiment","_rn_":"1"},{"1":"noP_101","2":"noP","3":"Strain101","4":"FirstExperiment","_rn_":"14"},{"1":"noP_230","2":"noP","3":"Strain230","4":"FirstExperiment","_rn_":"27"},{"1":"noP_28","2":"noP","3":"Strain28","4":"FirstExperiment","_rn_":"38"},{"1":"P","2":"P","3":"MgCl","4":"FirstExperiment","_rn_":"53"},{"1":"P_101","2":"P","3":"Strain101","4":"FirstExperiment","_rn_":"68"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Now we can use this mapping file and merge it with our growth data as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load our package</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co">#merge our mapping file with our growth data</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>new_data_frame <span class="ot">&lt;-</span> <span class="fu">merge</span>(growth_data, mapping_file, <span class="at">by =</span> <span class="st">"SampleID"</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_data_frame)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient.x"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition.x"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["Nutrient.y"],"name":[6],"type":["chr"],"align":["left"]},{"label":["Condition.y"],"name":[7],"type":["chr"],"align":["left"]},{"label":["Comment"],"name":[8],"type":["chr"],"align":["left"]}],"data":[{"1":"P","2":"P","3":"MgCl","4":"32.42","5":"11.286793","6":"P","7":"MgCl","8":"FirstExperiment","_rn_":"1"},{"1":"P","2":"P","3":"MgCl","4":"21.03","5":"10.456630","6":"P","7":"MgCl","8":"FirstExperiment","_rn_":"2"},{"1":"P","2":"P","3":"MgCl","4":"18.35","5":"11.106341","6":"P","7":"MgCl","8":"FirstExperiment","_rn_":"3"},{"1":"P","2":"P","3":"MgCl","4":"21.04","5":"10.816896","6":"P","7":"MgCl","8":"FirstExperiment","_rn_":"4"},{"1":"P","2":"P","3":"MgCl","4":"16.61","5":"10.608252","6":"P","7":"MgCl","8":"FirstExperiment","_rn_":"5"},{"1":"P","2":"P","3":"MgCl","4":"25.79","5":"7.121587","6":"P","7":"MgCl","8":"FirstExperiment","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>This now is a good example to check that all went fine and that the new dataframes has the same number of rows (=measurements) compared to the original dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co">#control that all went fine</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 105   5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(new_data_frame)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 105   8</code></pre>
</div>
</div>
<p>With dim we see that we still have 105 rows (i.e.&nbsp;measurements) and that we now added 3 new columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#if there is no match between dataframe 1 and dataframe 2 columns will by default be deleted. If you want to keep all columns do:</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co">#new_data_frame &lt;- merge(growth_data, mapping_file, by = "SampleID". all.x = T)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="combine-commands-into-one-line" class="level2" data-number="1.14">
<h2 data-number="1.14" class="anchored" data-anchor-id="combine-commands-into-one-line"><span class="header-section-number">1.14</span> Combine commands into one line</h2>
<p>While this gets more difficult to read, sometimes it might be useful to combine several commands into one go to condense code Generally, it is easier to just write line by line especially if you read your code months later.</p>
<p>What we want to do:</p>
<ul>
<li>in the example above we duplicate the columns for Nutrient and Condition and before merging we might first subset the mapping file to only include the info we want to merge.</li>
<li>So our two steps are:
<ul>
<li>trim the mapping file</li>
<li>merge</li>
</ul></li>
</ul>
<p>To do this, we use these two lines of code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make mapping file more simple</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>mapping_reduced <span class="ot">&lt;-</span> mapping_file[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Comment"</span>)]</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co">#merge</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>new_data_frame <span class="ot">&lt;-</span> <span class="fu">merge</span>(growth_data, mapping_reduced, <span class="at">by =</span> <span class="st">"SampleID"</span>)</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_data_frame)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["Comment"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"P","2":"P","3":"MgCl","4":"32.42","5":"11.286793","6":"FirstExperiment","_rn_":"1"},{"1":"P","2":"P","3":"MgCl","4":"21.03","5":"10.456630","6":"FirstExperiment","_rn_":"2"},{"1":"P","2":"P","3":"MgCl","4":"18.35","5":"11.106341","6":"FirstExperiment","_rn_":"3"},{"1":"P","2":"P","3":"MgCl","4":"21.04","5":"10.816896","6":"FirstExperiment","_rn_":"4"},{"1":"P","2":"P","3":"MgCl","4":"16.61","5":"10.608252","6":"FirstExperiment","_rn_":"5"},{"1":"P","2":"P","3":"MgCl","4":"25.79","5":"7.121587","6":"FirstExperiment","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Now, this worked fine but requires a bit more code and we need to create one more object.</p>
<p>We could also combine these two lines of code into one line by subsetting our mapping file INSIDE the merge function as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#clean mapping file and merge</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>new_data_frame <span class="ot">&lt;-</span> <span class="fu">merge</span>(growth_data, mapping_file[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Comment"</span>)], <span class="at">by =</span> <span class="st">"SampleID"</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_data_frame)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["Comment"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"P","2":"P","3":"MgCl","4":"32.42","5":"11.286793","6":"FirstExperiment","_rn_":"1"},{"1":"P","2":"P","3":"MgCl","4":"21.03","5":"10.456630","6":"FirstExperiment","_rn_":"2"},{"1":"P","2":"P","3":"MgCl","4":"18.35","5":"11.106341","6":"FirstExperiment","_rn_":"3"},{"1":"P","2":"P","3":"MgCl","4":"21.04","5":"10.816896","6":"FirstExperiment","_rn_":"4"},{"1":"P","2":"P","3":"MgCl","4":"16.61","5":"10.608252","6":"FirstExperiment","_rn_":"5"},{"1":"P","2":"P","3":"MgCl","4":"25.79","5":"7.121587","6":"FirstExperiment","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../index.html" class="pagination-link" aria-label="Preface">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../code/data_types.html" class="pagination-link" aria-label="Data objects">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data objects</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>