<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.251">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>R cookbook - 1&nbsp; Introduction into R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../code/2_misc.html" rel="next">
<link href="../index.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

<script src="../site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="../site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<link href="../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../site_libs/pagedtable-1.1/js/pagedtable.js"></script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction into R</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../" class="sidebar-logo-link">
      <img src="../images/tree_of_life.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="../">R cookbook</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/1_intro.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction into R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../code/2_misc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R basics</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#the-example-data" id="toc-the-example-data" class="nav-link active" data-scroll-target="#the-example-data"><span class="toc-section-number">1.1</span>  The example data</a></li>
  <li><a href="#rstudio-everything-in-one-place" id="toc-rstudio-everything-in-one-place" class="nav-link" data-scroll-target="#rstudio-everything-in-one-place"><span class="toc-section-number">1.2</span>  RStudio (everything in one place):</a></li>
  <li><a href="#opening-r-via-the-terminal" id="toc-opening-r-via-the-terminal" class="nav-link" data-scroll-target="#opening-r-via-the-terminal"><span class="toc-section-number">1.3</span>  Opening R via the terminal</a></li>
  <li><a href="#documenting-code-via-markdown" id="toc-documenting-code-via-markdown" class="nav-link" data-scroll-target="#documenting-code-via-markdown"><span class="toc-section-number">1.4</span>  Documenting code via markdown</a>
  <ul class="collapse">
  <li><a href="#software" id="toc-software" class="nav-link" data-scroll-target="#software"><span class="toc-section-number">1.4.1</span>  Software</a></li>
  <li><a href="#r-code-in-markdown" id="toc-r-code-in-markdown" class="nav-link" data-scroll-target="#r-code-in-markdown"><span class="toc-section-number">1.4.2</span>  R code in Markdown</a></li>
  <li><a href="#text-in-markdown" id="toc-text-in-markdown" class="nav-link" data-scroll-target="#text-in-markdown"><span class="toc-section-number">1.4.3</span>  Text in markdown</a></li>
  <li><a href="#execution-options" id="toc-execution-options" class="nav-link" data-scroll-target="#execution-options"><span class="toc-section-number">1.4.4</span>  Execution options</a></li>
  <li><a href="#using-languages-other-than-r" id="toc-using-languages-other-than-r" class="nav-link" data-scroll-target="#using-languages-other-than-r"><span class="toc-section-number">1.4.5</span>  Using languages other than R</a></li>
  </ul></li>
  <li><a href="#good-practices-for-coding" id="toc-good-practices-for-coding" class="nav-link" data-scroll-target="#good-practices-for-coding"><span class="toc-section-number">1.5</span>  Good practices for coding</a></li>
  <li><a href="#general-introduction-into-r" id="toc-general-introduction-into-r" class="nav-link" data-scroll-target="#general-introduction-into-r"><span class="toc-section-number">1.6</span>  General introduction into R</a>
  <ul class="collapse">
  <li><a href="#getting-help" id="toc-getting-help" class="nav-link" data-scroll-target="#getting-help"><span class="toc-section-number">1.6.1</span>  Getting help</a></li>
  <li><a href="#what-is-a-workspace" id="toc-what-is-a-workspace" class="nav-link" data-scroll-target="#what-is-a-workspace"><span class="toc-section-number">1.6.2</span>  What is a workspace?</a></li>
  <li><a href="#setting-your-working-directory-wdir" id="toc-setting-your-working-directory-wdir" class="nav-link" data-scroll-target="#setting-your-working-directory-wdir"><span class="toc-section-number">1.6.3</span>  Setting your working directory (wdir)</a></li>
  <li><a href="#packages" id="toc-packages" class="nav-link" data-scroll-target="#packages"><span class="toc-section-number">1.6.4</span>  Packages</a></li>
  <li><a href="#the-assignment-operator--" id="toc-the-assignment-operator--" class="nav-link" data-scroll-target="#the-assignment-operator--"><span class="toc-section-number">1.6.5</span>  The assignment operator <code>&lt;-</code></a></li>
  <li><a href="#use-build-in-functions" id="toc-use-build-in-functions" class="nav-link" data-scroll-target="#use-build-in-functions"><span class="toc-section-number">1.6.6</span>  Use build in functions</a></li>
  </ul></li>
  <li><a href="#read-data-into-r" id="toc-read-data-into-r" class="nav-link" data-scroll-target="#read-data-into-r"><span class="toc-section-number">1.7</span>  Read data into R</a></li>
  <li><a href="#store-data-from-r-on-your-computer-as-a-text-file" id="toc-store-data-from-r-on-your-computer-as-a-text-file" class="nav-link" data-scroll-target="#store-data-from-r-on-your-computer-as-a-text-file"><span class="toc-section-number">1.8</span>  Store data from R on your computer as a text file</a></li>
  <li><a href="#what-are-objects" id="toc-what-are-objects" class="nav-link" data-scroll-target="#what-are-objects"><span class="toc-section-number">1.9</span>  What are objects?</a></li>
  <li><a href="#data-classes" id="toc-data-classes" class="nav-link" data-scroll-target="#data-classes"><span class="toc-section-number">1.10</span>  Data classes</a></li>
  <li><a href="#data-structures" id="toc-data-structures" class="nav-link" data-scroll-target="#data-structures"><span class="toc-section-number">1.11</span>  Data structures</a>
  <ul class="collapse">
  <li><a href="#vectors" id="toc-vectors" class="nav-link" data-scroll-target="#vectors"><span class="toc-section-number">1.11.1</span>  Vectors</a></li>
  <li><a href="#matrix" id="toc-matrix" class="nav-link" data-scroll-target="#matrix"><span class="toc-section-number">1.11.2</span>  Matrix</a></li>
  <li><a href="#lists" id="toc-lists" class="nav-link" data-scroll-target="#lists"><span class="toc-section-number">1.11.3</span>  Lists</a></li>
  <li><a href="#dataframes" id="toc-dataframes" class="nav-link" data-scroll-target="#dataframes"><span class="toc-section-number">1.11.4</span>  Dataframes</a></li>
  </ul></li>
  <li><a href="#how-can-we-check-what-kind-of-structure-our-data-has" id="toc-how-can-we-check-what-kind-of-structure-our-data-has" class="nav-link" data-scroll-target="#how-can-we-check-what-kind-of-structure-our-data-has"><span class="toc-section-number">1.12</span>  How can we check what kind of structure our data has?</a></li>
  <li><a href="#factors" id="toc-factors" class="nav-link" data-scroll-target="#factors"><span class="toc-section-number">1.13</span>  Factors</a>
  <ul class="collapse">
  <li><a href="#checking-the-behaviour-of-factors" id="toc-checking-the-behaviour-of-factors" class="nav-link" data-scroll-target="#checking-the-behaviour-of-factors"><span class="toc-section-number">1.13.1</span>  Checking the behaviour of factors</a></li>
  <li><a href="#ordering-factor-levels" id="toc-ordering-factor-levels" class="nav-link" data-scroll-target="#ordering-factor-levels"><span class="toc-section-number">1.13.2</span>  Ordering factor levels</a></li>
  <li><a href="#converting-factors" id="toc-converting-factors" class="nav-link" data-scroll-target="#converting-factors"><span class="toc-section-number">1.13.3</span>  Converting factors</a></li>
  </ul></li>
  <li><a href="#dealing-with-nas-in-our-data" id="toc-dealing-with-nas-in-our-data" class="nav-link" data-scroll-target="#dealing-with-nas-in-our-data"><span class="toc-section-number">1.14</span>  Dealing with NAs in our data</a>
  <ul class="collapse">
  <li><a href="#removing-nas" id="toc-removing-nas" class="nav-link" data-scroll-target="#removing-nas"><span class="toc-section-number">1.14.1</span>  Removing NAs</a></li>
  <li><a href="#replacing-nas-in-our-dataframe" id="toc-replacing-nas-in-our-dataframe" class="nav-link" data-scroll-target="#replacing-nas-in-our-dataframe"><span class="toc-section-number">1.14.2</span>  Replacing NAs in our dataframe</a></li>
  </ul></li>
  <li><a href="#important-code-snippets" id="toc-important-code-snippets" class="nav-link" data-scroll-target="#important-code-snippets"><span class="toc-section-number">1.15</span>  Important code snippets</a>
  <ul class="collapse">
  <li><a href="#print-the-session-info" id="toc-print-the-session-info" class="nav-link" data-scroll-target="#print-the-session-info"><span class="toc-section-number">1.15.1</span>  Print the session info:</a></li>
  <li><a href="#clean-up-the-working-environment" id="toc-clean-up-the-working-environment" class="nav-link" data-scroll-target="#clean-up-the-working-environment"><span class="toc-section-number">1.15.2</span>  Clean-up the working environment</a></li>
  </ul></li>
  <li><a href="#basic-operations" id="toc-basic-operations" class="nav-link" data-scroll-target="#basic-operations"><span class="toc-section-number">1.16</span>  Basic operations</a></li>
  <li><a href="#logical-operators" id="toc-logical-operators" class="nav-link" data-scroll-target="#logical-operators"><span class="toc-section-number">1.17</span>  Logical operators</a></li>
  <li><a href="#examples-for-using-operators" id="toc-examples-for-using-operators" class="nav-link" data-scroll-target="#examples-for-using-operators"><span class="toc-section-number">1.18</span>  Examples for using operators:</a>
  <ul class="collapse">
  <li><a href="#basic-operators" id="toc-basic-operators" class="nav-link" data-scroll-target="#basic-operators"><span class="toc-section-number">1.18.1</span>  Basic operators</a></li>
  <li><a href="#using-operators-to-subset-dataframes" id="toc-using-operators-to-subset-dataframes" class="nav-link" data-scroll-target="#using-operators-to-subset-dataframes"><span class="toc-section-number">1.18.2</span>  Using operators to subset dataframes</a></li>
  </ul></li>
  <li><a href="#useful-base-functions-in-r" id="toc-useful-base-functions-in-r" class="nav-link" data-scroll-target="#useful-base-functions-in-r"><span class="toc-section-number">1.19</span>  Useful base functions in R</a></li>
  <li><a href="#subsetting-our-data" id="toc-subsetting-our-data" class="nav-link" data-scroll-target="#subsetting-our-data"><span class="toc-section-number">1.20</span>  Subsetting our data</a>
  <ul class="collapse">
  <li><a href="#grep-and-grepl" id="toc-grep-and-grepl" class="nav-link" data-scroll-target="#grep-and-grepl"><span class="toc-section-number">1.20.1</span>  grep and grepl</a></li>
  </ul></li>
  <li><a href="#the-unique-command" id="toc-the-unique-command" class="nav-link" data-scroll-target="#the-unique-command"><span class="toc-section-number">1.21</span>  The unique command</a></li>
  <li><a href="#the-merge-command" id="toc-the-merge-command" class="nav-link" data-scroll-target="#the-merge-command"><span class="toc-section-number">1.22</span>  The merge command</a></li>
  <li><a href="#combine-commands-into-one-line" id="toc-combine-commands-into-one-line" class="nav-link" data-scroll-target="#combine-commands-into-one-line"><span class="toc-section-number">1.23</span>  Combine commands into one line</a></li>
  <li><a href="#add-new-columns-to-our-data-and-combining-values-in-different-columns" id="toc-add-new-columns-to-our-data-and-combining-values-in-different-columns" class="nav-link" data-scroll-target="#add-new-columns-to-our-data-and-combining-values-in-different-columns"><span class="toc-section-number">1.24</span>  Add new columns to our data and combining values in different columns</a></li>
  <li><a href="#stats" id="toc-stats" class="nav-link" data-scroll-target="#stats"><span class="toc-section-number">1.25</span>  STATS</a>
  <ul class="collapse">
  <li><a href="#summary-stats" id="toc-summary-stats" class="nav-link" data-scroll-target="#summary-stats"><span class="toc-section-number">1.25.1</span>  Summary stats</a></li>
  <li><a href="#the-table-command" id="toc-the-table-command" class="nav-link" data-scroll-target="#the-table-command"><span class="toc-section-number">1.25.2</span>  The table command</a></li>
  <li><a href="#the-ddply-command" id="toc-the-ddply-command" class="nav-link" data-scroll-target="#the-ddply-command"><span class="toc-section-number">1.25.3</span>  The ddply command</a></li>
  <li><a href="#tidyr" id="toc-tidyr" class="nav-link" data-scroll-target="#tidyr"><span class="toc-section-number">1.25.4</span>  tidyr</a></li>
  <li><a href="#more-stats-tapply" id="toc-more-stats-tapply" class="nav-link" data-scroll-target="#more-stats-tapply"><span class="toc-section-number">1.25.5</span>  More stats: tapply</a></li>
  <li><a href="#linear-models" id="toc-linear-models" class="nav-link" data-scroll-target="#linear-models"><span class="toc-section-number">1.25.6</span>  Linear models</a></li>
  <li><a href="#basics" id="toc-basics" class="nav-link" data-scroll-target="#basics"><span class="toc-section-number">1.25.7</span>  Basics</a></li>
  <li><a href="#normal-distribution" id="toc-normal-distribution" class="nav-link" data-scroll-target="#normal-distribution"><span class="toc-section-number">1.25.8</span>  Normal distribution</a></li>
  <li><a href="#anova" id="toc-anova" class="nav-link" data-scroll-target="#anova"><span class="toc-section-number">1.25.9</span>  ANOVA</a></li>
  <li><a href="#tukey" id="toc-tukey" class="nav-link" data-scroll-target="#tukey"><span class="toc-section-number">1.25.10</span>  TUKEY</a></li>
  <li><a href="#check-the-homogeneity-of-variances" id="toc-check-the-homogeneity-of-variances" class="nav-link" data-scroll-target="#check-the-homogeneity-of-variances"><span class="toc-section-number">1.25.11</span>  Check the homogeneity of variances</a></li>
  <li><a href="#check-for-normality-v2" id="toc-check-for-normality-v2" class="nav-link" data-scroll-target="#check-for-normality-v2"><span class="toc-section-number">1.25.12</span>  Check for normality v2</a></li>
  </ul></li>
  <li><a href="#control-structures" id="toc-control-structures" class="nav-link" data-scroll-target="#control-structures"><span class="toc-section-number">1.26</span>  Control Structures</a>
  <ul class="collapse">
  <li><a href="#loops" id="toc-loops" class="nav-link" data-scroll-target="#loops"><span class="toc-section-number">1.26.1</span>  Loops</a></li>
  <li><a href="#if-else" id="toc-if-else" class="nav-link" data-scroll-target="#if-else"><span class="toc-section-number">1.26.2</span>  if-else</a></li>
  </ul></li>
  <li><a href="#plotting" id="toc-plotting" class="nav-link" data-scroll-target="#plotting"><span class="toc-section-number">1.27</span>  Plotting</a>
  <ul class="collapse">
  <li><a href="#basic-plots" id="toc-basic-plots" class="nav-link" data-scroll-target="#basic-plots"><span class="toc-section-number">1.27.1</span>  Basic Plots</a></li>
  <li><a href="#info-on-using-par" id="toc-info-on-using-par" class="nav-link" data-scroll-target="#info-on-using-par"><span class="toc-section-number">1.27.2</span>  info on using par</a></li>
  <li><a href="#histograms" id="toc-histograms" class="nav-link" data-scroll-target="#histograms"><span class="toc-section-number">1.27.3</span>  Histograms</a></li>
  <li><a href="#boxplots" id="toc-boxplots" class="nav-link" data-scroll-target="#boxplots"><span class="toc-section-number">1.27.4</span>  Boxplots</a></li>
  <li><a href="#ggplot2" id="toc-ggplot2" class="nav-link" data-scroll-target="#ggplot2"><span class="toc-section-number">1.27.5</span>  Ggplot2</a></li>
  </ul></li>
  <li><a href="#working-with-annotation-data" id="toc-working-with-annotation-data" class="nav-link" data-scroll-target="#working-with-annotation-data"><span class="toc-section-number">1.28</span>  Working with annotation data</a>
  <ul class="collapse">
  <li><a href="#load-essential-packages" id="toc-load-essential-packages" class="nav-link" data-scroll-target="#load-essential-packages"><span class="toc-section-number">1.28.1</span>  Load essential packages</a></li>
  <li><a href="#deal-with-the-mapping-file-for-our-genomes" id="toc-deal-with-the-mapping-file-for-our-genomes" class="nav-link" data-scroll-target="#deal-with-the-mapping-file-for-our-genomes"><span class="toc-section-number">1.28.2</span>  Deal with the mapping file for our genomes</a></li>
  <li><a href="#clean-mapping-file" id="toc-clean-mapping-file" class="nav-link" data-scroll-target="#clean-mapping-file"><span class="toc-section-number">1.28.3</span>  Clean mapping file</a></li>
  <li><a href="#generate-lists-to-order-dataframes" id="toc-generate-lists-to-order-dataframes" class="nav-link" data-scroll-target="#generate-lists-to-order-dataframes"><span class="toc-section-number">1.28.4</span>  Generate lists to order dataframes</a></li>
  </ul></li>
  <li><a href="#deal-with-mapping-files-for-the-annotations" id="toc-deal-with-mapping-files-for-the-annotations" class="nav-link" data-scroll-target="#deal-with-mapping-files-for-the-annotations"><span class="toc-section-number">1.29</span>  Deal with mapping files for the annotations</a>
  <ul class="collapse">
  <li><a href="#load-and-view-the-tables" id="toc-load-and-view-the-tables" class="nav-link" data-scroll-target="#load-and-view-the-tables"><span class="toc-section-number">1.29.1</span>  Load and view the tables</a></li>
  <li><a href="#make-a-vector-to-order-our-genes-of-interest" id="toc-make-a-vector-to-order-our-genes-of-interest" class="nav-link" data-scroll-target="#make-a-vector-to-order-our-genes-of-interest"><span class="toc-section-number">1.29.2</span>  Make a vector to order our genes of interest</a></li>
  </ul></li>
  <li><a href="#deal-with-annotation-file" id="toc-deal-with-annotation-file" class="nav-link" data-scroll-target="#deal-with-annotation-file"><span class="toc-section-number">1.30</span>  Deal with annotation file</a>
  <ul class="collapse">
  <li><a href="#read-in-table" id="toc-read-in-table" class="nav-link" data-scroll-target="#read-in-table"><span class="toc-section-number">1.30.1</span>  Read in table</a></li>
  <li><a href="#make-a-mapping-file-that-links-all-annotation-ids-to-their-descriptions" id="toc-make-a-mapping-file-that-links-all-annotation-ids-to-their-descriptions" class="nav-link" data-scroll-target="#make-a-mapping-file-that-links-all-annotation-ids-to-their-descriptions"><span class="toc-section-number">1.30.2</span>  Make a mapping file that links all annotation IDs to their descriptions</a></li>
  <li><a href="#parse-table-to-make-it-easier-to-work-with-it" id="toc-parse-table-to-make-it-easier-to-work-with-it" class="nav-link" data-scroll-target="#parse-table-to-make-it-easier-to-work-with-it"><span class="toc-section-number">1.30.3</span>  Parse table to make it easier to work with it</a></li>
  </ul></li>
  <li><a href="#make-count-tables" id="toc-make-count-tables" class="nav-link" data-scroll-target="#make-count-tables"><span class="toc-section-number">1.31</span>  Make count tables</a>
  <ul class="collapse">
  <li><a href="#generate-a-count-table-for-our-genomes-of-interest" id="toc-generate-a-count-table-for-our-genomes-of-interest" class="nav-link" data-scroll-target="#generate-a-count-table-for-our-genomes-of-interest"><span class="toc-section-number">1.31.1</span>  Generate a count table for our genomes of interest</a></li>
  <li><a href="#generate-a-count-table-for-our-clusters-of-interest" id="toc-generate-a-count-table-for-our-clusters-of-interest" class="nav-link" data-scroll-target="#generate-a-count-table-for-our-clusters-of-interest"><span class="toc-section-number">1.31.2</span>  Generate a count table for our clusters of interest</a></li>
  </ul></li>
  <li><a href="#merge-our-tables-with-the-mapping-data-we-have" id="toc-merge-our-tables-with-the-mapping-data-we-have" class="nav-link" data-scroll-target="#merge-our-tables-with-the-mapping-data-we-have"><span class="toc-section-number">1.32</span>  Merge our tables with the mapping data we have</a>
  <ul class="collapse">
  <li><a href="#for-the-bins" id="toc-for-the-bins" class="nav-link" data-scroll-target="#for-the-bins"><span class="toc-section-number">1.32.1</span>  For the bins</a></li>
  <li><a href="#for-the-clusters" id="toc-for-the-clusters" class="nav-link" data-scroll-target="#for-the-clusters"><span class="toc-section-number">1.32.2</span>  For the clusters</a></li>
  </ul></li>
  <li><a href="#plot-the-data-for-our-genes-of-interest" id="toc-plot-the-data-for-our-genes-of-interest" class="nav-link" data-scroll-target="#plot-the-data-for-our-genes-of-interest"><span class="toc-section-number">1.33</span>  Plot the data for our genes of interest</a>
  <ul class="collapse">
  <li><a href="#categorizing-data" id="toc-categorizing-data" class="nav-link" data-scroll-target="#categorizing-data"><span class="toc-section-number">1.33.1</span>  Categorizing data</a></li>
  <li><a href="#order-our-data" id="toc-order-our-data" class="nav-link" data-scroll-target="#order-our-data"><span class="toc-section-number">1.33.2</span>  Order our data</a></li>
  <li><a href="#plotting-1" id="toc-plotting-1" class="nav-link" data-scroll-target="#plotting-1"><span class="toc-section-number">1.33.3</span>  Plotting</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction into R</span></h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<p>R is a statistical programming language and environment and free, open source and in active development. This tutorial will introduce into the basic concepts of R.</p>
<p>This tutorial will work with example data for two datasets:</p>
<p><strong>1. Growth data</strong></p>
<p>We have two data files that work with a similar experimental setup:</p>
<p>1.1. <strong>Growth_Data.txt</strong></p>
<p>During this experiments, we are doing plant growth experiments and treated our plants with different microbes wondering if any microbe affects plant growth in a positive way.</p>
<p>This file contains measurements of root length and shoot fresh weight for plants grown under control treatments (=MgCL) or when treating with 4 different bacteria. For simplicity only 1 biological experiment with 7-10 individual measurements per treatment were included.</p>
<p>1.2. <strong>Timecourse.txt</strong></p>
<p>We found positive effects for some of the strains tested above and now we want to know how long it takes for this effect to appear. To answer this, we measured root length of our plants when adding our microbe and compared it to control treatments at 5 different time points. Something unique with this dataframe is that we have empty cells (=NAs) and we need to deal with them as some R functions don’t like empty cells.</p>
<section id="the-example-data" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="the-example-data"><span class="header-section-number">1.1</span> The example data</h2>
<p>Let’s have a look at the data structure for our first dataframe:</p>
<div class="cell">
<pre><code>## list()</code></pre>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["fct"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["fct"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["fct"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The timecourse data looks similar, we just have an extra column for the different timepoints and we only have measurements for the root length.</p>
<p><strong>2. Annotation data</strong></p>
<p>This file is specific to the output of the <a href="https://github.com/ndombrowski/Annotation_workflow">Spang_team annotation pipeline</a> but this workflow can be used for any type of categorical data one wants to summarize.</p>
<p>Using this workflow, we generated the file <strong>UAP2_Annotation_table_u.txt</strong>, which includes annotations for a set of 46 DPANN genomes. This includes annotations across several databases (arCOG, KO, PFAM, …) for each individual protein found across all these 46 genomes.</p>
<p>Specifically, we want to learn how to:</p>
<ul>
<li>Make a count table for each genome</li>
<li>Make a count table for clusters of interest</li>
<li>Make a heatmap for genes of interest</li>
<li>Merge our results with some pre-sorted tables</li>
</ul>
<p>For this to work we have some additional files to make our job easier:</p>
<ul>
<li><strong>mapping.txt</strong> = a list that defines to what cluster (i.e.&nbsp;grouping based on a phylogenetic tree) or bins belong to</li>
<li><strong>Genes_of_interest</strong> = a list of genes we are interested in and that we want to plot in a heatmap</li>
<li><strong>ar14_arCOGdef19.txt</strong> = metadata for the arCOG annotations</li>
<li><strong>Metabolism_Table_KO_Apr2020.txt</strong> = metadata for KOs and sorted by pathways</li>
</ul>
<p>The annotation table looks like this:</p>
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 4%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 4%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 5%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 3%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 3%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 6%">
<col style="width: 15%">
<col style="width: 1%">
<col style="width: 3%">
<col style="width: 7%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">accession</th>
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">TaxString</th>
<th style="text-align: left;">NewContigID</th>
<th style="text-align: left;">OldContigId</th>
<th style="text-align: left;">ContigIdMerge</th>
<th style="text-align: right;">ContigNewLength</th>
<th style="text-align: right;">GC</th>
<th style="text-align: left;">ProteinID</th>
<th style="text-align: right;">ProteinGC</th>
<th style="text-align: right;">ProteinLength</th>
<th style="text-align: left;">Prokka</th>
<th style="text-align: left;">arcogs</th>
<th style="text-align: left;">arcogs_geneID</th>
<th style="text-align: left;">arcogs_Description</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">arcogs_evalue</th>
<th style="text-align: left;">KO_hmm</th>
<th style="text-align: left;">e_value</th>
<th style="text-align: left;">bit_score</th>
<th style="text-align: left;">bit_score_cutoff</th>
<th style="text-align: left;">Definition</th>
<th style="text-align: left;">confidence</th>
<th style="text-align: left;">PFAM_hmm</th>
<th style="text-align: left;">PFAM_description</th>
<th style="text-align: left;">Pfam_Evalue</th>
<th style="text-align: left;">TIRGR</th>
<th style="text-align: left;">TIGR_description</th>
<th style="text-align: left;">EC</th>
<th style="text-align: left;">TIGR_Evalue</th>
<th style="text-align: left;">CAZy</th>
<th style="text-align: left;">CAZy_evalue</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">TDBD_ID</th>
<th style="text-align: left;">TPDB_evalue</th>
<th style="text-align: left;">HydDB</th>
<th style="text-align: left;">Description.1</th>
<th style="text-align: left;">HydDB_evalue</th>
<th style="text-align: left;">PFAM</th>
<th style="text-align: left;">PFAMdescription</th>
<th style="text-align: left;">IPR</th>
<th style="text-align: left;">IPRdescription</th>
<th style="text-align: left;">TopHit</th>
<th style="text-align: left;">E_value</th>
<th style="text-align: left;">PecID</th>
<th style="text-align: left;">TaxID</th>
<th style="text-align: left;">TaxString.1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00010</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00010</td>
<td style="text-align: right;">0.094</td>
<td style="text-align: right;">350</td>
<td style="text-align: left;">Digeranylgeranylglycerophospholipid reductase</td>
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Geranylgeranyl_reductase,_flavoprotein”</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">2.80E-75</td>
<td style="text-align: left;">K17830</td>
<td style="text-align: left;">2.10E-69</td>
<td style="text-align: left;">243.3</td>
<td style="text-align: left;">246.87</td>
<td style="text-align: left;">digeranylgeranylglycerophospholipid_reductase_[EC:1.3.1.101_1.3.7.11]</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">FAD_binding_domain</td>
<td style="text-align: left;">5.90E-09</td>
<td style="text-align: left;">TIGR02032</td>
<td style="text-align: left;">geranylgeranyl_reductase_family</td>
<td style="text-align: left;">1.3.1.-</td>
<td style="text-align: left;">1.40E-49</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR036188;_IPR002938</td>
<td style="text-align: left;">FAD/NAD(P)-binding_domain_superfamily;_FAD-binding_domain</td>
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">FAD_binding_domain</td>
<td style="text-align: left;">KYK22416.1_hypothetical_protein_AYK24_02275_[Thermoplasmatales_archaeon_SG8-52-4]</td>
<td style="text-align: left;">2.80E-58</td>
<td style="text-align: left;">36.5</td>
<td style="text-align: left;">1803819</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Thermoplasmata,Thermoplasmatales,none,none,Thermoplasmatales_archaeon_SG8-52-4”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00020</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00020</td>
<td style="text-align: right;">0.057</td>
<td style="text-align: right;">105</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00030</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00030</td>
<td style="text-align: right;">0.092</td>
<td style="text-align: right;">304</td>
<td style="text-align: left;">tRNA-2-methylthio-N(6)-dimethylallyladenosine synthase</td>
<td style="text-align: left;">arCOG01358</td>
<td style="text-align: left;">MiaB</td>
<td style="text-align: left;">2-methylthioadenine_synthetase</td>
<td style="text-align: left;">J</td>
<td style="text-align: left;">5.80E-90</td>
<td style="text-align: left;">K15865</td>
<td style="text-align: left;">1.20E-109</td>
<td style="text-align: left;">375.8</td>
<td style="text-align: left;">343.7</td>
<td style="text-align: left;">threonylcarbamoyladenosine_tRNA_methylthiotransferase_CDKAL1_[EC:2.8.4.5]</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF04055</td>
<td style="text-align: left;">Radical_SAM_superfamily</td>
<td style="text-align: left;">2.60E-21</td>
<td style="text-align: left;">TIGR01578</td>
<td style="text-align: left;">“MiaB-like_tRNA_modifying_enzyme,_archaeal-type”</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">1.70E-104</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR005839;_IPR002792;_IPR006466;_IPR006638;_IPR007197;_IPR020612;_IPR023404</td>
<td style="text-align: left;">“Methylthiotransferase;_TRAM_domain;_MiaB-like_tRNA_modifying_enzyme,_archaeal-type;_Elp3/MiaB/NifB;_Radical_SAM;_Methylthiotransferase,_conserved_site;_Radical_SAM,_alpha/beta_horseshoe”</td>
<td style="text-align: left;">PF01938;_PF04055</td>
<td style="text-align: left;">TRAM_domain;_Radical_SAM_superfamily</td>
<td style="text-align: left;">OIO63284.1_hypothetical_protein_AUJ83_01460_[Candidatus_Woesearchaeota_archaeon_CG1_02_33_12]</td>
<td style="text-align: left;">8.20E-91</td>
<td style="text-align: left;">55.6</td>
<td style="text-align: left;">1805422</td>
<td style="text-align: left;">“Archaea,Candidatus_Woesearchaeota,none,none,none,none,Candidatus_Woesearchaeota_archaeon_CG1_02_33_12”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00040</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00040</td>
<td style="text-align: right;">0.065</td>
<td style="text-align: right;">92</td>
<td style="text-align: left;">Enolase</td>
<td style="text-align: left;">arCOG01169</td>
<td style="text-align: left;">Eno</td>
<td style="text-align: left;">Enolase</td>
<td style="text-align: left;">G</td>
<td style="text-align: left;">4.00E-26</td>
<td style="text-align: left;">K01689</td>
<td style="text-align: left;">5.10E-20</td>
<td style="text-align: left;">81</td>
<td style="text-align: left;">48.73</td>
<td style="text-align: left;">enolase_[EC:4.2.1.11]</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain”</td>
<td style="text-align: left;">1.20E-20</td>
<td style="text-align: left;">TIGR01060</td>
<td style="text-align: left;">phosphopyruvate_hydratase</td>
<td style="text-align: left;">4.2.1.11</td>
<td style="text-align: left;">1.10E-25</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR020810;_IPR036849;_IPR000941;_IPR020809</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain;_Enolase-like,_C-terminal_domain_superfamily;_Enolase;_Enolase,_conserved_site”</td>
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain”</td>
<td style="text-align: left;">BAW30993.1_2-phosphoglycerate_dehydratase_[Methanothermobacter_sp._MT-2]</td>
<td style="text-align: left;">3.80E-23</td>
<td style="text-align: left;">70.7</td>
<td style="text-align: left;">1898379</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Methanobacteria,Methanobacteriales,Methanobacteriaceae,Methanothermobacter,Methanothermobacter_sp._MT-2”</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00050</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00050</td>
<td style="text-align: right;">0.059</td>
<td style="text-align: right;">218</td>
<td style="text-align: left;">30S ribosomal protein S2</td>
<td style="text-align: left;">arCOG04245</td>
<td style="text-align: left;">RpsB</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">J</td>
<td style="text-align: left;">8.90E-64</td>
<td style="text-align: left;">K02998</td>
<td style="text-align: left;">1.60E-46</td>
<td style="text-align: left;">167.8</td>
<td style="text-align: left;">210.97</td>
<td style="text-align: left;">small_subunit_ribosomal_protein_SAe</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">7.00E-24</td>
<td style="text-align: left;">TIGR01012</td>
<td style="text-align: left;">ribosomal_protein_uS2</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">1.00E-72</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR005707;_IPR023454;_IPR023591;_IPR018130;_IPR001865</td>
<td style="text-align: left;">“Ribosomal_protein_S2,_eukaryotic/archaeal;_Ribosomal_protein_S2,_archaeal;_Ribosomal_protein_S2,_flavodoxin-like_domain_superfamily;_Ribosomal_protein_S2,_conserved_site;_Ribosomal_protein_S2”</td>
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">A0B6E5.1_RecName:_Full=30S_ribosomal_protein_S2</td>
<td style="text-align: left;">1.20E-56</td>
<td style="text-align: left;">52.8</td>
<td style="text-align: left;">349307</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Methanomicrobia,Methanosarcinales,Methanotrichaceae,Methanothrix,Methanothrix_thermoacetophila”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00060</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00060</td>
<td style="text-align: right;">0.100</td>
<td style="text-align: right;">280</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">arCOG01728</td>
<td style="text-align: left;">Mho1</td>
<td style="text-align: left;">“Predicted_class_III_extradiol_dioxygenase,_MEMO1_family”</td>
<td style="text-align: left;">R</td>
<td style="text-align: left;">8.80E-74</td>
<td style="text-align: left;">K06990</td>
<td style="text-align: left;">1.40E-72</td>
<td style="text-align: left;">253.3</td>
<td style="text-align: left;">47</td>
<td style="text-align: left;">MEMO1_family_protein</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">Memo-like_protein</td>
<td style="text-align: left;">4.50E-77</td>
<td style="text-align: left;">TIGR04336</td>
<td style="text-align: left;">AmmeMemoRadiSam_system_protein_B</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">5.30E-91</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR002737</td>
<td style="text-align: left;">MEMO1_family</td>
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">Memo-like_protein</td>
<td style="text-align: left;">OYT50994.1_AmmeMemoRadiSam_system_protein_B_[Candidatus_Bathyarchaeota_archaeon_ex4484_135]</td>
<td style="text-align: left;">1.30E-66</td>
<td style="text-align: left;">47</td>
<td style="text-align: left;">2012509</td>
<td style="text-align: left;">“Archaea,Candidatus_Bathyarchaeota,none,none,none,none,Candidatus_Bathyarchaeota_archaeon_ex4484_135”</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="rstudio-everything-in-one-place" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="rstudio-everything-in-one-place"><span class="header-section-number">1.2</span> RStudio (everything in one place):</h2>
<p>R is command-line only while RStudio is a GUI ((graphical user interface)) version of R.</p>
<p>RStudio includes the following:</p>
<ul>
<li>Script separate from command-line (left-hand screen)</li>
<li>Lists your variables (upper, right-hand corner)</li>
<li>Manual and an extensive help function</li>
<li>Easy install of new packages</li>
<li>Plots are shown within RStudio</li>
</ul>
<p align="left">
<img height="400" src="../images/RStudio.png">
</p>
</section>
<section id="opening-r-via-the-terminal" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="opening-r-via-the-terminal"><span class="header-section-number">1.3</span> Opening R via the terminal</h2>
<p>If you work with linux or want to start R from the terminal then open your terminal, change your directory to the R_exercises folder and then just type <code>R</code>.</p>
<p>Then, you should see something like this:</p>
<p align="left">
<img height="300" src="../images/R.png">
</p>
<p>To check your R version, start and quit R you can do the following:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Ask what R version we have</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>R.version</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#start R</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>R</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co">#exit R</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">q</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="documenting-code-via-markdown" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="documenting-code-via-markdown"><span class="header-section-number">1.4</span> Documenting code via markdown</h2>
<section id="software" class="level3" data-number="1.4.1">
<h3 data-number="1.4.1" class="anchored" data-anchor-id="software"><span class="header-section-number">1.4.1</span> Software</h3>
<section id="rmarkdown" class="level4" data-number="1.4.1.1">
<h4 data-number="1.4.1.1" class="anchored" data-anchor-id="rmarkdown"><span class="header-section-number">1.4.1.1</span> Rmarkdown</h4>
<p>One nice way to commend code is combine R-code with informative text in R markdown format (rmd).</p>
<p>R Markdown supports dozens of static and dynamic output formats including: HTML, PDF, MS Word, Beamer, HTML5 slides, Tufte-style handouts, books, dashboards, shiny applications, scientific articles, websites, and more.</p>
<p>The R markdown file specifies code chunks which will be executed in R (or python or bash) and plain text which will be written to the report as is. A report is created by rendering the file in R, then the R-code is executed and the results are merged in a pdf or html output.</p>
<p>How you create a Rmarkdown document in R:</p>
<ul>
<li>Open RStudio</li>
<li>press <code>File/new File/R markdown</code></li>
</ul>
<p>This will create an R markdown file that already contains some R code and text. You can also open this document (the rmd file) in RStudio and see how the code looks.</p>
<p>You can also open a new file in any text editor and save it with the .rmd extension.</p>
</section>
<section id="quarto" class="level4" data-number="1.4.1.2">
<h4 data-number="1.4.1.2" class="anchored" data-anchor-id="quarto"><span class="header-section-number">1.4.1.2</span> Quarto</h4>
<p><a href="https://quarto.org">Quarto<c2><ae></ae></c2></a> is the predecessor of RMarkdown and is an open-source scientific and technical publishing system built on Pandoc and allows to:</p>
<ul>
<li>Create dynamic content with Python, R, Julia, and Observable.</li>
<li>Author documents as plain text markdown or Jupyter notebooks.</li>
<li>Publish high-quality articles, reports, presentations, websites, blogs, and books in HTML, PDF, MS Word, ePub, and more.</li>
<li>Author with scientific markdown, including equations, citations, crossrefs, figure panels, callouts, advanced layout, and more.</li>
</ul>
<p>If you installed the newest version of RStudio, Quarto is already installed and we can create a quarto document with</p>
<ul>
<li>Open RStudio</li>
<li>press <code>File/New File/Quarto document</code></li>
</ul>
<p>Same as with RMarkdown, we document with Markdown (and HTML if we want), so knowing some basics is very useful.</p>
</section>
</section>
<section id="r-code-in-markdown" class="level3" data-number="1.4.2">
<h3 data-number="1.4.2" class="anchored" data-anchor-id="r-code-in-markdown"><span class="header-section-number">1.4.2</span> R code in Markdown</h3>
<ul>
<li>The R-code is embedded in between the<br>
<code>```{r}</code> and <code>```</code> symbols.</li>
</ul>
<p>In Rstudio, on the top-right position of such a section you will find three symbols. Pressing the middle one will run all code chunks above, while the right symbol will run the current R-chunk.</p>
<p align="left">
<img width="800" src="../images/R2.png">
</p>
<p>An important menu-button is the <strong>Knit</strong> or <strong>Render</strong> button at the top, left-hand corner of RStudio. Pressing this button will create the final, rendered document (i.e.&nbsp;a HTML or PDF)</p>
</section>
<section id="text-in-markdown" class="level3" data-number="1.4.3">
<h3 data-number="1.4.3" class="anchored" data-anchor-id="text-in-markdown"><span class="header-section-number">1.4.3</span> Text in markdown</h3>
<ul>
<li>Headings are defined with ‘#’, or ‘##’, or ‘###’ for first, second and third level.</li>
<li>Lists are created by using ’*’ for (bullets) and ‘1’, <e2>&lt;80&gt;<a6> for numbered lists.</a6></e2></li>
<li>A lot of other options are possible, more information on these can be found in the following documents.
<ul>
<li>The Rmarkdown website: <strong>https://rmarkdown.rstudio.com</strong></li>
<li>You can find more R cheat sheets on <strong>https://rstudio.com/resources/cheatsheets/</strong></li>
<li>Another info sheet with info how to modify code: <strong>https://www.rstudio.com/wp-content/uploads/2016/03/rmarkdown-cheatsheet-2.0.pdf</strong></li>
</ul></li>
</ul>
</section>
<section id="execution-options" class="level3" data-number="1.4.4">
<h3 data-number="1.4.4" class="anchored" data-anchor-id="execution-options"><span class="header-section-number">1.4.4</span> Execution options</h3>
<p>There are a wide variety of options available for customizing output from executed code</p>
<ul>
<li>include = FALSE — prevents code and results from appearing in the finished file. R Markdown still runs the code in the chunk, and the results can be used by other chunks.</li>
<li>echo = FALSE — prevents code, but not the results from appearing in the finished file. This is a useful way to embed figures.</li>
<li>message = FALSE — prevents messages that are generated by code from appearing in the finished file.</li>
<li>warning = FALSE — prevents warnings that are generated by code from appearing in the finished.</li>
<li>fig.cap = “…” — adds a caption to graphical results.</li>
</ul>
<p>For setting these options inside a quarto document, see more <a href="https://quarto.org/docs/computations/execution-options.html">here</a>.</p>
</section>
<section id="using-languages-other-than-r" class="level3" data-number="1.4.5">
<h3 data-number="1.4.5" class="anchored" data-anchor-id="using-languages-other-than-r"><span class="header-section-number">1.4.5</span> Using languages other than R</h3>
<p>R Markdown/Quarto support several languages, such as bash and python, and you can call them in the same way as R code.</p>
<p>This is useful if you for example modify a dataframe in bash but then want continue to work on the data R. With proper documenting you can document the code in the same file.</p>
<p>Below is just an example, we see that we only need to “tell” R to use bash instead of R</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#run echo to print something to the screen</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'hello world'</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">#run echo and follow with a sed to modify text</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'a b c'</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/ /\|/g'</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co">#list qmd files we have in our directory</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> <span class="pp">*</span>qmd</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co">#show whats in our data</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> ../data/Growth_Data.txt </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co">## hello world</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co">## a|b|c</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co">## -rw-r--r--@ 1 ninadombrowski  staff  118003 Oct  5 12:26 1_intro.qmd</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co">## -rw-r--r--@ 1 ninadombrowski  staff     239 Sep 29 15:31 2_misc.qmd</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="co">## -rw-r--r--@ 1 ninadombrowski  staff    4015 Oct  5 10:20 test.qmd</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co">## SampleID Nutrient    Condition   FW_shoot_mg     Rootlength</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    10.26   5.931015152</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    6.52    5.74344697</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    12.17   6.834719697</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    11.37   6.742734848</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    9.8 6.736886364</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    3.75    4.236348485</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    5.38    4.753484848</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    4.53    5.532333333</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co">## noP  noP MgCl    7.75    5.484363636</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A general introduction into bash and awk is provided in two separate tutorials that are also available on github.</p>
</section>
</section>
<section id="good-practices-for-coding" class="level2" data-number="1.5">
<h2 data-number="1.5" class="anchored" data-anchor-id="good-practices-for-coding"><span class="header-section-number">1.5</span> Good practices for coding</h2>
<p>These practices are useful regardless of the computational language you use.</p>
<ul>
<li>Record what program versions you used</li>
<li>For each project, document who wrote the code, when you did it and why</li>
<li>Put dependencies in the beginning (ie packages)</li>
<li>Record the working directory (wdir)</li>
<li>Document ALL your code and comment it (using the # symbol)</li>
<li>Comment code in detail, so that you can still understand it after 5 years</li>
<li>Break code into smaller pieces for better readability</li>
<li>Test each line of code and build in control steps</li>
<li>If you work with random numbers, report the seed</li>
<li>Use sessionInfo() at the end of the script, which documents all the packages used within R for the current project</li>
<li>For larger files: save objects not workspaces (for space reasons)</li>
<li>Have descriptive names for objects, short and simple but easy enough to understand what they mean</li>
</ul>
</section>
<section id="general-introduction-into-r" class="level2" data-number="1.6">
<h2 data-number="1.6" class="anchored" data-anchor-id="general-introduction-into-r"><span class="header-section-number">1.6</span> General introduction into R</h2>
<section id="getting-help" class="level3" data-number="1.6.1">
<h3 data-number="1.6.1" class="anchored" data-anchor-id="getting-help"><span class="header-section-number">1.6.1</span> Getting help</h3>
<ol type="1">
<li>Some good places to check for things online are:</li>
</ol>
<ul>
<li>www.r-project.org <em>stack overflow </em>many more</li>
</ul>
<ol start="2" type="1">
<li>Inside of R, we can get help on functions and other things with:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help</span>(mean)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>?mean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="what-is-a-workspace" class="level3" data-number="1.6.2">
<h3 data-number="1.6.2" class="anchored" data-anchor-id="what-is-a-workspace"><span class="header-section-number">1.6.2</span> What is a workspace?</h3>
<p>The workspace is your current R working environment and includes any user-defined objects (i.e.&nbsp;vectors, matrices, data frames, lists, functions).</p>
<p>At the end of an R session, the user can save an image of the current workspace that is automatically reloaded the next time R is started.</p>
<ul>
<li>From now on the code is shown in the grey area.</li>
<li>If you see a line symbol, such as <code>[1]</code> then this shows the result of running our code.</li>
<li>For example, first we run the code <code>getwd()</code> and the second box shows us the results.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print the current working directory</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>() </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "/Users/ninadombrowski/Desktop/WorkingDir/Notebooks/Code_snippets/R/code"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co">#list the objects in the current workspace</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>()   </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "annotation_data" "growth_data"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="setting-your-working-directory-wdir" class="level3" data-number="1.6.3">
<h3 data-number="1.6.3" class="anchored" data-anchor-id="setting-your-working-directory-wdir"><span class="header-section-number">1.6.3</span> Setting your working directory (wdir)</h3>
<p>The directory from which you work is usually first set from where you start R. But it can be re-set to find your data more easily. Ideally, you make one wdir per project and define the path in the script (see later below). It is recommended to have a similar format for these project folders, i.e.&nbsp;consider to create subfolders for input and output files. From the wdir you set you can load files using absolute and relative paths.</p>
<p>An example would be something with a structure like this:</p>
<p align="left">
<img height="200" src="../images/FolderStructure.png">
</p>
<p>In this example you see that we have 4 projects, and in each folder we have the R_script and folders for the required input and required output files. Also useful to have is a text file with the session info and if you plan to share this with others it is also good to have a README file that provides some background on the analysis done.</p>
<p>Options to see the working dir and set the working directory in R are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print your wdir</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#setting your wdir</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="fu">getwd</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="packages" class="level3" data-number="1.6.4">
<h3 data-number="1.6.4" class="anchored" data-anchor-id="packages"><span class="header-section-number">1.6.4</span> Packages</h3>
<p>Packages are a collection of functions and tools that can be added to R that are often contributed by the community. Careful: there might be incompatibilities and packages are updated frequently but updating can break dependencies.</p>
<p>You need to install packages and load them EVERY TIME you want to use them. Therefore, ideally add them at the beginning of your scripts.</p>
<section id="installing-packages" class="level4" data-number="1.6.4.1">
<h4 data-number="1.6.4.1" class="anchored" data-anchor-id="installing-packages"><span class="header-section-number">1.6.4.1</span> Installing packages</h4>
<p>We have two ways to install packages:</p>
<ol type="1">
<li>Via the console by typing:</li>
</ol>
<p><code>install.packages("package-name")</code></p>
<p>This will download a package from one of the CRAN mirrors assuming that a binary is available for your operating system. If you have not set a preferred CRAN mirror in your options(), then a menu will pop up asking you to choose a location.</p>
<p>Use <code>old.packages()</code> to list all your locally installed packages that are now out of date. <code>update.packages()</code> will update all packages in the known libraries interactively. This can take a while if you haven<e2>&lt;80&gt;&lt;99&gt;t done it recently. To update everything without any user intervention, use the ask = FALSE argument.</e2></p>
<ol start="2" type="1">
<li>Using R studio:</li>
</ol>
<p>Go to the lower right hand-side window and click on packages and then install. Find the packages your are interested in.</p>
<p>Notice: If libraries come with their own data (i.e.&nbsp;example tables), then the data needs to be loaded separately. I.e. via <code>data(cars)</code> to load the cars data from the cars package.</p>
</section>
<section id="loading-packages-into-your-current-r-session" class="level4" data-number="1.6.4.2">
<h4 data-number="1.6.4.2" class="anchored" data-anchor-id="loading-packages-into-your-current-r-session"><span class="header-section-number">1.6.4.2</span> Loading packages into your current R session</h4>
<p>R will not remember what libraries you have loaded after you closed R. Therefore, you need to load necessary libraries every time. Here, we will load some libraries that are usually quite helpful and it is recommended to make the libraries you load part of each of your scripts. For example like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#some example packages needed for Rmarkdown</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(knitr)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kableExtra)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="the-assignment-operator--" class="level3" data-number="1.6.5">
<h3 data-number="1.6.5" class="anchored" data-anchor-id="the-assignment-operator--"><span class="header-section-number">1.6.5</span> The assignment operator <code>&lt;-</code></h3>
<p>The <code>&lt;-</code> symbol assigns a value to a variable,</p>
<p>General rules for the syntax R uses:</p>
<ul>
<li>R is case sensitive</li>
<li>If a variable exists, it will overwrite it with a new variable without asking</li>
<li>If you work with characters, i.e.&nbsp;words like ‘hello’, then this need to written with quotes around it: “hello” (this will become clearer below)</li>
<li>ls() shows all the variables that are known by the system at the moment</li>
<li>you can remove individual objects with <code>rm()</code> and remove all objects <code>rm(list=ls())</code></li>
</ul>
<p>We can store more or less everything in a variable and use it later. For example, we can store numbers and do some math with them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#store some numbers</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span><span class="dv">4</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#do some simple math with the numbers we have stored</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>x<span class="sc">+</span>y</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="use-build-in-functions" class="level3" data-number="1.6.6">
<h3 data-number="1.6.6" class="anchored" data-anchor-id="use-build-in-functions"><span class="header-section-number">1.6.6</span> Use build in functions</h3>
<p>Functions are build in code that we can use to make our life easier, i.e.&nbsp;we can calculate lengths of vectors, do math or do statistical analyses.</p>
<p>Base R already knows many useful functions but loading new packages greatly increases our repertoire.</p>
<p>A list of most used functions can be found <a href="**http://www.statmethods.net/management/functions.html">here</a></p>
<p>A function consists of:</p>
<ol type="1">
<li>Function name</li>
<li>Arguments (optional, some might be set with a default) = control how exactly the function behaves</li>
<li>Body of the function = defines what the function does</li>
</ol>
<p>As an example lets test some simple functions: print and log:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#use the print function</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="dv">3</span><span class="sc">+</span><span class="dv">5</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 8</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#use the log function</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">log</span>(<span class="dv">10</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2.302585</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="call-the-default-values-of-a-function" class="level4" data-number="1.6.6.1">
<h4 data-number="1.6.6.1" class="anchored" data-anchor-id="call-the-default-values-of-a-function"><span class="header-section-number">1.6.6.1</span> Call the default values of a function</h4>
<p>Every function comes with a set of arguments that you can set but that usually also have some default values. In R Studio you can easily access all those details with the help function.</p>
<ul>
<li><code>?</code> allows us to first of all check exactly what a function is doing. If you scroll down to the bottom of the help page you also get some examples on how to use a function.</li>
<li>More specifically the help function also allows us to get details on the arguments of a function.</li>
<li>For example, if we check the help page of <strong>read.table</strong> we see that by default this function does not read in a header and if we want to provide a header we have to change that argument.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#let's check what **log** is doing</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>?log</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#lets check the default arguments of a function</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>?read.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other useful functions:</p>
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">Name</th>
<th style="text-align: left;">Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ls()</td>
<td style="text-align: left;">List objects in your current workspace</td>
</tr>
<tr class="even">
<td style="text-align: left;">rm(object)</td>
<td style="text-align: left;">Remove object from your current workspace</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rm(list = ls())</td>
<td style="text-align: left;">Remove all objects from your current workspace</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
</section>
<section id="read-data-into-r" class="level2" data-number="1.7">
<h2 data-number="1.7" class="anchored" data-anchor-id="read-data-into-r"><span class="header-section-number">1.7</span> Read data into R</h2>
<p>To work with any kind of data we need to first read the data into R to be able to work with it.</p>
<p>For tables, there are some things to be aware of:</p>
<ul>
<li>It matters what separator our table uses to specify individual columns. I.e. some programs store data using commas while others use tab as default delimiter. By default R assume we use a tab, but we can change this behavior when we read in the table.</li>
<li>Do not have any hash symbols (#) in your table. R will read this as a commented cell and not read the table from that point onward</li>
<li>Avoid empty cells, as these sometimes can mess up your data.</li>
</ul>
<p>For now, let’s read in the table with our growth data and store it under the variable name <strong>growth_data</strong>. To read in this file we need to direct it to the correct path as we do not have the file in the working directory but in a subdirectory namd <code>data</code>.</p>
<p>Options that are good to keep in mind when reading in a table:</p>
<ul>
<li><code>sep</code> = define that our field separator is a tab. A tab is written like this <code>/t</code>. If your data is using a space or comma, you can change that here.</li>
<li><code>header</code> = tell R that our data comes with a custom header (the first row in your dataframe)</li>
<li><code>quote</code> = deals with some annoying issue with data formatting in excel files</li>
</ul>
<p>General notice:</p>
<ul>
<li>To view data, the <code>head()</code> command is extremely practical, use it always when you modify data to check if everything went alright</li>
<li><code>dim()</code> is another useful function that displays the dimensions of a table, i.e.&nbsp;how many rows and columns we have. Again, this is useful to verify our data after we have transformed it to check if everything went alright.</li>
<li><code>colnames()</code> allows to only we the column names</li>
<li><code>rownames()</code> allows to only we the row names. Usually these are numbers, but we can also add anything else into the rows.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#read in data</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>timecourse <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/Timecourse.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T,  <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>growth_data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/Growth_Data.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T,  <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#check the first few lines of our data</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the dimensions of our data</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 105   5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the column names</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(growth_data)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "SampleID"    "Nutrient"    "Condition"   "FW_shoot_mg" "Rootlength"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the row names</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(growth_data)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   [1] "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12" </span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  [13] "13"  "14"  "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24" </span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  [25] "25"  "26"  "27"  "28"  "29"  "30"  "31"  "32"  "33"  "34"  "35"  "36" </span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  [37] "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44"  "45"  "46"  "47"  "48" </span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  [49] "49"  "50"  "51"  "52"  "53"  "54"  "55"  "56"  "57"  "58"  "59"  "60" </span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="do">##  [61] "61"  "62"  "63"  "64"  "65"  "66"  "67"  "68"  "69"  "70"  "71"  "72" </span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="do">##  [73] "73"  "74"  "75"  "76"  "77"  "78"  "79"  "80"  "81"  "82"  "83"  "84" </span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="do">##  [85] "85"  "86"  "87"  "88"  "89"  "90"  "91"  "92"  "93"  "94"  "95"  "96" </span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="do">##  [97] "97"  "98"  "99"  "100" "101" "102" "103" "104" "105"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Useful comments:</p>
<p>Sometimes we have to deal with really large data that take long to load with read.table. The function <code>fread()</code> from the data.table package is a very nice alternative. Below, we will explain how to read in R libraries to use additional functions.</p>
<p><strong>This script always uses kable for plotting, which is just to make it visually attractive in html. Whenever you see a function using kable simply replace it with the <code>head()</code> function, i.e.&nbsp;write <code>head(growth_data)</code></strong></p>
</section>
<section id="store-data-from-r-on-your-computer-as-a-text-file" class="level2" data-number="1.8">
<h2 data-number="1.8" class="anchored" data-anchor-id="store-data-from-r-on-your-computer-as-a-text-file"><span class="header-section-number">1.8</span> Store data from R on your computer as a text file</h2>
<p>Now, if we would have modified the table we might want to store it on your computer. We can do this using <code>write.table()</code> and below we use a different output directory. Notice, we always start from the location we set as working directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(growth_data, <span class="st">"../output_examples/growth_data_changed.txt"</span>,  <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> T, <span class="at">quote =</span>F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Arguments:</p>
<ul>
<li><code>sep</code> –&gt; we define what delimiter we want to use</li>
<li><code>row.names</code> = T –&gt; we want to include whatever is in the rownames</li>
<li><code>quote</code> = F –&gt; we do not want R to add any quotes around our columns.</li>
</ul>
</section>
<section id="what-are-objects" class="level2" data-number="1.9">
<h2 data-number="1.9" class="anchored" data-anchor-id="what-are-objects"><span class="header-section-number">1.9</span> What are objects?</h2>
<p>In R, all types of data are treated as objects. As such objects are units that we work with, i.e.&nbsp;data and functions.</p>
<p>Roughly we distinguish between:</p>
<ul>
<li>Data types = For example: numeric, characters, factors, … You can identify your data type with the function <code>typeof()</code></li>
<li>Data structures = For example: vector, matrix, list. You can identify the structure with <code>class()</code> or <code>is()</code></li>
<li><code>str()</code> = defines what the indiv. components of your data are</li>
</ul>
<p>Everything that exists is an <strong>object</strong>. Everything that happens is a <strong>function call</strong>.</p>
<p>Below, we introduce all these different types of objects.</p>
</section>
<section id="data-classes" class="level2" data-number="1.10">
<h2 data-number="1.10" class="anchored" data-anchor-id="data-classes"><span class="header-section-number">1.10</span> Data classes</h2>
<p>When programming, data,values,etc. are stored in different ways:</p>
<p>R has 6 atomic classes. Below you can find each class with an example</p>
<ul>
<li>character = “hello”</li>
<li>numeric (real or decimal) = 3, 14, ….</li>
<li>logical = TRUE</li>
<li>complex = 1+4i</li>
<li>integer = 2 (Must add a L at end to denote integer)</li>
<li>double = a number class, like the integer but with double precision floating points</li>
</ul>
<p>Here, is a quick example, how we can find out some things about our objects using:</p>
<ul>
<li><ul>
<li><code>c()</code>= a function that will create a vector (a one dimensional array) and in our case store 3 numbers. We need to use this every time we deal with more than one number, character, etc….</li>
</ul></li>
<li><code>class()</code> = what class is our data?</li>
<li><code>length()</code> = how long is our data?</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create some objects</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>character_object <span class="ot">&lt;-</span> <span class="st">"dataset"</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>number_object <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">5</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co">#asking with what type we work</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(character_object)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "character"</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(number_object)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "numeric"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#ask how long our objects are</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(character_object)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 1</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(number_object)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-structures" class="level2" data-number="1.11">
<h2 data-number="1.11" class="anchored" data-anchor-id="data-structures"><span class="header-section-number">1.11</span> Data structures</h2>
<p>There are many types of data structures, the most frequently used ones being:</p>
<ul>
<li>Vectors</li>
<li>Factors</li>
<li>Matrices</li>
<li>Lists</li>
<li>Data frames</li>
</ul>
<p>Certain operations only work on certain kind of structures, therefore, it is important to know what kind of data we are working with.</p>
<p>In R, you do not need to specify the type of data a variable will receive beforehand. You simply do the assignment, R will create a so called R-Object and assign a data type automatically.</p>
<section id="vectors" class="level3" data-number="1.11.1">
<h3 data-number="1.11.1" class="anchored" data-anchor-id="vectors"><span class="header-section-number">1.11.1</span> Vectors</h3>
<p>A vector is a collection of items of the same type (i.e characters, numbers). You can read in numbers and characters into the same vector, however, the number will be then seen as a character if you mix different classes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#lets create a random vector</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>a_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">7</span>, <span class="dv">1</span>) </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">#show the vector we just created</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>a_vector</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2 3 5 7 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#asking how long your vector is</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(a_vector)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="vector-indexing" class="level4" data-number="1.11.1.1">
<h4 data-number="1.11.1.1" class="anchored" data-anchor-id="vector-indexing"><span class="header-section-number">1.11.1.1</span> Vector indexing</h4>
<p>If we want to only retrieve part of the data stored in a vector we can create a subset using the index as shown below.</p>
<ul>
<li>square brackets [] = allow us to retrieve certain elements of a vector, i.e.&nbsp;[3] retrieves the 3rd element</li>
<li>we can combine <code>c()</code> and [] if we want to retrieve several elements of a vector.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#retrieve the third element stored in a vector</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>a_vector[<span class="dv">3</span>]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#retrieve the 1st and 3rd element by combining ``c()`` and []</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>a_vector[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)]</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#retrieve the 1-3rd element</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>a_vector[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)]</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2 3 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we can also add vectors of the same length together</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co">#and now we can combine our vectors</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>x <span class="sc">+</span> y</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2 4 6 8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Beware: If we add two vectors of different length, the shorter vector is duplicated. This only works if the shorter vector is proportional to the longer one</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#adding vectors of different lengths</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#and now we can combine our vectors</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>x <span class="sc">+</span> y</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2 4 4 6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Another way to extend vectors is:</p>
<ul>
<li><code>append()</code> –&gt; Add elements to a vector.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#add another datapoint to our vector</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>a_vector <span class="ot">&lt;-</span> <span class="fu">append</span>(a_vector, <span class="dv">13</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>a_vector</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1]  2  3  5  7  1 13</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#add +1 to all our four numbers</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>a_vector <span class="ot">&lt;-</span> a_vector <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>a_vector</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1]  3  4  6  8  2 14</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#remove the first element of our vector</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>a_vector <span class="ot">&lt;-</span> a_vector[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>a_vector</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1]  4  6  8  2 14</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We not only can extract the nth element but if we have header names then we can also use these to retrieve data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create a vector and give it names (i.e. for counts from some microbes)</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">300</span>, <span class="dv">410</span>, <span class="dv">531</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(x) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ecoli"</span>,<span class="st">"Archaeoglobus"</span>,<span class="st">"Ignicoccus"</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co">#check how our data looks</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>x</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="do">##         Ecoli Archaeoglobus    Ignicoccus </span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="do">##           300           410           531</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#now we can retrieve part of the vector using the names</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>x[<span class="fu">c</span>(<span class="st">"Ecoli"</span>,<span class="st">"Ignicoccus"</span>)]</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="do">##      Ecoli Ignicoccus </span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="do">##        300        531</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="changing-vectors" class="level4" data-number="1.11.1.2">
<h4 data-number="1.11.1.2" class="anchored" data-anchor-id="changing-vectors"><span class="header-section-number">1.11.1.2</span> Changing vectors</h4>
<p>We can also change elements in our vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create a vector</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="co">#change the second last positions to 5 and 9</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>x[<span class="dv">9</span><span class="sc">:</span><span class="dv">10</span>] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">9</span>)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co">#check if this worked</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>x</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] 1 2 3 4 5 6 7 8 5 9</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we can not only add things, we can also remove this using the minus/**-** symbol</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>x[<span class="sc">-</span><span class="dv">3</span>]</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 1 2 4 5 6 7 8 5 9</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">#if we want to remove more than one thing we can use the **c()**</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>x[<span class="sc">-</span>(<span class="dv">4</span><span class="sc">:</span><span class="dv">10</span>)]</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 1 2 3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="matrix" class="level3" data-number="1.11.2">
<h3 data-number="1.11.2" class="anchored" data-anchor-id="matrix"><span class="header-section-number">1.11.2</span> Matrix</h3>
<p>Matrices are the R objects in which the elements are arranged in a two-dimensional rectangular layout. They contain elements of the same type. Although you can construct matrices with characters or logicals, matrices are generally used to store numeric data.</p>
<p>The basic syntax for creating a matrix is:</p>
<p><strong>matrix(data, nrow, ncol, byrow, dimnames)</strong></p>
<ul>
<li>data: input vector whose components become the data elements from the matrix.</li>
<li>dnrow: number of rows to be created.</li>
<li>dncol: number of columns to be created.</li>
<li>dbyrow: logical. If FALSE,(the default) the matrix is filled by columns, otherwise the matrix is filled by rows.</li>
<li>ddimnames: A <e2>&lt;80&gt;&lt;98&gt;dimnames<e2>&lt;80&gt;&lt;99&gt; attribute for the matrix: NULL or a list of length 2 giving the row and column names respectively.</e2></e2></li>
</ul>
<p>In contrast in a data frame (see below) the columns contain different types of data, while in a matrix all the elements are the same type of data. A matrix in R is like a mathematical matrix, containing all the same type of thing (usually numbers). R often but not always can use dataframes and a matrix used interchangably.</p>
<ul>
<li><p>Individual elements in a matrix can be printed using <strong>[row,column]</strong>. For example <strong>[2,3]</strong> would pull out the value in the 2nd ROW and third COLUMN.</p></li>
<li><p><code>dim()</code> is extremely useful to control whether our data was transformed correctly during different operations. For example, after we merge two files we would like to know that they still have the same number of rows as when we started the analysis. Same if we remove for example 10 samples, then we want to make sure that this is indeed what happened.</p></li>
<li><p><code>head()</code> is another useful function to check the first rows of a larger matrix (or dataframe)</p></li>
<li><p><code>tail()</code> same as head but showing the last rows</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co">#define our row and column names</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>row.names <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"row1"</span>, <span class="st">"row2"</span>, <span class="st">"row3"</span>, <span class="st">"row4"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>col.names <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"col1"</span>, <span class="st">"col2"</span>, <span class="st">"col3"</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co">#create our matrix (check the help function to see what is happening)</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>matrix_A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">14</span>), <span class="at">nrow =</span> <span class="dv">4</span>, <span class="at">byrow =</span> T, <span class="at">dimnames =</span> <span class="fu">list</span>(row.names,col.names))</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="co">#check how our matrix looks like</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>matrix_A</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="do">##      col1 col2 col3</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="do">## row1    3    4    5</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a><span class="do">## row2    6    7    8</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a><span class="do">## row3    9   10   11</span></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="do">## row4   12   13   14</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print the value in the 2row and 3rd column</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>matrix_A[<span class="dv">2</span>,<span class="dv">3</span>]</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print the values in the 3rd column</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>matrix_A[,<span class="dv">3</span>]</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="do">## row1 row2 row3 row4 </span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="do">##    5    8   11   14</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print everything except the 1st row</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>matrix_A[<span class="sc">-</span><span class="dv">1</span>,]</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="do">##      col1 col2 col3</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="do">## row2    6    7    8</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="do">## row3    9   10   11</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="do">## row4   12   13   14</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print everything except the 2nd column</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>matrix_A[,<span class="sc">-</span><span class="dv">2</span>]</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="do">##      col1 col3</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="do">## row1    3    5</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="do">## row2    6    8</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="do">## row3    9   11</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="do">## row4   12   14</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co">#see the dimensions of matrix, i.e. the nr of rows and columns</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(matrix_A)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 4 3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the first rows of our matrix, since our data is small, everything is shown</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(matrix_A)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="do">##      col1 col2 col3</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="do">## row1    3    4    5</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="do">## row2    6    7    8</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="do">## row3    9   10   11</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="do">## row4   12   13   14</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="lists" class="level3" data-number="1.11.3">
<h3 data-number="1.11.3" class="anchored" data-anchor-id="lists"><span class="header-section-number">1.11.3</span> Lists</h3>
<p>Sometimes you need to store data of different types. For example, if you are collecting cell counts, you might want to have cell counts (numeric), the microbes investigated (character), their status (logical, with TRUE for alive and FALSE for dead, …. This kind of data can be stored in lists. Lists are the R objects which contain elements of different types (numeric, strings, vectors, even another list, or a matrix).</p>
<p>A list is created using the list() function.</p>
<p>For example, the following variable x is a list containing copies of three vectors n, s, b.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#define our vectors</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">20</span>, <span class="dv">30</span>, <span class="dv">50</span>) </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Ecoli"</span>, <span class="st">"Archaeoglobus"</span>, <span class="st">"Bacillus"</span>) </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>b <span class="ot">=</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>) </span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="co">#combine the vectors in a list</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>our_list <span class="ot">=</span> <span class="fu">list</span>(<span class="at">counts=</span>n, <span class="at">strain=</span>s, <span class="at">status=</span>b) </span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="co">#show our list</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>our_list</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="do">## $counts</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 20 30 50</span></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a><span class="do">## $strain</span></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Ecoli"         "Archaeoglobus" "Bacillus"     </span></span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a><span class="do">## $status</span></span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a><span class="do">## [1]  TRUE FALSE  TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#sublist the second element in a list</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>our_list[<span class="dv">2</span>]</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="do">## $strain</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Ecoli"         "Archaeoglobus" "Bacillus"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#retrieve the 2nd and 3rd member of our list</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>our_list[<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>)] </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="do">## $strain</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Ecoli"         "Archaeoglobus" "Bacillus"     </span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="do">## $status</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1]  TRUE FALSE  TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we can also retrieve elements of a list if we know the name using two different ways:</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>our_list<span class="sc">$</span>strain</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Ecoli"         "Archaeoglobus" "Bacillus"</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>our_list[[<span class="st">"strain"</span>]]</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Ecoli"         "Archaeoglobus" "Bacillus"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the last example we use the <code>$</code> dollar symbol to extract data, i.e.&nbsp;to extract variables in a dataset (a matrix, list, dataframe). I.e. above the data we want to access is ‘our_list’ and the variable we want to extract is the strain.</p>
</section>
<section id="dataframes" class="level3" data-number="1.11.4">
<h3 data-number="1.11.4" class="anchored" data-anchor-id="dataframes"><span class="header-section-number">1.11.4</span> Dataframes</h3>
<p>Dataframes are tables in which each column contains values of one variable type and each row contains one set of values from each column. You can think of a data frame as a list of vectors of equal length. Most of our data very likely will be stored as dataframes.</p>
<p>A Dataframe usually follows these rules:</p>
<ul>
<li>The top line of the table, called the header, contains the column names.</li>
<li>Column names (i.e.&nbsp;the header of our data) should be non-empty (if they are, R provides the object with default values).</li>
<li>Row names should be unique</li>
<li>Each column should contain the same number of data items</li>
<li>Each horizontal line after the header is a data row, which begins with the name of the row, and then followed by the actual data.</li>
<li>Each data member of a row is called a cell.</li>
</ul>
<p>Importantly, most of the things we have learned before, i.e.&nbsp;how to subset data, apply here too.</p>
<p>The growth data that we have read into R will be used to explain how dataframes work.</p>
<section id="viewing-data-dataframes" class="level4" data-number="1.11.4.1">
<h4 data-number="1.11.4.1" class="anchored" data-anchor-id="viewing-data-dataframes"><span class="header-section-number">1.11.4.1</span> Viewing data Dataframes</h4>
<ul>
<li>We can use the brackets as before to extract certain rows or columns.</li>
<li>We can use the dollar sign to again extract information as long as we know the column names. I.e. now we want to access the shoot fresh weight (FW_shoot_mg) in our ‘growth_data’ dataframe.</li>
<li><code>colnames()</code> allows us to access the column names, i.e.&nbsp;the headers</li>
<li><code>rownames()</code> allows us to access the rownames of our data (usually these are numbered if not specified otherwise while reading the table)</li>
<li><code>dim()</code> allows us to check the dimensions (i.e.&nbsp;the number of rows and columns). This is useful to regullary check, especially if we modified our data somehow.</li>
<li><code>head()</code> shows the first rows of our dataframe</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co">#view our table</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check how many rows and columns our data has</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 105   5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">#extract the data from the 2nd row</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>growth_data[<span class="dv">2</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","_rn_":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co">#extract the first three columns</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co">#extract a column of our data using the column name</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data<span class="sc">$</span>Condition)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "MgCl" "MgCl" "MgCl" "MgCl" "MgCl" "MgCl"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print our headers</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(growth_data)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "SampleID"    "Nutrient"    "Condition"   "FW_shoot_mg" "Rootlength"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get the dimensions</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 105   5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When we print the rownames, we see that we have numbers from 1-105. When reading in a table into R it is the default behaviour how rownames are generated. As a general rule, if you want o have other rownames, these must be unique.</p>
</section>
<section id="adding-new-columns-to-dataframes" class="level4" data-number="1.11.4.2">
<h4 data-number="1.11.4.2" class="anchored" data-anchor-id="adding-new-columns-to-dataframes"><span class="header-section-number">1.11.4.2</span> Adding new columns to Dataframes</h4>
<p>Below is a very basic way to add a new column (we name it newColumn) and fill all rows with the word <code>comment</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co">#expand a dataframe, functions data.frame or cbind (or see below)</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>newColumn <span class="ot">&lt;-</span> <span class="st">"comment"</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check if that worked</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","6":"comment","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","6":"comment","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","6":"comment","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","6":"comment","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","6":"comment","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","6":"comment","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>There are more sophisticated ways to add columns based on conditions or even merge dataframes. Some of these we will discuss later.</p>
</section>
</section>
</section>
<section id="how-can-we-check-what-kind-of-structure-our-data-has" class="level2" data-number="1.12">
<h2 data-number="1.12" class="anchored" data-anchor-id="how-can-we-check-what-kind-of-structure-our-data-has"><span class="header-section-number">1.12</span> How can we check what kind of structure our data has?</h2>
<p>If we read in our own data, we should check as what type of class our table is stored. We have several ways to do this:</p>
<ul>
<li><code>class()</code> = determines as what kind of object is stored</li>
<li><code>str()</code> = display the internal structure of an R object.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check what kind of data we have:</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(growth_data)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "data.frame"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check how are different parts of our data stored?</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(growth_data)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="do">## 'data.frame':    105 obs. of  6 variables:</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  $ SampleID   : chr  "noP" "noP" "noP" "noP" ...</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  $ Nutrient   : chr  "noP" "noP" "noP" "noP" ...</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  $ Condition  : chr  "MgCl" "MgCl" "MgCl" "MgCl" ...</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  $ FW_shoot_mg: num  10.26 6.52 12.17 11.37 9.8 ...</span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="do">##  $ Rootlength : num  5.93 5.74 6.83 6.74 6.74 ...</span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="do">##  $ newColumn  : chr  "comment" "comment" "comment" "comment" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We see that</p>
<ul>
<li>our data is stored in a dataframe</li>
<li>that the data stored in different formats, i.e.&nbsp;numeric and characters</li>
<li>the first 3 columns are stored in something we call a factor.</li>
</ul>
</section>
<section id="factors" class="level2" data-number="1.13">
<h2 data-number="1.13" class="anchored" data-anchor-id="factors"><span class="header-section-number">1.13</span> Factors</h2>
<p>Factors are data objects that are used to represent categorical data and store it in its different levels. They are an important class for statistical analysis and for plotting. Factors are stored as integers, and have labels associated with these unique integers. Once created, factors can only contain a pre-defined set values, known as levels. By default, R always sorts levels in alphabetical order.</p>
<ul>
<li><code>factor()</code> allows us to create our own factor</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co">#lets make a vector</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>Nutrients <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"P"</span>, <span class="st">"P"</span>, <span class="st">"noP"</span>, <span class="st">"noP"</span>)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co">#lets make our own simple factor</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>Nutrients_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"P"</span>, <span class="st">"P"</span>, <span class="st">"noP"</span>, <span class="st">"noP"</span>))</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co">#lets compare the vector and factor we generated</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>Nutrients</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "P"   "P"   "noP" "noP"</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>Nutrients_factor</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] P   P   noP noP</span></span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Levels: P noP</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When we check our factor, we see that R assigns one level to <code>P</code> and another level to <code>noP</code>. We can also see, that R sorts the levels in an alphabetical way, even though in the initial code we first had P before noP.</p>
<section id="checking-the-behaviour-of-factors" class="level3" data-number="1.13.1">
<h3 data-number="1.13.1" class="anchored" data-anchor-id="checking-the-behaviour-of-factors"><span class="header-section-number">1.13.1</span> Checking the behaviour of factors</h3>
<p>Now, lets check how factors behave.</p>
<ul>
<li><code>levels()</code> = only prints the levels of a given factor. We can also run this on any column of our dataframe.</li>
<li><code>nlevels()</code> = check how many levels we have.</li>
<li>While factors look (and often behave) like character vectors, they are actually integers under the hood, and you need to be careful when treating them like strings. We can test this by looking at what type of object we generated.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co">#only print the levels</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(Nutrients_factor)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "P"   "noP"</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(growth_data<span class="sc">$</span>Condition)</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="do">## NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check how many levels we have</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">nlevels</span>(Nutrients_factor)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co">#what class do we have</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(Nutrients_factor)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "factor"</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="fu">typeof</span>(Nutrients_factor)</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "integer"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ordering-factor-levels" class="level3" data-number="1.13.2">
<h3 data-number="1.13.2" class="anchored" data-anchor-id="ordering-factor-levels"><span class="header-section-number">1.13.2</span> Ordering factor levels</h3>
<p>For some things, the order of things might matter and then we need to order the factors ourselves.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check our levels</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(Nutrients_factor)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "P"   "noP"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#reorder levels</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>Nutrients_factor_reordered <span class="ot">&lt;-</span> <span class="fu">factor</span>(Nutrients_factor, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"P"</span>, <span class="st">"noP"</span>))</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check our levels</span></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(Nutrients_factor_reordered)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "P"   "noP"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="converting-factors" class="level3" data-number="1.13.3">
<h3 data-number="1.13.3" class="anchored" data-anchor-id="converting-factors"><span class="header-section-number">1.13.3</span> Converting factors</h3>
<p>Sometimes you need to explicitly convert factors to either text or numbers. Or numbers to characters, etc. To do this, you use the functions <code>as.character()</code> or <code>as.numeric()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">#convert our factor to a character</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>Nutrients_characters <span class="ot">&lt;-</span> <span class="fu">as.character</span>(Nutrients_factor)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>Nutrients_characters</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "P"   "P"   "noP" "noP"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="dealing-with-nas-in-our-data" class="level2" data-number="1.14">
<h2 data-number="1.14" class="anchored" data-anchor-id="dealing-with-nas-in-our-data"><span class="header-section-number">1.14</span> Dealing with NAs in our data</h2>
<section id="removing-nas" class="level3" data-number="1.14.1">
<h3 data-number="1.14.1" class="anchored" data-anchor-id="removing-nas"><span class="header-section-number">1.14.1</span> Removing NAs</h3>
<p>NAs are generated when our data contains a missing value. This can become problematic for certain programs and we can decide to remove all NAs.</p>
<p>The function to do this is <code>is.na</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create a vector that includes an NA</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="cn">NA</span>,<span class="dv">5</span>)</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check whether we have NAs</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="fu">is.na</span>(y)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] FALSE FALSE FALSE  TRUE FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co">#remove NAs from our data</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>y[<span class="sc">!</span><span class="fu">is.na</span>(y)]</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 1 2 3 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Exclamation marks in R –&gt; <code>!=</code> is “not equal to.”</li>
<li>The function is.na(z)gives a logical vector of the same size as z with value TRUE if and only if the corre-sponding element in z is NA.</li>
<li>I.e. in this example we have FALSE FALSE FALSE TRUE FALSE</li>
<li>When using y[!is.na(y)], retain the columns were is.na is False</li>
</ul>
<section id="replacing-nas-with-something-else" class="level4" data-number="1.14.1.1">
<h4 data-number="1.14.1.1" class="anchored" data-anchor-id="replacing-nas-with-something-else"><span class="header-section-number">1.14.1.1</span> Replacing NAs with something else</h4>
<p>Another option might be to replace a NA with a 0 (or whatever else makes sense in a given context)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create a vector that includes an NA</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="cn">NA</span>,<span class="dv">5</span>)</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check whether we have NAs</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="fu">is.na</span>(x)</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] FALSE FALSE FALSE  TRUE FALSE</span></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="co">#remove NAs from our data</span></span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>x[<span class="fu">is.na</span>(x)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a><span class="co">#check data</span></span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>x</span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 1 2 3 0 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notice, that here we use a similar syntax compared to before. Now we do</p>
<ul>
<li>Find the NAs (I.e. in this example we have FALSE FALSE FALSE TRUE FALSE)</li>
<li>If NA is TRUE then replace it with a 0</li>
</ul>
</section>
</section>
<section id="replacing-nas-in-our-dataframe" class="level3" data-number="1.14.2">
<h3 data-number="1.14.2" class="anchored" data-anchor-id="replacing-nas-in-our-dataframe"><span class="header-section-number">1.14.2</span> Replacing NAs in our dataframe</h3>
<p>Remember, in the beginning we said that their are NAs in our timecourse data. Let’s have a look at this data, i.e.&nbsp;row 30, we will see that we have NAs.</p>
<p>Next, we want to remove all rows that have no values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check in which rows we have NAs inside the root length measurements</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(timecourse<span class="sc">$</span>Rootlength)) </span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   [1]  29  30 233 234 235 236 237 238 239 240 268 269 270 330 360 374 375 376</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  [19] 377 378 379 380 381 382 383 384 385 386 387 388 389 390 402 403 404 405</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  [37] 406 407 408 409 410 411 412 413 414 415 416 417 418 419 420 444 445 446</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  [55] 447 448 449 450 475 476 477 478 479 480 492 493 494 495 496 497 498 499</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  [73] 500 515 516 517 518 519 520 533 534 535 536 537 538 539 540 550 551 552</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a><span class="do">##  [91] 553 554 555 556 557 558 559 560 574 575 605</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check a column</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>timecourse[<span class="dv">29</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Timepoint"],"name":[4],"type":["chr"],"align":["left"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"NA","_rn_":"29"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co">#remove rows that contain NAs</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>timecourse_noNA <span class="ot">&lt;-</span> timecourse[<span class="sc">!</span><span class="fu">is.na</span>(timecourse<span class="sc">$</span>Rootlength), ]</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check if that worked</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>timecourse_noNA[<span class="dv">29</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Timepoint"],"name":[4],"type":["chr"],"align":["left"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP_MgCl_t0","2":"noP","3":"MgCl","4":"t0","5":"1.448","_rn_":"31"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
</section>
<section id="important-code-snippets" class="level2" data-number="1.15">
<h2 data-number="1.15" class="anchored" data-anchor-id="important-code-snippets"><span class="header-section-number">1.15</span> Important code snippets</h2>
<section id="print-the-session-info" class="level3" data-number="1.15.1">
<h3 data-number="1.15.1" class="anchored" data-anchor-id="print-the-session-info"><span class="header-section-number">1.15.1</span> Print the session info:</h3>
<p>This prints our session info (used R version, packages, etc) and is very important to store to reproduce our code for others</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co">#|eval: false</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="do">## R version 4.2.1 (2022-06-23)</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Running under: macOS Catalina 10.15.7</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Matrix products: default</span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a><span class="do">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="do">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a><span class="do">## locale:</span></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] C</span></span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a><span class="do">## attached base packages:</span></span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a><span class="do">## other attached packages:</span></span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] kableExtra_1.3.4 knitr_1.40      </span></span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a><span class="do">## loaded via a namespace (and not attached):</span></span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] rstudioapi_0.14   xml2_1.3.3        magrittr_2.0.3    rvest_1.0.3      </span></span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a><span class="do">##  [5] munsell_0.5.0     viridisLite_0.4.1 colorspace_2.0-3  R6_2.5.1         </span></span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a><span class="do">##  [9] rlang_1.0.5       fastmap_1.1.0     highr_0.9         stringr_1.4.1    </span></span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a><span class="do">## [13] httr_1.4.4        tools_4.2.1       webshot_0.5.3     xfun_0.33        </span></span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a><span class="do">## [17] cli_3.4.0         htmltools_0.5.3   systemfonts_1.0.4 yaml_2.3.5       </span></span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a><span class="do">## [21] digest_0.6.29     lifecycle_1.0.2   vctrs_0.4.1       htmlwidgets_1.5.4</span></span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a><span class="do">## [25] glue_1.6.2        evaluate_0.16     rmarkdown_2.16    stringi_1.7.8    </span></span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a><span class="do">## [29] compiler_4.2.1    scales_1.2.1      jsonlite_1.8.0    svglite_2.1.0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="clean-up-the-working-environment" class="level3" data-number="1.15.2">
<h3 data-number="1.15.2" class="anchored" data-anchor-id="clean-up-the-working-environment"><span class="header-section-number">1.15.2</span> Clean-up the working environment</h3>
<p>Notice: At the moment, we do not want to clean up the working directory, therefore, in our documentation we put a <code>#</code> in front of it. If you want to run the code just remove the <code>#</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co">#rm(list = ls()) </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="basic-operations" class="level2" data-number="1.16">
<h2 data-number="1.16" class="anchored" data-anchor-id="basic-operations"><span class="header-section-number">1.16</span> Basic operations</h2>
<ul>
<li>R can work as a calculator and knows all basic operations</li>
<li>In expressions and assignments, R does not care about spaces</li>
<li>Another important character is the semicolon. R evaluates code line by line. A line break tells R that a statement has to be evaluated. Instead, you can use a semicolon to tell R where a statement ends</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co">#add</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">+</span><span class="dv">2</span> </span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co">#multiply</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fl">3.5</span><span class="sc">*</span><span class="dv">3</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 10.5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co">#divide</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fl">2.1</span><span class="sc">/</span><span class="dv">3</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 0.7</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co">#exponentials</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span><span class="sc">^</span><span class="dv">10</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 1024</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co">#more complex functions also work</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">*</span>(<span class="dv">5</span><span class="sc">+</span><span class="dv">2</span><span class="sc">^</span><span class="dv">1</span>)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 21</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co">#combining two operations into one using a semicolon</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span><span class="sc">+</span><span class="dv">9</span> ; <span class="dv">4</span><span class="sc">+</span><span class="dv">5</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 14</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 9</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="logical-operators" class="level2" data-number="1.17">
<h2 data-number="1.17" class="anchored" data-anchor-id="logical-operators"><span class="header-section-number">1.17</span> Logical operators</h2>
<p>One important features of R (and of any programming language) are logical operators.</p>
<p>When R evaluates an expression containing logical operators, it will return either TRUE or FALSE. It follows a list of them:</p>
<div class="cell" data-warnings="false">
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">Operator</th>
<th style="text-align: left;">Meaning</th>
<th style="text-align: left;">Operator</th>
<th style="text-align: left;">Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">&lt;</td>
<td style="text-align: left;">less than</td>
<td style="text-align: left;">==</td>
<td style="text-align: left;">equal</td>
</tr>
<tr class="even">
<td style="text-align: left;">&gt;</td>
<td style="text-align: left;">greater than</td>
<td style="text-align: left;">!=</td>
<td style="text-align: left;">not equal</td>
</tr>
<tr class="odd">
<td style="text-align: left;">&lt;=</td>
<td style="text-align: left;">less than or equal to</td>
<td style="text-align: left;">&amp;</td>
<td style="text-align: left;">and</td>
</tr>
<tr class="even">
<td style="text-align: left;">&gt;=</td>
<td style="text-align: left;">greater than or equal to</td>
<td style="text-align: left;">|</td>
<td style="text-align: left;">or</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="examples-for-using-operators" class="level2" data-number="1.18">
<h2 data-number="1.18" class="anchored" data-anchor-id="examples-for-using-operators"><span class="header-section-number">1.18</span> Examples for using operators:</h2>
<section id="basic-operators" class="level3" data-number="1.18.1">
<h3 data-number="1.18.1" class="anchored" data-anchor-id="basic-operators"><span class="header-section-number">1.18.1</span> Basic operators</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co">#basic operators</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">==</span><span class="dv">1</span></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span><span class="sc">&gt;</span><span class="dv">1</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">==</span><span class="dv">1</span> <span class="sc">&amp;</span> <span class="dv">1</span><span class="sc">&gt;</span><span class="dv">2</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>(<span class="dv">1</span><span class="sc">==</span><span class="dv">1</span><span class="sc">&amp;</span> <span class="dv">1</span><span class="sc">&gt;</span><span class="dv">2</span>) <span class="sc">|</span> <span class="dv">5</span><span class="sc">&lt;</span><span class="dv">6</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="using-operators-to-subset-dataframes" class="level3" data-number="1.18.2">
<h3 data-number="1.18.2" class="anchored" data-anchor-id="using-operators-to-subset-dataframes"><span class="header-section-number">1.18.2</span> Using operators to subset dataframes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co">#only print data for the normal P conditions</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>P_data <span class="ot">&lt;-</span> growth_data[growth_data<span class="sc">$</span>Nutrient <span class="sc">==</span> <span class="st">"P"</span>, ]</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="co">#test that all went ok</span></span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data)</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 105   6</span></span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(P_data)</span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 53  6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we can also subset our data by numbers, i.e. only keep rows were roots are longer than 5cm</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>growth_data_10cm <span class="ot">&lt;-</span> growth_data[growth_data<span class="sc">$</span>Rootlength <span class="sc">&gt;</span> <span class="dv">5</span>, ]</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data_10cm)</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 101   6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co">#instead of keeping rows with roots longer than 5cm we can also remove them using *!*</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>growth_data_no10cm <span class="ot">&lt;-</span> growth_data[<span class="sc">!</span>growth_data<span class="sc">$</span>Rootlength <span class="sc">&gt;</span> <span class="dv">5</span>, ]</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data_no10cm)</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 4 6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Above you see a good example why dim() can very useful as it allows us to quickly check that what we do actually works. Always check afterwards that your function did what you expect you to do since as long as R is able to do something it will do it regardless if that might be different that you want R to do.</strong></p>
</section>
</section>
<section id="useful-base-functions-in-r" class="level2" data-number="1.19">
<h2 data-number="1.19" class="anchored" data-anchor-id="useful-base-functions-in-r"><span class="header-section-number">1.19</span> Useful base functions in R</h2>
<p>Apart from the elementary operations, all of the common arithmetic functions are available: log, exp, sin, cos, tan, sqrt, etc. Other useful functions one can use on vectors are:</p>
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">Name</th>
<th style="text-align: left;">Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">max</td>
<td style="text-align: left;">select smallest element</td>
</tr>
<tr class="even">
<td style="text-align: left;">min</td>
<td style="text-align: left;">select largest element</td>
</tr>
<tr class="odd">
<td style="text-align: left;">length</td>
<td style="text-align: left;">gives the number of elements</td>
</tr>
<tr class="even">
<td style="text-align: left;">sum</td>
<td style="text-align: left;">sums all elements</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mean</td>
<td style="text-align: left;">obtains the mean value</td>
</tr>
<tr class="even">
<td style="text-align: left;">var</td>
<td style="text-align: left;">unbiased sample variance</td>
</tr>
<tr class="odd">
<td style="text-align: left;">sort</td>
<td style="text-align: left;">see exercise 2c</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="subsetting-our-data" class="level2" data-number="1.20">
<h2 data-number="1.20" class="anchored" data-anchor-id="subsetting-our-data"><span class="header-section-number">1.20</span> Subsetting our data</h2>
<p>We already have seen important ways to subset data:</p>
<ol type="1">
<li>Use of the index by using the square brackets</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">#subsetting rows and columns using the index</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>growth_data[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["Nutrient"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[2],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"MgCl","3":"10.26","_rn_":"1"},{"1":"noP","2":"MgCl","3":"6.52","_rn_":"2"},{"1":"noP","2":"MgCl","3":"12.17","_rn_":"3"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<ol start="2" type="1">
<li>Use of operators</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co">#subset using operators (only print rows if the Nutrient column equals P)</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>P_data <span class="ot">&lt;-</span> growth_data[growth_data<span class="sc">$</span>Nutrient <span class="sc">==</span> <span class="st">"P"</span>, ]</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(P_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"P","2":"P","3":"MgCl","4":"32.42","5":"11.286793","6":"comment","_rn_":"53"},{"1":"P","2":"P","3":"MgCl","4":"21.03","5":"10.456630","6":"comment","_rn_":"54"},{"1":"P","2":"P","3":"MgCl","4":"18.35","5":"11.106341","6":"comment","_rn_":"55"},{"1":"P","2":"P","3":"MgCl","4":"21.04","5":"10.816896","6":"comment","_rn_":"56"},{"1":"P","2":"P","3":"MgCl","4":"16.61","5":"10.608252","6":"comment","_rn_":"57"},{"1":"P","2":"P","3":"MgCl","4":"25.79","5":"7.121587","6":"comment","_rn_":"58"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Now we want to look at some other examples:</p>
<section id="grep-and-grepl" class="level3" data-number="1.20.1">
<h3 data-number="1.20.1" class="anchored" data-anchor-id="grep-and-grepl"><span class="header-section-number">1.20.1</span> grep and grepl</h3>
<p>Oftentimes you may need to filter a data set based on a partial character string that is beyond the scope of comparison operators.</p>
<p>R provides such functions (grep and grepl) that match character patterns in specified vector. While both of these functions find patterns, they return different output types based on those patterns.</p>
<ul>
<li>grep returns numeric values that correspond to the indexed locations of the patterns</li>
<li>grepl returns a logical vector in which ‘TRUE’ represents a pattern match.</li>
</ul>
<p>In our growth data, we only want to print measurements of our strains.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co">#grep return the index value of each matched pattern</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">"MgCl"</span>, growth_data<span class="sc">$</span>Condition)</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 53 54 55 56 57 58 59 60 61 62 63 64</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [26] 65 66 67</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co">#grepl returns a logical output for each indices in the original vector </span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a><span class="co">#with "TRUE" representing matched patterns</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="fu">grepl</span>(<span class="st">"Strain"</span>, growth_data<span class="sc">$</span>Condition)</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a><span class="do">##   [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  [13] FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  [25]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  [37]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a><span class="do">##  [49]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a><span class="do">##  [61] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a><span class="do">##  [73]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a><span class="do">##  [85]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a><span class="do">##  [97]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now lets use this to actually filter our data table for a pattern.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>filter_for_value <span class="ot">&lt;-</span>timecourse_noNA[<span class="fu">grepl</span>(<span class="st">"MgC"</span>, timecourse_noNA<span class="sc">$</span>Condition),]</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filter_for_value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Timepoint"],"name":[4],"type":["chr"],"align":["left"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.542","_rn_":"1"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.849","_rn_":"2"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.209","_rn_":"3"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.834","_rn_":"4"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.586","_rn_":"5"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.683","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co">#filter data set based on values that do not match the specified pattern (by using the minus symbol)</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>filter_for_not_a_value <span class="ot">&lt;-</span> timecourse_noNA[<span class="sc">-</span><span class="fu">grepl</span>(<span class="st">"MgC"</span>, timecourse_noNA<span class="sc">$</span>Condition),]</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filter_for_not_a_value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Timepoint"],"name":[4],"type":["chr"],"align":["left"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.849","_rn_":"2"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.209","_rn_":"3"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.834","_rn_":"4"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.586","_rn_":"5"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.683","_rn_":"6"},{"1":"P_MgCl_t0","2":"P","3":"MgCl","4":"t0","5":"1.608","_rn_":"7"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Other comments:</p>
<ul>
<li>Using regular expressions (programming symbol pattern) will increase their functionality</li>
<li>Specified patterns are case sensitive (‘t’ does not equal ‘T’)</li>
<li>Any matching pattern will be returned despite the context in which that pattern is located (i.e., grep(‘the’, data) with return matches for ‘the’, ‘theme’, ‘heather’, ‘breathe’, and so on<e2>&lt;80&gt;&lt;93&gt;this is where regular expressions are useful for specifying where in a string the pattern should appear.</e2></li>
</ul>
<p>Regular expressions are explained in the AWK and General notebook. But just to give an example lets just grep Strains that have a 3 letter number</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>filter_3letters <span class="ot">&lt;-</span> growth_data[<span class="fu">grepl</span>(<span class="st">"[0-9][0-9][0-9]"</span>, growth_data<span class="sc">$</span>Condition),]</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="co">#check the structure of our data</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filter_3letters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"noP_101","2":"noP","3":"Strain101","4":"10.49","5":"7.143667","6":"comment","_rn_":"14"},{"1":"noP_101","2":"noP","3":"Strain101","4":"8.91","5":"7.669841","6":"comment","_rn_":"15"},{"1":"noP_101","2":"noP","3":"Strain101","4":"9.32","5":"7.807710","6":"comment","_rn_":"16"},{"1":"noP_101","2":"noP","3":"Strain101","4":"6.76","5":"7.508370","6":"comment","_rn_":"17"},{"1":"noP_101","2":"noP","3":"Strain101","4":"5.99","5":"6.607630","6":"comment","_rn_":"18"},{"1":"noP_101","2":"noP","3":"Strain101","4":"8.26","5":"7.269267","6":"comment","_rn_":"19"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Here, <code>[0-9]</code> searches for every number from 0-9 and we look for three numbers. One important thing with subsetting is that the levels are still kept. So with the command above we remove every row that is not Strain101 and Strain230. However, the levels are still kept. Let’s check this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(filter_3letters<span class="sc">$</span>Condition)</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="do">## NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So we see that MgCl and Strain28 are still in the levels even if they do not occur in our table itself. Sometimes when working with subsetted dataframes, i.e.&nbsp;when doing stats or plotting, this can interfere with our analysis. Here, it is useful to drop empty levels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>filter_3letters_clean <span class="ot">&lt;-</span> <span class="fu">droplevels</span>(filter_3letters)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(filter_3letters_clean<span class="sc">$</span>Condition)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="do">## NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="the-unique-command" class="level2" data-number="1.21">
<h2 data-number="1.21" class="anchored" data-anchor-id="the-unique-command"><span class="header-section-number">1.21</span> The unique command</h2>
<p><code>unique()</code> allows to determine duplicate rows and allows us to subset our data for certain categories. For example, for very large dataframes we often can simplify things.</p>
<p>Here, if we have a lot of treatments and did the experiment a long time ago, we might want to ask for a table that lists the treatments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make unique contig list that still contains info of our bin ID</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>mapping_file <span class="ot">&lt;-</span> <span class="fu">unique</span>(growth_data[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Nutrient"</span>, <span class="st">"Condition"</span>)])</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mapping_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","_rn_":"1"},{"1":"noP_101","2":"noP","3":"Strain101","_rn_":"14"},{"1":"noP_230","2":"noP","3":"Strain230","_rn_":"27"},{"1":"noP_28","2":"noP","3":"Strain28","_rn_":"38"},{"1":"P","2":"P","3":"MgCl","_rn_":"53"},{"1":"P_101","2":"P","3":"Strain101","_rn_":"68"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
<section id="the-merge-command" class="level2" data-number="1.22">
<h2 data-number="1.22" class="anchored" data-anchor-id="the-merge-command"><span class="header-section-number">1.22</span> The merge command</h2>
<p>We can also add additional metadata to our growth data.</p>
<p>One way to do this is the <strong>cbind()</strong> or <strong>rbind()</strong> functions. However, these functions require the two dataframes to have the exact number of columns or rows, which we do not have.</p>
<p>Here, the <code>merge()</code> function of the data.table package is very useful to merge data with different dimensions as long as they have a common pattern (i.e.&nbsp;the SampleID).</p>
<p>First lets build an artificial mapping file that incldues the number of days we grew our plants:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make mapping that contains our basic sample info</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>mapping_file <span class="ot">&lt;-</span> <span class="fu">unique</span>(growth_data[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Nutrient"</span>, <span class="st">"Condition"</span>)])</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="co">#add a new column, where we list our experiment ID</span></span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>mapping_file<span class="sc">$</span>Comment <span class="ot">&lt;-</span> <span class="st">"FirstExperiment"</span></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mapping_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Comment"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"FirstExperiment","_rn_":"1"},{"1":"noP_101","2":"noP","3":"Strain101","4":"FirstExperiment","_rn_":"14"},{"1":"noP_230","2":"noP","3":"Strain230","4":"FirstExperiment","_rn_":"27"},{"1":"noP_28","2":"noP","3":"Strain28","4":"FirstExperiment","_rn_":"38"},{"1":"P","2":"P","3":"MgCl","4":"FirstExperiment","_rn_":"53"},{"1":"P_101","2":"P","3":"Strain101","4":"FirstExperiment","_rn_":"68"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Now we can use this mapping file and merge it with our growth data as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load our package</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a><span class="co">#merge our mapping file with our growth data</span></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>new_data_frame <span class="ot">&lt;-</span> <span class="fu">merge</span>(growth_data, mapping_file, <span class="at">by =</span> <span class="st">"SampleID"</span>)</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_data_frame)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient.x"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition.x"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]},{"label":["Nutrient.y"],"name":[7],"type":["chr"],"align":["left"]},{"label":["Condition.y"],"name":[8],"type":["chr"],"align":["left"]},{"label":["Comment"],"name":[9],"type":["chr"],"align":["left"]}],"data":[{"1":"P","2":"P","3":"MgCl","4":"32.42","5":"11.286793","6":"comment","7":"P","8":"MgCl","9":"FirstExperiment","_rn_":"1"},{"1":"P","2":"P","3":"MgCl","4":"21.03","5":"10.456630","6":"comment","7":"P","8":"MgCl","9":"FirstExperiment","_rn_":"2"},{"1":"P","2":"P","3":"MgCl","4":"18.35","5":"11.106341","6":"comment","7":"P","8":"MgCl","9":"FirstExperiment","_rn_":"3"},{"1":"P","2":"P","3":"MgCl","4":"21.04","5":"10.816896","6":"comment","7":"P","8":"MgCl","9":"FirstExperiment","_rn_":"4"},{"1":"P","2":"P","3":"MgCl","4":"16.61","5":"10.608252","6":"comment","7":"P","8":"MgCl","9":"FirstExperiment","_rn_":"5"},{"1":"P","2":"P","3":"MgCl","4":"25.79","5":"7.121587","6":"comment","7":"P","8":"MgCl","9":"FirstExperiment","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>This now is a good example to check that all went fine and that the new dataframes has the same number of rows (=measurements) compared to the original dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co">#control that all went fine</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(growth_data)</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 105   6</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(new_data_frame)</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 105   9</span></span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a><span class="co">#if there is no match between dataframe 1 and dataframe 2 columns will by default be deleted. If you want to keep all columns do:</span></span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a><span class="co">#new_data_frame &lt;- merge(growth_data, mapping_file, by = "SampleID". all.x = T)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With dim we see that we still have 105 rows (i.e.&nbsp;measurements) and that we now added 3 new columns.</p>
</section>
<section id="combine-commands-into-one-line" class="level2" data-number="1.23">
<h2 data-number="1.23" class="anchored" data-anchor-id="combine-commands-into-one-line"><span class="header-section-number">1.23</span> Combine commands into one line</h2>
<p>While this gets more difficult to read, sometimes it might be useful to combine several commands into one go to condense code Generally, it is easier to just write line by line especially if you read your code months later.</p>
<p>What we want to do: - in the example above we duplicate the columns for Nutrient and Condition and before merging we might first subset the mapping file to only include the info we want to merge. - so our two steps are: a) trim the mapping file and b) merge</p>
<p>To do this, we use these two lines of code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make mapping file more simple</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>mapping_reduced <span class="ot">&lt;-</span> mapping_file[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Comment"</span>)]</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a><span class="co">#merge</span></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>new_data_frame <span class="ot">&lt;-</span> <span class="fu">merge</span>(growth_data, mapping_reduced, <span class="at">by =</span> <span class="st">"SampleID"</span>)</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_data_frame)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]},{"label":["Comment"],"name":[7],"type":["chr"],"align":["left"]}],"data":[{"1":"P","2":"P","3":"MgCl","4":"32.42","5":"11.286793","6":"comment","7":"FirstExperiment","_rn_":"1"},{"1":"P","2":"P","3":"MgCl","4":"21.03","5":"10.456630","6":"comment","7":"FirstExperiment","_rn_":"2"},{"1":"P","2":"P","3":"MgCl","4":"18.35","5":"11.106341","6":"comment","7":"FirstExperiment","_rn_":"3"},{"1":"P","2":"P","3":"MgCl","4":"21.04","5":"10.816896","6":"comment","7":"FirstExperiment","_rn_":"4"},{"1":"P","2":"P","3":"MgCl","4":"16.61","5":"10.608252","6":"comment","7":"FirstExperiment","_rn_":"5"},{"1":"P","2":"P","3":"MgCl","4":"25.79","5":"7.121587","6":"comment","7":"FirstExperiment","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Now, this worked fine but requires a bit more code and we need to create one more object.</p>
<p>We could also combine these two kines of code into one line by subsetting our mapping file INSIDE the merge function as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>new_data_frame <span class="ot">&lt;-</span> <span class="fu">merge</span>(growth_data, mapping_file[,<span class="fu">c</span>(<span class="st">"SampleID"</span>, <span class="st">"Comment"</span>)], <span class="at">by =</span> <span class="st">"SampleID"</span>)</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_data_frame)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]},{"label":["Comment"],"name":[7],"type":["chr"],"align":["left"]}],"data":[{"1":"P","2":"P","3":"MgCl","4":"32.42","5":"11.286793","6":"comment","7":"FirstExperiment","_rn_":"1"},{"1":"P","2":"P","3":"MgCl","4":"21.03","5":"10.456630","6":"comment","7":"FirstExperiment","_rn_":"2"},{"1":"P","2":"P","3":"MgCl","4":"18.35","5":"11.106341","6":"comment","7":"FirstExperiment","_rn_":"3"},{"1":"P","2":"P","3":"MgCl","4":"21.04","5":"10.816896","6":"comment","7":"FirstExperiment","_rn_":"4"},{"1":"P","2":"P","3":"MgCl","4":"16.61","5":"10.608252","6":"comment","7":"FirstExperiment","_rn_":"5"},{"1":"P","2":"P","3":"MgCl","4":"25.79","5":"7.121587","6":"comment","7":"FirstExperiment","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
<section id="add-new-columns-to-our-data-and-combining-values-in-different-columns" class="level2" data-number="1.24">
<h2 data-number="1.24" class="anchored" data-anchor-id="add-new-columns-to-our-data-and-combining-values-in-different-columns"><span class="header-section-number">1.24</span> Add new columns to our data and combining values in different columns</h2>
<p>We can also add new columns into original table, i.e.&nbsp;if we want to not show the fresh weight in mg but in g.</p>
<p>Below you can see that if we have numerical data in a column, we can use normal math operators (like +,-,/)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co">#convert mg to g</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>FW_shoot_g <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>FW_shoot_mg<span class="sc">/</span><span class="dv">10</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]},{"label":["FW_shoot_g"],"name":[7],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","6":"comment","7":"1.026","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","6":"comment","7":"0.652","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","6":"comment","7":"1.217","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","6":"comment","7":"1.137","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","6":"comment","7":"0.980","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","6":"comment","7":"0.375","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we can also round our data</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>FW_shoot_g <span class="ot">&lt;-</span> <span class="fu">round</span>(growth_data<span class="sc">$</span>FW_shoot_mg<span class="sc">/</span><span class="dv">10</span>, <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]},{"label":["FW_shoot_g"],"name":[7],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","6":"comment","7":"1.03","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","6":"comment","7":"0.65","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","6":"comment","7":"1.22","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","6":"comment","7":"1.14","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","6":"comment","7":"0.98","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","6":"comment","7":"0.38","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>We can also add (or substract, etc …) values from different columns. I.e. here we could calculate the ratios.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we can also do math with the values in two columns, i.e. if we want to calculate the ration between root length and fresh weight</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>ratio <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>Rootlength<span class="sc">/</span>growth_data<span class="sc">$</span>FW_shoot_mg</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["newColumn"],"name":[6],"type":["chr"],"align":["left"]},{"label":["FW_shoot_g"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["ratio"],"name":[8],"type":["dbl"],"align":["right"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","6":"comment","7":"1.03","8":"0.5780717","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","6":"comment","7":"0.65","8":"0.8808968","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","6":"comment","7":"1.22","8":"0.5616039","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","6":"comment","7":"1.14","8":"0.5930286","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","6":"comment","7":"0.98","8":"0.6874374","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","6":"comment","7":"0.38","8":"1.1296929","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
<section id="stats" class="level2" data-number="1.25">
<h2 data-number="1.25" class="anchored" data-anchor-id="stats"><span class="header-section-number">1.25</span> STATS</h2>
<p>A lot of helpful information that that was used here was inspired by a more detailed site that can be found <a href="http://www.learnbymarketing.com/tutorials/linear-regression-in-r/">here</a> . And please take note, that the writer of this tutorial is no statistician and if you spot issues feel free to let me know!</p>
<section id="summary-stats" class="level3" data-number="1.25.1">
<h3 data-number="1.25.1" class="anchored" data-anchor-id="summary-stats"><span class="header-section-number">1.25.1</span> Summary stats</h3>
<p>To do some basic stats we use the nutrient growth data set, which you find in the R_exercises folder.</p>
<ul>
<li><code>means()</code> = print the mean of a certain column</li>
<li><code>summary()</code>= print the mean, median, max, min, … of a certain column.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get the menas of our root length</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(growth_data<span class="sc">$</span>Rootlength)</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 9.046117</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co">#return summary stats for the root length</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(growth_data<span class="sc">$</span>Rootlength)</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="do">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a><span class="do">##   4.236   6.721   8.099   9.046  11.793  14.716</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also summarize data of several columns</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co">#return summary stats for the root length</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(growth_data[,<span class="fu">c</span>(<span class="st">"Rootlength"</span>, <span class="st">"FW_shoot_mg"</span>)])</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a><span class="do">##    Rootlength      FW_shoot_mg   </span></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  Min.   : 4.236   Min.   : 3.75  </span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  1st Qu.: 6.721   1st Qu.: 7.69  </span></span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  Median : 8.099   Median :11.62  </span></span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  Mean   : 9.046   Mean   :15.23  </span></span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a><span class="do">##  3rd Qu.:11.793   3rd Qu.:22.05  </span></span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a><span class="do">##  Max.   :14.716   Max.   :39.39</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also run <code>summary()</code> on our complete dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co">#run summary on all our data</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(growth_data)</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a><span class="do">##    SampleID           Nutrient          Condition          FW_shoot_mg   </span></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  Length:105         Length:105         Length:105         Min.   : 3.75  </span></span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  Class :character   Class :character   Class :character   1st Qu.: 7.69  </span></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  Mode  :character   Mode  :character   Mode  :character   Median :11.62  </span></span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a><span class="do">##                                                           Mean   :15.23  </span></span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a><span class="do">##                                                           3rd Qu.:22.05  </span></span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a><span class="do">##                                                           Max.   :39.39  </span></span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a><span class="do">##    Rootlength      newColumn           FW_shoot_g        ratio       </span></span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a><span class="do">##  Min.   : 4.236   Length:105         Min.   :0.380   Min.   :0.2761  </span></span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a><span class="do">##  1st Qu.: 6.721   Class :character   1st Qu.:0.770   1st Qu.:0.5141  </span></span>
<span id="cb102-13"><a href="#cb102-13" aria-hidden="true" tabindex="-1"></a><span class="do">##  Median : 8.099   Mode  :character   Median :1.160   Median :0.6874  </span></span>
<span id="cb102-14"><a href="#cb102-14" aria-hidden="true" tabindex="-1"></a><span class="do">##  Mean   : 9.046                      Mean   :1.523   Mean   :0.7381  </span></span>
<span id="cb102-15"><a href="#cb102-15" aria-hidden="true" tabindex="-1"></a><span class="do">##  3rd Qu.:11.793                      3rd Qu.:2.200   3rd Qu.:0.8801  </span></span>
<span id="cb102-16"><a href="#cb102-16" aria-hidden="true" tabindex="-1"></a><span class="do">##  Max.   :14.716                      Max.   :3.940   Max.   :1.6901</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This way we do not only get summary stats for our values, but also we get some counts on how many measurements we have for different conditions.</p>
</section>
<section id="the-table-command" class="level3" data-number="1.25.2">
<h3 data-number="1.25.2" class="anchored" data-anchor-id="the-table-command"><span class="header-section-number">1.25.2</span> The table command</h3>
<p>The function table builds a contingency table of the counts at each factor level, or combinations of factor levels. This is quite useful to count the number of data points with certain metadata associated. So for example for our genome data we can ask how many measurements we made for each condition.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co">#summarize how many measurements we have for each treatment</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(growth_data<span class="sc">$</span>SampleID)</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="do">##       P   P_101   P_230    P_28     noP noP_101 noP_230  noP_28 </span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a><span class="do">##      15      12      13      13      13      13      11      15</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we see that we have a slightly different number of measurements for each condition and timepoint. For this tutorial we can ignore them, but this might be relevant if we have huge differences for some statistical analyses. This approach allows you to for larger datasets to easily check for samples that might be outliers in terms of measurements per sample.</p>
</section>
<section id="the-ddply-command" class="level3" data-number="1.25.3">
<h3 data-number="1.25.3" class="anchored" data-anchor-id="the-ddply-command"><span class="header-section-number">1.25.3</span> The ddply command</h3>
<p>The table command can get rather slow and there are some useful packages to speed up things and run more complicated mathematical operations. One example is the <code>ddply()</code> function of the plyr package. A useful feature of ddply is that this tool stores the output as a dataframe instead of a table.</p>
<p>Below we see examples were we summarize the data for the root length across different nutrient conditions. I.e. we do not want to have the mean for all root lengths but we want to see if roots have different lengths under low and high P conditions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load our package</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a><span class="co">#calculate the mean root length across our Sample IDs</span></span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>growth_data_summarized <span class="ot">&lt;-</span> <span class="fu">ddply</span>(growth_data, .(Nutrient), summarize, <span class="at">Mean_RootLength =</span> <span class="fu">mean</span>(Rootlength))</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data_summarized)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["Nutrient"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Mean_RootLength"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"P","2":"11.513931","_rn_":"1"},{"1":"noP","2":"6.530845","_rn_":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The structure of the command is as follows:</p>
<p><strong>ddply(Input_Data, .(Colum_we_want_to_sum_arcoss), summarize, New_Column_name = mean(Column_we_want_to_calc_the_mean_for))</strong></p>
<p>We can also summarize the data combining different conditions (i.e.&nbsp;nutrient and condition).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarize our data across Nutrients and Conditions</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>growth_data_summarized <span class="ot">&lt;-</span> <span class="fu">ddply</span>(growth_data, .(Nutrient, Condition), summarize, <span class="at">Mean_RootLength =</span> <span class="fu">mean</span>(Rootlength))</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data_summarized)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["Nutrient"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Mean_RootLength"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"P","2":"MgCl","3":"9.853457","_rn_":"1"},{"1":"P","2":"Strain101","3":"12.729133","_rn_":"2"},{"1":"P","2":"Strain230","3":"11.814678","_rn_":"3"},{"1":"P","2":"Strain28","3":"12.007390","_rn_":"4"},{"1":"noP","2":"MgCl","3":"5.805198","_rn_":"5"},{"1":"noP","2":"Strain101","3":"7.094616","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>We can also calculate the mean, sd and se in one line of code</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co">#and we can use even fancier functions (i.e. get the se and sd), check the plyr package for more details</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>growth_data_summarized <span class="ot">&lt;-</span> <span class="fu">ddply</span>(growth_data, .(Nutrient, Condition), summarize, <span class="at">RootLength =</span> <span class="fu">mean</span>(Rootlength), <span class="at">sd =</span> <span class="fu">sd</span> (Rootlength), <span class="at">se =</span> <span class="fu">sd</span>(Rootlength) <span class="sc">/</span> <span class="fu">sqrt</span>((<span class="fu">length</span>(Rootlength))))</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data_summarized)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["Nutrient"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[2],"type":["chr"],"align":["left"]},{"label":["RootLength"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["sd"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["se"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"P","2":"MgCl","3":"9.853457","4":"1.2981376","5":"0.3351777","_rn_":"1"},{"1":"P","2":"Strain101","3":"12.729133","4":"0.9813173","5":"0.2832819","_rn_":"2"},{"1":"P","2":"Strain230","3":"11.814678","4":"0.9128499","5":"0.2531790","_rn_":"3"},{"1":"P","2":"Strain28","3":"12.007390","4":"0.5812310","5":"0.1612045","_rn_":"4"},{"1":"noP","2":"MgCl","3":"5.805198","4":"0.7738124","5":"0.2146170","_rn_":"5"},{"1":"noP","2":"Strain101","3":"7.094616","4":"0.5939804","5":"0.1647405","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
<section id="tidyr" class="level3" data-number="1.25.4">
<h3 data-number="1.25.4" class="anchored" data-anchor-id="tidyr"><span class="header-section-number">1.25.4</span> tidyr</h3>
<p>Tydir is another package to summarize data but also to transform data. For now we will just discuss the basics to summarize data but we will try to add an extended section on this package later.</p>
<p>You notice here that the syntax is a bit unusual and we use the <code>%&gt;%</code> symbol (a so-called forward pipe operator). This symbol is commonly used in the <strong>dplyr</strong> and <strong>tidyr</strong> packages. This function passes the left hand side of the operator to the first argument of the right hand side of the operator.</p>
<p>In the following example, the a subset of the growth_data data (only 3 columns, not the whole dataframe) gets passed to <code>count()</code></p>
<p>New functions:</p>
<ul>
<li><code>count()</code> = A function of the dplyr package. Here, we count the unique protein IDs grouped by BinID and gene (i.e.&nbsp;roughly equivalent to the columns we want to keep)</li>
<li><code>summarize()</code> creates new data frame into one (or more) rows for each combination of grouping variables; if there are no grouping variables, the output will have a single row summarising all observations in the input. It will contain one column for each grouping variable and one column for each of the summary statistics that you have specified.</li>
<li><code>mean()</code> = calculate the arithmetic mean.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># | warning: false</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="co">#read in package</span></span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a><span class="do">## -- Attaching packages --------------------------------------- tidyverse 1.3.2 --</span></span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a><span class="do">## v ggplot2 3.3.6      v purrr   0.3.4 </span></span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a><span class="do">## v tibble  3.1.8      v dplyr   1.0.10</span></span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a><span class="do">## v tidyr   1.2.1      v stringr 1.4.1 </span></span>
<span id="cb107-9"><a href="#cb107-9" aria-hidden="true" tabindex="-1"></a><span class="do">## v readr   2.1.2      v forcats 0.5.2 </span></span>
<span id="cb107-10"><a href="#cb107-10" aria-hidden="true" tabindex="-1"></a><span class="do">## -- Conflicts ------------------------------------------ tidyverse_conflicts() --</span></span>
<span id="cb107-11"><a href="#cb107-11" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::arrange()    masks plyr::arrange()</span></span>
<span id="cb107-12"><a href="#cb107-12" aria-hidden="true" tabindex="-1"></a><span class="do">## x purrr::compact()    masks plyr::compact()</span></span>
<span id="cb107-13"><a href="#cb107-13" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::count()      masks plyr::count()</span></span>
<span id="cb107-14"><a href="#cb107-14" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::failwith()   masks plyr::failwith()</span></span>
<span id="cb107-15"><a href="#cb107-15" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::filter()     masks stats::filter()</span></span>
<span id="cb107-16"><a href="#cb107-16" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::group_rows() masks kableExtra::group_rows()</span></span>
<span id="cb107-17"><a href="#cb107-17" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::id()         masks plyr::id()</span></span>
<span id="cb107-18"><a href="#cb107-18" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::lag()        masks stats::lag()</span></span>
<span id="cb107-19"><a href="#cb107-19" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::mutate()     masks plyr::mutate()</span></span>
<span id="cb107-20"><a href="#cb107-20" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::rename()     masks plyr::rename()</span></span>
<span id="cb107-21"><a href="#cb107-21" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::summarise()  masks plyr::summarise()</span></span>
<span id="cb107-22"><a href="#cb107-22" aria-hidden="true" tabindex="-1"></a><span class="do">## x dplyr::summarize()  masks plyr::summarize()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co">#count how many measurements we have per conditions</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>growth_data <span class="sc">%&gt;%</span> <span class="fu">count</span>(Condition, <span class="at">sort =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Condition"],"name":[1],"type":["chr"],"align":["left"]},{"label":["n"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"MgCl","2":"28"},{"1":"Strain101","2":"25"},{"1":"Strain230","2":"24"},{"1":"Strain28","2":"28"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co">#count how many measurements we have per nutrient and conditions</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>growth_data <span class="sc">%&gt;%</span> <span class="fu">count</span>(Nutrient, Condition, <span class="at">sort =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Nutrient"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[2],"type":["chr"],"align":["left"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"P","2":"MgCl","3":"15"},{"1":"P","2":"Strain101","3":"12"},{"1":"P","2":"Strain230","3":"13"},{"1":"P","2":"Strain28","3":"13"},{"1":"noP","2":"MgCl","3":"13"},{"1":"noP","2":"Strain101","3":"13"},{"1":"noP","2":"Strain230","3":"11"},{"1":"noP","2":"Strain28","3":"15"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get more comprehensive data stats and summarize for the whole dataset</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>growth_data <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">count =</span> <span class="fu">n</span>(),</span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">mean_root =</span> <span class="fu">mean</span>(Rootlength, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["count"],"name":[1],"type":["int"],"align":["right"]},{"label":["mean_root"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"105","2":"9.046117"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Good things about tydir are:</p>
<ul>
<li>it is extremely fast, so is important for dealing with larger datasets</li>
<li>we can combine several compands by using the pipe</li>
</ul>
<p>I.e. below we can see that we first group the input data and then summarized the group data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get more comprehensive data stats and summarize and grouping for the different conditions</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>growth_data <span class="sc">%&gt;%</span> </span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(Nutrient, Condition) <span class="sc">%&gt;%</span>  </span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarise</span>(</span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">count =</span> <span class="fu">n</span>(),</span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">mean_root =</span> <span class="fu">mean</span>(Rootlength, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) ,<span class="at">.groups =</span> <span class="st">'drop'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Nutrient"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[2],"type":["chr"],"align":["left"]},{"label":["count"],"name":[3],"type":["int"],"align":["right"]},{"label":["mean_root"],"name":[4],"type":["dbl"],"align":["right"]}],"data":[{"1":"P","2":"MgCl","3":"15","4":"9.853457"},{"1":"P","2":"Strain101","3":"12","4":"12.729133"},{"1":"P","2":"Strain230","3":"13","4":"11.814678"},{"1":"P","2":"Strain28","3":"13","4":"12.007390"},{"1":"noP","2":"MgCl","3":"13","4":"5.805198"},{"1":"noP","2":"Strain101","3":"13","4":"7.094616"},{"1":"noP","2":"Strain230","3":"11","4":"7.172883"},{"1":"noP","2":"Strain28","3":"15","4":"6.200309"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>We can also do other stats than just calculating the mean:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get more comprehensive data stats and summarize and grouping for the different conditions</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>growth_data <span class="sc">%&gt;%</span> </span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(Nutrient, Condition) <span class="sc">%&gt;%</span>  </span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarise</span>(</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">count =</span> <span class="fu">n</span>(),</span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">mean_root =</span> <span class="fu">mean</span>(Rootlength, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">sd_root =</span> <span class="fu">sd</span>(Rootlength),<span class="at">.groups =</span> <span class="st">'drop'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Nutrient"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[2],"type":["chr"],"align":["left"]},{"label":["count"],"name":[3],"type":["int"],"align":["right"]},{"label":["mean_root"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["sd_root"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"P","2":"MgCl","3":"15","4":"9.853457","5":"1.2981376"},{"1":"P","2":"Strain101","3":"12","4":"12.729133","5":"0.9813173"},{"1":"P","2":"Strain230","3":"13","4":"11.814678","5":"0.9128499"},{"1":"P","2":"Strain28","3":"13","4":"12.007390","5":"0.5812310"},{"1":"noP","2":"MgCl","3":"13","4":"5.805198","5":"0.7738124"},{"1":"noP","2":"Strain101","3":"13","4":"7.094616","5":"0.5939804"},{"1":"noP","2":"Strain230","3":"11","4":"7.172883","5":"0.6117203"},{"1":"noP","2":"Strain28","3":"15","4":"6.200309","5":"0.9988989"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Other useful summary functions:</p>
<ul>
<li><p><code>mean(x)</code>: sum of x divided by the length</p></li>
<li><p><code>median(x)</code>: 50% of x is above and 50% is below</p></li>
<li><p><code>sd(x)</code>: standard deviation</p></li>
<li><p><code>IQR(x)</code>: interquartile range (robust equivalent of sd when outliers are present in the data)</p></li>
<li><p><code>mad(x)</code>: median absolute deviation (robust equivalent of sd when outliers are present in the data)</p></li>
<li><p><code>min(x)</code>: minimum value of x</p></li>
<li><p><code>max(x)</code>: maximum value of x</p></li>
<li><p><code>quantile(x, 0.25)</code>: 25% of x is below this value</p></li>
<li><p><code>first(x)</code>: equivalent to x[1]</p></li>
<li><p><code>nth(x, 2)</code>: equivalent to n&lt;-2; x[n]</p></li>
<li><p><code>last(x)</code>: equivalent to x[length(x)]</p></li>
<li><p><code>n(x)</code>: the number of elements in x</p></li>
<li><p><code>sum(!is.na(x))</code>: count non-missing values</p></li>
<li><p><code>n_distinct(x)</code>: count the number of unique value</p></li>
<li><p><code>sum(x &gt; 10)</code>: count the number of elements where x &gt; 10</p></li>
<li><p><code>mean(y == 0)</code>: proportion of elements where y = 0</p></li>
</ul>
</section>
<section id="more-stats-tapply" class="level3" data-number="1.25.5">
<h3 data-number="1.25.5" class="anchored" data-anchor-id="more-stats-tapply"><span class="header-section-number">1.25.5</span> More stats: tapply</h3>
<p>Tapply is a base R function and allows to apply a function to each cell of a ragged array, that is to each (non-empty) group of values given by a unique combination of the levels of certain factors. It can be used as an alternative to ddply.</p>
<p>Suppose now we want to estimate the mean root length for each growth condition. Notice, how we can vary the index?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co">#index = factors we want to select</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(<span class="at">X =</span> growth_data<span class="sc">$</span>Rootlength, <span class="at">INDEX =</span> growth_data<span class="sc">$</span>Nutrient, <span class="at">FUN =</span> mean,<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a><span class="do">##         P       noP </span></span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a><span class="do">## 11.513931  6.530845</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co">#same but for two-way tables (this is not so useful here, but might be handy when you have different conditions for the same organism or a timecourse)</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(growth_data<span class="sc">$</span>Rootlength, <span class="at">INDEX =</span> <span class="fu">list</span>(growth_data<span class="sc">$</span>Nutrient, growth_data<span class="sc">$</span>Condition),<span class="at">FUN =</span> mean, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a><span class="do">##         MgCl Strain101 Strain230  Strain28</span></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a><span class="do">## P   9.853457 12.729133 11.814678 12.007390</span></span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a><span class="do">## noP 5.805198  7.094616  7.172883  6.200309</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="linear-models" class="level3" data-number="1.25.6">
<h3 data-number="1.25.6" class="anchored" data-anchor-id="linear-models"><span class="header-section-number">1.25.6</span> Linear models</h3>
</section>
<section id="basics" class="level3" data-number="1.25.7">
<h3 data-number="1.25.7" class="anchored" data-anchor-id="basics"><span class="header-section-number">1.25.7</span> Basics</h3>
<p>It is very simple to investigate linear relationships among variables in R. We want to estimate how a quantitative dependent variable changes according to the levels of one or more categorical independent variables. In the command below, the linear relationship between Rootlength (the <strong>dependent</strong> variable, i.e.&nbsp;the one we<e2>&lt;80&gt;&lt;99&gt;re trying to predict) and FW_shoot_mg (the <strong>independent</strong> variable or the predictor) is calculated.</e2></p>
<p>We need to use the function summary to see the results of that command; coef extracts the best-fit coefficients, anova performs an analysis of variance; there are many other extractor functions.</p>
<p>In this case lets work with the data were we do not have to deal with the different time points to simplify things.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="co">#read in data were we have several measurements</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>growth_data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/Growth_Data.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a><span class="co">#is there a correlation between freshweight and root length?</span></span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>linearmodel <span class="ot">&lt;-</span> <span class="fu">lm</span>(Rootlength <span class="sc">~</span> FW_shoot_mg, <span class="at">data =</span> growth_data)</span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a>linearmodel</span>
<span id="cb115-7"><a href="#cb115-7" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb115-8"><a href="#cb115-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Call:</span></span>
<span id="cb115-9"><a href="#cb115-9" aria-hidden="true" tabindex="-1"></a><span class="do">## lm(formula = Rootlength ~ FW_shoot_mg, data = growth_data)</span></span>
<span id="cb115-10"><a href="#cb115-10" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb115-11"><a href="#cb115-11" aria-hidden="true" tabindex="-1"></a><span class="do">## Coefficients:</span></span>
<span id="cb115-12"><a href="#cb115-12" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept)  FW_shoot_mg  </span></span>
<span id="cb115-13"><a href="#cb115-13" aria-hidden="true" tabindex="-1"></a><span class="do">##      5.2912       0.2465</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co">#let's extract the entire t-table</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(linearmodel) </span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Call:</span></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a><span class="do">## lm(formula = Rootlength ~ FW_shoot_mg, data = growth_data)</span></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Residuals:</span></span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a><span class="do">##     Min      1Q  Median      3Q     Max </span></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a><span class="do">## -4.5263 -0.9978 -0.0181  1.0094  3.9292 </span></span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a><span class="do">## Coefficients:</span></span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a><span class="do">##             Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept)  5.29124    0.30469   17.37   &lt;2e-16 ***</span></span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a><span class="do">## FW_shoot_mg  0.24648    0.01712   14.40   &lt;2e-16 ***</span></span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a><span class="do">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span id="cb116-17"><a href="#cb116-17" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb116-18"><a href="#cb116-18" aria-hidden="true" tabindex="-1"></a><span class="do">## Residual standard error: 1.615 on 103 degrees of freedom</span></span>
<span id="cb116-19"><a href="#cb116-19" aria-hidden="true" tabindex="-1"></a><span class="do">## Multiple R-squared:  0.6681, Adjusted R-squared:  0.6649 </span></span>
<span id="cb116-20"><a href="#cb116-20" aria-hidden="true" tabindex="-1"></a><span class="do">## F-statistic: 207.4 on 1 and 103 DF,  p-value: &lt; 2.2e-16</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here one of the values is the model p-Value (bottom last line) and the p-Value of individual predictor variables (extreme right column under <e2>&lt;80&gt;&lt;98&gt;Coefficients<e2>&lt;80&gt;&lt;99&gt;). The p-Values are very important because, we can consider a linear model to be statistically significant only when both these p-Values are less that the pre-determined statistical significance level, which is ideally 0.05. This is visually interpreted by the significance stars at the end of the row. The more the stars beside the variable<e2>&lt;80&gt;&lt;99&gt;s p-Value, the more significant the variable.</e2></e2></e2></p>
<ul>
<li>Residuals: The section summarizes the residuals, the error between the prediction of the model and the actual results. Smaller residuals are better.</li>
<li>Coefficients: For each variable and the intercept, a weight is produced and that weight has other attributes like the standard error, a t-test value and significance.</li>
<li>Estimate: This is the weight given to the variable. In the simple regression case (one variable plus the intercept), for every increase in root length, the model predicts an increase of 0.24.</li>
<li>Std. Error: Tells you how precisely was the estimate measured. It<e2>&lt;80&gt;&lt;99&gt;s really only useful for calculating the t-value.</e2></li>
<li>t-value and Pr(&gt;[t]): The t-value is calculated by taking the coefficient divided by the Std. Error. It is then used to test whether or not the coefficient is significantly different from zero. If it isn<e2>&lt;80&gt;&lt;99&gt;t significant, then the coefficient really isn<e2>&lt;80&gt;&lt;99&gt;t adding anything to the model and could be dropped or investigated further. Pr(&gt;|t|) is the significance level.</e2></e2></li>
</ul>
<p>Performance Measures:</p>
<p>Three sets of measurements are provided.</p>
<ul>
<li>Residual Standard Error: This is the standard deviation of the residuals. Smaller is better.</li>
<li>Multiple / Adjusted R-Square: For one variable, the distinction doesn<e2>&lt;80&gt;&lt;99&gt;t really matter. R-squared shows the amount of variance explained by the model. Adjusted R-Square takes into account the number of variables and is most useful for multiple-regression.</e2></li>
<li>F-Statistic: The F-test checks if at least one variable<e2>&lt;80&gt;&lt;99&gt;s weight is significantly different than zero. This is a global test to help asses a model. If the p-value is not significant (e.g.&nbsp;greater than 0.05) than your model is essentially not doing anything.</e2></li>
</ul>
<p>We also can just print parts of the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print only the coefficients </span></span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="fu">coef</span>(<span class="fu">summary</span>(linearmodel))</span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a><span class="do">##              Estimate Std. Error  t value     Pr(&gt;|t|)</span></span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept) 5.2912436 0.30469198 17.36588 2.278273e-32</span></span>
<span id="cb117-5"><a href="#cb117-5" aria-hidden="true" tabindex="-1"></a><span class="do">## FW_shoot_mg 0.2464783 0.01711628 14.40023 2.039703e-26</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print only the anova stats</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(linearmodel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["Df"],"name":[1],"type":["int"],"align":["right"]},{"label":["Sum Sq"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["Mean Sq"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["F value"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Pr(>F)"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"1","2":"540.9865","3":"540.986514","4":"207.3665","5":"2.039703e-26","_rn_":"FW_shoot_mg"},{"1":"103","2":"268.7108","3":"2.608842","4":"NA","5":"NA","_rn_":"Residuals"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot add the best line to a plot</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(growth_data, <span class="fu">plot</span>(Rootlength <span class="sc">~</span> FW_shoot_mg, <span class="at">col =</span> <span class="dv">2</span>))</span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(linearmodel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-121-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>If we look at the stats and the p value we see a nice correlation but also that we have two distinct clusters as well as more spread in the cluster that is more to the right. These clusters likely are the two different nutrient conditions and sometimes it might make sense to separate data to get a clearer picture. Something else to consider is to ask whether the data is normally distributed and based on that what statistical test to choose.</p>
<section id="analysing-residuals" class="level4" data-number="1.25.7.1">
<h4 data-number="1.25.7.1" class="anchored" data-anchor-id="analysing-residuals"><span class="header-section-number">1.25.7.1</span> Analysing residuals</h4>
<p>Anyone can fit a linear model in R. The real test is analyzing the residuals (the error or the difference between actual and predicted results).</p>
<p>There are four things we are looking for when analyzing residuals.</p>
<ul>
<li>The mean of the errors is zero (and the sum of the errors is zero)</li>
<li>The distribution of the errors are normal.</li>
<li>All of the errors are independent.</li>
<li>Variance of errors is constant (Homoscedastic)</li>
</ul>
<p>In R, you pull out the residuals by referencing the model and then the resid variable inside the model. Using the simple linear regression model (simple.fit) we<e2>&lt;80&gt;&lt;99&gt;ll plot a few graphs to help illustrate any problems with the model.</e2></p>
<p>Below some examples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>simple.fit <span class="ot">&lt;-</span> linearmodel</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a><span class="fu">layout</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>),<span class="dv">2</span>,<span class="dv">2</span>,<span class="at">byrow=</span>T))</span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Rootlength x Residuals Plot</span></span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(simple.fit<span class="sc">$</span>resid<span class="sc">~</span>growth_data<span class="sc">$</span>Rootlength[<span class="fu">order</span>(growth_data<span class="sc">$</span>Rootlength)],</span>
<span id="cb120-7"><a href="#cb120-7" aria-hidden="true" tabindex="-1"></a> <span class="at">main=</span><span class="st">"Rootlength x Residuals</span><span class="sc">\n</span><span class="st">for Simple Regression"</span>,</span>
<span id="cb120-8"><a href="#cb120-8" aria-hidden="true" tabindex="-1"></a> <span class="at">xlab=</span><span class="st">"Marketing Rootlength"</span>, <span class="at">ylab=</span><span class="st">"Residuals"</span>)</span>
<span id="cb120-9"><a href="#cb120-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">0</span>,<span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb120-10"><a href="#cb120-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-11"><a href="#cb120-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Histogram of Residuals</span></span>
<span id="cb120-12"><a href="#cb120-12" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(simple.fit<span class="sc">$</span>resid, <span class="at">main=</span><span class="st">"Histogram of Residuals"</span>,</span>
<span id="cb120-13"><a href="#cb120-13" aria-hidden="true" tabindex="-1"></a> <span class="at">ylab=</span><span class="st">"Residuals"</span>)</span>
<span id="cb120-14"><a href="#cb120-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-15"><a href="#cb120-15" aria-hidden="true" tabindex="-1"></a><span class="co">#Q-Q Plot</span></span>
<span id="cb120-16"><a href="#cb120-16" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(simple.fit<span class="sc">$</span>resid)</span>
<span id="cb120-17"><a href="#cb120-17" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(simple.fit<span class="sc">$</span>resid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-122-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The histogram and QQ-plot are the ways to visually evaluate if the residual fit a normal distribution.</p>
<ul>
<li>If the histogram looks like a bell-curve it might be normally distributed.</li>
<li>If the QQ-plot has the vast majority of points on or very near the line, the residuals may be normally distributed.</li>
</ul>
</section>
</section>
<section id="normal-distribution" class="level3" data-number="1.25.8">
<h3 data-number="1.25.8" class="anchored" data-anchor-id="normal-distribution"><span class="header-section-number">1.25.8</span> Normal distribution</h3>
<section id="visualize-our-data-via-density-plots" class="level5" data-number="1.25.8.0.1">
<h5 data-number="1.25.8.0.1" class="anchored" data-anchor-id="visualize-our-data-via-density-plots"><span class="header-section-number">1.25.8.0.1</span> Visualize our data via density plots</h5>
<p>There are different ways to visualize this, one example is ggdensity of the ggpubr package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggpubr"</span>)</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'ggpubr'</span></span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:plyr':</span></span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a><span class="do">##     mutate</span></span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a><span class="co">#is the data for my different variables normally distributed</span></span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggdensity</span>(growth_data<span class="sc">$</span>Rootlength)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-123-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We see nicely that we have two tails that likely represent the two nutrient conditions. To test this, we can simply subset the data as we have done before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggdensity</span>(growth_data, <span class="at">x =</span> <span class="st">"Rootlength"</span>,</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>   <span class="at">add =</span> <span class="st">"mean"</span>, <span class="at">rug =</span> <span class="cn">TRUE</span>,</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">color =</span> <span class="st">"Nutrient"</span>, <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-124-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>No we see that indeed the tool tails we see are seem to be due to our two nutrient conditions.</p>
</section>
<section id="visualize-our-data-via-q-q-plots" class="level4" data-number="1.25.8.1">
<h4 data-number="1.25.8.1" class="anchored" data-anchor-id="visualize-our-data-via-q-q-plots"><span class="header-section-number">1.25.8.1</span> Visualize our data via Q-Q plots</h4>
<p>Another way to represent data is in a Q-Q plot: Q-Q plot (or quantile-quantile plot) draws the correlation between a given sample and the normal distribution. A 45-degree reference line is also plotted.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot all data</span></span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggqqplot</span>(growth_data<span class="sc">$</span>Rootlength)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-125-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot by group</span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggqqplot</span>(growth_data, <span class="at">x =</span> <span class="st">"Rootlength"</span>,</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">color =</span> <span class="st">"Nutrient"</span>,  <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-126-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Again, here we see that our data for the indivdual growth conditions fit quite nicely into normal distribution.</p>
</section>
<section id="test-for-normality" class="level4" data-number="1.25.8.2">
<h4 data-number="1.25.8.2" class="anchored" data-anchor-id="test-for-normality"><span class="header-section-number">1.25.8.2</span> Test for normality</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co">#for all data</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(growth_data<span class="sc">$</span>Rootlength)</span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  Shapiro-Wilk normality test</span></span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a><span class="do">## data:  growth_data$Rootlength</span></span>
<span id="cb125-7"><a href="#cb125-7" aria-hidden="true" tabindex="-1"></a><span class="do">## W = 0.91773, p-value = 6.744e-06</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="co">#separate nutrient conditions</span></span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>noP_data <span class="ot">&lt;-</span> growth_data[growth_data<span class="sc">$</span>Nutrient <span class="sc">==</span> <span class="st">"noP"</span>, ]</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a><span class="co">#test for noP only</span></span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(noP_data<span class="sc">$</span>Rootlength)</span>
<span id="cb126-6"><a href="#cb126-6" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb126-7"><a href="#cb126-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  Shapiro-Wilk normality test</span></span>
<span id="cb126-8"><a href="#cb126-8" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb126-9"><a href="#cb126-9" aria-hidden="true" tabindex="-1"></a><span class="do">## data:  noP_data$Rootlength</span></span>
<span id="cb126-10"><a href="#cb126-10" aria-hidden="true" tabindex="-1"></a><span class="do">## W = 0.96403, p-value = 0.1171</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can use ddplyr and the dlookr package for doing a group-wise comparison</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dlookr)</span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'dlookr'</span></span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:tidyr':</span></span>
<span id="cb127-6"><a href="#cb127-6" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb127-7"><a href="#cb127-7" aria-hidden="true" tabindex="-1"></a><span class="do">##     extract</span></span>
<span id="cb127-8"><a href="#cb127-8" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:base':</span></span>
<span id="cb127-9"><a href="#cb127-9" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb127-10"><a href="#cb127-10" aria-hidden="true" tabindex="-1"></a><span class="do">##     transform</span></span>
<span id="cb127-11"><a href="#cb127-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-12"><a href="#cb127-12" aria-hidden="true" tabindex="-1"></a>growth_data <span class="sc">%&gt;%</span></span>
<span id="cb127-13"><a href="#cb127-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Nutrient) <span class="sc">%&gt;%</span></span>
<span id="cb127-14"><a href="#cb127-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normality</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["variable"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["statistic"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["p_value"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["sample"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"FW_shoot_mg","2":"P","3":"0.9681759","4":"0.16818081","5":"53"},{"1":"FW_shoot_mg","2":"noP","3":"0.9831431","4":"0.66649085","5":"52"},{"1":"Rootlength","2":"P","3":"0.9515682","4":"0.03144205","5":"53"},{"1":"Rootlength","2":"noP","3":"0.9640334","4":"0.11709068","5":"52"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The shapiro.test tests the NULL hypothesis that the samples came from a Normal distribution. This means that if your p-value &lt;= 0.05, then you would reject the NULL hypothesis that the samples came from a normal distribution.</p>
<p>From the output, the p-value &lt; 0.05 for our complete dataset implies that the distribution of the data is significantly different from normal distribution. In other words, we can not assume the normality. However, we expect quite some differences dependent on the growth pattern and once we only look at our low P data we see that our indeed is normally distributed.</p>
<p>Notice: Shapiro works only for sample sizes between 3-5000 numbers since when you feed it more data, the chances of the null hypothesis being rejected becomes larger. An alternative is the Anderson-Darling test that however has a similar problem with the Shapiro Wilk test. For large samples, you are most likely to reject the null hypothesis, so be aware of this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(nortest)</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ad.test</span>(noP_data<span class="sc">$</span>Rootlength)</span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  Anderson-Darling normality test</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb128-6"><a href="#cb128-6" aria-hidden="true" tabindex="-1"></a><span class="do">## data:  noP_data$Rootlength</span></span>
<span id="cb128-7"><a href="#cb128-7" aria-hidden="true" tabindex="-1"></a><span class="do">## A = 0.58838, p-value = 0.1187</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="anova" class="level3" data-number="1.25.9">
<h3 data-number="1.25.9" class="anchored" data-anchor-id="anova"><span class="header-section-number">1.25.9</span> ANOVA</h3>
<p>When we visualize the data, we see that there is a difference between the nutrient conditions but we want to know whether it is significant and more importantly, whether there is also a difference based on our treatments with different strains of microbes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co">#lets visually compare our data with ggpubr again</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggpubr"</span>)</span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggboxplot</span>(growth_data, <span class="at">x =</span> <span class="st">"Condition"</span>, <span class="at">y =</span> <span class="st">"Rootlength"</span>, <span class="at">color =</span> <span class="st">"Nutrient"</span>,</span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-131-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We want to know whether root length depends on nutrient treatment</span></span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>aov <span class="ot">&lt;-</span> <span class="fu">aov</span>(Rootlength <span class="sc">~</span> Nutrient, <span class="at">data =</span> growth_data)</span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(aov)</span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a><span class="do">##              Df Sum Sq Mean Sq F value Pr(&gt;F)    </span></span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Nutrient      1  651.8   651.8     425 &lt;2e-16 ***</span></span>
<span id="cb130-6"><a href="#cb130-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Residuals   103  157.9     1.5                   </span></span>
<span id="cb130-7"><a href="#cb130-7" aria-hidden="true" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb130-8"><a href="#cb130-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we see that there are significant differences based on our nutrient treatments. Now lets see how we can look at both the nutrient treatment and growth conditions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We want to know if root length depends on condition and nutrient</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>aov <span class="ot">&lt;-</span> <span class="fu">aov</span>(Rootlength <span class="sc">~</span> Nutrient <span class="sc">+</span> Condition, <span class="at">data =</span> growth_data)</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(aov)</span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a><span class="do">##              Df Sum Sq Mean Sq F value   Pr(&gt;F)    </span></span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Nutrient      1  651.8   651.8  712.58  &lt; 2e-16 ***</span></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Condition     3   66.5    22.2   24.23 7.27e-12 ***</span></span>
<span id="cb131-7"><a href="#cb131-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Residuals   100   91.5     0.9                     </span></span>
<span id="cb131-8"><a href="#cb131-8" aria-hidden="true" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb131-9"><a href="#cb131-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>From the ANOVA table we can conclude that both nutrient condition and treatment are statistically significant. Nutrient treatment is the most significant factor variable.</p>
<p>Not the above fitted model is called additive model. It makes an assumption that the two factor variables are independent. If you think that these two variables might interact to create an synergistic effect, replace the plus symbol (+) by an asterisk (*), as follows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Two-way ANOVA with interaction effect</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a><span class="co"># These two calls are equivalent</span></span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>aov <span class="ot">&lt;-</span> <span class="fu">aov</span>(Rootlength <span class="sc">~</span> Nutrient <span class="sc">*</span> Condition, <span class="at">data =</span> growth_data)</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>aov <span class="ot">&lt;-</span> <span class="fu">aov</span>(Rootlength <span class="sc">~</span> Nutrient <span class="sc">+</span> Condition <span class="sc">+</span> Nutrient<span class="sc">:</span>Condition, <span class="at">data =</span> growth_data)</span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a><span class="co">#summarize the aov</span></span>
<span id="cb132-8"><a href="#cb132-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(aov)</span>
<span id="cb132-9"><a href="#cb132-9" aria-hidden="true" tabindex="-1"></a><span class="do">##                    Df Sum Sq Mean Sq F value   Pr(&gt;F)    </span></span>
<span id="cb132-10"><a href="#cb132-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Nutrient            1  651.8   651.8 817.131  &lt; 2e-16 ***</span></span>
<span id="cb132-11"><a href="#cb132-11" aria-hidden="true" tabindex="-1"></a><span class="do">## Condition           3   66.5    22.2  27.780 4.72e-13 ***</span></span>
<span id="cb132-12"><a href="#cb132-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Nutrient:Condition  3   14.1     4.7   5.891 0.000976 ***</span></span>
<span id="cb132-13"><a href="#cb132-13" aria-hidden="true" tabindex="-1"></a><span class="do">## Residuals          97   77.4     0.8                     </span></span>
<span id="cb132-14"><a href="#cb132-14" aria-hidden="true" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb132-15"><a href="#cb132-15" aria-hidden="true" tabindex="-1"></a><span class="do">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It can be seen that the two main effects (Nutrient and Condition) are statistically significant, as well as their interaction.</p>
<p><strong>Note that, in the situation where the interaction is not significant you should use the additive model.</strong></p>
</section>
<section id="tukey" class="level3" data-number="1.25.10">
<h3 data-number="1.25.10" class="anchored" data-anchor-id="tukey"><span class="header-section-number">1.25.10</span> TUKEY</h3>
<p>In ANOVA test, a significant p-value indicates that some of the group means are different, but we don<e2>&lt;80&gt;&lt;99&gt;t know which pairs of groups are different. It<e2>&lt;80&gt;&lt;99&gt;s possible to perform multiple pairwise-comparison, to determine if the mean difference between specific pairs of group are statistically significant.</e2></e2></p>
<p>As the ANOVA test is significant, we can compute Tukey HSD (Tukey Honest Significant Differences). Tukey test is a single-step multiple comparison procedure and statistical test. It is a post-hoc analysis, what means that it is used in conjunction with an ANOVA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co">#test with anova</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>aov <span class="ot">&lt;-</span> <span class="fu">aov</span>(Rootlength <span class="sc">~</span> Nutrient <span class="sc">*</span> Condition, <span class="at">data =</span> growth_data)</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a><span class="co">#run tukey</span></span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a><span class="fu">TukeyHSD</span>(aov)</span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a><span class="do">##   Tukey multiple comparisons of means</span></span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a><span class="do">##     95% family-wise confidence level</span></span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-9"><a href="#cb133-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Fit: aov(formula = Rootlength ~ Nutrient * Condition, data = growth_data)</span></span>
<span id="cb133-10"><a href="#cb133-10" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-11"><a href="#cb133-11" aria-hidden="true" tabindex="-1"></a><span class="do">## $Nutrient</span></span>
<span id="cb133-12"><a href="#cb133-12" aria-hidden="true" tabindex="-1"></a><span class="do">##            diff       lwr       upr p adj</span></span>
<span id="cb133-13"><a href="#cb133-13" aria-hidden="true" tabindex="-1"></a><span class="do">## noP-P -4.983086 -5.329067 -4.637105     0</span></span>
<span id="cb133-14"><a href="#cb133-14" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-15"><a href="#cb133-15" aria-hidden="true" tabindex="-1"></a><span class="do">## $Condition</span></span>
<span id="cb133-16"><a href="#cb133-16" aria-hidden="true" tabindex="-1"></a><span class="do">##                           diff        lwr        upr     p adj</span></span>
<span id="cb133-17"><a href="#cb133-17" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain101-MgCl       2.1029052  1.4604930  2.7453173 0.0000000</span></span>
<span id="cb133-18"><a href="#cb133-18" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain230-MgCl       1.6836191  1.0341749  2.3330633 0.0000000</span></span>
<span id="cb133-19"><a href="#cb133-19" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain28-MgCl        1.2784796  0.6545138  1.9024455 0.0000034</span></span>
<span id="cb133-20"><a href="#cb133-20" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain230-Strain101 -0.4192861 -1.0864726  0.2479004 0.3597470</span></span>
<span id="cb133-21"><a href="#cb133-21" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain28-Strain101  -0.8244256 -1.4668377 -0.1820134 0.0061445</span></span>
<span id="cb133-22"><a href="#cb133-22" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain28-Strain230  -0.4051395 -1.0545837  0.2443048 0.3663501</span></span>
<span id="cb133-23"><a href="#cb133-23" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-24"><a href="#cb133-24" aria-hidden="true" tabindex="-1"></a><span class="do">## $`Nutrient:Condition`</span></span>
<span id="cb133-25"><a href="#cb133-25" aria-hidden="true" tabindex="-1"></a><span class="do">##                                   diff        lwr        upr     p adj</span></span>
<span id="cb133-26"><a href="#cb133-26" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:MgCl-P:MgCl             -4.0482585 -5.0967311 -2.9997859 0.0000000</span></span>
<span id="cb133-27"><a href="#cb133-27" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain101-P:MgCl           2.8756763  1.8040558  3.9472967 0.0000000</span></span>
<span id="cb133-28"><a href="#cb133-28" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain101-P:MgCl        -2.7588405 -3.8073131 -1.7103679 0.0000000</span></span>
<span id="cb133-29"><a href="#cb133-29" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain230-P:MgCl           1.9612207  0.9127482  3.0096933 0.0000023</span></span>
<span id="cb133-30"><a href="#cb133-30" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain230-P:MgCl        -2.6805735 -3.7789218 -1.5822252 0.0000000</span></span>
<span id="cb133-31"><a href="#cb133-31" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain28-P:MgCl            2.1539326  1.1054600  3.2024052 0.0000002</span></span>
<span id="cb133-32"><a href="#cb133-32" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain28-P:MgCl         -3.6531484 -4.6634819 -2.6428149 0.0000000</span></span>
<span id="cb133-33"><a href="#cb133-33" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain101-noP:MgCl         6.9239348  5.8162834  8.0315861 0.0000000</span></span>
<span id="cb133-34"><a href="#cb133-34" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain101-noP:MgCl       1.2894180  0.2041458  2.3746903 0.0087961</span></span>
<span id="cb133-35"><a href="#cb133-35" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain230-noP:MgCl         6.0094793  4.9242070  7.0947515 0.0000000</span></span>
<span id="cb133-36"><a href="#cb133-36" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain230-noP:MgCl       1.3676850  0.2341551  2.5012149 0.0073003</span></span>
<span id="cb133-37"><a href="#cb133-37" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain28-noP:MgCl          6.2021911  5.1169189  7.2874633 0.0000000</span></span>
<span id="cb133-38"><a href="#cb133-38" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain28-noP:MgCl        0.3951101 -0.6533625  1.4435827 0.9391499</span></span>
<span id="cb133-39"><a href="#cb133-39" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain101-P:Strain101   -5.6345168 -6.7421681 -4.5268654 0.0000000</span></span>
<span id="cb133-40"><a href="#cb133-40" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain230-P:Strain101     -0.9144555 -2.0221069  0.1931958 0.1845494</span></span>
<span id="cb133-41"><a href="#cb133-41" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain230-P:Strain101   -5.5562498 -6.7112241 -4.4012755 0.0000000</span></span>
<span id="cb133-42"><a href="#cb133-42" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain28-P:Strain101      -0.7217437 -1.8293950  0.3859077 0.4749256</span></span>
<span id="cb133-43"><a href="#cb133-43" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain28-P:Strain101    -6.5288247 -7.6004451 -5.4572042 0.0000000</span></span>
<span id="cb133-44"><a href="#cb133-44" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain230-noP:Strain101    4.7200613  3.6347890  5.8053335 0.0000000</span></span>
<span id="cb133-45"><a href="#cb133-45" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain230-noP:Strain101  0.0782670 -1.0552629  1.2117969 0.9999989</span></span>
<span id="cb133-46"><a href="#cb133-46" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain28-noP:Strain101     4.9127731  3.8275008  5.9980453 0.0000000</span></span>
<span id="cb133-47"><a href="#cb133-47" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain28-noP:Strain101  -0.8943079 -1.9427805  0.1541647 0.1537008</span></span>
<span id="cb133-48"><a href="#cb133-48" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain230-P:Strain230   -4.6417943 -5.7753242 -3.5082644 0.0000000</span></span>
<span id="cb133-49"><a href="#cb133-49" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain28-P:Strain230       0.1927118 -0.8925604  1.2779841 0.9993212</span></span>
<span id="cb133-50"><a href="#cb133-50" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain28-P:Strain230    -5.6143692 -6.6628417 -4.5658966 0.0000000</span></span>
<span id="cb133-51"><a href="#cb133-51" aria-hidden="true" tabindex="-1"></a><span class="do">## P:Strain28-noP:Strain230     4.8345061  3.7009762  5.9680360 0.0000000</span></span>
<span id="cb133-52"><a href="#cb133-52" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain28-noP:Strain230  -0.9725749 -2.0709232  0.1257734 0.1223122</span></span>
<span id="cb133-53"><a href="#cb133-53" aria-hidden="true" tabindex="-1"></a><span class="do">## noP:Strain28-P:Strain28     -5.8070810 -6.8555536 -4.7586084 0.0000000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that most differences are significant, with the exception of Strain28, which in most cases does not show an effect.</p>
<p>For some representations it is useful to plot significant letters. We can do this using some extra packages as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load library</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(agricolae)</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'agricolae'</span></span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a><span class="do">## The following objects are masked from 'package:dlookr':</span></span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a><span class="do">##     kurtosis, skewness</span></span>
<span id="cb134-8"><a href="#cb134-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-9"><a href="#cb134-9" aria-hidden="true" tabindex="-1"></a><span class="co">#separate nutrient conditions</span></span>
<span id="cb134-10"><a href="#cb134-10" aria-hidden="true" tabindex="-1"></a>noP_data <span class="ot">&lt;-</span> growth_data[growth_data<span class="sc">$</span>Nutrient <span class="sc">==</span> <span class="st">"noP"</span>, ]</span>
<span id="cb134-11"><a href="#cb134-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-12"><a href="#cb134-12" aria-hidden="true" tabindex="-1"></a><span class="co">#run an anova</span></span>
<span id="cb134-13"><a href="#cb134-13" aria-hidden="true" tabindex="-1"></a>aov_noP <span class="ot">&lt;-</span> <span class="fu">aov</span>(Rootlength <span class="sc">~</span> Condition, <span class="at">data =</span> noP_data)</span>
<span id="cb134-14"><a href="#cb134-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-15"><a href="#cb134-15" aria-hidden="true" tabindex="-1"></a><span class="co">#run test</span></span>
<span id="cb134-16"><a href="#cb134-16" aria-hidden="true" tabindex="-1"></a><span class="fu">HSD.test</span>(aov_noP,<span class="st">"Condition"</span>, <span class="at">group=</span><span class="cn">TRUE</span>,<span class="at">console=</span><span class="cn">TRUE</span>)</span>
<span id="cb134-17"><a href="#cb134-17" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-18"><a href="#cb134-18" aria-hidden="true" tabindex="-1"></a><span class="do">## Study: aov_noP ~ "Condition"</span></span>
<span id="cb134-19"><a href="#cb134-19" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-20"><a href="#cb134-20" aria-hidden="true" tabindex="-1"></a><span class="do">## HSD Test for Rootlength </span></span>
<span id="cb134-21"><a href="#cb134-21" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-22"><a href="#cb134-22" aria-hidden="true" tabindex="-1"></a><span class="do">## Mean Square Error:  0.606883 </span></span>
<span id="cb134-23"><a href="#cb134-23" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-24"><a href="#cb134-24" aria-hidden="true" tabindex="-1"></a><span class="do">## Condition,  means</span></span>
<span id="cb134-25"><a href="#cb134-25" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-26"><a href="#cb134-26" aria-hidden="true" tabindex="-1"></a><span class="do">##           Rootlength       std  r      Min      Max</span></span>
<span id="cb134-27"><a href="#cb134-27" aria-hidden="true" tabindex="-1"></a><span class="do">## MgCl        5.805198 0.7738124 13 4.236348 6.834720</span></span>
<span id="cb134-28"><a href="#cb134-28" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain101   7.094616 0.5939804 13 5.947965 7.973207</span></span>
<span id="cb134-29"><a href="#cb134-29" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain230   7.172883 0.6117203 11 6.120089 8.197116</span></span>
<span id="cb134-30"><a href="#cb134-30" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain28    6.200309 0.9988989 15 4.384756 7.428674</span></span>
<span id="cb134-31"><a href="#cb134-31" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-32"><a href="#cb134-32" aria-hidden="true" tabindex="-1"></a><span class="do">## Alpha: 0.05 ; DF Error: 48 </span></span>
<span id="cb134-33"><a href="#cb134-33" aria-hidden="true" tabindex="-1"></a><span class="do">## Critical Value of Studentized Range: 3.763749 </span></span>
<span id="cb134-34"><a href="#cb134-34" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-35"><a href="#cb134-35" aria-hidden="true" tabindex="-1"></a><span class="do">## Groups according to probability of means differences and alpha level( 0.05 )</span></span>
<span id="cb134-36"><a href="#cb134-36" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-37"><a href="#cb134-37" aria-hidden="true" tabindex="-1"></a><span class="do">## Treatments with the same letter are not significantly different.</span></span>
<span id="cb134-38"><a href="#cb134-38" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb134-39"><a href="#cb134-39" aria-hidden="true" tabindex="-1"></a><span class="do">##           Rootlength groups</span></span>
<span id="cb134-40"><a href="#cb134-40" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain230   7.172883      a</span></span>
<span id="cb134-41"><a href="#cb134-41" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain101   7.094616      a</span></span>
<span id="cb134-42"><a href="#cb134-42" aria-hidden="true" tabindex="-1"></a><span class="do">## Strain28    6.200309      b</span></span>
<span id="cb134-43"><a href="#cb134-43" aria-hidden="true" tabindex="-1"></a><span class="do">## MgCl        5.805198      b</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="check-the-homogeneity-of-variances" class="level3" data-number="1.25.11">
<h3 data-number="1.25.11" class="anchored" data-anchor-id="check-the-homogeneity-of-variances"><span class="header-section-number">1.25.11</span> Check the homogeneity of variances</h3>
<p>The residuals versus fits plot is used to check the homogeneity of variances. In the plot below, there is no evident relationships between residuals and fitted values (the mean of each groups), which is good. So, we can assume the homogeneity of variances. Only a few points (41, 58 and 77 are detected as outliers, which can severely normality and homogeneity of variance. It can be useful to remove outliers to meet the test assumptions.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check for homogeneity</span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(aov, <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-137-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Use the Levene&lt;e2&gt;&lt;80&gt;&lt;99&gt;s test to check the homogeneity of variances. </span></span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Loading required package: carData</span></span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'car'</span></span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:dplyr':</span></span>
<span id="cb136-8"><a href="#cb136-8" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb136-9"><a href="#cb136-9" aria-hidden="true" tabindex="-1"></a><span class="do">##     recode</span></span>
<span id="cb136-10"><a href="#cb136-10" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:purrr':</span></span>
<span id="cb136-11"><a href="#cb136-11" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb136-12"><a href="#cb136-12" aria-hidden="true" tabindex="-1"></a><span class="do">##     some</span></span>
<span id="cb136-13"><a href="#cb136-13" aria-hidden="true" tabindex="-1"></a><span class="fu">leveneTest</span>(Rootlength <span class="sc">~</span> Nutrient <span class="sc">*</span> Condition, <span class="at">data =</span> growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["Df"],"name":[1],"type":["int"],"align":["right"]},{"label":["F value"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["Pr(>F)"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"7","2":"1.862279","3":"0.08411691","_rn_":"group"},{"1":"97","2":"NA","3":"NA","_rn_":""}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>From the output above we can see that the p-value is not less than the significance level of 0.05. This means that there is no evidence to suggest that the variance across groups is statistically significantly different. Therefore, we can assume the homogeneity of variances in the different treatment groups.</p>
</section>
<section id="check-for-normality-v2" class="level3" data-number="1.25.12">
<h3 data-number="1.25.12" class="anchored" data-anchor-id="check-for-normality-v2"><span class="header-section-number">1.25.12</span> Check for normality v2</h3>
<p>Normality plot of the residuals. In the plot below, the quantiles of the residuals are plotted against the quantiles of the normal distribution. A 45-degree reference line is also plotted. The normal probability plot of residuals is used to verify the assumption that the residuals are normally distributed.</p>
<p>The normal probability plot of the residuals should approximately follow a straight line. which we can see below. Again, we see the points marked as potential outliers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="do">## plot</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(aov, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-138-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="control-structures" class="level2" data-number="1.26">
<h2 data-number="1.26" class="anchored" data-anchor-id="control-structures"><span class="header-section-number">1.26</span> Control Structures</h2>
<section id="loops" class="level3" data-number="1.26.1">
<h3 data-number="1.26.1" class="anchored" data-anchor-id="loops"><span class="header-section-number">1.26.1</span> Loops</h3>
<p>Sometimes it is necessary to repeat a calculation multiple times, e.g.&nbsp;calculate the sum of each row of a matrix. You can use for loops to do this.</p>
<p>In the first loop, we:</p>
<ul>
<li>define that we want to loop through rows 1 to 5 (1:5)</li>
<li>for (i in 1:5) –&gt; we say that we want to run a for loop, in which we name our variable i</li>
<li>{} –&gt; defines the operation we want to loop through</li>
<li>in our case we want to sum a specific row (rows 1 trhough 5)</li>
<li>so in i we store the numbers 1,2,3,4 and 5 and then run through the sum functions 5x</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="co">#build a dataframe</span></span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="dv">5</span>)</span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>m</span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3]</span></span>
<span id="cb138-5"><a href="#cb138-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1,]    1    6   11</span></span>
<span id="cb138-6"><a href="#cb138-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [2,]    2    7   12</span></span>
<span id="cb138-7"><a href="#cb138-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [3,]    3    8   13</span></span>
<span id="cb138-8"><a href="#cb138-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [4,]    4    9   14</span></span>
<span id="cb138-9"><a href="#cb138-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [5,]    5   10   15</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co">#sum rows without a loop --&gt; sum row 1</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(m[<span class="dv">1</span>, ])</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 18</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co">#sum rows without a loop --&gt; sum row 2, ...</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(m[<span class="dv">2</span>, ])</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 21</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="co">#run a loop to get the sum for all five rows</span></span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) {</span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sum</span>(m[i, ]))</span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 18</span></span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 21</span></span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 24</span></span>
<span id="cb141-8"><a href="#cb141-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 27</span></span>
<span id="cb141-9"><a href="#cb141-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 30</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the second loop we store our results in a new dataframe and do not just print them to the screen. Therefore, we:</p>
<ul>
<li>define a variable (named results) in which we store an empty vector</li>
<li>we need this empty vector to have something to store our results in</li>
<li>start the foor loop and define i as before</li>
<li>run the sum as before but now store the results for each interation in <code>results</code></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="co">#store the results, Note that here, the variable results had to be created before as an empty vector.</span></span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) {</span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>results[i] <span class="ot">&lt;-</span> <span class="fu">sum</span>(m[i, ])</span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a>results</span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 18 21 24 27 30</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Very often, there are some R packages or functions that are faster than loops. The <code>apply()</code> function is such an example. <code>apply()</code> applies a function, i.e.&nbsp;sum, across a matrix.</p>
<p>However, since loops are also a very useful feature in bash or python it is useful to understand their general concept.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="co">#way to do the same using the apply function (faster than loops)</span></span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a>results2 <span class="ot">&lt;-</span> <span class="fu">apply</span>(m, <span class="dv">1</span>, sum)</span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a>results2</span>
<span id="cb143-4"><a href="#cb143-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 18 21 24 27 30</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The first argument of <code>apply()</code> has to be a matrix, the second argument specifies the dimension of the matrix (1 means rows and 2 means columns) on which the function should be applied, and the third argument has to be a function that takes a vector as argument. Here, the function sum is applied to each row of m.</p>
<p>Actually, there is a family of apply functions, depending on the object you pass as an input, and/or the object you want as output. You can read a brief tutorial under this <a href="http://datascienceplus.com/using-the-apply-family-of-functions-in-r/">link</a>.</p>
<p>I.e. apply(data, 1, mean) = apply mean on each row apply(data, 2, mean) = apply mean on each column</p>
<p>Two other examples are sapply and lapply, which work on lists.</p>
</section>
<section id="if-else" class="level3" data-number="1.26.2">
<h3 data-number="1.26.2" class="anchored" data-anchor-id="if-else"><span class="header-section-number">1.26.2</span> if-else</h3>
<p>One can also create a decision making structure in R. A decision making structure has at least one condition to be evaluated together with a statement or statements to be evaluated if the condition is TRUE, and optionally, other statements to be executed if the condition is FALSE, as we can see in the figure.</p>
<p>The test_expression has to be logical, i.e., it has to be a expression that, when evaluated, is either TRUE or FALSE. The logical operators listed above can be used to construct them. For example, we can use an if-else statement to check if a number is positive or negative,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="co">#store a random number</span></span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">5</span></span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a><span class="co">#write a loop and ask if our number is positive or negative</span></span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb144-6"><a href="#cb144-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Positive number"</span>)</span>
<span id="cb144-7"><a href="#cb144-7" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (x <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb144-8"><a href="#cb144-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Number is zero"</span>)</span>
<span id="cb144-9"><a href="#cb144-9" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb144-10"><a href="#cb144-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Negative number"</span>)</span>
<span id="cb144-11"><a href="#cb144-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb144-12"><a href="#cb144-12" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Negative number"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that in the example there was an else if. In that way we can check more than 2 conditions.</p>
<p>This of course is a very simplistic example. But if-else statements can be useful if we want to deal with larger dataframes, i.e.&nbsp;our growth data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co">#apply ifelse</span></span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>category <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(growth_data<span class="sc">$</span>Rootlength<span class="sc">&lt;=</span><span class="fl">10.0</span>, <span class="st">"small"</span>, <span class="st">"large"</span>)</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check the structure of our data</span></span>
<span id="cb145-5"><a href="#cb145-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(growth_data) <span class="sc">%&gt;%</span></span>
<span id="cb145-6"><a href="#cb145-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>() <span class="sc">%&gt;%</span></span>
<span id="cb145-7"><a href="#cb145-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scroll_box</span>(<span class="at">width =</span> <span class="st">"700px"</span>, <span class="at">height =</span> <span class="st">"400px"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:400px; overflow-x: scroll; width:700px; "><table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> SampleID </th>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> Nutrient </th>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> Condition </th>
   <th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;"> FW_shoot_mg </th>
   <th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;"> Rootlength </th>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> category </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 10.26 </td>
   <td style="text-align:right;"> 5.931015 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 6.52 </td>
   <td style="text-align:right;"> 5.743447 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 12.17 </td>
   <td style="text-align:right;"> 6.834720 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 11.37 </td>
   <td style="text-align:right;"> 6.742735 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 9.80 </td>
   <td style="text-align:right;"> 6.736886 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 3.75 </td>
   <td style="text-align:right;"> 4.236349 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 5.38 </td>
   <td style="text-align:right;"> 4.753485 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 4.53 </td>
   <td style="text-align:right;"> 5.532333 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 7.75 </td>
   <td style="text-align:right;"> 5.484364 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 8.64 </td>
   <td style="text-align:right;"> 5.963254 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 10.36 </td>
   <td style="text-align:right;"> 5.359044 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 7.69 </td>
   <td style="text-align:right;"> 5.725348 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 8.57 </td>
   <td style="text-align:right;"> 6.424601 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 10.49 </td>
   <td style="text-align:right;"> 7.143667 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.91 </td>
   <td style="text-align:right;"> 7.669841 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 9.32 </td>
   <td style="text-align:right;"> 7.807710 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 6.76 </td>
   <td style="text-align:right;"> 7.508370 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 5.99 </td>
   <td style="text-align:right;"> 6.607630 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.26 </td>
   <td style="text-align:right;"> 7.269267 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 7.61 </td>
   <td style="text-align:right;"> 7.973207 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 7.56 </td>
   <td style="text-align:right;"> 7.399504 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 7.90 </td>
   <td style="text-align:right;"> 6.717792 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 6.69 </td>
   <td style="text-align:right;"> 6.721007 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.14 </td>
   <td style="text-align:right;"> 7.070333 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 6.07 </td>
   <td style="text-align:right;"> 5.947965 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.19 </td>
   <td style="text-align:right;"> 6.393722 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 4.96 </td>
   <td style="text-align:right;"> 7.166174 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 6.20 </td>
   <td style="text-align:right;"> 7.515659 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.97 </td>
   <td style="text-align:right;"> 7.250036 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.32 </td>
   <td style="text-align:right;"> 7.134681 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.45 </td>
   <td style="text-align:right;"> 6.917319 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 6.03 </td>
   <td style="text-align:right;"> 6.120089 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.70 </td>
   <td style="text-align:right;"> 7.665526 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 6.04 </td>
   <td style="text-align:right;"> 7.809111 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.22 </td>
   <td style="text-align:right;"> 6.601296 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 3.99 </td>
   <td style="text-align:right;"> 6.524710 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 4.85 </td>
   <td style="text-align:right;"> 8.197116 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 8.63 </td>
   <td style="text-align:right;"> 6.160052 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.27 </td>
   <td style="text-align:right;"> 4.720711 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.51 </td>
   <td style="text-align:right;"> 6.917185 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.46 </td>
   <td style="text-align:right;"> 4.384756 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.91 </td>
   <td style="text-align:right;"> 6.069185 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.40 </td>
   <td style="text-align:right;"> 7.113879 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.36 </td>
   <td style="text-align:right;"> 5.428766 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.74 </td>
   <td style="text-align:right;"> 6.694142 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 4.15 </td>
   <td style="text-align:right;"> 5.270298 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 8.60 </td>
   <td style="text-align:right;"> 6.811773 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 5.41 </td>
   <td style="text-align:right;"> 5.106644 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 8.30 </td>
   <td style="text-align:right;"> 7.413519 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.93 </td>
   <td style="text-align:right;"> 7.112385 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.44 </td>
   <td style="text-align:right;"> 7.428674 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.85 </td>
   <td style="text-align:right;"> 6.372659 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 32.42 </td>
   <td style="text-align:right;"> 11.286793 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.03 </td>
   <td style="text-align:right;"> 10.456630 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 18.35 </td>
   <td style="text-align:right;"> 11.106341 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.04 </td>
   <td style="text-align:right;"> 10.816896 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 16.61 </td>
   <td style="text-align:right;"> 10.608252 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 25.79 </td>
   <td style="text-align:right;"> 7.121587 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 31.00 </td>
   <td style="text-align:right;"> 9.165891 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 36.13 </td>
   <td style="text-align:right;"> 11.053978 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 17.50 </td>
   <td style="text-align:right;"> 8.271196 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.42 </td>
   <td style="text-align:right;"> 8.649225 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 19.60 </td>
   <td style="text-align:right;"> 10.313985 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 11.62 </td>
   <td style="text-align:right;"> 8.098859 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 18.76 </td>
   <td style="text-align:right;"> 10.061778 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.55 </td>
   <td style="text-align:right;"> 11.048822 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 19.05 </td>
   <td style="text-align:right;"> 9.741622 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 20.91 </td>
   <td style="text-align:right;"> 12.119304 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 23.48 </td>
   <td style="text-align:right;"> 12.057252 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 17.47 </td>
   <td style="text-align:right;"> 12.064659 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 24.49 </td>
   <td style="text-align:right;"> 13.576118 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 25.70 </td>
   <td style="text-align:right;"> 10.983965 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 28.24 </td>
   <td style="text-align:right;"> 12.935397 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 17.70 </td>
   <td style="text-align:right;"> 11.921333 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 32.90 </td>
   <td style="text-align:right;"> 13.030630 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 22.80 </td>
   <td style="text-align:right;"> 12.644756 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 32.47 </td>
   <td style="text-align:right;"> 14.715830 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 22.05 </td>
   <td style="text-align:right;"> 13.186593 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 23.47 </td>
   <td style="text-align:right;"> 13.513763 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 23.46 </td>
   <td style="text-align:right;"> 10.981978 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 22.93 </td>
   <td style="text-align:right;"> 12.563616 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 17.08 </td>
   <td style="text-align:right;"> 12.174456 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 15.95 </td>
   <td style="text-align:right;"> 13.151797 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 15.03 </td>
   <td style="text-align:right;"> 12.072456 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 13.00 </td>
   <td style="text-align:right;"> 10.692603 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 18.30 </td>
   <td style="text-align:right;"> 12.236482 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 17.80 </td>
   <td style="text-align:right;"> 11.792879 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 17.43 </td>
   <td style="text-align:right;"> 11.914695 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 14.93 </td>
   <td style="text-align:right;"> 11.281731 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 22.75 </td>
   <td style="text-align:right;"> 11.502507 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 37.90 </td>
   <td style="text-align:right;"> 13.170210 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 13.27 </td>
   <td style="text-align:right;"> 10.055399 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 37.14 </td>
   <td style="text-align:right;"> 11.908378 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 39.39 </td>
   <td style="text-align:right;"> 12.612785 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 28.27 </td>
   <td style="text-align:right;"> 12.434356 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 29.50 </td>
   <td style="text-align:right;"> 12.559904 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 27.29 </td>
   <td style="text-align:right;"> 11.585637 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 22.82 </td>
   <td style="text-align:right;"> 11.925326 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 25.05 </td>
   <td style="text-align:right;"> 12.831222 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 17.48 </td>
   <td style="text-align:right;"> 11.644037 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 15.85 </td>
   <td style="text-align:right;"> 11.900193 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.54 </td>
   <td style="text-align:right;"> 10.698752 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 27.91 </td>
   <td style="text-align:right;"> 11.668794 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 26.63 </td>
   <td style="text-align:right;"> 12.555404 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 29.96 </td>
   <td style="text-align:right;"> 11.771277 </td>
   <td style="text-align:left;"> large </td>
  </tr>
</tbody>
</table></div>

</div>
</div>
<p>Here, we:</p>
<ul>
<li>Create a new column, with the name category</li>
<li>If a value in our Rootlength column is equal to or smaller than 10, we want to say this category is small.</li>
<li>If that is not the case, we want to say it is large</li>
</ul>
<p>So the function works like this:</p>
<p><strong>ifelse(our_test, value_to_return_if_test_is_true, value_to_return_if_test_is_false)</strong></p>
<p>We can also combine diffent statements with the &amp; (AND) or | (OR) symbol. I.e. we only send things in the big category if the roots are longer than 10cm AND the shoot weight is larger than 15mg</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co">#apply ifelse</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>category <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(growth_data<span class="sc">$</span>Rootlength<span class="sc">&gt;</span><span class="dv">10</span> <span class="sc">&amp;</span> growth_data<span class="sc">$</span>FW_shoot_mg<span class="sc">&gt;</span><span class="dv">15</span>, <span class="st">"large"</span>, <span class="st">"small"</span>)</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check the structure of our data</span></span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(growth_data) <span class="sc">%&gt;%</span></span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>() <span class="sc">%&gt;%</span></span>
<span id="cb146-7"><a href="#cb146-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scroll_box</span>(<span class="at">width =</span> <span class="st">"700px"</span>, <span class="at">height =</span> <span class="st">"400px"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:400px; overflow-x: scroll; width:700px; "><table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> SampleID </th>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> Nutrient </th>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> Condition </th>
   <th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;"> FW_shoot_mg </th>
   <th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;"> Rootlength </th>
   <th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;"> category </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 10.26 </td>
   <td style="text-align:right;"> 5.931015 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 6.52 </td>
   <td style="text-align:right;"> 5.743447 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 12.17 </td>
   <td style="text-align:right;"> 6.834720 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 11.37 </td>
   <td style="text-align:right;"> 6.742735 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 9.80 </td>
   <td style="text-align:right;"> 6.736886 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 3.75 </td>
   <td style="text-align:right;"> 4.236349 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 5.38 </td>
   <td style="text-align:right;"> 4.753485 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 4.53 </td>
   <td style="text-align:right;"> 5.532333 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 7.75 </td>
   <td style="text-align:right;"> 5.484364 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 8.64 </td>
   <td style="text-align:right;"> 5.963254 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 10.36 </td>
   <td style="text-align:right;"> 5.359044 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 7.69 </td>
   <td style="text-align:right;"> 5.725348 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 8.57 </td>
   <td style="text-align:right;"> 6.424601 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 10.49 </td>
   <td style="text-align:right;"> 7.143667 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.91 </td>
   <td style="text-align:right;"> 7.669841 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 9.32 </td>
   <td style="text-align:right;"> 7.807710 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 6.76 </td>
   <td style="text-align:right;"> 7.508370 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 5.99 </td>
   <td style="text-align:right;"> 6.607630 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.26 </td>
   <td style="text-align:right;"> 7.269267 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 7.61 </td>
   <td style="text-align:right;"> 7.973207 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 7.56 </td>
   <td style="text-align:right;"> 7.399504 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 7.90 </td>
   <td style="text-align:right;"> 6.717792 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 6.69 </td>
   <td style="text-align:right;"> 6.721007 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.14 </td>
   <td style="text-align:right;"> 7.070333 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 6.07 </td>
   <td style="text-align:right;"> 5.947965 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_101 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 8.19 </td>
   <td style="text-align:right;"> 6.393722 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 4.96 </td>
   <td style="text-align:right;"> 7.166174 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 6.20 </td>
   <td style="text-align:right;"> 7.515659 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.97 </td>
   <td style="text-align:right;"> 7.250036 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.32 </td>
   <td style="text-align:right;"> 7.134681 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.45 </td>
   <td style="text-align:right;"> 6.917319 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 6.03 </td>
   <td style="text-align:right;"> 6.120089 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.70 </td>
   <td style="text-align:right;"> 7.665526 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 6.04 </td>
   <td style="text-align:right;"> 7.809111 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 5.22 </td>
   <td style="text-align:right;"> 6.601296 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 3.99 </td>
   <td style="text-align:right;"> 6.524710 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_230 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 4.85 </td>
   <td style="text-align:right;"> 8.197116 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 8.63 </td>
   <td style="text-align:right;"> 6.160052 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.27 </td>
   <td style="text-align:right;"> 4.720711 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.51 </td>
   <td style="text-align:right;"> 6.917185 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.46 </td>
   <td style="text-align:right;"> 4.384756 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.91 </td>
   <td style="text-align:right;"> 6.069185 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.40 </td>
   <td style="text-align:right;"> 7.113879 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.36 </td>
   <td style="text-align:right;"> 5.428766 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.74 </td>
   <td style="text-align:right;"> 6.694142 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 4.15 </td>
   <td style="text-align:right;"> 5.270298 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 8.60 </td>
   <td style="text-align:right;"> 6.811773 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 5.41 </td>
   <td style="text-align:right;"> 5.106644 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 8.30 </td>
   <td style="text-align:right;"> 7.413519 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.93 </td>
   <td style="text-align:right;"> 7.112385 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.44 </td>
   <td style="text-align:right;"> 7.428674 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> noP_28 </td>
   <td style="text-align:left;"> noP </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 7.85 </td>
   <td style="text-align:right;"> 6.372659 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 32.42 </td>
   <td style="text-align:right;"> 11.286793 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.03 </td>
   <td style="text-align:right;"> 10.456630 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 18.35 </td>
   <td style="text-align:right;"> 11.106341 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.04 </td>
   <td style="text-align:right;"> 10.816896 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 16.61 </td>
   <td style="text-align:right;"> 10.608252 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 25.79 </td>
   <td style="text-align:right;"> 7.121587 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 31.00 </td>
   <td style="text-align:right;"> 9.165891 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 36.13 </td>
   <td style="text-align:right;"> 11.053978 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 17.50 </td>
   <td style="text-align:right;"> 8.271196 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.42 </td>
   <td style="text-align:right;"> 8.649225 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 19.60 </td>
   <td style="text-align:right;"> 10.313985 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 11.62 </td>
   <td style="text-align:right;"> 8.098859 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 18.76 </td>
   <td style="text-align:right;"> 10.061778 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 21.55 </td>
   <td style="text-align:right;"> 11.048822 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> MgCl </td>
   <td style="text-align:right;"> 19.05 </td>
   <td style="text-align:right;"> 9.741622 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 20.91 </td>
   <td style="text-align:right;"> 12.119304 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 23.48 </td>
   <td style="text-align:right;"> 12.057252 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 17.47 </td>
   <td style="text-align:right;"> 12.064659 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 24.49 </td>
   <td style="text-align:right;"> 13.576118 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 25.70 </td>
   <td style="text-align:right;"> 10.983965 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 28.24 </td>
   <td style="text-align:right;"> 12.935397 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 17.70 </td>
   <td style="text-align:right;"> 11.921333 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 32.90 </td>
   <td style="text-align:right;"> 13.030630 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 22.80 </td>
   <td style="text-align:right;"> 12.644756 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 32.47 </td>
   <td style="text-align:right;"> 14.715830 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 22.05 </td>
   <td style="text-align:right;"> 13.186593 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_101 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain101 </td>
   <td style="text-align:right;"> 23.47 </td>
   <td style="text-align:right;"> 13.513763 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 23.46 </td>
   <td style="text-align:right;"> 10.981978 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 22.93 </td>
   <td style="text-align:right;"> 12.563616 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 17.08 </td>
   <td style="text-align:right;"> 12.174456 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 15.95 </td>
   <td style="text-align:right;"> 13.151797 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 15.03 </td>
   <td style="text-align:right;"> 12.072456 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 13.00 </td>
   <td style="text-align:right;"> 10.692603 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 18.30 </td>
   <td style="text-align:right;"> 12.236482 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 17.80 </td>
   <td style="text-align:right;"> 11.792879 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 17.43 </td>
   <td style="text-align:right;"> 11.914695 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 14.93 </td>
   <td style="text-align:right;"> 11.281731 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 22.75 </td>
   <td style="text-align:right;"> 11.502507 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 37.90 </td>
   <td style="text-align:right;"> 13.170210 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_230 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain230 </td>
   <td style="text-align:right;"> 13.27 </td>
   <td style="text-align:right;"> 10.055399 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 37.14 </td>
   <td style="text-align:right;"> 11.908378 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 39.39 </td>
   <td style="text-align:right;"> 12.612785 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 28.27 </td>
   <td style="text-align:right;"> 12.434356 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 29.50 </td>
   <td style="text-align:right;"> 12.559904 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 27.29 </td>
   <td style="text-align:right;"> 11.585637 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 22.82 </td>
   <td style="text-align:right;"> 11.925326 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 25.05 </td>
   <td style="text-align:right;"> 12.831222 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 17.48 </td>
   <td style="text-align:right;"> 11.644037 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 15.85 </td>
   <td style="text-align:right;"> 11.900193 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 9.54 </td>
   <td style="text-align:right;"> 10.698752 </td>
   <td style="text-align:left;"> small </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 27.91 </td>
   <td style="text-align:right;"> 11.668794 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 26.63 </td>
   <td style="text-align:right;"> 12.555404 </td>
   <td style="text-align:left;"> large </td>
  </tr>
  <tr>
   <td style="text-align:left;"> P_28 </td>
   <td style="text-align:left;"> P </td>
   <td style="text-align:left;"> Strain28 </td>
   <td style="text-align:right;"> 29.96 </td>
   <td style="text-align:right;"> 11.771277 </td>
   <td style="text-align:left;"> large </td>
  </tr>
</tbody>
</table></div>

</div>
</div>
</section>
</section>
<section id="plotting" class="level2" data-number="1.27">
<h2 data-number="1.27" class="anchored" data-anchor-id="plotting"><span class="header-section-number">1.27</span> Plotting</h2>
<section id="basic-plots" class="level3" data-number="1.27.1">
<h3 data-number="1.27.1" class="anchored" data-anchor-id="basic-plots"><span class="header-section-number">1.27.1</span> Basic Plots</h3>
<section id="scatter-plots" class="level4" data-number="1.27.1.1">
<h4 data-number="1.27.1.1" class="anchored" data-anchor-id="scatter-plots"><span class="header-section-number">1.27.1.1</span> Scatter plots</h4>
<p>The examples below are part of base R, i.e.&nbsp;we can plot without using any packages. However, there are some nice packages that let you control a lot of parameters, which are good to learn for more sophisticates plots.</p>
<p>Plotting in base R is simple, we just need to define, what we plot against the x- and y-axis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># lets plot our root length against our shoot weight</span></span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>FW_shoot_mg</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>Rootlength</span>
<span id="cb147-4"><a href="#cb147-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb147-5"><a href="#cb147-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-148-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now, lets add some more axis labels to make it more informative and lets startt the plot at 0:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># lets plot our root length against our shoot weight</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>FW_shoot_mg</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>Rootlength</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">xlab =</span> <span class="st">"Shoot_weight_mg"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength_cm"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">15</span>), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">50</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-149-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>A useful thing to know is that you can add plots together, i.e.&nbsp;we can add a regression line:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co">#lets do the stats running a linear model (lm)</span></span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lm</span>(growth_data<span class="sc">$</span>Rootlength <span class="sc">~</span> growth_data<span class="sc">$</span>FW_shoot_mg)</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Call:</span></span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a><span class="do">## lm(formula = growth_data$Rootlength ~ growth_data$FW_shoot_mg)</span></span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb149-7"><a href="#cb149-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Coefficients:</span></span>
<span id="cb149-8"><a href="#cb149-8" aria-hidden="true" tabindex="-1"></a><span class="do">##             (Intercept)  growth_data$FW_shoot_mg  </span></span>
<span id="cb149-9"><a href="#cb149-9" aria-hidden="true" tabindex="-1"></a><span class="do">##                  5.2912                   0.2465</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, the intercept is 5.2912 and the slope is 0.2465 .</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="co"># lets plot our root length against our shoot weight</span></span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>FW_shoot_mg</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> growth_data<span class="sc">$</span>Rootlength</span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb150-6"><a href="#cb150-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">xlab =</span> <span class="st">"Shoot_weight_mg"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength_cm"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">15</span>), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">50</span>))</span>
<span id="cb150-7"><a href="#cb150-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb150-8"><a href="#cb150-8" aria-hidden="true" tabindex="-1"></a><span class="co">#add the info from ln (both lines of code do the same thing)</span></span>
<span id="cb150-9"><a href="#cb150-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="fu">lm</span>(growth_data<span class="sc">$</span>Rootlength <span class="sc">~</span> growth_data<span class="sc">$</span>FW_shoot_mg))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-151-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can even add the stats, but therefore we need to prepare the stats a bit better:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co">#to stats</span></span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a>modl <span class="ot">=</span> <span class="fu">lm</span>(growth_data<span class="sc">$</span>Rootlength <span class="sc">~</span> growth_data<span class="sc">$</span>FW_shoot_mg)</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a><span class="co">#get summary</span></span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>modsum <span class="ot">=</span> <span class="fu">summary</span>(modl)</span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a><span class="co">#get R2</span></span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>r2 <span class="ot">=</span> modsum<span class="sc">$</span>adj.r.squared</span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a><span class="co">#look at the coefficients</span></span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a>modsum<span class="sc">$</span>coefficients</span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a><span class="do">##                          Estimate Std. Error  t value     Pr(&gt;|t|)</span></span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept)             5.2912436 0.30469198 17.36588 2.278273e-32</span></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a><span class="do">## growth_data$FW_shoot_mg 0.2464783 0.01711628 14.40023 2.039703e-26</span></span>
<span id="cb151-15"><a href="#cb151-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-16"><a href="#cb151-16" aria-hidden="true" tabindex="-1"></a><span class="co">#get the p-value (its in the coefficient table, in the 2nd row and 4th column)</span></span>
<span id="cb151-17"><a href="#cb151-17" aria-hidden="true" tabindex="-1"></a>my.p <span class="ot">=</span> modsum<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can plot this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="co">#add a label, in which we store the text we want to add</span></span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>mylabel <span class="ot">=</span> <span class="fu">bquote</span>(<span class="fu">italic</span>(R)<span class="sc">^</span><span class="dv">2</span> <span class="sc">==</span> .(<span class="fu">format</span>(r2, <span class="at">digits =</span> <span class="dv">3</span>)))</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">xlab =</span> <span class="st">"Shoot_weight_mg"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength_cm"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">15</span>), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">50</span>))</span>
<span id="cb152-6"><a href="#cb152-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-7"><a href="#cb152-7" aria-hidden="true" tabindex="-1"></a><span class="co">#add the info from ln (both lines of code do the same thing)</span></span>
<span id="cb152-8"><a href="#cb152-8" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="fu">lm</span>(growth_data<span class="sc">$</span>Rootlength <span class="sc">~</span> growth_data<span class="sc">$</span>FW_shoot_mg))</span>
<span id="cb152-9"><a href="#cb152-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-10"><a href="#cb152-10" aria-hidden="true" tabindex="-1"></a><span class="co">#add the text</span></span>
<span id="cb152-11"><a href="#cb152-11" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="dv">45</span>, <span class="at">y =</span><span class="dv">14</span>, <span class="at">labels =</span> mylabel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-153-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>If we want to add the other value, it gets a bit more complicated esp.&nbsp;if we want to have first the R2 and then, in a new line, the p-value.</p>
<p>For this, lets first prepare a new label</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make an empty vecotr</span></span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a>rp <span class="ot">=</span> <span class="fu">vector</span>(<span class="st">'expression'</span>,<span class="dv">2</span>)</span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>rp</span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a><span class="do">## expression(NULL, NULL)</span></span>
<span id="cb153-5"><a href="#cb153-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-6"><a href="#cb153-6" aria-hidden="true" tabindex="-1"></a><span class="co">#add our two values into the vector</span></span>
<span id="cb153-7"><a href="#cb153-7" aria-hidden="true" tabindex="-1"></a>rp[<span class="dv">1</span>] <span class="ot">=</span> <span class="fu">substitute</span>(<span class="fu">expression</span>(<span class="fu">italic</span>(R)<span class="sc">^</span><span class="dv">2</span> <span class="sc">==</span> MYVALUE), </span>
<span id="cb153-8"><a href="#cb153-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">MYVALUE =</span> <span class="fu">format</span>(r2,<span class="at">dig=</span><span class="dv">3</span>)))[<span class="dv">2</span>]</span>
<span id="cb153-9"><a href="#cb153-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-10"><a href="#cb153-10" aria-hidden="true" tabindex="-1"></a>rp[<span class="dv">2</span>] <span class="ot">=</span> <span class="fu">substitute</span>(<span class="fu">expression</span>(<span class="fu">italic</span>(p) <span class="sc">==</span> MYOTHERVALUE), </span>
<span id="cb153-11"><a href="#cb153-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">MYOTHERVALUE =</span> <span class="fu">format</span>(my.p, <span class="at">digits =</span> <span class="dv">2</span>)))[<span class="dv">2</span>]</span>
<span id="cb153-12"><a href="#cb153-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-13"><a href="#cb153-13" aria-hidden="true" tabindex="-1"></a>rp</span>
<span id="cb153-14"><a href="#cb153-14" aria-hidden="true" tabindex="-1"></a><span class="do">## expression(italic(R)^2 == "0.665", italic(p) == "2e-26")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lets plot this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">xlab =</span> <span class="st">"Shoot_weight_mg"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength_cm"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">15</span>), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">50</span>))</span>
<span id="cb154-3"><a href="#cb154-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-4"><a href="#cb154-4" aria-hidden="true" tabindex="-1"></a><span class="co">#add the info from ln (both lines of code do the same thing)</span></span>
<span id="cb154-5"><a href="#cb154-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="fu">lm</span>(growth_data<span class="sc">$</span>Rootlength <span class="sc">~</span> growth_data<span class="sc">$</span>FW_shoot_mg))</span>
<span id="cb154-6"><a href="#cb154-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-7"><a href="#cb154-7" aria-hidden="true" tabindex="-1"></a><span class="co">#add the text as a legend and remove the border with bty</span></span>
<span id="cb154-8"><a href="#cb154-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> rp, <span class="at">bty =</span> <span class="st">'n'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-155-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="lineplots" class="level4" data-number="1.27.1.2">
<h4 data-number="1.27.1.2" class="anchored" data-anchor-id="lineplots"><span class="header-section-number">1.27.1.2</span> Lineplots</h4>
<p>Now, lets work with our time course data to draw some line plots. I.e. lineplots are ideal if we have measurements over time.</p>
<p>Lets first summarize our time course data to make it easier to plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a><span class="co">#first filter to only print P data, then summarize the data by condition and timepoint and calculate the mean</span></span>
<span id="cb155-4"><a href="#cb155-4" aria-hidden="true" tabindex="-1"></a>timecourse_summary <span class="ot">&lt;-</span> timecourse_noNA <span class="sc">%&gt;%</span> </span>
<span id="cb155-5"><a href="#cb155-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(Nutrient <span class="sc">==</span> <span class="st">"P"</span>)  <span class="sc">%&gt;%</span> </span>
<span id="cb155-6"><a href="#cb155-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(Condition, Timepoint) <span class="sc">%&gt;%</span>  </span>
<span id="cb155-7"><a href="#cb155-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarise</span>(<span class="at">mean_root =</span> <span class="fu">mean</span>(Rootlength))</span>
<span id="cb155-8"><a href="#cb155-8" aria-hidden="true" tabindex="-1"></a><span class="do">## `summarise()` has grouped output by 'Condition'. You can override using the</span></span>
<span id="cb155-9"><a href="#cb155-9" aria-hidden="true" tabindex="-1"></a><span class="do">## `.groups` argument.</span></span>
<span id="cb155-10"><a href="#cb155-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb155-11"><a href="#cb155-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(timecourse_summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Condition"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Timepoint"],"name":[2],"type":["chr"],"align":["left"]},{"label":["mean_root"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"350","2":"t0","3":"1.521533"},{"1":"350","2":"t1","3":"2.328333"},{"1":"350","2":"t2","3":"3.851552"},{"1":"350","2":"t3","3":"5.717783"},{"1":"350","2":"t4","3":"9.387750"},{"1":"350","2":"t5","3":"11.221909"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a><span class="co">#split the data into our two categories</span></span>
<span id="cb156-3"><a href="#cb156-3" aria-hidden="true" tabindex="-1"></a>timecourse_summary_Control <span class="ot">&lt;-</span> timecourse_summary[timecourse_summary<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">'MgCl'</span>,]</span>
<span id="cb156-4"><a href="#cb156-4" aria-hidden="true" tabindex="-1"></a>timecourse_summary_350 <span class="ot">&lt;-</span> timecourse_summary[timecourse_summary<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">'350'</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Useful comments:</p>
<ul>
<li>type= It controls the type (p for points, l for lines, b for both,<e2>&lt;80&gt;<a6>).</a6></e2></li>
<li>pch= integer [0,25]. Controls the plot symbol.</li>
<li>log= It causes the x axis x, y axis y or both xy to be logarithmic.</li>
<li>xlab=, ylab= string, labels for the x and y axis, respectively.</li>
<li>xlim=, ylim= length 2 vector, x-axis, y-axis limits.</li>
<li>main= string, title of the plot.</li>
<li>col = hexadecimal or string, colour of the points/lines.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> timecourse_summary_Control<span class="sc">$</span>mean_root</span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"Timepoints"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength (cm)"</span>, <span class="at">col =</span> <span class="st">"#7700BB"</span>, <span class="at">main =</span> <span class="st">"Growth measurements"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-157-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>plot() always overwrites the content for the current graphics window.</p>
<p>To add elements to an existing plot, one can use points, lines, abline, text, <e2>&lt;80&gt;<a6> We can also add a legend to the plot.</a6></e2></p>
<p>In the previous graph eith out growth measurements let us add the root growth of our microbe treatment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> timecourse_summary_Control<span class="sc">$</span>mean_root</span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> timecourse_summary_350<span class="sc">$</span>mean_root</span>
<span id="cb158-4"><a href="#cb158-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb158-5"><a href="#cb158-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"Timepoints"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength (cm)"</span>,<span class="at">col =</span> <span class="st">"#7700BB"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">15</span>), <span class="at">main =</span> <span class="st">"Growth measurements"</span>)</span>
<span id="cb158-6"><a href="#cb158-6" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(x, z, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">col =</span> <span class="st">"#5555DD"</span>)</span>
<span id="cb158-7"><a href="#cb158-7" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"microbe"</span>), <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#7700BB"</span>, <span class="st">"#5555DD"</span>),<span class="at">pch =</span> <span class="fu">c</span>(<span class="dv">19</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-158-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can see nicely, that our roots grow longer past the 3rd timepoint.</p>
</section>
</section>
<section id="info-on-using-par" class="level3" data-number="1.27.2">
<h3 data-number="1.27.2" class="anchored" data-anchor-id="info-on-using-par"><span class="header-section-number">1.27.2</span> info on using par</h3>
<p>To change graphical parameters globally you can use par. R allows for n <c3>&lt;97&gt; m figures in a single page, by adjusting the parameter mfrow:</c3></p>
<ul>
<li>mfrow = c(3, 1) –&gt; we have 3 plots distributed across 3 rows and one column</li>
<li>par(mai = c(2, 0.82, 0.82, 0.42)) –&gt; sets the bottom, left, top and right margins respectively of the plot region in number of lines of text. If we change the margins it is recommended to reset them to the default after plotting with <code>par(mai = c(1.02, 0.82, 0.82, 0.42))</code>.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(n, m)) <span class="co"># n: number of figures in rows, m: ... in columns.</span></span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x1, y1)</span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x1, y2)</span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a>...</span>
<span id="cb159-5"><a href="#cb159-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(xn, ym)</span>
<span id="cb159-6"><a href="#cb159-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb159-7"><a href="#cb159-7" aria-hidden="true" tabindex="-1"></a><span class="co">#save a file</span></span>
<span id="cb159-8"><a href="#cb159-8" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"../output_examples/filename.png"</span>)</span>
<span id="cb159-9"><a href="#cb159-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y)</span>
<span id="cb159-10"><a href="#cb159-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the above code chunk, you first open a png file, then plot directly to that file, and finally explicitly close the plotting device with dev.off(). Thus, you do not see the plot on the graphics window. The Cairo package is supposed to be superior to these basic plotting functions, but it does not come with the base installation of R, therefore you will have to install it to try it out (if you are interested, or at a later time).</p>
<section id="printing-our-results-using-par" class="level4" data-number="1.27.2.1">
<h4 data-number="1.27.2.1" class="anchored" data-anchor-id="printing-our-results-using-par"><span class="header-section-number">1.27.2.1</span> Printing our results using par</h4>
<p>Now lets generate a png file plot.png containing three plots in three different rows.</p>
<ul>
<li>The data for our control treatment</li>
<li>The data for our microbe treatment</li>
<li>The data combined in one plot</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb160-3"><a href="#cb160-3" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> timecourse_summary_Control<span class="sc">$</span>mean_root</span>
<span id="cb160-4"><a href="#cb160-4" aria-hidden="true" tabindex="-1"></a>Microbe <span class="ot">&lt;-</span> timecourse_summary_350<span class="sc">$</span>mean_root</span>
<span id="cb160-5"><a href="#cb160-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-6"><a href="#cb160-6" aria-hidden="true" tabindex="-1"></a><span class="co">#define what we want to print</span></span>
<span id="cb160-7"><a href="#cb160-7" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"../output_examples/filename2.png"</span>, <span class="at">width =</span> <span class="dv">240</span>, <span class="at">height =</span> <span class="dv">480</span>)</span>
<span id="cb160-8"><a href="#cb160-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-9"><a href="#cb160-9" aria-hidden="true" tabindex="-1"></a><span class="co">#define how we want to order our 3 plots into rows and columns</span></span>
<span id="cb160-10"><a href="#cb160-10" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>)) </span>
<span id="cb160-11"><a href="#cb160-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-12"><a href="#cb160-12" aria-hidden="true" tabindex="-1"></a><span class="co">#build plot1</span></span>
<span id="cb160-13"><a href="#cb160-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, control, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"Timepoints"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength (cm)"</span>,<span class="at">col =</span> <span class="st">"#7700BB"</span>, <span class="at">main =</span> <span class="st">"Growth measurements"</span>)</span>
<span id="cb160-14"><a href="#cb160-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-15"><a href="#cb160-15" aria-hidden="true" tabindex="-1"></a><span class="co">#build plot2</span></span>
<span id="cb160-16"><a href="#cb160-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, Microbe, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"Timepoints"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength (cm)"</span>,<span class="at">col =</span> <span class="st">"#5555DD"</span>, <span class="at">main =</span> <span class="st">"Growth measurements"</span>)</span>
<span id="cb160-17"><a href="#cb160-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-18"><a href="#cb160-18" aria-hidden="true" tabindex="-1"></a><span class="co">#build plot3</span></span>
<span id="cb160-19"><a href="#cb160-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, control, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"Timepoints"</span>, <span class="at">ylab =</span> <span class="st">"Rootlength (cm)"</span>,<span class="at">col =</span> <span class="st">"#7700BB"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">15</span>), <span class="at">main =</span> <span class="st">"Growth measurements"</span>)</span>
<span id="cb160-20"><a href="#cb160-20" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(x, Microbe, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">col =</span> <span class="st">"#5555DD"</span>)</span>
<span id="cb160-21"><a href="#cb160-21" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"microbe"</span>), <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#7700BB"</span>, <span class="st">"#5555DD"</span>),<span class="at">pch =</span> <span class="fu">c</span>(<span class="dv">19</span>, <span class="dv">1</span>))</span>
<span id="cb160-22"><a href="#cb160-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-23"><a href="#cb160-23" aria-hidden="true" tabindex="-1"></a><span class="co">#closes the specified plot</span></span>
<span id="cb160-24"><a href="#cb160-24" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="histograms" class="level3" data-number="1.27.3">
<h3 data-number="1.27.3" class="anchored" data-anchor-id="histograms"><span class="header-section-number">1.27.3</span> Histograms</h3>
<p>A histogram shows the frequency of data values in equally sized intervals. Density plots are an alternative, but because of the smoothing between data points, histograms provide a more <e2>&lt;80&gt;&lt;98&gt;natural<e2>&lt;80&gt;&lt;99&gt; look at your data. If you are interested in how to make a density plot, look at the help page of density.</e2></e2></p>
<p>As an example, lets plot the distribution of our root length measurements across our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(growth_data[, <span class="st">"Rootlength"</span>], <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="at">main =</span> <span class="st">"Data distribution"</span>, <span class="at">breaks =</span> <span class="dv">10</span>, <span class="at">density =</span> <span class="dv">100</span>, <span class="at">col =</span> <span class="st">"lightblue"</span>, <span class="at">border =</span> <span class="st">"darkblue"</span>, <span class="at">xlab =</span> <span class="st">"Rootlength"</span>, <span class="at">labels =</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-161-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="boxplots" class="level3" data-number="1.27.4">
<h3 data-number="1.27.4" class="anchored" data-anchor-id="boxplots"><span class="header-section-number">1.27.4</span> Boxplots</h3>
<p>Boxplots represent a compact summary of a data vector in graphical form.</p>
<p>As we<e2>&lt;80&gt;&lt;99&gt;ve already seen above, the function summary returns summary statistics on the command line: the minimum, first quartile, mean, median, third quartile and the maximum.</e2></p>
<p>The boxplot displays these values graphically (except the mean) as follows:</p>
<ul>
<li>the thick line in the middle of the box represents the median,</li>
<li>the lower bound of the box the first quartile and the upper bound the third quartile. Thus, 50% of the data are within the range of the box.</li>
<li>The whiskers (thin lines below and above the box) represent the minimum and maximum.</li>
<li>Points more extreme than the min. and max. are considered outliers and the help page describes how they are defined.</li>
</ul>
<p>We will first make a boxplot of all measurements and then check for differences between the two nutrient conditions.. Play around with the parameters and add colors and labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="co"># all data</span></span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(growth_data<span class="sc">$</span>Rootlength, <span class="at">cex =</span> <span class="fl">0.8</span>,  <span class="at">ylab =</span> <span class="st">"Root length (cm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-162-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># data by nutrient condition</span></span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(growth_data<span class="sc">$</span>Rootlength <span class="sc">~</span> growth_data<span class="sc">$</span>Nutrient, <span class="at">las =</span> <span class="dv">2</span>, <span class="at">cex =</span> <span class="fl">0.8</span>, <span class="at">ylab =</span> <span class="st">"Root length (cm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-163-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="co"># data by nutrient and growth condition</span></span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(growth_data<span class="sc">$</span>Rootlength <span class="sc">~</span> growth_data<span class="sc">$</span>Nutrient <span class="sc">*</span> growth_data<span class="sc">$</span>Condition, <span class="at">las =</span> <span class="dv">2</span>, <span class="at">cex =</span> <span class="fl">0.8</span>, <span class="at">ylab =</span> <span class="st">"Root length (cm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-164-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="ggplot2" class="level3" data-number="1.27.5">
<h3 data-number="1.27.5" class="anchored" data-anchor-id="ggplot2"><span class="header-section-number">1.27.5</span> Ggplot2</h3>
<p>Gplot2 is a system for declaratively creating graphics, based on The Grammar of Graphics. You provide the data, tell ggplot2 how to map variables to aesthetics, what graphical primitives to use, and it takes care of the details.</p>
<p>Detailed info can be found here: https://ggplot2.tidyverse.org</p>
<p>One important difference to basic plots is that argument can be given in separate blocks that are separated by a <code>+</code></p>
<section id="start-with-a-basic-bargraph" class="level4" data-number="1.27.5.1">
<h4 data-number="1.27.5.1" class="anchored" data-anchor-id="start-with-a-basic-bargraph"><span class="header-section-number">1.27.5.1</span> Start with a basic bargraph</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-3"><a href="#cb165-3" aria-hidden="true" tabindex="-1"></a><span class="co">#make a bargraph</span></span>
<span id="cb165-4"><a href="#cb165-4" aria-hidden="true" tabindex="-1"></a>myplot <span class="ot">&lt;-</span></span>
<span id="cb165-5"><a href="#cb165-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>Nutrient, <span class="at">y =</span> Rootlength)) <span class="sc">+</span>  <span class="co">#here we provide the dimensions of our plot</span></span>
<span id="cb165-6"><a href="#cb165-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>)                                   <span class="co">#here, we say what kind of plot we want </span></span>
<span id="cb165-7"><a href="#cb165-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-8"><a href="#cb165-8" aria-hidden="true" tabindex="-1"></a>myplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-165-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We see that the default behavior is to sum everything, which is not what we want. Luckily switching different graph types is very quick</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make a boxplot instead of bargrpah</span></span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>myplot <span class="ot">&lt;-</span></span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>Nutrient, <span class="at">y =</span> Rootlength)) <span class="sc">+</span>  </span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()                                   </span>
<span id="cb166-5"><a href="#cb166-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-6"><a href="#cb166-6" aria-hidden="true" tabindex="-1"></a>myplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-166-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="co">#do a histogram</span></span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>myplot <span class="ot">&lt;-</span></span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(Rootlength)) <span class="sc">+</span>  </span>
<span id="cb167-4"><a href="#cb167-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()                                   </span>
<span id="cb167-5"><a href="#cb167-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-6"><a href="#cb167-6" aria-hidden="true" tabindex="-1"></a>myplot</span>
<span id="cb167-7"><a href="#cb167-7" aria-hidden="true" tabindex="-1"></a><span class="do">## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-167-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The only thing we might want to watch out for is that depending on what data we plot the astethics might need to be adopted. I.e. for a histogram there is no need to provide a x and y value, but we only need to define for what data we want to plot a histogram.</p>
<p>Another useful feature is to add colors by groupings, i.e.&nbsp;nutrient conditions, using the fill option.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a>myplot <span class="ot">&lt;-</span></span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>Nutrient, <span class="at">y =</span> Rootlength, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()                                   </span>
<span id="cb168-4"><a href="#cb168-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb168-5"><a href="#cb168-5" aria-hidden="true" tabindex="-1"></a>myplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-168-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="prettify" class="level4" data-number="1.27.5.2">
<h4 data-number="1.27.5.2" class="anchored" data-anchor-id="prettify"><span class="header-section-number">1.27.5.2</span> Prettify</h4>
<p>By default the basic design of a ggplot2 is not ready for publication but we can control every aspects to make it look nicer. A cheat sheet for all the options can be found <a href="http://zevross.com/blog/2014/08/04/beautiful-plotting-in-r-a-ggplot2-cheatsheet-3/">here</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>myplot <span class="ot">&lt;-</span></span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>Nutrient, <span class="at">y =</span> Rootlength, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()  <span class="sc">+</span>                                 </span>
<span id="cb169-4"><a href="#cb169-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"grey"</span>)) <span class="sc">+</span>                                          <span class="co">#add some other colors</span></span>
<span id="cb169-5"><a href="#cb169-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">16</span>)) <span class="sc">+</span>                               <span class="co">#make the axis start at 0 </span></span>
<span id="cb169-6"><a href="#cb169-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span>                                                                            <span class="co">#remove the grey background</span></span>
<span id="cb169-7"><a href="#cb169-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Root length (cm)"</span>) <span class="sc">+</span>                                                              <span class="co">#change the label for the y-axis</span></span>
<span id="cb169-8"><a href="#cb169-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span>                                                                              <span class="co">#remove the label for the x axis</span></span>
<span id="cb169-9"><a href="#cb169-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(), <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),           <span class="co">#modify the grid lines</span></span>
<span id="cb169-10"><a href="#cb169-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(), <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb169-11"><a href="#cb169-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"right"</span>,                                                          <span class="co">#move the legend around</span></span>
<span id="cb169-12"><a href="#cb169-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>,<span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size=</span><span class="dv">12</span>))                            <span class="co">#control size and orientation of the axis labels</span></span>
<span id="cb169-13"><a href="#cb169-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb169-14"><a href="#cb169-14" aria-hidden="true" tabindex="-1"></a>myplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-169-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="sort-data" class="level4" data-number="1.27.5.3">
<h4 data-number="1.27.5.3" class="anchored" data-anchor-id="sort-data"><span class="header-section-number">1.27.5.3</span> Sort data</h4>
<p>By default most programs sort alpha in an alphabetical way. We can reorder this using vectors (which we can write ourselves or use a mapping file to create them)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="co">#reorder the nutrient levels in our factors</span></span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>Nutrient2 <span class="ot">&lt;-</span>  <span class="fu">factor</span>(growth_data<span class="sc">$</span>Nutrient, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"P"</span>, <span class="st">"noP"</span>)) </span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-4"><a href="#cb170-4" aria-hidden="true" tabindex="-1"></a><span class="co">#plot in our new order</span></span>
<span id="cb170-5"><a href="#cb170-5" aria-hidden="true" tabindex="-1"></a>myplot <span class="ot">&lt;-</span></span>
<span id="cb170-6"><a href="#cb170-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>Nutrient2, <span class="at">y =</span> Rootlength, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb170-7"><a href="#cb170-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb170-8"><a href="#cb170-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-9"><a href="#cb170-9" aria-hidden="true" tabindex="-1"></a>myplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-170-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="printing-data" class="level4" data-number="1.27.5.4">
<h4 data-number="1.27.5.4" class="anchored" data-anchor-id="printing-data"><span class="header-section-number">1.27.5.4</span> Printing data</h4>
<p>With <code>pdf()</code> we tell R that we want to print something to our computer. Inside the function we can define the name of the pdf to generate, the size and other things. After adding the plot we want to print it is important to run <code>dev.off()</code> in order to tell R to stop the “printing mode” and go back to the default mode.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="co">#lets first generate two plots</span></span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a>myplot_root <span class="ot">&lt;-</span></span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>Nutrient2, <span class="at">y =</span> Rootlength, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb171-4"><a href="#cb171-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb171-5"><a href="#cb171-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-6"><a href="#cb171-6" aria-hidden="true" tabindex="-1"></a>myplot_shoot <span class="ot">&lt;-</span></span>
<span id="cb171-7"><a href="#cb171-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>Nutrient2, <span class="at">y =</span> FW_shoot_mg, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb171-8"><a href="#cb171-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb171-9"><a href="#cb171-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-10"><a href="#cb171-10" aria-hidden="true" tabindex="-1"></a><span class="co">#plot one graph</span></span>
<span id="cb171-11"><a href="#cb171-11" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"../output_examples/Plot_root_length.pdf"</span>, <span class="at">width=</span><span class="dv">3</span>, <span class="at">height=</span><span class="dv">3</span>, <span class="at">family  =</span> <span class="st">"Times"</span>)</span>
<span id="cb171-12"><a href="#cb171-12" aria-hidden="true" tabindex="-1"></a>myplot_root</span>
<span id="cb171-13"><a href="#cb171-13" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-14"><a href="#cb171-14" aria-hidden="true" tabindex="-1"></a><span class="do">## quartz_off_screen </span></span>
<span id="cb171-15"><a href="#cb171-15" aria-hidden="true" tabindex="-1"></a><span class="do">##                 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="combining-and-printing-multiple-plots" class="level4" data-number="1.27.5.5">
<h4 data-number="1.27.5.5" class="anchored" data-anchor-id="combining-and-printing-multiple-plots"><span class="header-section-number">1.27.5.5</span> Combining and printing multiple plots</h4>
<p>The ggpubr package is nice to combine multiple plots onto one plot. Some more information on this package can be found <a href="https://rpkgs.datanovia.com/ggpubr/index.html">here</a></p>
<p>How it works:</p>
<ul>
<li>First list the plots we want to print</li>
<li>the labels argument allows us to add labels</li>
<li>the ncol and nrow arguments allow us to control how many plots go into a row and a column</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load package</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-4"><a href="#cb172-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view how ggarrange deals with two plots</span></span>
<span id="cb172-5"><a href="#cb172-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(myplot_root, myplot_shoot, </span>
<span id="cb172-6"><a href="#cb172-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>),</span>
<span id="cb172-7"><a href="#cb172-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-172-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we can also plot two graphs and print them in one pdf</span></span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"../output_examples/Two_plots.pdf"</span>, <span class="at">width=</span><span class="dv">3</span>, <span class="at">height=</span><span class="dv">3</span>, <span class="at">family  =</span> <span class="st">"Times"</span>)</span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(myplot_root, myplot_shoot, </span>
<span id="cb173-4"><a href="#cb173-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>),</span>
<span id="cb173-5"><a href="#cb173-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb173-6"><a href="#cb173-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb173-7"><a href="#cb173-7" aria-hidden="true" tabindex="-1"></a><span class="do">## quartz_off_screen </span></span>
<span id="cb173-8"><a href="#cb173-8" aria-hidden="true" tabindex="-1"></a><span class="do">##                 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sorting-data-v2" class="level4" data-number="1.27.5.6">
<h4 data-number="1.27.5.6" class="anchored" data-anchor-id="sorting-data-v2"><span class="header-section-number">1.27.5.6</span> Sorting data v2</h4>
<p>If we have multiple conditions, i.e.&nbsp;nutrient conditions and other treatments there are several ways to plot these</p>
<ol type="1">
<li>Plot them side by side and color by nutrient conditions.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="co">#pdf("Plot_root_length.pdf", width=3, height=3, family  = "Times")</span></span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>SampleID, <span class="at">y =</span> Rootlength, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb174-3"><a href="#cb174-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-174-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="co">#dev.off()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Change the order.</li>
</ol>
<p>Now, here the order is not nice, but as mentioned we can use mapping files to sort our data. Lets try.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="co">#lets check how our mapping file looked like</span></span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((mapping_file), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">SampleID</th>
<th style="text-align: left;">Nutrient</th>
<th style="text-align: left;">Condition</th>
<th style="text-align: left;">Comment</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1</td>
<td style="text-align: left;">noP</td>
<td style="text-align: left;">noP</td>
<td style="text-align: left;">MgCl</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
<tr class="even">
<td style="text-align: left;">14</td>
<td style="text-align: left;">noP_101</td>
<td style="text-align: left;">noP</td>
<td style="text-align: left;">Strain101</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
<tr class="odd">
<td style="text-align: left;">27</td>
<td style="text-align: left;">noP_230</td>
<td style="text-align: left;">noP</td>
<td style="text-align: left;">Strain230</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
<tr class="even">
<td style="text-align: left;">38</td>
<td style="text-align: left;">noP_28</td>
<td style="text-align: left;">noP</td>
<td style="text-align: left;">Strain28</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
<tr class="odd">
<td style="text-align: left;">53</td>
<td style="text-align: left;">P</td>
<td style="text-align: left;">P</td>
<td style="text-align: left;">MgCl</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
<tr class="even">
<td style="text-align: left;">68</td>
<td style="text-align: left;">P_101</td>
<td style="text-align: left;">P</td>
<td style="text-align: left;">Strain101</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
<tr class="odd">
<td style="text-align: left;">80</td>
<td style="text-align: left;">P_230</td>
<td style="text-align: left;">P</td>
<td style="text-align: left;">Strain230</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
<tr class="even">
<td style="text-align: left;">93</td>
<td style="text-align: left;">P_28</td>
<td style="text-align: left;">P</td>
<td style="text-align: left;">Strain28</td>
<td style="text-align: left;">FirstExperiment</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can use this simpler table to define how we want to resort our growth data. First, lets reorder the metadata first by nutrient and then condition:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="co">#lets sort the file first conditon an the nutrient (in reverse order, by using the rev() function )</span></span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>mapping_file <span class="ot">&lt;-</span> mapping_file[<span class="fu">with</span>(mapping_file, <span class="fu">order</span>(Condition, <span class="fu">rev</span>(Nutrient))), ]</span>
<span id="cb177-3"><a href="#cb177-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-4"><a href="#cb177-4" aria-hidden="true" tabindex="-1"></a><span class="co">#check whether we like how things are ordered (look at the order of the first line)</span></span>
<span id="cb177-5"><a href="#cb177-5" aria-hidden="true" tabindex="-1"></a>mapping_file<span class="sc">$</span>SampleID</span>
<span id="cb177-6"><a href="#cb177-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "noP"     "P"       "noP_101" "P_101"   "noP_230" "P_230"   "noP_28" </span></span>
<span id="cb177-7"><a href="#cb177-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [8] "P_28"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can use the order of this file to re-order our larger dataframe with the growth data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="co">#reorder the levels of our growth data using the mapping file</span></span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a>growth_data<span class="sc">$</span>SampleID2 <span class="ot">&lt;-</span>  <span class="fu">factor</span>(growth_data<span class="sc">$</span>SampleID, <span class="at">levels =</span> <span class="fu">as.character</span>(mapping_file<span class="sc">$</span>SampleID)) </span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(growth_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SampleID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Nutrient"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Condition"],"name":[3],"type":["chr"],"align":["left"]},{"label":["FW_shoot_mg"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Rootlength"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["category"],"name":[6],"type":["chr"],"align":["left"]},{"label":["Nutrient2"],"name":[7],"type":["fct"],"align":["left"]},{"label":["SampleID2"],"name":[8],"type":["fct"],"align":["left"]}],"data":[{"1":"noP","2":"noP","3":"MgCl","4":"10.26","5":"5.931015","6":"small","7":"noP","8":"noP","_rn_":"1"},{"1":"noP","2":"noP","3":"MgCl","4":"6.52","5":"5.743447","6":"small","7":"noP","8":"noP","_rn_":"2"},{"1":"noP","2":"noP","3":"MgCl","4":"12.17","5":"6.834720","6":"small","7":"noP","8":"noP","_rn_":"3"},{"1":"noP","2":"noP","3":"MgCl","4":"11.37","5":"6.742735","6":"small","7":"noP","8":"noP","_rn_":"4"},{"1":"noP","2":"noP","3":"MgCl","4":"9.80","5":"6.736886","6":"small","7":"noP","8":"noP","_rn_":"5"},{"1":"noP","2":"noP","3":"MgCl","4":"3.75","5":"4.236348","6":"small","7":"noP","8":"noP","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb179-2"><a href="#cb179-2" aria-hidden="true" tabindex="-1"></a><span class="co">#plot (for now lets do this side by side)</span></span>
<span id="cb179-3"><a href="#cb179-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>SampleID2, <span class="at">y =</span> Rootlength, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb179-4"><a href="#cb179-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-177-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="bargraphs-with-error-bars" class="level4" data-number="1.27.5.7">
<h4 data-number="1.27.5.7" class="anchored" data-anchor-id="bargraphs-with-error-bars"><span class="header-section-number">1.27.5.7</span> Bargraphs with error bars</h4>
<p>For bargraphs we can also make them nice looking with errorbars, however, the values for the mean, sd and so one ideally should be listed in a summary table.</p>
<p>Luckily we have learned before how we can use ddply to create such a table again and then blot it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="co">#summarize</span></span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a>growth_data_summarized <span class="ot">&lt;-</span> <span class="fu">ddply</span>(growth_data, .(SampleID, Nutrient), summarize, <span class="at">RootLength =</span> <span class="fu">mean</span>(Rootlength), <span class="at">sd =</span> <span class="fu">sd</span> (Rootlength), <span class="at">se =</span> <span class="fu">sd</span>(Rootlength) <span class="sc">/</span> <span class="fu">sqrt</span>((<span class="fu">length</span>(Rootlength))))</span>
<span id="cb180-3"><a href="#cb180-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-4"><a href="#cb180-4" aria-hidden="true" tabindex="-1"></a><span class="co">#order levels</span></span>
<span id="cb180-5"><a href="#cb180-5" aria-hidden="true" tabindex="-1"></a>growth_data_summarized<span class="sc">$</span>SampleID2 <span class="ot">&lt;-</span>  <span class="fu">factor</span>(growth_data_summarized<span class="sc">$</span>SampleID, <span class="at">levels =</span> <span class="fu">as.character</span>(mapping_file<span class="sc">$</span>SampleID)) </span>
<span id="cb180-6"><a href="#cb180-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-7"><a href="#cb180-7" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb180-8"><a href="#cb180-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data_summarized, <span class="fu">aes</span>(<span class="at">x=</span>SampleID2, <span class="at">y=</span>RootLength, <span class="at">fill=</span>Nutrient)) <span class="sc">+</span> </span>
<span id="cb180-9"><a href="#cb180-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">color=</span><span class="st">"black"</span>, </span>
<span id="cb180-10"><a href="#cb180-10" aria-hidden="true" tabindex="-1"></a>           <span class="at">position=</span><span class="fu">position_dodge</span>()) <span class="sc">+</span></span>
<span id="cb180-11"><a href="#cb180-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin=</span>RootLength<span class="sc">-</span>sd, <span class="at">ymax=</span>RootLength<span class="sc">+</span>sd), <span class="at">width=</span>.<span class="dv">2</span>,</span>
<span id="cb180-12"><a href="#cb180-12" aria-hidden="true" tabindex="-1"></a>                 <span class="at">position=</span><span class="fu">position_dodge</span>(.<span class="dv">9</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-178-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="faceting-data" class="level4" data-number="1.27.5.8">
<h4 data-number="1.27.5.8" class="anchored" data-anchor-id="faceting-data"><span class="header-section-number">1.27.5.8</span> Faceting data</h4>
<p>Sometimes, we might to plot data into separate plots. This can be done in ggplot with one extra command. Facetting can do this for you.</p>
<p>Options: - Scales “free” tells ggplot that the scales can be different between plots. I.e. axis height. - ncol = allows to control the dimensions of the plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot horizontal</span></span>
<span id="cb181-2"><a href="#cb181-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(growth_data, <span class="fu">aes</span>(<span class="at">x =</span>SampleID2, <span class="at">y =</span> Rootlength, <span class="at">fill =</span> Nutrient)) <span class="sc">+</span>  </span>
<span id="cb181-3"><a href="#cb181-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb181-4"><a href="#cb181-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(. <span class="sc">~</span> Nutrient, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-179-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
</section>
<section id="working-with-annotation-data" class="level2" data-number="1.28">
<h2 data-number="1.28" class="anchored" data-anchor-id="working-with-annotation-data"><span class="header-section-number">1.28</span> Working with annotation data</h2>
<p>As mentioned in the beginning section, if working with genomes we often have large lists of proteins and their annotations, often from different databases.</p>
<p>Here, we might be interested in condensing this information for each genome or for each phylum etc . OR we want to merge our data with other tables that are ordered differently. We might also want to plot the data to compare the presence absence levels of different genomes. This section with discuss how to do this in a step by step manner.</p>
<section id="load-essential-packages" class="level3" data-number="1.28.1">
<h3 data-number="1.28.1" class="anchored" data-anchor-id="load-essential-packages"><span class="header-section-number">1.28.1</span> Load essential packages</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="co"># | warning: false</span></span>
<span id="cb182-2"><a href="#cb182-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb182-3"><a href="#cb182-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"plyr"</span>)</span>
<span id="cb182-4"><a href="#cb182-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb182-5"><a href="#cb182-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"grid"</span>)</span>
<span id="cb182-6"><a href="#cb182-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"gplots"</span>)</span>
<span id="cb182-7"><a href="#cb182-7" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-8"><a href="#cb182-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'gplots'</span></span>
<span id="cb182-9"><a href="#cb182-9" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:stats':</span></span>
<span id="cb182-10"><a href="#cb182-10" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-11"><a href="#cb182-11" aria-hidden="true" tabindex="-1"></a><span class="do">##     lowess</span></span>
<span id="cb182-12"><a href="#cb182-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"gridExtra"</span>)</span>
<span id="cb182-13"><a href="#cb182-13" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-14"><a href="#cb182-14" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'gridExtra'</span></span>
<span id="cb182-15"><a href="#cb182-15" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:dplyr':</span></span>
<span id="cb182-16"><a href="#cb182-16" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-17"><a href="#cb182-17" aria-hidden="true" tabindex="-1"></a><span class="do">##     combine</span></span>
<span id="cb182-18"><a href="#cb182-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"multcomp"</span>)</span>
<span id="cb182-19"><a href="#cb182-19" aria-hidden="true" tabindex="-1"></a><span class="do">## Loading required package: mvtnorm</span></span>
<span id="cb182-20"><a href="#cb182-20" aria-hidden="true" tabindex="-1"></a><span class="do">## Loading required package: survival</span></span>
<span id="cb182-21"><a href="#cb182-21" aria-hidden="true" tabindex="-1"></a><span class="do">## Loading required package: TH.data</span></span>
<span id="cb182-22"><a href="#cb182-22" aria-hidden="true" tabindex="-1"></a><span class="do">## Loading required package: MASS</span></span>
<span id="cb182-23"><a href="#cb182-23" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-24"><a href="#cb182-24" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'MASS'</span></span>
<span id="cb182-25"><a href="#cb182-25" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:dplyr':</span></span>
<span id="cb182-26"><a href="#cb182-26" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-27"><a href="#cb182-27" aria-hidden="true" tabindex="-1"></a><span class="do">##     select</span></span>
<span id="cb182-28"><a href="#cb182-28" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-29"><a href="#cb182-29" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'TH.data'</span></span>
<span id="cb182-30"><a href="#cb182-30" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:MASS':</span></span>
<span id="cb182-31"><a href="#cb182-31" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-32"><a href="#cb182-32" aria-hidden="true" tabindex="-1"></a><span class="do">##     geyser</span></span>
<span id="cb182-33"><a href="#cb182-33" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"reshape2"</span>)</span>
<span id="cb182-34"><a href="#cb182-34" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-35"><a href="#cb182-35" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'reshape2'</span></span>
<span id="cb182-36"><a href="#cb182-36" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:tidyr':</span></span>
<span id="cb182-37"><a href="#cb182-37" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-38"><a href="#cb182-38" aria-hidden="true" tabindex="-1"></a><span class="do">##     smiths</span></span>
<span id="cb182-39"><a href="#cb182-39" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"RColorBrewer"</span>)</span>
<span id="cb182-40"><a href="#cb182-40" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">'tidyr'</span>)</span>
<span id="cb182-41"><a href="#cb182-41" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">'tidyverse'</span>)</span>
<span id="cb182-42"><a href="#cb182-42" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb182-43"><a href="#cb182-43" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-44"><a href="#cb182-44" aria-hidden="true" tabindex="-1"></a><span class="do">## Attaching package: 'data.table'</span></span>
<span id="cb182-45"><a href="#cb182-45" aria-hidden="true" tabindex="-1"></a><span class="do">## The following objects are masked from 'package:reshape2':</span></span>
<span id="cb182-46"><a href="#cb182-46" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-47"><a href="#cb182-47" aria-hidden="true" tabindex="-1"></a><span class="do">##     dcast, melt</span></span>
<span id="cb182-48"><a href="#cb182-48" aria-hidden="true" tabindex="-1"></a><span class="do">## The following objects are masked from 'package:dplyr':</span></span>
<span id="cb182-49"><a href="#cb182-49" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-50"><a href="#cb182-50" aria-hidden="true" tabindex="-1"></a><span class="do">##     between, first, last</span></span>
<span id="cb182-51"><a href="#cb182-51" aria-hidden="true" tabindex="-1"></a><span class="do">## The following object is masked from 'package:purrr':</span></span>
<span id="cb182-52"><a href="#cb182-52" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb182-53"><a href="#cb182-53" aria-hidden="true" tabindex="-1"></a><span class="do">##     transpose</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="deal-with-the-mapping-file-for-our-genomes" class="level3" data-number="1.28.2">
<h3 data-number="1.28.2" class="anchored" data-anchor-id="deal-with-the-mapping-file-for-our-genomes"><span class="header-section-number">1.28.2</span> Deal with the mapping file for our genomes</h3>
<section id="load-mapping-file" class="level4" data-number="1.28.2.1">
<h4 data-number="1.28.2.1" class="anchored" data-anchor-id="load-mapping-file"><span class="header-section-number">1.28.2.1</span> Load mapping file</h4>
<p>mapping file = a table that has the metadata for our genomes. In this specific example it lists our 46 DPANN genomes into clusters based on their phylogenetic placement.</p>
<p>We later want to summarize our data by individual bins and the phylogenetic clusters defined in the mapping file.</p>
<p>Notice: We can also sort our mapping file beforehand and use this order to order our tables, graph, etc…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb183-2"><a href="#cb183-2" aria-hidden="true" tabindex="-1"></a><span class="co">#read in the mapping file</span></span>
<span id="cb183-3"><a href="#cb183-3" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/mapping.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb183-4"><a href="#cb183-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb183-5"><a href="#cb183-5" aria-hidden="true" tabindex="-1"></a><span class="co">#check the structure of the mapping file</span></span>
<span id="cb183-6"><a href="#cb183-6" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(design), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">Cluster</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">NIOZ132_cc_b51_2</td>
<td style="text-align: left;">Basal</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ134_mx_b281_2</td>
<td style="text-align: left;">Basal</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SRR2090153_bin1042</td>
<td style="text-align: left;">Aquifer</td>
</tr>
<tr class="even">
<td style="text-align: left;">SRR2090153_bin461</td>
<td style="text-align: left;">Aquifer</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SRR2090159_bin1129</td>
<td style="text-align: left;">Aquifer</td>
</tr>
<tr class="even">
<td style="text-align: left;">SRR2090159_bin1288</td>
<td style="text-align: left;">Aquifer</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="clean-mapping-file" class="level3" data-number="1.28.3">
<h3 data-number="1.28.3" class="anchored" data-anchor-id="clean-mapping-file"><span class="header-section-number">1.28.3</span> Clean mapping file</h3>
<p>Here, we want to add an extra column that summarizes how many genomes are in each phylogenetic cluster.</p>
<p>First lets summarize how many bins we hav for each cluster:</p>
<ul>
<li><code>ddply()</code> = summarize our data by counting the number of bins we have in each cluster</li>
<li><code>order()</code> = next, we order our data based on the clusters with the greated number of bins</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="co">#add a new column that links the BinID and the taxonomic cluster</span></span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a>design<span class="sc">$</span>NewName2 <span class="ot">&lt;-</span><span class="fu">paste</span>(design<span class="sc">$</span>BinID, design<span class="sc">$</span>Cluster, <span class="at">sep =</span> <span class="st">"-"</span>)</span>
<span id="cb184-3"><a href="#cb184-3" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((design), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">Cluster</th>
<th style="text-align: left;">NewName2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">NIOZ132_cc_b51_2</td>
<td style="text-align: left;">Basal</td>
<td style="text-align: left;">NIOZ132_cc_b51_2-Basal</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ134_mx_b281_2</td>
<td style="text-align: left;">Basal</td>
<td style="text-align: left;">NIOZ134_mx_b281_2-Basal</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SRR2090153_bin1042</td>
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090153_bin1042-Aquifer</td>
</tr>
<tr class="even">
<td style="text-align: left;">SRR2090153_bin461</td>
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090153_bin461-Aquifer</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SRR2090159_bin1129</td>
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090159_bin1129-Aquifer</td>
</tr>
<tr class="even">
<td style="text-align: left;">SRR2090159_bin1288</td>
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090159_bin1288-Aquifer</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">Marine</td>
<td style="text-align: left;">GCA_002494525-Marine</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002495465</td>
<td style="text-align: left;">Marine</td>
<td style="text-align: left;">GCA_002495465-Marine</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002501985</td>
<td style="text-align: left;">Marine</td>
<td style="text-align: left;">GCA_002501985-Marine</td>
</tr>
<tr class="even">
<td style="text-align: left;">SRR4028224_bin17</td>
<td style="text-align: left;">Marine</td>
<td style="text-align: left;">SRR4028224_bin17-Marine</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SRR5007147_bin71</td>
<td style="text-align: left;">Marine</td>
<td style="text-align: left;">SRR5007147_bin71-Marine</td>
</tr>
<tr class="even">
<td style="text-align: left;">U_76725</td>
<td style="text-align: left;">Marine</td>
<td style="text-align: left;">U_76725-Marine</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ124_cc_b247_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ124_cc_b247_2-Clade1</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ125_mb_b254_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ125_mb_b254_2-Clade1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ129_mb_b262_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ129_mb_b262_2-Clade1</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ126_mb_b401_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ126_mb_b401_2-Clade1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ122_mb_b319_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ122_mb_b319_2-Clade1</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ134_mb_b41_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ134_mb_b41_2-Clade1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ132_mb_b282_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ132_mb_b282_2-Clade1</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ127_mb_b161_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ127_mb_b161_2-Clade1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ123_bs_b5_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ123_bs_b5_2-Clade1</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ121_cc_b94_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ121_cc_b94_2-Clade1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ136_cc_b15_2</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: left;">NIOZ136_cc_b15_2-Clade1</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002502135</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">GCA_002502135-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ134_mb_b293_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ134_mb_b293_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ132_mb_b260_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ132_mb_b260_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ126_mb_b137_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ126_mb_b137_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ129_mb_b254_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ129_mb_b254_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ127_mb_b282_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ127_mb_b282_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ122_mb_b305_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ122_mb_b305_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ136_mb_b335_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ136_mb_b335_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ132_mb_b198_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ132_mb_b198_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ136_mb_b104_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ136_mb_b104_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ125_mb_b178_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ125_mb_b178_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ121_mb_b48_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ121_mb_b48_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ134_cc_b149_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ134_cc_b149_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ134_mb_b361_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ134_mb_b361_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ136_mb_b226_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ136_mb_b226_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ123_bs_b392_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ123_bs_b392_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ124_mb_b130_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ124_mb_b130_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ132_cc_b149_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ132_cc_b149_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ125_cc_b75_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ125_cc_b75_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ132_mx_b314_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ132_mx_b314_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ120_mb_b229_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ120_mb_b229_2-Clade2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ126_mb_b304_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ126_mb_b304_2-Clade2</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: left;">NIOZ119_mb_b5_2-Clade2</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a><span class="co">#transform mapping file and summarize how many genomes we have/cluster</span></span>
<span id="cb185-3"><a href="#cb185-3" aria-hidden="true" tabindex="-1"></a>Number_of_taxa <span class="ot">&lt;-</span> <span class="fu">ddply</span>(design, .(Cluster), summarize, <span class="at">NrGenomes =</span> <span class="fu">length</span>(Cluster))</span>
<span id="cb185-4"><a href="#cb185-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-5"><a href="#cb185-5" aria-hidden="true" tabindex="-1"></a><span class="co">#order our data</span></span>
<span id="cb185-6"><a href="#cb185-6" aria-hidden="true" tabindex="-1"></a>Number_of_taxa <span class="ot">&lt;-</span> Number_of_taxa[<span class="fu">order</span>(Number_of_taxa<span class="sc">$</span>NrGenomes, <span class="at">decreasing =</span> <span class="cn">TRUE</span>),] </span>
<span id="cb185-7"><a href="#cb185-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-8"><a href="#cb185-8" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb185-9"><a href="#cb185-9" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((Number_of_taxa), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Cluster</th>
<th style="text-align: right;">NrGenomes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">4</td>
<td style="text-align: left;">Clade2</td>
<td style="text-align: right;">23</td>
</tr>
<tr class="even">
<td style="text-align: left;">3</td>
<td style="text-align: left;">Clade1</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">5</td>
<td style="text-align: left;">Marine</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">1</td>
<td style="text-align: left;">Aquifer</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">2</td>
<td style="text-align: left;">Basal</td>
<td style="text-align: right;">2</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Next, add this information into our original mapping file:</p>
<ul>
<li><code>paste()</code> = paste together different columns. With sep we can control what delimiter we want to use when combining info</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb186"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb186-2"><a href="#cb186-2" aria-hidden="true" tabindex="-1"></a><span class="co">#add a new column with the cluster count info into mapping file (then it can be added as a label into figures)</span></span>
<span id="cb186-3"><a href="#cb186-3" aria-hidden="true" tabindex="-1"></a>design_2 <span class="ot">&lt;-</span> <span class="fu">merge</span>(design, Number_of_taxa, <span class="at">by =</span> <span class="st">"Cluster"</span> )</span>
<span id="cb186-4"><a href="#cb186-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb186-5"><a href="#cb186-5" aria-hidden="true" tabindex="-1"></a><span class="co">#view wheter all went fine</span></span>
<span id="cb186-6"><a href="#cb186-6" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(design_2), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">Cluster</th>
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">NewName2</th>
<th style="text-align: right;">NrGenomes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090153_bin1042</td>
<td style="text-align: left;">SRR2090153_bin1042-Aquifer</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090153_bin461</td>
<td style="text-align: left;">SRR2090153_bin461-Aquifer</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090159_bin1129</td>
<td style="text-align: left;">SRR2090159_bin1129-Aquifer</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090159_bin1288</td>
<td style="text-align: left;">SRR2090159_bin1288-Aquifer</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Basal</td>
<td style="text-align: left;">NIOZ132_cc_b51_2</td>
<td style="text-align: left;">NIOZ132_cc_b51_2-Basal</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Basal</td>
<td style="text-align: left;">NIOZ134_mx_b281_2</td>
<td style="text-align: left;">NIOZ134_mx_b281_2-Basal</td>
<td style="text-align: right;">2</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Next, let’s merge all relevant into one column using <code>paste()</code> again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-2"><a href="#cb187-2" aria-hidden="true" tabindex="-1"></a><span class="co">#generate a new column into the mapping file that links the cluster name with the number of genomes of clusters in each cluster</span></span>
<span id="cb187-3"><a href="#cb187-3" aria-hidden="true" tabindex="-1"></a>design_2<span class="sc">$</span>ClusterName <span class="ot">&lt;-</span> <span class="fu">paste</span>(design_2<span class="sc">$</span>Cluster, <span class="st">" ("</span>, design_2<span class="sc">$</span>NrGenomes, <span class="st">")"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb187-4"><a href="#cb187-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-5"><a href="#cb187-5" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb187-6"><a href="#cb187-6" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(design_2), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 10%">
<col style="width: 25%">
<col style="width: 35%">
<col style="width: 13%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Cluster</th>
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">NewName2</th>
<th style="text-align: right;">NrGenomes</th>
<th style="text-align: left;">ClusterName</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090153_bin1042</td>
<td style="text-align: left;">SRR2090153_bin1042-Aquifer</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">Aquifer (4)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090153_bin461</td>
<td style="text-align: left;">SRR2090153_bin461-Aquifer</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">Aquifer (4)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090159_bin1129</td>
<td style="text-align: left;">SRR2090159_bin1129-Aquifer</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">Aquifer (4)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer</td>
<td style="text-align: left;">SRR2090159_bin1288</td>
<td style="text-align: left;">SRR2090159_bin1288-Aquifer</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">Aquifer (4)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Basal</td>
<td style="text-align: left;">NIOZ132_cc_b51_2</td>
<td style="text-align: left;">NIOZ132_cc_b51_2-Basal</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">Basal (2)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Basal</td>
<td style="text-align: left;">NIOZ134_mx_b281_2</td>
<td style="text-align: left;">NIOZ134_mx_b281_2-Basal</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">Basal (2)</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="generate-lists-to-order-dataframes" class="level3" data-number="1.28.4">
<h3 data-number="1.28.4" class="anchored" data-anchor-id="generate-lists-to-order-dataframes"><span class="header-section-number">1.28.4</span> Generate lists to order dataframes</h3>
<p>The default behaviour of a lot of R functions is to order data alphabetically but this is not what we might want. I.e. we want to order by different columns or simply order by the initial order in our mapping file. To do this it is useful to have vectors that are order our bins, clusters, etc… like we want them.</p>
<p>In this example the mapping file was ordered like this</p>
<ul>
<li>The basal group</li>
<li>Groups from an older dataset (marine and aquifer)</li>
<li>two black sea clades</li>
</ul>
<p>Lets first check, whether our new mapping file still has that order</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check the old data</span></span>
<span id="cb188-2"><a href="#cb188-2" aria-hidden="true" tabindex="-1"></a>design<span class="sc">$</span>Cluster</span>
<span id="cb188-3"><a href="#cb188-3" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] "Basal"   "Basal"   "Aquifer" "Aquifer" "Aquifer" "Aquifer" "Marine" </span></span>
<span id="cb188-4"><a href="#cb188-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  [8] "Marine"  "Marine"  "Marine"  "Marine"  "Marine"  "Clade1"  "Clade1" </span></span>
<span id="cb188-5"><a href="#cb188-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [15] "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1" </span></span>
<span id="cb188-6"><a href="#cb188-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [22] "Clade1"  "Clade1"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb188-7"><a href="#cb188-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [29] "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb188-8"><a href="#cb188-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [36] "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb188-9"><a href="#cb188-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [43] "Clade2"  "Clade2"  "Clade2"  "Clade2"</span></span>
<span id="cb188-10"><a href="#cb188-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-11"><a href="#cb188-11" aria-hidden="true" tabindex="-1"></a><span class="co">#check the new data</span></span>
<span id="cb188-12"><a href="#cb188-12" aria-hidden="true" tabindex="-1"></a>design_2<span class="sc">$</span>Cluster</span>
<span id="cb188-13"><a href="#cb188-13" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] "Aquifer" "Aquifer" "Aquifer" "Aquifer" "Basal"   "Basal"   "Clade1" </span></span>
<span id="cb188-14"><a href="#cb188-14" aria-hidden="true" tabindex="-1"></a><span class="do">##  [8] "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1" </span></span>
<span id="cb188-15"><a href="#cb188-15" aria-hidden="true" tabindex="-1"></a><span class="do">## [15] "Clade1"  "Clade1"  "Clade1"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb188-16"><a href="#cb188-16" aria-hidden="true" tabindex="-1"></a><span class="do">## [22] "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb188-17"><a href="#cb188-17" aria-hidden="true" tabindex="-1"></a><span class="do">## [29] "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb188-18"><a href="#cb188-18" aria-hidden="true" tabindex="-1"></a><span class="do">## [36] "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Marine"  "Marine" </span></span>
<span id="cb188-19"><a href="#cb188-19" aria-hidden="true" tabindex="-1"></a><span class="do">## [43] "Marine"  "Marine"  "Marine"  "Marine"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we check the individual factors, we can see that the new dataframe is order by alphabet. Since this is not what we want, lets correct this</p>
<ul>
<li><code>match(design_2$BinID, design$BinID)</code> = We check what positions in design_2 match with design. Here, the first argument are the values to be matched and the second are the values to be matched against.</li>
<li><code>order()</code> = we reorder our dataframe based on the results from <code>match()</code></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb189"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-2"><a href="#cb189-2" aria-hidden="true" tabindex="-1"></a><span class="co">#reorder our old dataframe by the original one</span></span>
<span id="cb189-3"><a href="#cb189-3" aria-hidden="true" tabindex="-1"></a>design_2 <span class="ot">&lt;-</span> design_2[ <span class="fu">order</span>(<span class="fu">match</span>(design_2<span class="sc">$</span>BinID, design<span class="sc">$</span>BinID)), ]</span>
<span id="cb189-4"><a href="#cb189-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-5"><a href="#cb189-5" aria-hidden="true" tabindex="-1"></a><span class="co">#check, if the basal clade is now in the correct order</span></span>
<span id="cb189-6"><a href="#cb189-6" aria-hidden="true" tabindex="-1"></a>design_2<span class="sc">$</span>Cluster</span>
<span id="cb189-7"><a href="#cb189-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] "Basal"   "Basal"   "Aquifer" "Aquifer" "Aquifer" "Aquifer" "Marine" </span></span>
<span id="cb189-8"><a href="#cb189-8" aria-hidden="true" tabindex="-1"></a><span class="do">##  [8] "Marine"  "Marine"  "Marine"  "Marine"  "Marine"  "Clade1"  "Clade1" </span></span>
<span id="cb189-9"><a href="#cb189-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [15] "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1"  "Clade1" </span></span>
<span id="cb189-10"><a href="#cb189-10" aria-hidden="true" tabindex="-1"></a><span class="do">## [22] "Clade1"  "Clade1"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb189-11"><a href="#cb189-11" aria-hidden="true" tabindex="-1"></a><span class="do">## [29] "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb189-12"><a href="#cb189-12" aria-hidden="true" tabindex="-1"></a><span class="do">## [36] "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2"  "Clade2" </span></span>
<span id="cb189-13"><a href="#cb189-13" aria-hidden="true" tabindex="-1"></a><span class="do">## [43] "Clade2"  "Clade2"  "Clade2"  "Clade2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, that the basal clade is in the first position, lets make vectors of the bins and clusters in order this we use:</p>
<ul>
<li><code>unique()</code> = we use unique on all ClusterNames in our mapping file. That way instead of having repeated names, we only have the unique ones</li>
<li><code>as.character()</code> = we make sure that our R object we generated is a character (so has words).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="co">#make a list to order our bins</span></span>
<span id="cb190-2"><a href="#cb190-2" aria-hidden="true" tabindex="-1"></a>Bin_order <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(design_2<span class="sc">$</span>BinID))</span>
<span id="cb190-3"><a href="#cb190-3" aria-hidden="true" tabindex="-1"></a>Bin_order</span>
<span id="cb190-4"><a href="#cb190-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] "NIOZ132_cc_b51_2"   "NIOZ134_mx_b281_2"  "SRR2090153_bin1042"</span></span>
<span id="cb190-5"><a href="#cb190-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  [4] "SRR2090153_bin461"  "SRR2090159_bin1129" "SRR2090159_bin1288"</span></span>
<span id="cb190-6"><a href="#cb190-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  [7] "GCA_002494525"      "GCA_002495465"      "GCA_002501985"     </span></span>
<span id="cb190-7"><a href="#cb190-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [10] "SRR4028224_bin17"   "SRR5007147_bin71"   "U_76725"           </span></span>
<span id="cb190-8"><a href="#cb190-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [13] "NIOZ124_cc_b247_2"  "NIOZ125_mb_b254_2"  "NIOZ129_mb_b262_2" </span></span>
<span id="cb190-9"><a href="#cb190-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [16] "NIOZ126_mb_b401_2"  "NIOZ122_mb_b319_2"  "NIOZ134_mb_b41_2"  </span></span>
<span id="cb190-10"><a href="#cb190-10" aria-hidden="true" tabindex="-1"></a><span class="do">## [19] "NIOZ132_mb_b282_2"  "NIOZ127_mb_b161_2"  "NIOZ123_bs_b5_2"   </span></span>
<span id="cb190-11"><a href="#cb190-11" aria-hidden="true" tabindex="-1"></a><span class="do">## [22] "NIOZ121_cc_b94_2"   "NIOZ136_cc_b15_2"   "GCA_002502135"     </span></span>
<span id="cb190-12"><a href="#cb190-12" aria-hidden="true" tabindex="-1"></a><span class="do">## [25] "NIOZ134_mb_b293_2"  "NIOZ132_mb_b260_2"  "NIOZ126_mb_b137_2" </span></span>
<span id="cb190-13"><a href="#cb190-13" aria-hidden="true" tabindex="-1"></a><span class="do">## [28] "NIOZ129_mb_b254_2"  "NIOZ127_mb_b282_2"  "NIOZ122_mb_b305_2" </span></span>
<span id="cb190-14"><a href="#cb190-14" aria-hidden="true" tabindex="-1"></a><span class="do">## [31] "NIOZ136_mb_b335_2"  "NIOZ132_mb_b198_2"  "NIOZ136_mb_b104_2" </span></span>
<span id="cb190-15"><a href="#cb190-15" aria-hidden="true" tabindex="-1"></a><span class="do">## [34] "NIOZ125_mb_b178_2"  "NIOZ121_mb_b48_2"   "NIOZ134_cc_b149_2" </span></span>
<span id="cb190-16"><a href="#cb190-16" aria-hidden="true" tabindex="-1"></a><span class="do">## [37] "NIOZ134_mb_b361_2"  "NIOZ136_mb_b226_2"  "NIOZ123_bs_b392_2" </span></span>
<span id="cb190-17"><a href="#cb190-17" aria-hidden="true" tabindex="-1"></a><span class="do">## [40] "NIOZ124_mb_b130_2"  "NIOZ132_cc_b149_2"  "NIOZ125_cc_b75_2"  </span></span>
<span id="cb190-18"><a href="#cb190-18" aria-hidden="true" tabindex="-1"></a><span class="do">## [43] "NIOZ132_mx_b314_2"  "NIOZ120_mb_b229_2"  "NIOZ126_mb_b304_2" </span></span>
<span id="cb190-19"><a href="#cb190-19" aria-hidden="true" tabindex="-1"></a><span class="do">## [46] "NIOZ119_mb_b5_2"</span></span>
<span id="cb190-20"><a href="#cb190-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-21"><a href="#cb190-21" aria-hidden="true" tabindex="-1"></a><span class="co">#make cluster order</span></span>
<span id="cb190-22"><a href="#cb190-22" aria-hidden="true" tabindex="-1"></a>Cluster_order <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(design_2<span class="sc">$</span>ClusterName))</span>
<span id="cb190-23"><a href="#cb190-23" aria-hidden="true" tabindex="-1"></a>Cluster_order</span>
<span id="cb190-24"><a href="#cb190-24" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Basal (2)"   "Aquifer (4)" "Marine (6)"  "Clade1 (11)" "Clade2 (23)"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="deal-with-mapping-files-for-the-annotations" class="level2" data-number="1.29">
<h2 data-number="1.29" class="anchored" data-anchor-id="deal-with-mapping-files-for-the-annotations"><span class="header-section-number">1.29</span> Deal with mapping files for the annotations</h2>
<p>The mapping files (Arcog_mapping and Metabolism_file_KEGG) that are provided with this tutorial give extra information and order our KO and arCOG annotations. Genes_of_interest is a list of key metabolic genes we want to look at more closely.</p>
<section id="load-and-view-the-tables" class="level3" data-number="1.29.1">
<h3 data-number="1.29.1" class="anchored" data-anchor-id="load-and-view-the-tables"><span class="header-section-number">1.29.1</span> Load and view the tables</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb191-2"><a href="#cb191-2" aria-hidden="true" tabindex="-1"></a><span class="co">#general mapping file for arcog IDs</span></span>
<span id="cb191-3"><a href="#cb191-3" aria-hidden="true" tabindex="-1"></a>Arcog_mapping <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/ar14_arCOGdef19.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb191-4"><a href="#cb191-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Arcog_mapping)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 11%">
<col style="width: 8%">
<col style="width: 7%">
<col style="width: 41%">
<col style="width: 9%">
<col style="width: 10%">
<col style="width: 3%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">arcog</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">GeneID</th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">COG</th>
<th style="text-align: left;">PFAM</th>
<th style="text-align: left;">CD</th>
<th style="text-align: left;">TIGR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">arCOG00001</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00002</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">TIGR03433</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00004</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00005</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00006</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00007</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb192"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-2"><a href="#cb192-2" aria-hidden="true" tabindex="-1"></a><span class="co">#pathway mapping file</span></span>
<span id="cb192-3"><a href="#cb192-3" aria-hidden="true" tabindex="-1"></a>Metabolism_file_KEGG <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/Metabolism_Table_KO_Apr2020.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb192-4"><a href="#cb192-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Metabolism_file_KEGG)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 3%">
<col style="width: 9%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 50%">
<col style="width: 21%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">IDs</th>
<th style="text-align: left;">KO</th>
<th style="text-align: left;">pathway_A</th>
<th style="text-align: left;">order</th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Comment.Uniprot.comment</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ID1</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">ID2</td>
<td style="text-align: left;">K02567</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N1</td>
<td style="text-align: left;">napA; periplasmic nitrate reductase NapA [EC:1.7.99.4]</td>
<td style="text-align: left;">Dissimilatory</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID3</td>
<td style="text-align: left;">TIGR01706</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N1</td>
<td style="text-align: left;">“napA, periplasmic nitrate reductase”</td>
<td style="text-align: left;">Dissimilatory</td>
</tr>
<tr class="even">
<td style="text-align: left;">ID4</td>
<td style="text-align: left;">K02573</td>
<td style="text-align: left;">no_pathway</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">napG; ferredoxin-type protein NapG</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID5</td>
<td style="text-align: left;">K02574</td>
<td style="text-align: left;">no_pathway</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">ID6</td>
<td style="text-align: left;">K00367</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N1</td>
<td style="text-align: left;">narB; ferredoxin-nitrate reductase [EC:1.7.7.2]</td>
<td style="text-align: left;">Assimilatory</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb193"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb193-1"><a href="#cb193-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb193-2"><a href="#cb193-2" aria-hidden="true" tabindex="-1"></a><span class="co">#load the genes of interest</span></span>
<span id="cb193-3"><a href="#cb193-3" aria-hidden="true" tabindex="-1"></a>Genes_of_interest <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/Genes_of_interest.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb193-4"><a href="#cb193-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Arcog_mapping)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 11%">
<col style="width: 8%">
<col style="width: 7%">
<col style="width: 41%">
<col style="width: 9%">
<col style="width: 10%">
<col style="width: 3%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">arcog</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">GeneID</th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">COG</th>
<th style="text-align: left;">PFAM</th>
<th style="text-align: left;">CD</th>
<th style="text-align: left;">TIGR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">arCOG00001</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00002</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">TIGR03433</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00004</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00005</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00006</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;">pfam03551</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00007</td>
<td style="text-align: left;">K</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Transcriptional regulator, PadR family”</td>
<td style="text-align: left;">COG01695</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="make-a-vector-to-order-our-genes-of-interest" class="level3" data-number="1.29.2">
<h3 data-number="1.29.2" class="anchored" data-anchor-id="make-a-vector-to-order-our-genes-of-interest"><span class="header-section-number">1.29.2</span> Make a vector to order our genes of interest</h3>
<ul>
<li>We use <code>unique()</code> to remove any duplicate gene names</li>
<li><code>arrange()</code> can be used to order our dataframe by more than one column.</li>
</ul>
<p>You notice here that the syntax for <code>arrange()</code> is a bit unusual and we use the <code>%&gt;%</code> symbol (a so-called forward pipe operator). This symbol is commonly used in the <strong>dplyr</strong> and <strong>tidyr</strong> packages which are extremely useful to summarize data. This function passes the left hand side of the operator to the first argument of the right hand side of the operator. In the following example, the data frame Genes_of_interest gets passed to <code>arrange()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb194"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb194-1"><a href="#cb194-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-2"><a href="#cb194-2" aria-hidden="true" tabindex="-1"></a><span class="co">#order metabolic genes</span></span>
<span id="cb194-3"><a href="#cb194-3" aria-hidden="true" tabindex="-1"></a>Genes_Metabolism_order_temp <span class="ot">&lt;-</span> Genes_of_interest <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(Order, Order2)</span>
<span id="cb194-4"><a href="#cb194-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-5"><a href="#cb194-5" aria-hidden="true" tabindex="-1"></a><span class="co">#make a unique vector for our genes of interest</span></span>
<span id="cb194-6"><a href="#cb194-6" aria-hidden="true" tabindex="-1"></a>Genes_Metabolism_order <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(Genes_Metabolism_order_temp<span class="sc">$</span>Gene))</span>
<span id="cb194-7"><a href="#cb194-7" aria-hidden="true" tabindex="-1"></a>Genes_Metabolism_order</span>
<span id="cb194-8"><a href="#cb194-8" aria-hidden="true" tabindex="-1"></a><span class="do">##   [1] "HmgB"                     "HmgA"                    </span></span>
<span id="cb194-9"><a href="#cb194-9" aria-hidden="true" tabindex="-1"></a><span class="do">##   [3] "MvK"                      "M3k"                     </span></span>
<span id="cb194-10"><a href="#cb194-10" aria-hidden="true" tabindex="-1"></a><span class="do">##   [5] "M3P5K"                    "mvaD"                    </span></span>
<span id="cb194-11"><a href="#cb194-11" aria-hidden="true" tabindex="-1"></a><span class="do">##   [7] "Ipk"                      "GFPS"                    </span></span>
<span id="cb194-12"><a href="#cb194-12" aria-hidden="true" tabindex="-1"></a><span class="do">##   [9] "EgsA"                     "GGGP_synthase"           </span></span>
<span id="cb194-13"><a href="#cb194-13" aria-hidden="true" tabindex="-1"></a><span class="do">##  [11] "UbiA"                     "GGR"                     </span></span>
<span id="cb194-14"><a href="#cb194-14" aria-hidden="true" tabindex="-1"></a><span class="do">##  [13] "CarS"                     "PgsA"                    </span></span>
<span id="cb194-15"><a href="#cb194-15" aria-hidden="true" tabindex="-1"></a><span class="do">##  [15] "ASS"                      "Asd"                     </span></span>
<span id="cb194-16"><a href="#cb194-16" aria-hidden="true" tabindex="-1"></a><span class="do">##  [17] "Pol-B1"                   "Pol-B2"                  </span></span>
<span id="cb194-17"><a href="#cb194-17" aria-hidden="true" tabindex="-1"></a><span class="do">##  [19] "Pol-B3"                   "PolB-casposon-associated"</span></span>
<span id="cb194-18"><a href="#cb194-18" aria-hidden="true" tabindex="-1"></a><span class="do">##  [21] "PolD_small"               "PolD_large"              </span></span>
<span id="cb194-19"><a href="#cb194-19" aria-hidden="true" tabindex="-1"></a><span class="do">##  [23] "Pol4"                     "PolX"                    </span></span>
<span id="cb194-20"><a href="#cb194-20" aria-hidden="true" tabindex="-1"></a><span class="do">##  [25] "Orc1"                     "SSB/RPA1"                </span></span>
<span id="cb194-21"><a href="#cb194-21" aria-hidden="true" tabindex="-1"></a><span class="do">##  [27] "RPA2"                     "RPA3"                    </span></span>
<span id="cb194-22"><a href="#cb194-22" aria-hidden="true" tabindex="-1"></a><span class="do">##  [29] "GinS"                     "PriS"                    </span></span>
<span id="cb194-23"><a href="#cb194-23" aria-hidden="true" tabindex="-1"></a><span class="do">##  [31] "PriL"                     "DnaG"                    </span></span>
<span id="cb194-24"><a href="#cb194-24" aria-hidden="true" tabindex="-1"></a><span class="do">##  [33] "Lig "                     "LigN"                    </span></span>
<span id="cb194-25"><a href="#cb194-25" aria-hidden="true" tabindex="-1"></a><span class="do">##  [35] "Pcn1"                     "RfcS"                    </span></span>
<span id="cb194-26"><a href="#cb194-26" aria-hidden="true" tabindex="-1"></a><span class="do">##  [37] "RfcL"                     "FEN1"                    </span></span>
<span id="cb194-27"><a href="#cb194-27" aria-hidden="true" tabindex="-1"></a><span class="do">##  [39] "RnhA"                     "RnhB"                    </span></span>
<span id="cb194-28"><a href="#cb194-28" aria-hidden="true" tabindex="-1"></a><span class="do">##  [41] "MCM2"                     "MPH1"                    </span></span>
<span id="cb194-29"><a href="#cb194-29" aria-hidden="true" tabindex="-1"></a><span class="do">##  [43] "BRR2"                     "MUS81"                   </span></span>
<span id="cb194-30"><a href="#cb194-30" aria-hidden="true" tabindex="-1"></a><span class="do">##  [45] "SrmB"                     "Dna2"                    </span></span>
<span id="cb194-31"><a href="#cb194-31" aria-hidden="true" tabindex="-1"></a><span class="do">##  [47] "TopA"                     "GyrA"                    </span></span>
<span id="cb194-32"><a href="#cb194-32" aria-hidden="true" tabindex="-1"></a><span class="do">##  [49] "GyrB"                     "Rgy"                     </span></span>
<span id="cb194-33"><a href="#cb194-33" aria-hidden="true" tabindex="-1"></a><span class="do">##  [51] "TopB"                     "Top6A"                   </span></span>
<span id="cb194-34"><a href="#cb194-34" aria-hidden="true" tabindex="-1"></a><span class="do">##  [53] "Top6B"                    "RecJ/Cdc45"              </span></span>
<span id="cb194-35"><a href="#cb194-35" aria-hidden="true" tabindex="-1"></a><span class="do">##  [55] "SbcC"                     "SbcD"                    </span></span>
<span id="cb194-36"><a href="#cb194-36" aria-hidden="true" tabindex="-1"></a><span class="do">##  [57] "NurA"                     "Nth"                     </span></span>
<span id="cb194-37"><a href="#cb194-37" aria-hidden="true" tabindex="-1"></a><span class="do">##  [59] "Nfo"                      "Nfi"                     </span></span>
<span id="cb194-38"><a href="#cb194-38" aria-hidden="true" tabindex="-1"></a><span class="do">##  [61] "NucS"                     "XthA"                    </span></span>
<span id="cb194-39"><a href="#cb194-39" aria-hidden="true" tabindex="-1"></a><span class="do">##  [63] "DinG"                     "SSO0112"                 </span></span>
<span id="cb194-40"><a href="#cb194-40" aria-hidden="true" tabindex="-1"></a><span class="do">##  [65] "HerA"                     "RadA"                    </span></span>
<span id="cb194-41"><a href="#cb194-41" aria-hidden="true" tabindex="-1"></a><span class="do">##  [67] "RadB"                     "Udg1 "                   </span></span>
<span id="cb194-42"><a href="#cb194-42" aria-hidden="true" tabindex="-1"></a><span class="do">##  [69] "AlkD"                     "Ogt"                     </span></span>
<span id="cb194-43"><a href="#cb194-43" aria-hidden="true" tabindex="-1"></a><span class="do">##  [71] "Ogg"                      "hjc"                     </span></span>
<span id="cb194-44"><a href="#cb194-44" aria-hidden="true" tabindex="-1"></a><span class="do">##  [73] "UvrA"                     "UvrB"                    </span></span>
<span id="cb194-45"><a href="#cb194-45" aria-hidden="true" tabindex="-1"></a><span class="do">##  [75] "UvrC"                     "MthTI"                   </span></span>
<span id="cb194-46"><a href="#cb194-46" aria-hidden="true" tabindex="-1"></a><span class="do">##  [77] "MthZI"                    "MjaII"                   </span></span>
<span id="cb194-47"><a href="#cb194-47" aria-hidden="true" tabindex="-1"></a><span class="do">##  [79] "MjaIII"                   "MjaV"                    </span></span>
<span id="cb194-48"><a href="#cb194-48" aria-hidden="true" tabindex="-1"></a><span class="do">##  [81] "Dcm"                      "DNA_met"                 </span></span>
<span id="cb194-49"><a href="#cb194-49" aria-hidden="true" tabindex="-1"></a><span class="do">##  [83] "HmfAB"                    "HmvA"                    </span></span>
<span id="cb194-50"><a href="#cb194-50" aria-hidden="true" tabindex="-1"></a><span class="do">##  [85] "ScpA"                     "ScpB"                    </span></span>
<span id="cb194-51"><a href="#cb194-51" aria-hidden="true" tabindex="-1"></a><span class="do">##  [87] "MC1"                      "Smc"                     </span></span>
<span id="cb194-52"><a href="#cb194-52" aria-hidden="true" tabindex="-1"></a><span class="do">##  [89] "Alba1"                    "ELP3"                    </span></span>
<span id="cb194-53"><a href="#cb194-53" aria-hidden="true" tabindex="-1"></a><span class="do">##  [91] "FtsZ"                     "MinD"                    </span></span>
<span id="cb194-54"><a href="#cb194-54" aria-hidden="true" tabindex="-1"></a><span class="do">##  [93] "Cren-1"                   "actin_like"              </span></span>
<span id="cb194-55"><a href="#cb194-55" aria-hidden="true" tabindex="-1"></a><span class="do">##  [95] "Rkd1"                     "Rkd2"                    </span></span>
<span id="cb194-56"><a href="#cb194-56" aria-hidden="true" tabindex="-1"></a><span class="do">##  [97] "CetZ1"                    "CdvA"                    </span></span>
<span id="cb194-57"><a href="#cb194-57" aria-hidden="true" tabindex="-1"></a><span class="do">##  [99] "CdvB"                     "Vps4"                    </span></span>
<span id="cb194-58"><a href="#cb194-58" aria-hidden="true" tabindex="-1"></a><span class="do">## [101] "RpoC/Rpo3"                "RpoC/Rpo11"              </span></span>
<span id="cb194-59"><a href="#cb194-59" aria-hidden="true" tabindex="-1"></a><span class="do">## [103] "RpoB/Rpo2"                "RpoA/Rpo1"               </span></span>
<span id="cb194-60"><a href="#cb194-60" aria-hidden="true" tabindex="-1"></a><span class="do">## [105] "Rpo6/RpoZ"                "RPB11"                   </span></span>
<span id="cb194-61"><a href="#cb194-61" aria-hidden="true" tabindex="-1"></a><span class="do">## [107] "Rpo4"                     "RPB5"                    </span></span>
<span id="cb194-62"><a href="#cb194-62" aria-hidden="true" tabindex="-1"></a><span class="do">## [109] "RPB7"                     "RPB8"                    </span></span>
<span id="cb194-63"><a href="#cb194-63" aria-hidden="true" tabindex="-1"></a><span class="do">## [111] "RPB10"                    "RPC12"                   </span></span>
<span id="cb194-64"><a href="#cb194-64" aria-hidden="true" tabindex="-1"></a><span class="do">## [113] "Rpo13"                    "TBP"                     </span></span>
<span id="cb194-65"><a href="#cb194-65" aria-hidden="true" tabindex="-1"></a><span class="do">## [115] "TFB"                      "TFE"                     </span></span>
<span id="cb194-66"><a href="#cb194-66" aria-hidden="true" tabindex="-1"></a><span class="do">## [117] "TFS"                      "LysM"                    </span></span>
<span id="cb194-67"><a href="#cb194-67" aria-hidden="true" tabindex="-1"></a><span class="do">## [119] "Spt4"                     "NusG"                    </span></span>
<span id="cb194-68"><a href="#cb194-68" aria-hidden="true" tabindex="-1"></a><span class="do">## [121] "NusA"                     "EGD2"                    </span></span>
<span id="cb194-69"><a href="#cb194-69" aria-hidden="true" tabindex="-1"></a><span class="do">## [123] "L1"                       "L2"                      </span></span>
<span id="cb194-70"><a href="#cb194-70" aria-hidden="true" tabindex="-1"></a><span class="do">## [125] "L3"                       "L4"                      </span></span>
<span id="cb194-71"><a href="#cb194-71" aria-hidden="true" tabindex="-1"></a><span class="do">## [127] "L5"                       "L6P"                     </span></span>
<span id="cb194-72"><a href="#cb194-72" aria-hidden="true" tabindex="-1"></a><span class="do">## [129] "L7AE"                     "L10"                     </span></span>
<span id="cb194-73"><a href="#cb194-73" aria-hidden="true" tabindex="-1"></a><span class="do">## [131] "L10AE/L16"                "L11"                     </span></span>
<span id="cb194-74"><a href="#cb194-74" aria-hidden="true" tabindex="-1"></a><span class="do">## [133] "L12E/L44/L45/RPP1/RPP2"   "L13"                     </span></span>
<span id="cb194-75"><a href="#cb194-75" aria-hidden="true" tabindex="-1"></a><span class="do">## [135] "L13E"                     "L14"                     </span></span>
<span id="cb194-76"><a href="#cb194-76" aria-hidden="true" tabindex="-1"></a><span class="do">## [137] "L14E/L6E/L27E"            "L15"                     </span></span>
<span id="cb194-77"><a href="#cb194-77" aria-hidden="true" tabindex="-1"></a><span class="do">## [139] "L15E"                     "L18"                     </span></span>
<span id="cb194-78"><a href="#cb194-78" aria-hidden="true" tabindex="-1"></a><span class="do">## [141] "L18E"                     "L19E"                    </span></span>
<span id="cb194-79"><a href="#cb194-79" aria-hidden="true" tabindex="-1"></a><span class="do">## [143] "L20A"                     "L21E"                    </span></span>
<span id="cb194-80"><a href="#cb194-80" aria-hidden="true" tabindex="-1"></a><span class="do">## [145] "L22"                      "L23"                     </span></span>
<span id="cb194-81"><a href="#cb194-81" aria-hidden="true" tabindex="-1"></a><span class="do">## [147] "L24"                      "L24E"                    </span></span>
<span id="cb194-82"><a href="#cb194-82" aria-hidden="true" tabindex="-1"></a><span class="do">## [149] "L29"                      "L30"                     </span></span>
<span id="cb194-83"><a href="#cb194-83" aria-hidden="true" tabindex="-1"></a><span class="do">## [151] "L30E"                     "L31E"                    </span></span>
<span id="cb194-84"><a href="#cb194-84" aria-hidden="true" tabindex="-1"></a><span class="do">## [153] "L32E"                     "L34E"                    </span></span>
<span id="cb194-85"><a href="#cb194-85" aria-hidden="true" tabindex="-1"></a><span class="do">## [155] "L35AE/L33A"               "L37AE/L43A"              </span></span>
<span id="cb194-86"><a href="#cb194-86" aria-hidden="true" tabindex="-1"></a><span class="do">## [157] "L37E"                     "L38E"                    </span></span>
<span id="cb194-87"><a href="#cb194-87" aria-hidden="true" tabindex="-1"></a><span class="do">## [159] "L39E"                     "L40E"                    </span></span>
<span id="cb194-88"><a href="#cb194-88" aria-hidden="true" tabindex="-1"></a><span class="do">## [161] "L41E"                     "L44E"                    </span></span>
<span id="cb194-89"><a href="#cb194-89" aria-hidden="true" tabindex="-1"></a><span class="do">## [163] "S2"                       "S3"                      </span></span>
<span id="cb194-90"><a href="#cb194-90" aria-hidden="true" tabindex="-1"></a><span class="do">## [165] "S3AE"                     "S4"                      </span></span>
<span id="cb194-91"><a href="#cb194-91" aria-hidden="true" tabindex="-1"></a><span class="do">## [167] "S4E"                      "S5"                      </span></span>
<span id="cb194-92"><a href="#cb194-92" aria-hidden="true" tabindex="-1"></a><span class="do">## [169] "S6E/S10"                  "S7"                      </span></span>
<span id="cb194-93"><a href="#cb194-93" aria-hidden="true" tabindex="-1"></a><span class="do">## [171] "S8"                       "S8E"                     </span></span>
<span id="cb194-94"><a href="#cb194-94" aria-hidden="true" tabindex="-1"></a><span class="do">## [173] "S9"                       "S10"                     </span></span>
<span id="cb194-95"><a href="#cb194-95" aria-hidden="true" tabindex="-1"></a><span class="do">## [175] "S11"                      "S12"                     </span></span>
<span id="cb194-96"><a href="#cb194-96" aria-hidden="true" tabindex="-1"></a><span class="do">## [177] "S13"                      "S14"                     </span></span>
<span id="cb194-97"><a href="#cb194-97" aria-hidden="true" tabindex="-1"></a><span class="do">## [179] "S15P"                     "S17"                     </span></span>
<span id="cb194-98"><a href="#cb194-98" aria-hidden="true" tabindex="-1"></a><span class="do">## [181] "S17E"                     "S19"                     </span></span>
<span id="cb194-99"><a href="#cb194-99" aria-hidden="true" tabindex="-1"></a><span class="do">## [183] "S19E"                     "S24E"                    </span></span>
<span id="cb194-100"><a href="#cb194-100" aria-hidden="true" tabindex="-1"></a><span class="do">## [185] "S25"                      "S26"                     </span></span>
<span id="cb194-101"><a href="#cb194-101" aria-hidden="true" tabindex="-1"></a><span class="do">## [187] "S27AE"                    "S27E"                    </span></span>
<span id="cb194-102"><a href="#cb194-102" aria-hidden="true" tabindex="-1"></a><span class="do">## [189] "S28E/S33"                 "S30"                     </span></span>
<span id="cb194-103"><a href="#cb194-103" aria-hidden="true" tabindex="-1"></a><span class="do">## [191] "PelA"                     "RsgA"                    </span></span>
<span id="cb194-104"><a href="#cb194-104" aria-hidden="true" tabindex="-1"></a><span class="do">## [193] "HflX"                     "AlaS"                    </span></span>
<span id="cb194-105"><a href="#cb194-105" aria-hidden="true" tabindex="-1"></a><span class="do">## [195] "ArgS"                     "AsnS"                    </span></span>
<span id="cb194-106"><a href="#cb194-106" aria-hidden="true" tabindex="-1"></a><span class="do">## [197] "CysS"                     "GlnS"                    </span></span>
<span id="cb194-107"><a href="#cb194-107" aria-hidden="true" tabindex="-1"></a><span class="do">## [199] "GRS1"                     "HisS"                    </span></span>
<span id="cb194-108"><a href="#cb194-108" aria-hidden="true" tabindex="-1"></a><span class="do">## [201] "IleS"                     "LeuS"                    </span></span>
<span id="cb194-109"><a href="#cb194-109" aria-hidden="true" tabindex="-1"></a><span class="do">## [203] "LysS"                     "LysU"                    </span></span>
<span id="cb194-110"><a href="#cb194-110" aria-hidden="true" tabindex="-1"></a><span class="do">## [205] "MetG"                     "PheS"                    </span></span>
<span id="cb194-111"><a href="#cb194-111" aria-hidden="true" tabindex="-1"></a><span class="do">## [207] "SerS"                     "ProS"                    </span></span>
<span id="cb194-112"><a href="#cb194-112" aria-hidden="true" tabindex="-1"></a><span class="do">## [209] "ThrS"                     "TrpS"                    </span></span>
<span id="cb194-113"><a href="#cb194-113" aria-hidden="true" tabindex="-1"></a><span class="do">## [211] "TyrS"                     "ValS"                    </span></span>
<span id="cb194-114"><a href="#cb194-114" aria-hidden="true" tabindex="-1"></a><span class="do">## [213] "EIF1A"                    "InfB"                    </span></span>
<span id="cb194-115"><a href="#cb194-115" aria-hidden="true" tabindex="-1"></a><span class="do">## [215] "SUI1"                     "EIF2A"                   </span></span>
<span id="cb194-116"><a href="#cb194-116" aria-hidden="true" tabindex="-1"></a><span class="do">## [217] "EIF2B "                   "EIF2B-like"              </span></span>
<span id="cb194-117"><a href="#cb194-117" aria-hidden="true" tabindex="-1"></a><span class="do">## [219] "EIF2G"                    "EIF4A"                   </span></span>
<span id="cb194-118"><a href="#cb194-118" aria-hidden="true" tabindex="-1"></a><span class="do">## [221] "EIF5a"                    "EIF6"                    </span></span>
<span id="cb194-119"><a href="#cb194-119" aria-hidden="true" tabindex="-1"></a><span class="do">## [223] "Tuf"                      "EFB1"                    </span></span>
<span id="cb194-120"><a href="#cb194-120" aria-hidden="true" tabindex="-1"></a><span class="do">## [225] "FusA"                     "eRF1"                    </span></span>
<span id="cb194-121"><a href="#cb194-121" aria-hidden="true" tabindex="-1"></a><span class="do">## [227] "Rli1"                     "Csl4"                    </span></span>
<span id="cb194-122"><a href="#cb194-122" aria-hidden="true" tabindex="-1"></a><span class="do">## [229] "Rrp4"                     "Rrp41"                   </span></span>
<span id="cb194-123"><a href="#cb194-123" aria-hidden="true" tabindex="-1"></a><span class="do">## [231] "Rrp42"                    "Dph2"                    </span></span>
<span id="cb194-124"><a href="#cb194-124" aria-hidden="true" tabindex="-1"></a><span class="do">## [233] "Dph5"                     "Dph6"                    </span></span>
<span id="cb194-125"><a href="#cb194-125" aria-hidden="true" tabindex="-1"></a><span class="do">## [235] "Trm5"                     "TYW1"                    </span></span>
<span id="cb194-126"><a href="#cb194-126" aria-hidden="true" tabindex="-1"></a><span class="do">## [237] "TYW2"                     "TYW3"                    </span></span>
<span id="cb194-127"><a href="#cb194-127" aria-hidden="true" tabindex="-1"></a><span class="do">## [239] "EndA"                     "Fau1"                    </span></span>
<span id="cb194-128"><a href="#cb194-128" aria-hidden="true" tabindex="-1"></a><span class="do">## [241] "LigT"                     "Nob1"                    </span></span>
<span id="cb194-129"><a href="#cb194-129" aria-hidden="true" tabindex="-1"></a><span class="do">## [243] "RtcB"                     "Archaease"               </span></span>
<span id="cb194-130"><a href="#cb194-130" aria-hidden="true" tabindex="-1"></a><span class="do">## [245] "RCL1"                     "RnjA"                    </span></span>
<span id="cb194-131"><a href="#cb194-131" aria-hidden="true" tabindex="-1"></a><span class="do">## [247] "Rnz"                      "Rnp1"                    </span></span>
<span id="cb194-132"><a href="#cb194-132" aria-hidden="true" tabindex="-1"></a><span class="do">## [249] "Rnp2"                     "Rnp3"                    </span></span>
<span id="cb194-133"><a href="#cb194-133" aria-hidden="true" tabindex="-1"></a><span class="do">## [251] "Rnp4"                     "TgtA"                    </span></span>
<span id="cb194-134"><a href="#cb194-134" aria-hidden="true" tabindex="-1"></a><span class="do">## [253] "DnaJ"                     "DnaK"                    </span></span>
<span id="cb194-135"><a href="#cb194-135" aria-hidden="true" tabindex="-1"></a><span class="do">## [255] "GrpE"                     "GroL"                    </span></span>
<span id="cb194-136"><a href="#cb194-136" aria-hidden="true" tabindex="-1"></a><span class="do">## [257] "HtpX"                     "Hsp16"                   </span></span>
<span id="cb194-137"><a href="#cb194-137" aria-hidden="true" tabindex="-1"></a><span class="do">## [259] "Hsp19"                    "PsmA"                    </span></span>
<span id="cb194-138"><a href="#cb194-138" aria-hidden="true" tabindex="-1"></a><span class="do">## [261] "PRE1"                     "Pan1"                    </span></span>
<span id="cb194-139"><a href="#cb194-139" aria-hidden="true" tabindex="-1"></a><span class="do">## [263] "Pcm"                      "ThsA"</span></span>
<span id="cb194-140"><a href="#cb194-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-141"><a href="#cb194-141" aria-hidden="true" tabindex="-1"></a><span class="co">#define a order for metabolic pathways</span></span>
<span id="cb194-142"><a href="#cb194-142" aria-hidden="true" tabindex="-1"></a>Pathway_order <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(Genes_of_interest<span class="sc">$</span>pathway_2))</span>
<span id="cb194-143"><a href="#cb194-143" aria-hidden="true" tabindex="-1"></a>Pathway_order</span>
<span id="cb194-144"><a href="#cb194-144" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] "DNA_Polymerase"       "RFs"                  "Helicase"            </span></span>
<span id="cb194-145"><a href="#cb194-145" aria-hidden="true" tabindex="-1"></a><span class="do">##  [4] "Topoisomerase"        "Repair_Recombination" "Methylase"           </span></span>
<span id="cb194-146"><a href="#cb194-146" aria-hidden="true" tabindex="-1"></a><span class="do">##  [7] "Chromatin"            "Cell_cycle"           "RNA_Polymerase"      </span></span>
<span id="cb194-147"><a href="#cb194-147" aria-hidden="true" tabindex="-1"></a><span class="do">## [10] "TFs"                  "Transcription"        "Ribosome"            </span></span>
<span id="cb194-148"><a href="#cb194-148" aria-hidden="true" tabindex="-1"></a><span class="do">## [13] "tRNA_synthetase"      "Exosome"              "Diphthamide_BS"      </span></span>
<span id="cb194-149"><a href="#cb194-149" aria-hidden="true" tabindex="-1"></a><span class="do">## [16] "Wybutosine_BS"        "Translation"          "Posttranslation"     </span></span>
<span id="cb194-150"><a href="#cb194-150" aria-hidden="true" tabindex="-1"></a><span class="do">## [19] "Mevalonate"           "Lipids"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="deal-with-annotation-file" class="level2" data-number="1.30">
<h2 data-number="1.30" class="anchored" data-anchor-id="deal-with-annotation-file"><span class="header-section-number">1.30</span> Deal with annotation file</h2>
<section id="read-in-table" class="level3" data-number="1.30.1">
<h3 data-number="1.30.1" class="anchored" data-anchor-id="read-in-table"><span class="header-section-number">1.30.1</span> Read in table</h3>
<p>You already here notice that this takes a bit longer and we just work with 46 bins. This is a good reason to keep python in mind as depending on your computer the more memory heavy operations might get challenging. Another alternative would be to run R on the clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a><span class="co">#read in data and view it</span></span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a>Input <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../data/UAP2_Annotation_table_u.txt"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span>T, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb195-4"><a href="#cb195-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Input)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 4%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 4%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 5%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 3%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 3%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 6%">
<col style="width: 15%">
<col style="width: 1%">
<col style="width: 3%">
<col style="width: 7%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">accession</th>
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">TaxString</th>
<th style="text-align: left;">NewContigID</th>
<th style="text-align: left;">OldContigId</th>
<th style="text-align: left;">ContigIdMerge</th>
<th style="text-align: right;">ContigNewLength</th>
<th style="text-align: right;">GC</th>
<th style="text-align: left;">ProteinID</th>
<th style="text-align: right;">ProteinGC</th>
<th style="text-align: right;">ProteinLength</th>
<th style="text-align: left;">Prokka</th>
<th style="text-align: left;">arcogs</th>
<th style="text-align: left;">arcogs_geneID</th>
<th style="text-align: left;">arcogs_Description</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">arcogs_evalue</th>
<th style="text-align: left;">KO_hmm</th>
<th style="text-align: left;">e_value</th>
<th style="text-align: left;">bit_score</th>
<th style="text-align: left;">bit_score_cutoff</th>
<th style="text-align: left;">Definition</th>
<th style="text-align: left;">confidence</th>
<th style="text-align: left;">PFAM_hmm</th>
<th style="text-align: left;">PFAM_description</th>
<th style="text-align: left;">Pfam_Evalue</th>
<th style="text-align: left;">TIRGR</th>
<th style="text-align: left;">TIGR_description</th>
<th style="text-align: left;">EC</th>
<th style="text-align: left;">TIGR_Evalue</th>
<th style="text-align: left;">CAZy</th>
<th style="text-align: left;">CAZy_evalue</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">TDBD_ID</th>
<th style="text-align: left;">TPDB_evalue</th>
<th style="text-align: left;">HydDB</th>
<th style="text-align: left;">Description.1</th>
<th style="text-align: left;">HydDB_evalue</th>
<th style="text-align: left;">PFAM</th>
<th style="text-align: left;">PFAMdescription</th>
<th style="text-align: left;">IPR</th>
<th style="text-align: left;">IPRdescription</th>
<th style="text-align: left;">TopHit</th>
<th style="text-align: left;">E_value</th>
<th style="text-align: left;">PecID</th>
<th style="text-align: left;">TaxID</th>
<th style="text-align: left;">TaxString.1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00010</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00010</td>
<td style="text-align: right;">0.094</td>
<td style="text-align: right;">350</td>
<td style="text-align: left;">Digeranylgeranylglycerophospholipid reductase</td>
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">“Geranylgeranyl_reductase,_flavoprotein”</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">2.80E-75</td>
<td style="text-align: left;">K17830</td>
<td style="text-align: left;">2.10E-69</td>
<td style="text-align: left;">243.3</td>
<td style="text-align: left;">246.87</td>
<td style="text-align: left;">digeranylgeranylglycerophospholipid_reductase_[EC:1.3.1.101_1.3.7.11]</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">FAD_binding_domain</td>
<td style="text-align: left;">5.90E-09</td>
<td style="text-align: left;">TIGR02032</td>
<td style="text-align: left;">geranylgeranyl_reductase_family</td>
<td style="text-align: left;">1.3.1.-</td>
<td style="text-align: left;">1.40E-49</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR036188;_IPR002938</td>
<td style="text-align: left;">FAD/NAD(P)-binding_domain_superfamily;_FAD-binding_domain</td>
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">FAD_binding_domain</td>
<td style="text-align: left;">KYK22416.1_hypothetical_protein_AYK24_02275_[Thermoplasmatales_archaeon_SG8-52-4]</td>
<td style="text-align: left;">2.80E-58</td>
<td style="text-align: left;">36.5</td>
<td style="text-align: left;">1803819</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Thermoplasmata,Thermoplasmatales,none,none,Thermoplasmatales_archaeon_SG8-52-4”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00020</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00020</td>
<td style="text-align: right;">0.057</td>
<td style="text-align: right;">105</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00030</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_1</td>
<td style="text-align: left;">NIOZ119_sc1610046_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_1</td>
<td style="text-align: right;">2539</td>
<td style="text-align: right;">0.432</td>
<td style="text-align: left;">PJFDGLDN_00030</td>
<td style="text-align: right;">0.092</td>
<td style="text-align: right;">304</td>
<td style="text-align: left;">tRNA-2-methylthio-N(6)-dimethylallyladenosine synthase</td>
<td style="text-align: left;">arCOG01358</td>
<td style="text-align: left;">MiaB</td>
<td style="text-align: left;">2-methylthioadenine_synthetase</td>
<td style="text-align: left;">J</td>
<td style="text-align: left;">5.80E-90</td>
<td style="text-align: left;">K15865</td>
<td style="text-align: left;">1.20E-109</td>
<td style="text-align: left;">375.8</td>
<td style="text-align: left;">343.7</td>
<td style="text-align: left;">threonylcarbamoyladenosine_tRNA_methylthiotransferase_CDKAL1_[EC:2.8.4.5]</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF04055</td>
<td style="text-align: left;">Radical_SAM_superfamily</td>
<td style="text-align: left;">2.60E-21</td>
<td style="text-align: left;">TIGR01578</td>
<td style="text-align: left;">“MiaB-like_tRNA_modifying_enzyme,_archaeal-type”</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">1.70E-104</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR005839;_IPR002792;_IPR006466;_IPR006638;_IPR007197;_IPR020612;_IPR023404</td>
<td style="text-align: left;">“Methylthiotransferase;_TRAM_domain;_MiaB-like_tRNA_modifying_enzyme,_archaeal-type;_Elp3/MiaB/NifB;_Radical_SAM;_Methylthiotransferase,_conserved_site;_Radical_SAM,_alpha/beta_horseshoe”</td>
<td style="text-align: left;">PF01938;_PF04055</td>
<td style="text-align: left;">TRAM_domain;_Radical_SAM_superfamily</td>
<td style="text-align: left;">OIO63284.1_hypothetical_protein_AUJ83_01460_[Candidatus_Woesearchaeota_archaeon_CG1_02_33_12]</td>
<td style="text-align: left;">8.20E-91</td>
<td style="text-align: left;">55.6</td>
<td style="text-align: left;">1805422</td>
<td style="text-align: left;">“Archaea,Candidatus_Woesearchaeota,none,none,none,none,Candidatus_Woesearchaeota_archaeon_CG1_02_33_12”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00040</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00040</td>
<td style="text-align: right;">0.065</td>
<td style="text-align: right;">92</td>
<td style="text-align: left;">Enolase</td>
<td style="text-align: left;">arCOG01169</td>
<td style="text-align: left;">Eno</td>
<td style="text-align: left;">Enolase</td>
<td style="text-align: left;">G</td>
<td style="text-align: left;">4.00E-26</td>
<td style="text-align: left;">K01689</td>
<td style="text-align: left;">5.10E-20</td>
<td style="text-align: left;">81</td>
<td style="text-align: left;">48.73</td>
<td style="text-align: left;">enolase_[EC:4.2.1.11]</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain”</td>
<td style="text-align: left;">1.20E-20</td>
<td style="text-align: left;">TIGR01060</td>
<td style="text-align: left;">phosphopyruvate_hydratase</td>
<td style="text-align: left;">4.2.1.11</td>
<td style="text-align: left;">1.10E-25</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR020810;_IPR036849;_IPR000941;_IPR020809</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain;_Enolase-like,_C-terminal_domain_superfamily;_Enolase;_Enolase,_conserved_site”</td>
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain”</td>
<td style="text-align: left;">BAW30993.1_2-phosphoglycerate_dehydratase_[Methanothermobacter_sp._MT-2]</td>
<td style="text-align: left;">3.80E-23</td>
<td style="text-align: left;">70.7</td>
<td style="text-align: left;">1898379</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Methanobacteria,Methanobacteriales,Methanobacteriaceae,Methanothermobacter,Methanothermobacter_sp._MT-2”</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00050</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00050</td>
<td style="text-align: right;">0.059</td>
<td style="text-align: right;">218</td>
<td style="text-align: left;">30S ribosomal protein S2</td>
<td style="text-align: left;">arCOG04245</td>
<td style="text-align: left;">RpsB</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">J</td>
<td style="text-align: left;">8.90E-64</td>
<td style="text-align: left;">K02998</td>
<td style="text-align: left;">1.60E-46</td>
<td style="text-align: left;">167.8</td>
<td style="text-align: left;">210.97</td>
<td style="text-align: left;">small_subunit_ribosomal_protein_SAe</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">7.00E-24</td>
<td style="text-align: left;">TIGR01012</td>
<td style="text-align: left;">ribosomal_protein_uS2</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">1.00E-72</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR005707;_IPR023454;_IPR023591;_IPR018130;_IPR001865</td>
<td style="text-align: left;">“Ribosomal_protein_S2,_eukaryotic/archaeal;_Ribosomal_protein_S2,_archaeal;_Ribosomal_protein_S2,_flavodoxin-like_domain_superfamily;_Ribosomal_protein_S2,_conserved_site;_Ribosomal_protein_S2”</td>
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: left;">A0B6E5.1_RecName:_Full=30S_ribosomal_protein_S2</td>
<td style="text-align: left;">1.20E-56</td>
<td style="text-align: left;">52.8</td>
<td style="text-align: left;">349307</td>
<td style="text-align: left;">“Archaea,Euryarchaeota,Methanomicrobia,Methanosarcinales,Methanotrichaceae,Methanothrix,Methanothrix_thermoacetophila”</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00060</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">UAP2</td>
<td style="text-align: left;">PJFDGLDN_2</td>
<td style="text-align: left;">NIOZ119_sc560284_1</td>
<td style="text-align: left;">NIOZ119_mb_b5_2_contig_2</td>
<td style="text-align: right;">4191</td>
<td style="text-align: right;">0.456</td>
<td style="text-align: left;">PJFDGLDN_00060</td>
<td style="text-align: right;">0.100</td>
<td style="text-align: right;">280</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">arCOG01728</td>
<td style="text-align: left;">Mho1</td>
<td style="text-align: left;">“Predicted_class_III_extradiol_dioxygenase,_MEMO1_family”</td>
<td style="text-align: left;">R</td>
<td style="text-align: left;">8.80E-74</td>
<td style="text-align: left;">K06990</td>
<td style="text-align: left;">1.40E-72</td>
<td style="text-align: left;">253.3</td>
<td style="text-align: left;">47</td>
<td style="text-align: left;">MEMO1_family_protein</td>
<td style="text-align: left;">high_score</td>
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">Memo-like_protein</td>
<td style="text-align: left;">4.50E-77</td>
<td style="text-align: left;">TIGR04336</td>
<td style="text-align: left;">AmmeMemoRadiSam_system_protein_B</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">5.30E-91</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">IPR002737</td>
<td style="text-align: left;">MEMO1_family</td>
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">Memo-like_protein</td>
<td style="text-align: left;">OYT50994.1_AmmeMemoRadiSam_system_protein_B_[Candidatus_Bathyarchaeota_archaeon_ex4484_135]</td>
<td style="text-align: left;">1.30E-66</td>
<td style="text-align: left;">47</td>
<td style="text-align: left;">2012509</td>
<td style="text-align: left;">“Archaea,Candidatus_Bathyarchaeota,none,none,none,none,Candidatus_Bathyarchaeota_archaeon_ex4484_135”</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="make-a-mapping-file-that-links-all-annotation-ids-to-their-descriptions" class="level3" data-number="1.30.2">
<h3 data-number="1.30.2" class="anchored" data-anchor-id="make-a-mapping-file-that-links-all-annotation-ids-to-their-descriptions"><span class="header-section-number">1.30.2</span> Make a mapping file that links all annotation IDs to their descriptions</h3>
<p>What we do:</p>
<ul>
<li>separate columns we are interested in for each Database of interest, i.e.&nbsp;arCOGs, and remove duplicate rows by using <code>unique()</code></li>
<li>change the column names using <code>colnames()</code>. Here, we want to make sure that all the 6 new objects we generate have the same columns</li>
<li>combine our 6 dataframes using <code>rbind()</code>. For this to work we need the same number of columns.</li>
</ul>
<p>In theory that would be a nice example for a loop as well, since we do exactly the same thing for 6x.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb196"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb196-1"><a href="#cb196-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb196-2"><a href="#cb196-2" aria-hidden="true" tabindex="-1"></a><span class="co">#generate Description table for all DBs of interest</span></span>
<span id="cb196-3"><a href="#cb196-3" aria-hidden="true" tabindex="-1"></a>Arcogs_Description <span class="ot">&lt;-</span> <span class="fu">unique</span>(Input[,<span class="fu">c</span>(<span class="st">"arcogs"</span>,<span class="st">"arcogs_Description"</span> )])</span>
<span id="cb196-4"><a href="#cb196-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Arcogs_Description) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene"</span>, <span class="st">"Description"</span>)</span>
<span id="cb196-5"><a href="#cb196-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Arcogs_Description)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: left;">“Geranylgeranyl_reductase,_flavoprotein”</td>
</tr>
<tr class="even">
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG01358</td>
<td style="text-align: left;">2-methylthioadenine_synthetase</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG01169</td>
<td style="text-align: left;">Enolase</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG04245</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG01728</td>
<td style="text-align: left;">“Predicted_class_III_extradiol_dioxygenase,_MEMO1_family”</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb197"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb197-1"><a href="#cb197-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb197-2"><a href="#cb197-2" aria-hidden="true" tabindex="-1"></a>KOs_Description <span class="ot">&lt;-</span> Input[,<span class="fu">c</span>(<span class="st">"KO_hmm"</span>,<span class="st">"Definition"</span> )]</span>
<span id="cb197-3"><a href="#cb197-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(KOs_Description) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene"</span>, <span class="st">"Description"</span>)</span>
<span id="cb197-4"><a href="#cb197-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(KOs_Description)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 8%">
<col style="width: 91%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">K17830</td>
<td style="text-align: left;">digeranylgeranylglycerophospholipid_reductase_[EC:1.3.1.101_1.3.7.11]</td>
</tr>
<tr class="even">
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">K15865</td>
<td style="text-align: left;">threonylcarbamoyladenosine_tRNA_methylthiotransferase_CDKAL1_[EC:2.8.4.5]</td>
</tr>
<tr class="even">
<td style="text-align: left;">K01689</td>
<td style="text-align: left;">enolase_[EC:4.2.1.11]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">K02998</td>
<td style="text-align: left;">small_subunit_ribosomal_protein_SAe</td>
</tr>
<tr class="even">
<td style="text-align: left;">K06990</td>
<td style="text-align: left;">MEMO1_family_protein</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a>Pfam_Description <span class="ot">&lt;-</span> <span class="fu">unique</span>(Input[,<span class="fu">c</span>(<span class="st">"PFAM_hmm"</span>,<span class="st">"PFAM_description"</span> )])</span>
<span id="cb198-3"><a href="#cb198-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Pfam_Description) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene"</span>, <span class="st">"Description"</span>)</span>
<span id="cb198-4"><a href="#cb198-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Pfam_Description)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">FAD_binding_domain</td>
</tr>
<tr class="even">
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PF04055</td>
<td style="text-align: left;">Radical_SAM_superfamily</td>
</tr>
<tr class="even">
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">“Enolase,_C-terminal_TIM_barrel_domain”</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
</tr>
<tr class="even">
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">Memo-like_protein</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb199"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb199-1"><a href="#cb199-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-2"><a href="#cb199-2" aria-hidden="true" tabindex="-1"></a>TIRGR_Description <span class="ot">&lt;-</span> <span class="fu">unique</span>(Input[,<span class="fu">c</span>(<span class="st">"TIRGR"</span>,<span class="st">"TIGR_description"</span> )])</span>
<span id="cb199-3"><a href="#cb199-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TIRGR_Description) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene"</span>, <span class="st">"Description"</span>)</span>
<span id="cb199-4"><a href="#cb199-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(TIRGR_Description)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">TIGR02032</td>
<td style="text-align: left;">geranylgeranyl_reductase_family</td>
</tr>
<tr class="even">
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TIGR01578</td>
<td style="text-align: left;">“MiaB-like_tRNA_modifying_enzyme,_archaeal-type”</td>
</tr>
<tr class="even">
<td style="text-align: left;">TIGR01060</td>
<td style="text-align: left;">phosphopyruvate_hydratase</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TIGR01012</td>
<td style="text-align: left;">ribosomal_protein_uS2</td>
</tr>
<tr class="even">
<td style="text-align: left;">TIGR04336</td>
<td style="text-align: left;">AmmeMemoRadiSam_system_protein_B</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb200-2"><a href="#cb200-2" aria-hidden="true" tabindex="-1"></a>Cazy_Description <span class="ot">&lt;-</span> <span class="fu">unique</span>(Input[,<span class="fu">c</span>(<span class="st">"CAZy"</span>,<span class="st">"Description"</span> )])</span>
<span id="cb200-3"><a href="#cb200-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Cazy_Description) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene"</span>, <span class="st">"Description"</span>)</span>
<span id="cb200-4"><a href="#cb200-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Cazy_Description)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 98%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;">81</td>
<td style="text-align: left;">GH119</td>
<td style="text-align: left;">__alpha-amylase_(EC_3.2.1.1)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">245</td>
<td style="text-align: left;">GT4</td>
<td style="text-align: left;">“__sucrose_synthase_(EC_2.4.1.13);<em>sucrose-phosphate_synthase</em>(EC_2.4.1.14);<em>alpha-glucosyltransferase</em>(EC_2.4.1.52);<em>lipopolysaccharide_N-acetylglucosaminyltransferase</em>(EC_2.4.1.56);<em>phosphatidylinositol_alpha-mannosyltransferase</em>(EC_2.4.1.57);_GDP-Man:<em>Man1GlcNAc2-PP-dolichol_alpha-1,3-mannosyltransferase</em>(EC_2.4.1.132);_GDP-Man:<em>Man3GlcNAc2-PP-dolichol/Man4GlcNAc2-PP-dolichol_alpha-1,2-mannosyltransferase</em>(EC_2.4.1.131);<em>digalactosyldiacylglycerol_synthase</em>(EC_2.4.1.141);<em>1,2-diacylglycerol_3-glucosyltransferase</em>(EC_2.4.1.157);__diglucosyl_diacylglycerol_synthase_(EC_2.4.1.208);<em>trehalose_phosphorylase</em>(EC_2.4.1.231);_NDP-Glc:<em>alpha-glucose_alpha-glucosyltransferase</em>/<em>alpha,alpha-trehalose_synthase</em>(EC_2.4.1.245);_GDP-Man:<em>Man2GlcNAc2-PP-dolichol_alpha-1,6-mannosyltransferase</em>(EC_2.4.1.257);_UDP-GlcNAc:<em>2-deoxystreptamine_alpha-N-acetylglucosaminyltransferase</em>(EC_2.4.1.283);_UDP-GlcNAc:<em>ribostamycin_alpha-N-acetylglucosaminyltransferase</em>(EC_2.4.1.285);<em>UDP-Gal_alpha-galactosyltransferase</em>(EC_2.4.1.-);<em>UDP-Xyl_alpha-xylosyltransferase</em>(EC_2.4.2.-);<em>UDP-GlcA_alpha-glucuronyltransferase</em>(EC_2.4.1.-);<em>UDP-Glc_alpha-glucosyltransferase</em>(EC_2.4.1.-);_UDP-GalNAc:<em>GalNAc-PP-Und_alpha-1,3-N-acetylgalactosaminyltransferase</em>(EC_2.4.1.306);_UDP-GalNAc:<em>N,N’-diacetylbacillosaminyl-PP-Und_alpha-1,3-N-acetylgalactosaminyltransferase</em>(EC_2.4.1.290);<em>ADP-dependent_alpha-maltose-1-phosphate_synthase</em>(2.4.1.342)”</td>
</tr>
<tr class="even">
<td style="text-align: left;">291</td>
<td style="text-align: left;">GT2_Glycos_transf_2</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">613</td>
<td style="text-align: left;">GT83</td>
<td style="text-align: left;">__undecaprenyl_phosphate-alpha-L-Ara4N:<em>4-amino-4-deoxy-beta-L-arabinosyltransferase</em>(EC_2.4.2.43);_dodecaprenyl_phosphate-beta-galacturonic_acid:<em>lipopolysaccharide_core_alpha-galacturonosyl_transferase</em>(EC_2.4.1.-)</td>
</tr>
<tr class="even">
<td style="text-align: left;">1357</td>
<td style="text-align: left;">GT66</td>
<td style="text-align: left;">__dolichyl-diphosphooligosaccharide<c3>&lt;83&gt;<c2><a2><c3><a2><e2>&lt;80&gt;&lt;9a&gt;<c2><ac><c3><a2><e2>&lt;82&gt;<ac><c2>&lt;9d&gt;protein_glycotransferase_(EC_2.4.99.18);<em>undecaprenyl-diphosphooligosaccharide<c3>&lt;83&gt;<c2><a2><c3><a2><e2>&lt;80&gt;&lt;9a&gt;<c2><ac><c3><a2><e2>&lt;82&gt;<ac><c2>&lt;9d&gt;protein_glycotransferase</c2></ac></e2></a2></c3></ac></c2></e2></a2></c3></a2></c2></c3></em>(EC_2.4.99.19)_</c2></ac></e2></a2></c3></ac></c2></e2></a2></c3></a2></c2></c3></td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb201"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb201-1"><a href="#cb201-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb201-2"><a href="#cb201-2" aria-hidden="true" tabindex="-1"></a>HydDB_Description <span class="ot">&lt;-</span> <span class="fu">unique</span>(Input[,<span class="fu">c</span>(<span class="st">"Description.1"</span>,<span class="st">"Description.1"</span> )])</span>
<span id="cb201-3"><a href="#cb201-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HydDB_Description) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene"</span>, <span class="st">"Description"</span>)</span>
<span id="cb201-4"><a href="#cb201-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(HydDB_Description)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;">4787</td>
<td style="text-align: left;">[FeFe]_Group_A1</td>
<td style="text-align: left;">[FeFe]_Group_A1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">18935</td>
<td style="text-align: left;">[NiFe]_Group_4g</td>
<td style="text-align: left;">[NiFe]_Group_4g</td>
</tr>
<tr class="even">
<td style="text-align: left;">19355</td>
<td style="text-align: left;">[NiFe]_Group_3b</td>
<td style="text-align: left;">[NiFe]_Group_3b</td>
</tr>
<tr class="odd">
<td style="text-align: left;">37081</td>
<td style="text-align: left;">[FeFe]_Group_C3</td>
<td style="text-align: left;">[FeFe]_Group_C3</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb202"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb202-1"><a href="#cb202-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb202-2"><a href="#cb202-2" aria-hidden="true" tabindex="-1"></a><span class="co">#make a file with a description of all the ids for each search</span></span>
<span id="cb202-3"><a href="#cb202-3" aria-hidden="true" tabindex="-1"></a>All_Genes_Description <span class="ot">&lt;-</span> <span class="fu">rbind</span>(Arcogs_Description,KOs_Description,Pfam_Description,TIRGR_Description, Cazy_Description, HydDB_Description)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="parse-table-to-make-it-easier-to-work-with-it" class="level3" data-number="1.30.3">
<h3 data-number="1.30.3" class="anchored" data-anchor-id="parse-table-to-make-it-easier-to-work-with-it"><span class="header-section-number">1.30.3</span> Parse table to make it easier to work with it</h3>
<p>Here we:</p>
<ul>
<li>Subset the data for the columns we are interested in. Esp. for larger dataframes this will make the operations a bit quicker. For very large dataframes, i.e.&nbsp;5000 genomes, it might be better to switch to python</li>
<li>Convert data from wide to long format</li>
<li>Clean factors. After subsetting often factors are not removed, we clean them up in that step</li>
</ul>
<p>Info:</p>
<p><strong>Converting a wide to a long dataframe</strong></p>
<ul>
<li>Wide dataframe: The Input data in this example is considered as a wide dataframe. I.e. all the gene IDs we are interested in are spread out into different columns</li>
<li>Long dataframe: The gene IDs we are interested in are found all in the same column. Important, most R functions work with long dataframes.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb203"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb203-2"><a href="#cb203-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print the column names to subset our datatable</span></span>
<span id="cb203-3"><a href="#cb203-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Input)</span>
<span id="cb203-4"><a href="#cb203-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] "accession"          "BinID"              "TaxString"         </span></span>
<span id="cb203-5"><a href="#cb203-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  [4] "NewContigID"        "OldContigId"        "ContigIdMerge"     </span></span>
<span id="cb203-6"><a href="#cb203-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  [7] "ContigNewLength"    "GC"                 "ProteinID"         </span></span>
<span id="cb203-7"><a href="#cb203-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [10] "ProteinGC"          "ProteinLength"      "Prokka"            </span></span>
<span id="cb203-8"><a href="#cb203-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [13] "arcogs"             "arcogs_geneID"      "arcogs_Description"</span></span>
<span id="cb203-9"><a href="#cb203-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [16] "Pathway"            "arcogs_evalue"      "KO_hmm"            </span></span>
<span id="cb203-10"><a href="#cb203-10" aria-hidden="true" tabindex="-1"></a><span class="do">## [19] "e_value"            "bit_score"          "bit_score_cutoff"  </span></span>
<span id="cb203-11"><a href="#cb203-11" aria-hidden="true" tabindex="-1"></a><span class="do">## [22] "Definition"         "confidence"         "PFAM_hmm"          </span></span>
<span id="cb203-12"><a href="#cb203-12" aria-hidden="true" tabindex="-1"></a><span class="do">## [25] "PFAM_description"   "Pfam_Evalue"        "TIRGR"             </span></span>
<span id="cb203-13"><a href="#cb203-13" aria-hidden="true" tabindex="-1"></a><span class="do">## [28] "TIGR_description"   "EC"                 "TIGR_Evalue"       </span></span>
<span id="cb203-14"><a href="#cb203-14" aria-hidden="true" tabindex="-1"></a><span class="do">## [31] "CAZy"               "CAZy_evalue"        "Description"       </span></span>
<span id="cb203-15"><a href="#cb203-15" aria-hidden="true" tabindex="-1"></a><span class="do">## [34] "TDBD_ID"            "TPDB_evalue"        "HydDB"             </span></span>
<span id="cb203-16"><a href="#cb203-16" aria-hidden="true" tabindex="-1"></a><span class="do">## [37] "Description.1"      "HydDB_evalue"       "PFAM"              </span></span>
<span id="cb203-17"><a href="#cb203-17" aria-hidden="true" tabindex="-1"></a><span class="do">## [40] "PFAMdescription"    "IPR"                "IPRdescription"    </span></span>
<span id="cb203-18"><a href="#cb203-18" aria-hidden="true" tabindex="-1"></a><span class="do">## [43] "TopHit"             "E_value"            "PecID"             </span></span>
<span id="cb203-19"><a href="#cb203-19" aria-hidden="true" tabindex="-1"></a><span class="do">## [46] "TaxID"              "TaxString.1"</span></span>
<span id="cb203-20"><a href="#cb203-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb203-21"><a href="#cb203-21" aria-hidden="true" tabindex="-1"></a><span class="co">#only keep the columns we actually want to work with</span></span>
<span id="cb203-22"><a href="#cb203-22" aria-hidden="true" tabindex="-1"></a>Input_subset <span class="ot">=</span> Input[,<span class="fu">c</span>(<span class="st">'BinID'</span>,<span class="st">'accession'</span>,<span class="st">'arcogs'</span>,<span class="st">'KO_hmm'</span>,<span class="st">'PFAM_hmm'</span>,<span class="st">'TIRGR'</span>,<span class="st">'CAZy'</span>,<span class="st">'Description.1'</span> )]</span>
<span id="cb203-23"><a href="#cb203-23" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Input_subset)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 15%">
<col style="width: 30%">
<col style="width: 10%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 9%">
<col style="width: 4%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">accession</th>
<th style="text-align: left;">arcogs</th>
<th style="text-align: left;">KO_hmm</th>
<th style="text-align: left;">PFAM_hmm</th>
<th style="text-align: left;">TIRGR</th>
<th style="text-align: left;">CAZy</th>
<th style="text-align: left;">Description.1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00010</td>
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: left;">K17830</td>
<td style="text-align: left;">PF01494</td>
<td style="text-align: left;">TIGR02032</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00020</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00030</td>
<td style="text-align: left;">arCOG01358</td>
<td style="text-align: left;">K15865</td>
<td style="text-align: left;">PF04055</td>
<td style="text-align: left;">TIGR01578</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00040</td>
<td style="text-align: left;">arCOG01169</td>
<td style="text-align: left;">K01689</td>
<td style="text-align: left;">PF00113</td>
<td style="text-align: left;">TIGR01060</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00050</td>
<td style="text-align: left;">arCOG04245</td>
<td style="text-align: left;">K02998</td>
<td style="text-align: left;">PF00318</td>
<td style="text-align: left;">TIGR01012</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00060</td>
<td style="text-align: left;">arCOG01728</td>
<td style="text-align: left;">K06990</td>
<td style="text-align: left;">PF01875</td>
<td style="text-align: left;">TIGR04336</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">-</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb204"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb204-1"><a href="#cb204-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb204-2"><a href="#cb204-2" aria-hidden="true" tabindex="-1"></a><span class="co">#convert dataframe from wide to long</span></span>
<span id="cb204-3"><a href="#cb204-3" aria-hidden="true" tabindex="-1"></a>Input_long <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(Input_subset,  <span class="at">id=</span><span class="fu">c</span>(<span class="st">"accession"</span>,<span class="st">"BinID"</span>))</span>
<span id="cb204-4"><a href="#cb204-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb204-5"><a href="#cb204-5" aria-hidden="true" tabindex="-1"></a><span class="co">#give informative headers</span></span>
<span id="cb204-6"><a href="#cb204-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Input_long) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"accession"</span>, <span class="st">"BinID"</span>, <span class="st">"DB"</span>, <span class="st">"gene"</span>)</span>
<span id="cb204-7"><a href="#cb204-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb204-8"><a href="#cb204-8" aria-hidden="true" tabindex="-1"></a><span class="co">#clean factors, to remove issues when counting</span></span>
<span id="cb204-9"><a href="#cb204-9" aria-hidden="true" tabindex="-1"></a>Input_long<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(Input_long<span class="sc">$</span>gene)</span>
<span id="cb204-10"><a href="#cb204-10" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Input_long)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">accession</th>
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">DB</th>
<th style="text-align: left;">gene</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00010</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">arcogs</td>
<td style="text-align: left;">arCOG00570</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00020</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">arcogs</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00030</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">arcogs</td>
<td style="text-align: left;">arCOG01358</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00040</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">arcogs</td>
<td style="text-align: left;">arCOG01169</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00050</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">arcogs</td>
<td style="text-align: left;">arCOG04245</td>
</tr>
<tr class="even">
<td style="text-align: left;">NIOZ119_mb_b5_2-PJFDGLDN_00060</td>
<td style="text-align: left;">NIOZ119_mb_b5_2</td>
<td style="text-align: left;">arcogs</td>
<td style="text-align: left;">arCOG01728</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="make-count-tables" class="level2" data-number="1.31">
<h2 data-number="1.31" class="anchored" data-anchor-id="make-count-tables"><span class="header-section-number">1.31</span> Make count tables</h2>
<section id="generate-a-count-table-for-our-genomes-of-interest" class="level3" data-number="1.31.1">
<h3 data-number="1.31.1" class="anchored" data-anchor-id="generate-a-count-table-for-our-genomes-of-interest"><span class="header-section-number">1.31.1</span> Generate a count table for our genomes of interest</h3>
<p>Now we want to count, how often does a genome (i.e.&nbsp;NIOZ134_mb_b41_2) have a gene. I.e. how often do we want arCOG00570, arCOG01358, …</p>
<section id="do-this-via-a-loop-not-executed-just-an-example" class="level5" data-number="1.31.1.0.1">
<h5 data-number="1.31.1.0.1" class="anchored" data-anchor-id="do-this-via-a-loop-not-executed-just-an-example"><span class="header-section-number">1.31.1.0.1</span> Do this via a loop (not executed, just an example)</h5>
<p>Notice:</p>
<p>Since we run this chunk with <code>, eval = FALSE</code> we can still see the code but it is not executed. This is done because some computations take some time, which we do not want to spend, but I still want to show the code to give some alternative examples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb205"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb205-1"><a href="#cb205-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb205-2"><a href="#cb205-2" aria-hidden="true" tabindex="-1"></a><span class="co">#count the number of proteins for each genome of interest</span></span>
<span id="cb205-3"><a href="#cb205-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb205-4"><a href="#cb205-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> Bin_order) {</span>
<span id="cb205-5"><a href="#cb205-5" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span>  <span class="fu">table</span>(<span class="fu">subset</span>(Input_long, BinID <span class="sc">%in%</span> <span class="fu">paste</span>(i))<span class="sc">$</span>gene)</span>
<span id="cb205-6"><a href="#cb205-6" aria-hidden="true" tabindex="-1"></a>  y <span class="ot">&lt;-</span> <span class="fu">cbind</span> (y,x)</span>
<span id="cb205-7"><a href="#cb205-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb205-8"><a href="#cb205-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb205-9"><a href="#cb205-9" aria-hidden="true" tabindex="-1"></a><span class="co">#clean-up the table</span></span>
<span id="cb205-10"><a href="#cb205-10" aria-hidden="true" tabindex="-1"></a>Counts_Table_loop <span class="ot">&lt;-</span> y</span>
<span id="cb205-11"><a href="#cb205-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Counts_Table_loop) <span class="ot">&lt;-</span> Bin_order</span>
<span id="cb205-12"><a href="#cb205-12" aria-hidden="true" tabindex="-1"></a>Counts_Table_loop <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(Counts_Table_loop)</span>
<span id="cb205-13"><a href="#cb205-13" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_loop)), <span class="at">format=</span><span class="st">'markdown'</span>)</span>
<span id="cb205-14"><a href="#cb205-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb205-15"><a href="#cb205-15" aria-hidden="true" tabindex="-1"></a><span class="co">#the '-' (=not identified genes) is also counted and listed in the first column and removed at this step</span></span>
<span id="cb205-16"><a href="#cb205-16" aria-hidden="true" tabindex="-1"></a>Counts_Table_loop <span class="ot">&lt;-</span> Counts_Table_loop[<span class="sc">-</span><span class="dv">1</span>,]</span>
<span id="cb205-17"><a href="#cb205-17" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_loop)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="do-this-via-ddply-not-executed-just-an-example" class="level5" data-number="1.31.1.0.2">
<h5 data-number="1.31.1.0.2" class="anchored" data-anchor-id="do-this-via-ddply-not-executed-just-an-example"><span class="header-section-number">1.31.1.0.2</span> Do this via ddply ((not executed, just an example))</h5>
<p>New functions:</p>
<ul>
<li><code>spread()</code> = converts our long to a wide dataframe by using the BinIDs as new column names, the count table as values to populate our dataframe and with missing values we print a 0.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb206"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-2"><a href="#cb206-2" aria-hidden="true" tabindex="-1"></a><span class="co">#count data and clean header</span></span>
<span id="cb206-3"><a href="#cb206-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long <span class="ot">&lt;-</span> <span class="fu">ddply</span>(Input_long, .(BinID, gene), summarize, <span class="at">GeneCount =</span> <span class="fu">length</span>(gene))</span>
<span id="cb206-4"><a href="#cb206-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Counts_Table_long) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BinID"</span>, <span class="st">"geneID"</span>, <span class="st">"count"</span>)</span>
<span id="cb206-5"><a href="#cb206-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long)), <span class="at">format=</span><span class="st">'markdown'</span>)</span>
<span id="cb206-6"><a href="#cb206-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-7"><a href="#cb206-7" aria-hidden="true" tabindex="-1"></a><span class="co">#transform to wide format, with fill = 0 instead of a NA we add a 0</span></span>
<span id="cb206-8"><a href="#cb206-8" aria-hidden="true" tabindex="-1"></a>Counts_Table_wide <span class="ot">&lt;-</span> <span class="fu">spread</span>(Counts_Table_long, BinID, count, <span class="at">fill =</span> <span class="dv">0</span> )</span>
<span id="cb206-9"><a href="#cb206-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb206-10"><a href="#cb206-10" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb206-11"><a href="#cb206-11" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="do-this-via-tidyr-usually-a-bit-faster-than-ddplyr-which-is-why-we-use-this-way" class="level5" data-number="1.31.1.0.3">
<h5 data-number="1.31.1.0.3" class="anchored" data-anchor-id="do-this-via-tidyr-usually-a-bit-faster-than-ddplyr-which-is-why-we-use-this-way"><span class="header-section-number">1.31.1.0.3</span> Do this via tidyr (usually a bit faster than ddplyr, which is why we use this way)</h5>
<p>Here, we use the <code>%&gt;%</code> symbol again: In the following example, the a subset of the Input_long data (only 3 columns, not the whole dataframe) gets passed to <code>count()</code></p>
<p>New functions:</p>
<ul>
<li><code>count()</code> = A function of the dplyr package. Here, we count the unique protein IDs grouped by BinID and gene (i.e.&nbsp;roughly equivalent to the columns we want to keep)</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb207"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb207-1"><a href="#cb207-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb207-2"><a href="#cb207-2" aria-hidden="true" tabindex="-1"></a><span class="co">#count data and clean header</span></span>
<span id="cb207-3"><a href="#cb207-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long <span class="ot">&lt;-</span> Input_long[,<span class="fu">c</span>(<span class="st">'accession'</span>, <span class="st">'BinID'</span>,<span class="st">'gene'</span>)] <span class="sc">%&gt;%</span> <span class="fu">count</span>(BinID, gene, <span class="at">sort =</span> <span class="cn">FALSE</span>)</span>
<span id="cb207-4"><a href="#cb207-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Counts_Table_long) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BinID"</span>, <span class="st">"geneID"</span>, <span class="st">"count"</span>)</span>
<span id="cb207-5"><a href="#cb207-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">-</td>
<td style="text-align: right;">2350</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glycos_transf_2</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT4</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT66</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT83</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>When viewing the data we also see that proteins with no annotations are counted (the minus symbol), since we do not care about this at this stage, lets remove everything with a minus symbol</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb208"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb208-1"><a href="#cb208-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb208-2"><a href="#cb208-2" aria-hidden="true" tabindex="-1"></a><span class="co">#delete rows with a minus symbol</span></span>
<span id="cb208-3"><a href="#cb208-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long <span class="ot">&lt;-</span> Counts_Table_long[Counts_Table_long<span class="sc">$</span>geneID<span class="sc">!=</span> <span class="st">"-"</span>, ]</span>
<span id="cb208-4"><a href="#cb208-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb208-5"><a href="#cb208-5" aria-hidden="true" tabindex="-1"></a><span class="co">#clean factors</span></span>
<span id="cb208-6"><a href="#cb208-6" aria-hidden="true" tabindex="-1"></a>Counts_Table_long<span class="sc">$</span>geneID <span class="ot">&lt;-</span> <span class="fu">factor</span>(Counts_Table_long<span class="sc">$</span>geneID)</span>
<span id="cb208-7"><a href="#cb208-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb208-8"><a href="#cb208-8" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb208-9"><a href="#cb208-9" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">2</td>
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">3</td>
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glycos_transf_2</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">4</td>
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT4</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">5</td>
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT66</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">6</td>
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT83</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">7</td>
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">K00012</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now, we can convert the long to a wide table, since this format is a bit easier to read in excel later.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb209"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb209-1"><a href="#cb209-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb209-2"><a href="#cb209-2" aria-hidden="true" tabindex="-1"></a><span class="co">#transform to wide format, with fill = 0 instead of a NA we add a 0</span></span>
<span id="cb209-3"><a href="#cb209-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_wide <span class="ot">&lt;-</span> <span class="fu">spread</span>(Counts_Table_long, BinID, count, <span class="at">fill =</span> <span class="dv">0</span> )</span>
<span id="cb209-4"><a href="#cb209-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 0%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">GCA_002494525</th>
<th style="text-align: right;">GCA_002495465</th>
<th style="text-align: right;">GCA_002501985</th>
<th style="text-align: right;">GCA_002502135</th>
<th style="text-align: right;">NIOZ119_mb_b5_2</th>
<th style="text-align: right;">NIOZ120_mb_b229_2</th>
<th style="text-align: right;">NIOZ121_cc_b94_2</th>
<th style="text-align: right;">NIOZ121_mb_b48_2</th>
<th style="text-align: right;">NIOZ122_mb_b305_2</th>
<th style="text-align: right;">NIOZ122_mb_b319_2</th>
<th style="text-align: right;">NIOZ123_bs_b392_2</th>
<th style="text-align: right;">NIOZ123_bs_b5_2</th>
<th style="text-align: right;">NIOZ124_cc_b247_2</th>
<th style="text-align: right;">NIOZ124_mb_b130_2</th>
<th style="text-align: right;">NIOZ125_cc_b75_2</th>
<th style="text-align: right;">NIOZ125_mb_b178_2</th>
<th style="text-align: right;">NIOZ125_mb_b254_2</th>
<th style="text-align: right;">NIOZ126_mb_b137_2</th>
<th style="text-align: right;">NIOZ126_mb_b304_2</th>
<th style="text-align: right;">NIOZ126_mb_b401_2</th>
<th style="text-align: right;">NIOZ127_mb_b161_2</th>
<th style="text-align: right;">NIOZ127_mb_b282_2</th>
<th style="text-align: right;">NIOZ129_mb_b254_2</th>
<th style="text-align: right;">NIOZ129_mb_b262_2</th>
<th style="text-align: right;">NIOZ132_cc_b149_2</th>
<th style="text-align: right;">NIOZ132_cc_b51_2</th>
<th style="text-align: right;">NIOZ132_mb_b198_2</th>
<th style="text-align: right;">NIOZ132_mb_b260_2</th>
<th style="text-align: right;">NIOZ132_mb_b282_2</th>
<th style="text-align: right;">NIOZ132_mx_b314_2</th>
<th style="text-align: right;">NIOZ134_cc_b149_2</th>
<th style="text-align: right;">NIOZ134_mb_b293_2</th>
<th style="text-align: right;">NIOZ134_mb_b361_2</th>
<th style="text-align: right;">NIOZ134_mb_b41_2</th>
<th style="text-align: right;">NIOZ134_mx_b281_2</th>
<th style="text-align: right;">NIOZ136_cc_b15_2</th>
<th style="text-align: right;">NIOZ136_mb_b104_2</th>
<th style="text-align: right;">NIOZ136_mb_b226_2</th>
<th style="text-align: right;">NIOZ136_mb_b335_2</th>
<th style="text-align: right;">SRR2090153_bin1042</th>
<th style="text-align: right;">SRR2090153_bin461</th>
<th style="text-align: right;">SRR2090159_bin1129</th>
<th style="text-align: right;">SRR2090159_bin1288</th>
<th style="text-align: right;">SRR4028224_bin17</th>
<th style="text-align: right;">SRR5007147_bin71</th>
<th style="text-align: right;">U_76725</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Also, we want our geneIDs to be the new rownames and we do this by using the <code>rownames()</code> functions. We do this since some functions do not like to have characters in their dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb210"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb210-1"><a href="#cb210-1" aria-hidden="true" tabindex="-1"></a><span class="co">#change the rownames</span></span>
<span id="cb210-2"><a href="#cb210-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Counts_Table_wide) <span class="ot">&lt;-</span> Counts_Table_wide<span class="sc">$</span>geneID</span>
<span id="cb210-3"><a href="#cb210-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb210-4"><a href="#cb210-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb210-5"><a href="#cb210-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 0%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">GCA_002494525</th>
<th style="text-align: right;">GCA_002495465</th>
<th style="text-align: right;">GCA_002501985</th>
<th style="text-align: right;">GCA_002502135</th>
<th style="text-align: right;">NIOZ119_mb_b5_2</th>
<th style="text-align: right;">NIOZ120_mb_b229_2</th>
<th style="text-align: right;">NIOZ121_cc_b94_2</th>
<th style="text-align: right;">NIOZ121_mb_b48_2</th>
<th style="text-align: right;">NIOZ122_mb_b305_2</th>
<th style="text-align: right;">NIOZ122_mb_b319_2</th>
<th style="text-align: right;">NIOZ123_bs_b392_2</th>
<th style="text-align: right;">NIOZ123_bs_b5_2</th>
<th style="text-align: right;">NIOZ124_cc_b247_2</th>
<th style="text-align: right;">NIOZ124_mb_b130_2</th>
<th style="text-align: right;">NIOZ125_cc_b75_2</th>
<th style="text-align: right;">NIOZ125_mb_b178_2</th>
<th style="text-align: right;">NIOZ125_mb_b254_2</th>
<th style="text-align: right;">NIOZ126_mb_b137_2</th>
<th style="text-align: right;">NIOZ126_mb_b304_2</th>
<th style="text-align: right;">NIOZ126_mb_b401_2</th>
<th style="text-align: right;">NIOZ127_mb_b161_2</th>
<th style="text-align: right;">NIOZ127_mb_b282_2</th>
<th style="text-align: right;">NIOZ129_mb_b254_2</th>
<th style="text-align: right;">NIOZ129_mb_b262_2</th>
<th style="text-align: right;">NIOZ132_cc_b149_2</th>
<th style="text-align: right;">NIOZ132_cc_b51_2</th>
<th style="text-align: right;">NIOZ132_mb_b198_2</th>
<th style="text-align: right;">NIOZ132_mb_b260_2</th>
<th style="text-align: right;">NIOZ132_mb_b282_2</th>
<th style="text-align: right;">NIOZ132_mx_b314_2</th>
<th style="text-align: right;">NIOZ134_cc_b149_2</th>
<th style="text-align: right;">NIOZ134_mb_b293_2</th>
<th style="text-align: right;">NIOZ134_mb_b361_2</th>
<th style="text-align: right;">NIOZ134_mb_b41_2</th>
<th style="text-align: right;">NIOZ134_mx_b281_2</th>
<th style="text-align: right;">NIOZ136_cc_b15_2</th>
<th style="text-align: right;">NIOZ136_mb_b104_2</th>
<th style="text-align: right;">NIOZ136_mb_b226_2</th>
<th style="text-align: right;">NIOZ136_mb_b335_2</th>
<th style="text-align: right;">SRR2090153_bin1042</th>
<th style="text-align: right;">SRR2090153_bin461</th>
<th style="text-align: right;">SRR2090159_bin1129</th>
<th style="text-align: right;">SRR2090159_bin1288</th>
<th style="text-align: right;">SRR4028224_bin17</th>
<th style="text-align: right;">SRR5007147_bin71</th>
<th style="text-align: right;">U_76725</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>When we change the rownames and view the data, we see that the geneID is now both in the rownames as well as the first column. Since that is a bit messy, we next remove the first column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb211"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb211-1"><a href="#cb211-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb211-2"><a href="#cb211-2" aria-hidden="true" tabindex="-1"></a><span class="co">#delete the first column</span></span>
<span id="cb211-3"><a href="#cb211-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_wide <span class="ot">&lt;-</span> Counts_Table_wide[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb211-4"><a href="#cb211-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 0%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">GCA_002494525</th>
<th style="text-align: right;">GCA_002495465</th>
<th style="text-align: right;">GCA_002501985</th>
<th style="text-align: right;">GCA_002502135</th>
<th style="text-align: right;">NIOZ119_mb_b5_2</th>
<th style="text-align: right;">NIOZ120_mb_b229_2</th>
<th style="text-align: right;">NIOZ121_cc_b94_2</th>
<th style="text-align: right;">NIOZ121_mb_b48_2</th>
<th style="text-align: right;">NIOZ122_mb_b305_2</th>
<th style="text-align: right;">NIOZ122_mb_b319_2</th>
<th style="text-align: right;">NIOZ123_bs_b392_2</th>
<th style="text-align: right;">NIOZ123_bs_b5_2</th>
<th style="text-align: right;">NIOZ124_cc_b247_2</th>
<th style="text-align: right;">NIOZ124_mb_b130_2</th>
<th style="text-align: right;">NIOZ125_cc_b75_2</th>
<th style="text-align: right;">NIOZ125_mb_b178_2</th>
<th style="text-align: right;">NIOZ125_mb_b254_2</th>
<th style="text-align: right;">NIOZ126_mb_b137_2</th>
<th style="text-align: right;">NIOZ126_mb_b304_2</th>
<th style="text-align: right;">NIOZ126_mb_b401_2</th>
<th style="text-align: right;">NIOZ127_mb_b161_2</th>
<th style="text-align: right;">NIOZ127_mb_b282_2</th>
<th style="text-align: right;">NIOZ129_mb_b254_2</th>
<th style="text-align: right;">NIOZ129_mb_b262_2</th>
<th style="text-align: right;">NIOZ132_cc_b149_2</th>
<th style="text-align: right;">NIOZ132_cc_b51_2</th>
<th style="text-align: right;">NIOZ132_mb_b198_2</th>
<th style="text-align: right;">NIOZ132_mb_b260_2</th>
<th style="text-align: right;">NIOZ132_mb_b282_2</th>
<th style="text-align: right;">NIOZ132_mx_b314_2</th>
<th style="text-align: right;">NIOZ134_cc_b149_2</th>
<th style="text-align: right;">NIOZ134_mb_b293_2</th>
<th style="text-align: right;">NIOZ134_mb_b361_2</th>
<th style="text-align: right;">NIOZ134_mb_b41_2</th>
<th style="text-align: right;">NIOZ134_mx_b281_2</th>
<th style="text-align: right;">NIOZ136_cc_b15_2</th>
<th style="text-align: right;">NIOZ136_mb_b104_2</th>
<th style="text-align: right;">NIOZ136_mb_b226_2</th>
<th style="text-align: right;">NIOZ136_mb_b335_2</th>
<th style="text-align: right;">SRR2090153_bin1042</th>
<th style="text-align: right;">SRR2090153_bin461</th>
<th style="text-align: right;">SRR2090159_bin1129</th>
<th style="text-align: right;">SRR2090159_bin1288</th>
<th style="text-align: right;">SRR4028224_bin17</th>
<th style="text-align: right;">SRR5007147_bin71</th>
<th style="text-align: right;">U_76725</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb212"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb212-1"><a href="#cb212-1" aria-hidden="true" tabindex="-1"></a><span class="co">#order our data so that the bins start first with the bins from the basal group</span></span>
<span id="cb212-2"><a href="#cb212-2" aria-hidden="true" tabindex="-1"></a>Counts_Table_wide <span class="ot">&lt;-</span> Counts_Table_wide[,Bin_order]</span>
<span id="cb212-3"><a href="#cb212-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb212-4"><a href="#cb212-4" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb212-5"><a href="#cb212-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 0%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">NIOZ132_cc_b51_2</th>
<th style="text-align: right;">NIOZ134_mx_b281_2</th>
<th style="text-align: right;">SRR2090153_bin1042</th>
<th style="text-align: right;">SRR2090153_bin461</th>
<th style="text-align: right;">SRR2090159_bin1129</th>
<th style="text-align: right;">SRR2090159_bin1288</th>
<th style="text-align: right;">GCA_002494525</th>
<th style="text-align: right;">GCA_002495465</th>
<th style="text-align: right;">GCA_002501985</th>
<th style="text-align: right;">SRR4028224_bin17</th>
<th style="text-align: right;">SRR5007147_bin71</th>
<th style="text-align: right;">U_76725</th>
<th style="text-align: right;">NIOZ124_cc_b247_2</th>
<th style="text-align: right;">NIOZ125_mb_b254_2</th>
<th style="text-align: right;">NIOZ129_mb_b262_2</th>
<th style="text-align: right;">NIOZ126_mb_b401_2</th>
<th style="text-align: right;">NIOZ122_mb_b319_2</th>
<th style="text-align: right;">NIOZ134_mb_b41_2</th>
<th style="text-align: right;">NIOZ132_mb_b282_2</th>
<th style="text-align: right;">NIOZ127_mb_b161_2</th>
<th style="text-align: right;">NIOZ123_bs_b5_2</th>
<th style="text-align: right;">NIOZ121_cc_b94_2</th>
<th style="text-align: right;">NIOZ136_cc_b15_2</th>
<th style="text-align: right;">GCA_002502135</th>
<th style="text-align: right;">NIOZ134_mb_b293_2</th>
<th style="text-align: right;">NIOZ132_mb_b260_2</th>
<th style="text-align: right;">NIOZ126_mb_b137_2</th>
<th style="text-align: right;">NIOZ129_mb_b254_2</th>
<th style="text-align: right;">NIOZ127_mb_b282_2</th>
<th style="text-align: right;">NIOZ122_mb_b305_2</th>
<th style="text-align: right;">NIOZ136_mb_b335_2</th>
<th style="text-align: right;">NIOZ132_mb_b198_2</th>
<th style="text-align: right;">NIOZ136_mb_b104_2</th>
<th style="text-align: right;">NIOZ125_mb_b178_2</th>
<th style="text-align: right;">NIOZ121_mb_b48_2</th>
<th style="text-align: right;">NIOZ134_cc_b149_2</th>
<th style="text-align: right;">NIOZ134_mb_b361_2</th>
<th style="text-align: right;">NIOZ136_mb_b226_2</th>
<th style="text-align: right;">NIOZ123_bs_b392_2</th>
<th style="text-align: right;">NIOZ124_mb_b130_2</th>
<th style="text-align: right;">NIOZ132_cc_b149_2</th>
<th style="text-align: right;">NIOZ125_cc_b75_2</th>
<th style="text-align: right;">NIOZ132_mx_b314_2</th>
<th style="text-align: right;">NIOZ120_mb_b229_2</th>
<th style="text-align: right;">NIOZ126_mb_b304_2</th>
<th style="text-align: right;">NIOZ119_mb_b5_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>If you run these three examples yourself, take not how different the speed is.</p>
</section>
</section>
<section id="generate-a-count-table-for-our-clusters-of-interest" class="level3" data-number="1.31.2">
<h3 data-number="1.31.2" class="anchored" data-anchor-id="generate-a-count-table-for-our-clusters-of-interest"><span class="header-section-number">1.31.2</span> Generate a count table for our clusters of interest</h3>
<p>Same as above, but now we want to know for our 4 aquifer genomes, how many have Gene Xx and show this as percent. I.e. if 1/4 genomes have a gene, then 25% have it.</p>
<p>First, lets merge in our taxa info into our count table, we need this to summarize our data by clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb213"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb213-2"><a href="#cb213-2" aria-hidden="true" tabindex="-1"></a><span class="co">#merge the count table with mapping file to add in the taxa info (might take a while depending on size)</span></span>
<span id="cb213-3"><a href="#cb213-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax <span class="ot">&lt;-</span> <span class="fu">merge</span>(Counts_Table_long, design_2[,<span class="fu">c</span>(<span class="st">"BinID"</span>, <span class="st">"ClusterName"</span>, <span class="st">"NrGenomes"</span>)], <span class="at">by =</span> <span class="st">"BinID"</span>)</span>
<span id="cb213-4"><a href="#cb213-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">count</th>
<th style="text-align: left;">ClusterName</th>
<th style="text-align: right;">NrGenomes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glycos_transf_2</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT4</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT66</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT83</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">K00012</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Next, whenever we have a value higher than one, we replace it with 1. That way we deal with our data like it is a presence/absence data. I.e. 0 = no genes present &amp; 1 = gene present</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb214"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb214-1"><a href="#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="co">#convert counts to presence/absence matrix (just using 0/1) (this is needed to calculate the percentage across clusters)</span></span>
<span id="cb214-2"><a href="#cb214-2" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax<span class="sc">$</span>count[Counts_Table_long_Tax<span class="sc">$</span>count <span class="sc">&gt;</span> <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb214-3"><a href="#cb214-3" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">BinID</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">count</th>
<th style="text-align: left;">ClusterName</th>
<th style="text-align: right;">NrGenomes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT2_Glycos_transf_2</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT4</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT66</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">GT83</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">GCA_002494525</td>
<td style="text-align: left;">K00012</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now, we can use tidyr to count of how many genomes in a cluster have a gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb215"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb215-2"><a href="#cb215-2" aria-hidden="true" tabindex="-1"></a><span class="co">#count data and clean header</span></span>
<span id="cb215-3"><a href="#cb215-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax_sum <span class="ot">&lt;-</span> Counts_Table_long_Tax[,<span class="fu">c</span>(<span class="st">'ClusterName'</span>, <span class="st">'geneID'</span>,<span class="st">'NrGenomes'</span>, <span class="st">'count'</span>)] <span class="sc">%&gt;%</span> <span class="fu">count</span>(ClusterName, geneID, NrGenomes, <span class="at">sort =</span> <span class="cn">FALSE</span>)</span>
<span id="cb215-4"><a href="#cb215-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Counts_Table_long_Tax_sum) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ClusterName"</span>, <span class="st">"geneID"</span>, <span class="st">"NrGenomes"</span>, <span class="st">"quantity"</span>)</span>
<span id="cb215-5"><a href="#cb215-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax_sum)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">ClusterName</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">NrGenomes</th>
<th style="text-align: right;">quantity</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT2_Glycos_transf_2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT66</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT83</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">K00003</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">K00008</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Next, we want to calculate the percentage. I.e. in the first example, we have 4 aquifer genomes, two of which [FeFe]_Group_C3 (=50%).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb216"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb216-1"><a href="#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="co">#calculate of the percentage to answer of the total genomes per cluster how many have a certain gene</span></span>
<span id="cb216-2"><a href="#cb216-2" aria-hidden="true" tabindex="-1"></a><span class="co">#notice: if running for your own data check here that your percentage makes sense. I.e. we do not want values above 100</span></span>
<span id="cb216-3"><a href="#cb216-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax_sum<span class="sc">$</span>percentage <span class="ot">&lt;-</span> <span class="fu">round</span>(Counts_Table_long_Tax_sum<span class="sc">$</span>quantity<span class="sc">/</span>Counts_Table_long_Tax_sum<span class="sc">$</span>NrGenomes<span class="sc">*</span><span class="dv">100</span>, <span class="at">digits =</span> <span class="dv">0</span>)</span>
<span id="cb216-4"><a href="#cb216-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax_sum)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;">ClusterName</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">NrGenomes</th>
<th style="text-align: right;">quantity</th>
<th style="text-align: right;">percentage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT2_Glycos_transf_2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">100</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">75</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT66</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">50</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">GT83</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">75</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">K00003</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">50</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">K00008</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">75</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>For printing this, we want to convert our long to a wide table by using <code>spread()</code>. Also, we want our geneIDs to be the new rownames and we do this by using the <code>rownames()</code> functions. We do this since some functions do not like to have characters in their dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb217"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1" aria-hidden="true" tabindex="-1"></a><span class="co">#convert long to wide format and clean table (i.e. place the rownames)</span></span>
<span id="cb217-2"><a href="#cb217-2" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax_sum_wide <span class="ot">&lt;-</span> <span class="fu">spread</span>(Counts_Table_long_Tax_sum[,<span class="fu">c</span>(<span class="st">"geneID"</span>, <span class="st">"ClusterName"</span>, <span class="st">"percentage"</span>)], ClusterName, percentage)</span>
<span id="cb217-3"><a href="#cb217-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-4"><a href="#cb217-4" aria-hidden="true" tabindex="-1"></a><span class="co">#change the rownames</span></span>
<span id="cb217-5"><a href="#cb217-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Counts_Table_long_Tax_sum_wide) <span class="ot">&lt;-</span> Counts_Table_long_Tax_sum_wide<span class="sc">$</span>geneID</span>
<span id="cb217-6"><a href="#cb217-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-7"><a href="#cb217-7" aria-hidden="true" tabindex="-1"></a><span class="co">#view data</span></span>
<span id="cb217-8"><a href="#cb217-8" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax_sum_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 12%">
<col style="width: 10%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 11%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">Aquifer (4)</th>
<th style="text-align: right;">Basal (2)</th>
<th style="text-align: right;">Clade1 (11)</th>
<th style="text-align: right;">Clade2 (23)</th>
<th style="text-align: right;">Marine (6)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">83</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>When we change the rownames and view the data, we see that the geneID is now both in the rownames as well as the first column. Since that is a bit messy, we next remove the first column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb218"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb218-1"><a href="#cb218-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb218-2"><a href="#cb218-2" aria-hidden="true" tabindex="-1"></a><span class="co">#delete the first column</span></span>
<span id="cb218-3"><a href="#cb218-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax_sum_wide <span class="ot">&lt;-</span> Counts_Table_long_Tax_sum_wide[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb218-4"><a href="#cb218-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax_sum_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 25%">
<col style="width: 15%">
<col style="width: 12%">
<col style="width: 15%">
<col style="width: 15%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Aquifer (4)</th>
<th style="text-align: right;">Basal (2)</th>
<th style="text-align: right;">Clade1 (11)</th>
<th style="text-align: right;">Clade2 (23)</th>
<th style="text-align: right;">Marine (6)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">83</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now we see that we have NAs for genes that are not present in some of our clades. If we want to do more math then NAs are not helpful and we instead want to have a 0 there instead.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb219"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb219-1"><a href="#cb219-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb219-2"><a href="#cb219-2" aria-hidden="true" tabindex="-1"></a><span class="co">#replace NAs with 0</span></span>
<span id="cb219-3"><a href="#cb219-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax_sum_wide[<span class="fu">is.na</span>(Counts_Table_long_Tax_sum_wide)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb219-4"><a href="#cb219-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax_sum_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 25%">
<col style="width: 15%">
<col style="width: 12%">
<col style="width: 15%">
<col style="width: 15%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Aquifer (4)</th>
<th style="text-align: right;">Basal (2)</th>
<th style="text-align: right;">Clade1 (11)</th>
<th style="text-align: right;">Clade2 (23)</th>
<th style="text-align: right;">Marine (6)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">83</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Finally, we want to sort our data, starting with the Basal clade and ending with the Black Sea Clades</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb220"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb220-1"><a href="#cb220-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb220-2"><a href="#cb220-2" aria-hidden="true" tabindex="-1"></a><span class="co">#sort by cluster order (defined by the order of the mapping file)</span></span>
<span id="cb220-3"><a href="#cb220-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_long_Tax_sum_wide <span class="ot">&lt;-</span> Counts_Table_long_Tax_sum_wide[,Cluster_order]</span>
<span id="cb220-4"><a href="#cb220-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_long_Tax_sum_wide)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 25%">
<col style="width: 12%">
<col style="width: 15%">
<col style="width: 14%">
<col style="width: 15%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Basal (2)</th>
<th style="text-align: right;">Aquifer (4)</th>
<th style="text-align: right;">Marine (6)</th>
<th style="text-align: right;">Clade1 (11)</th>
<th style="text-align: right;">Clade2 (23)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CBM44</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">CE4</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH119</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">52</td>
</tr>
<tr class="even">
<td style="text-align: left;">GH130</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GH95</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GT2_Glyco_tranf_2_3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">83</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="merge-our-tables-with-the-mapping-data-we-have" class="level2" data-number="1.32">
<h2 data-number="1.32" class="anchored" data-anchor-id="merge-our-tables-with-the-mapping-data-we-have"><span class="header-section-number">1.32</span> Merge our tables with the mapping data we have</h2>
<p>Now, that we have our count tables both for the bins as well as for all the clusters, we now want to add some gene description and subset the data based on different categores.</p>
<section id="for-the-bins" class="level3" data-number="1.32.1">
<h3 data-number="1.32.1" class="anchored" data-anchor-id="for-the-bins"><span class="header-section-number">1.32.1</span> For the bins</h3>
<section id="add-gene-descriptions" class="level4" data-number="1.32.1.1">
<h4 data-number="1.32.1.1" class="anchored" data-anchor-id="add-gene-descriptions"><span class="header-section-number">1.32.1.1</span> Add gene descriptions</h4>
<p>Remember above, we made a list of descriptions that links all geneIDs with what is behind all the gene IDs? Now we want to add this info back in in order to print all the counts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb221"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb221-1"><a href="#cb221-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-2"><a href="#cb221-2" aria-hidden="true" tabindex="-1"></a><span class="co">#merge</span></span>
<span id="cb221-3"><a href="#cb221-3" aria-hidden="true" tabindex="-1"></a>Counts_Table_final <span class="ot">&lt;-</span> <span class="fu">merge</span>(All_Genes_Description, Counts_Table_wide, <span class="at">by.x=</span><span class="st">"Gene"</span>, <span class="at">by.y=</span><span class="st">"row.names"</span>, <span class="at">all.x =</span> T, <span class="at">sort =</span> F)</span>
<span id="cb221-4"><a href="#cb221-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Counts_Table_final)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 6%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Description</th>
<th style="text-align: right;">NIOZ132_cc_b51_2</th>
<th style="text-align: right;">NIOZ134_mx_b281_2</th>
<th style="text-align: right;">SRR2090153_bin1042</th>
<th style="text-align: right;">SRR2090153_bin461</th>
<th style="text-align: right;">SRR2090159_bin1129</th>
<th style="text-align: right;">SRR2090159_bin1288</th>
<th style="text-align: right;">GCA_002494525</th>
<th style="text-align: right;">GCA_002495465</th>
<th style="text-align: right;">GCA_002501985</th>
<th style="text-align: right;">SRR4028224_bin17</th>
<th style="text-align: right;">SRR5007147_bin71</th>
<th style="text-align: right;">U_76725</th>
<th style="text-align: right;">NIOZ124_cc_b247_2</th>
<th style="text-align: right;">NIOZ125_mb_b254_2</th>
<th style="text-align: right;">NIOZ129_mb_b262_2</th>
<th style="text-align: right;">NIOZ126_mb_b401_2</th>
<th style="text-align: right;">NIOZ122_mb_b319_2</th>
<th style="text-align: right;">NIOZ134_mb_b41_2</th>
<th style="text-align: right;">NIOZ132_mb_b282_2</th>
<th style="text-align: right;">NIOZ127_mb_b161_2</th>
<th style="text-align: right;">NIOZ123_bs_b5_2</th>
<th style="text-align: right;">NIOZ121_cc_b94_2</th>
<th style="text-align: right;">NIOZ136_cc_b15_2</th>
<th style="text-align: right;">GCA_002502135</th>
<th style="text-align: right;">NIOZ134_mb_b293_2</th>
<th style="text-align: right;">NIOZ132_mb_b260_2</th>
<th style="text-align: right;">NIOZ126_mb_b137_2</th>
<th style="text-align: right;">NIOZ129_mb_b254_2</th>
<th style="text-align: right;">NIOZ127_mb_b282_2</th>
<th style="text-align: right;">NIOZ122_mb_b305_2</th>
<th style="text-align: right;">NIOZ136_mb_b335_2</th>
<th style="text-align: right;">NIOZ132_mb_b198_2</th>
<th style="text-align: right;">NIOZ136_mb_b104_2</th>
<th style="text-align: right;">NIOZ125_mb_b178_2</th>
<th style="text-align: right;">NIOZ121_mb_b48_2</th>
<th style="text-align: right;">NIOZ134_cc_b149_2</th>
<th style="text-align: right;">NIOZ134_mb_b361_2</th>
<th style="text-align: right;">NIOZ136_mb_b226_2</th>
<th style="text-align: right;">NIOZ123_bs_b392_2</th>
<th style="text-align: right;">NIOZ124_mb_b130_2</th>
<th style="text-align: right;">NIOZ132_cc_b149_2</th>
<th style="text-align: right;">NIOZ125_cc_b75_2</th>
<th style="text-align: right;">NIOZ132_mx_b314_2</th>
<th style="text-align: right;">NIOZ120_mb_b229_2</th>
<th style="text-align: right;">NIOZ126_mb_b304_2</th>
<th style="text-align: right;">NIOZ119_mb_b5_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: left;">“Geranylgeranyl_reductase,_flavoprotein”</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG01358</td>
<td style="text-align: left;">2-methylthioadenine_synthetase</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG01169</td>
<td style="text-align: left;">Enolase</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG04245</td>
<td style="text-align: left;">Ribosomal_protein_S2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG01728</td>
<td style="text-align: left;">“Predicted_class_III_extradiol_dioxygenase,_MEMO1_family”</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00546</td>
<td style="text-align: left;">mRNA_degradation_ribonuclease_J1/J2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb222"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb222-1"><a href="#cb222-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb222-2"><a href="#cb222-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print (and beautify elsewhere)</span></span>
<span id="cb222-3"><a href="#cb222-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(Counts_Table_final, <span class="st">"../output_examples/Counts_Table_final.txt"</span>,  <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> F, <span class="at">row.names =</span> T, <span class="at">na =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="merging-with-the-arcog_table" class="level4" data-number="1.32.1.2">
<h4 data-number="1.32.1.2" class="anchored" data-anchor-id="merging-with-the-arcog_table"><span class="header-section-number">1.32.1.2</span> Merging with the arcog_table</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb223"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb223-1"><a href="#cb223-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb223-2"><a href="#cb223-2" aria-hidden="true" tabindex="-1"></a><span class="co">#merge</span></span>
<span id="cb223-3"><a href="#cb223-3" aria-hidden="true" tabindex="-1"></a>Arcog_Data <span class="ot">&lt;-</span> <span class="fu">merge</span>(Arcog_mapping, Counts_Table_wide, <span class="at">by.x=</span><span class="st">"arcog"</span>, <span class="at">by.y=</span><span class="st">"row.names"</span>, <span class="at">all.x =</span> T, <span class="at">sort =</span> F)</span>
<span id="cb223-4"><a href="#cb223-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Arcog_Data)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 7%">
<col style="width: 0%">
<col style="width: 2%">
<col style="width: 0%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">arcog</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">GeneID</th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">COG</th>
<th style="text-align: left;">PFAM</th>
<th style="text-align: left;">CD</th>
<th style="text-align: left;">TIGR</th>
<th style="text-align: right;">NIOZ132_cc_b51_2</th>
<th style="text-align: right;">NIOZ134_mx_b281_2</th>
<th style="text-align: right;">SRR2090153_bin1042</th>
<th style="text-align: right;">SRR2090153_bin461</th>
<th style="text-align: right;">SRR2090159_bin1129</th>
<th style="text-align: right;">SRR2090159_bin1288</th>
<th style="text-align: right;">GCA_002494525</th>
<th style="text-align: right;">GCA_002495465</th>
<th style="text-align: right;">GCA_002501985</th>
<th style="text-align: right;">SRR4028224_bin17</th>
<th style="text-align: right;">SRR5007147_bin71</th>
<th style="text-align: right;">U_76725</th>
<th style="text-align: right;">NIOZ124_cc_b247_2</th>
<th style="text-align: right;">NIOZ125_mb_b254_2</th>
<th style="text-align: right;">NIOZ129_mb_b262_2</th>
<th style="text-align: right;">NIOZ126_mb_b401_2</th>
<th style="text-align: right;">NIOZ122_mb_b319_2</th>
<th style="text-align: right;">NIOZ134_mb_b41_2</th>
<th style="text-align: right;">NIOZ132_mb_b282_2</th>
<th style="text-align: right;">NIOZ127_mb_b161_2</th>
<th style="text-align: right;">NIOZ123_bs_b5_2</th>
<th style="text-align: right;">NIOZ121_cc_b94_2</th>
<th style="text-align: right;">NIOZ136_cc_b15_2</th>
<th style="text-align: right;">GCA_002502135</th>
<th style="text-align: right;">NIOZ134_mb_b293_2</th>
<th style="text-align: right;">NIOZ132_mb_b260_2</th>
<th style="text-align: right;">NIOZ126_mb_b137_2</th>
<th style="text-align: right;">NIOZ129_mb_b254_2</th>
<th style="text-align: right;">NIOZ127_mb_b282_2</th>
<th style="text-align: right;">NIOZ122_mb_b305_2</th>
<th style="text-align: right;">NIOZ136_mb_b335_2</th>
<th style="text-align: right;">NIOZ132_mb_b198_2</th>
<th style="text-align: right;">NIOZ136_mb_b104_2</th>
<th style="text-align: right;">NIOZ125_mb_b178_2</th>
<th style="text-align: right;">NIOZ121_mb_b48_2</th>
<th style="text-align: right;">NIOZ134_cc_b149_2</th>
<th style="text-align: right;">NIOZ134_mb_b361_2</th>
<th style="text-align: right;">NIOZ136_mb_b226_2</th>
<th style="text-align: right;">NIOZ123_bs_b392_2</th>
<th style="text-align: right;">NIOZ124_mb_b130_2</th>
<th style="text-align: right;">NIOZ132_cc_b149_2</th>
<th style="text-align: right;">NIOZ125_cc_b75_2</th>
<th style="text-align: right;">NIOZ132_mx_b314_2</th>
<th style="text-align: right;">NIOZ120_mb_b229_2</th>
<th style="text-align: right;">NIOZ126_mb_b304_2</th>
<th style="text-align: right;">NIOZ119_mb_b5_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">arCOG00009</td>
<td style="text-align: left;">E</td>
<td style="text-align: left;">PotE</td>
<td style="text-align: left;">Amino acid transporter</td>
<td style="text-align: left;">COG00531</td>
<td style="text-align: left;">pfam00324</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">TIGR00909</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00014</td>
<td style="text-align: left;">G</td>
<td style="text-align: left;">RbsK</td>
<td style="text-align: left;">“Sugar kinase, ribokinase family”</td>
<td style="text-align: left;">COG00524</td>
<td style="text-align: left;">pfam00294</td>
<td style="text-align: left;">cd01174</td>
<td style="text-align: left;">TIGR02152</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00017</td>
<td style="text-align: left;">R</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">Predicted transcriptional regulator</td>
<td style="text-align: left;">COG02522</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00018</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Nnr2</td>
<td style="text-align: left;">“NAD(P)H-hydrate repair enzyme Nnr, NAD(P)H-hydrate dehydratase domain”</td>
<td style="text-align: left;">COG00063</td>
<td style="text-align: left;">“pfam03853,pfam01256”</td>
<td style="text-align: left;">cd01171</td>
<td style="text-align: left;">“TIGR00197,TIGR00196”</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00029</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">PyrE</td>
<td style="text-align: left;">Orotate phosphoribosyltransferase</td>
<td style="text-align: left;">COG00461</td>
<td style="text-align: left;">pfam00156</td>
<td style="text-align: left;">cd06223</td>
<td style="text-align: left;">TIGR00336</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00030</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Apt</td>
<td style="text-align: left;">Adenine/guanine phosphoribosyltransferase or related PRPP-binding protein</td>
<td style="text-align: left;">COG00503</td>
<td style="text-align: left;">pfam00156</td>
<td style="text-align: left;">cd06223</td>
<td style="text-align: left;">TIGR01090</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb224"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb224-1"><a href="#cb224-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb224-2"><a href="#cb224-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print (and beautify elsewhere)</span></span>
<span id="cb224-3"><a href="#cb224-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(Arcog_Data, <span class="st">"../output_examples/ArCOG_Data.txt"</span>,  <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> F, <span class="at">row.names =</span> T, <span class="at">na =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="merging-with-the-metabolism-metadata-file" class="level4" data-number="1.32.1.3">
<h4 data-number="1.32.1.3" class="anchored" data-anchor-id="merging-with-the-metabolism-metadata-file"><span class="header-section-number">1.32.1.3</span> Merging with the metabolism metadata file</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb225"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb225-1"><a href="#cb225-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb225-2"><a href="#cb225-2" aria-hidden="true" tabindex="-1"></a><span class="co">#merge</span></span>
<span id="cb225-3"><a href="#cb225-3" aria-hidden="true" tabindex="-1"></a>KEGG_Metabolism <span class="ot">&lt;-</span> <span class="fu">merge</span>(Metabolism_file_KEGG, Counts_Table_wide, <span class="at">by.x=</span><span class="st">"KO"</span>, <span class="at">by.y=</span><span class="st">"row.names"</span>, <span class="at">all.x =</span> T, <span class="at">sort =</span> F)</span>
<span id="cb225-4"><a href="#cb225-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(KEGG_Metabolism)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 6%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">KO</th>
<th style="text-align: left;">IDs</th>
<th style="text-align: left;">pathway_A</th>
<th style="text-align: left;">order</th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Comment.Uniprot.comment</th>
<th style="text-align: right;">NIOZ132_cc_b51_2</th>
<th style="text-align: right;">NIOZ134_mx_b281_2</th>
<th style="text-align: right;">SRR2090153_bin1042</th>
<th style="text-align: right;">SRR2090153_bin461</th>
<th style="text-align: right;">SRR2090159_bin1129</th>
<th style="text-align: right;">SRR2090159_bin1288</th>
<th style="text-align: right;">GCA_002494525</th>
<th style="text-align: right;">GCA_002495465</th>
<th style="text-align: right;">GCA_002501985</th>
<th style="text-align: right;">SRR4028224_bin17</th>
<th style="text-align: right;">SRR5007147_bin71</th>
<th style="text-align: right;">U_76725</th>
<th style="text-align: right;">NIOZ124_cc_b247_2</th>
<th style="text-align: right;">NIOZ125_mb_b254_2</th>
<th style="text-align: right;">NIOZ129_mb_b262_2</th>
<th style="text-align: right;">NIOZ126_mb_b401_2</th>
<th style="text-align: right;">NIOZ122_mb_b319_2</th>
<th style="text-align: right;">NIOZ134_mb_b41_2</th>
<th style="text-align: right;">NIOZ132_mb_b282_2</th>
<th style="text-align: right;">NIOZ127_mb_b161_2</th>
<th style="text-align: right;">NIOZ123_bs_b5_2</th>
<th style="text-align: right;">NIOZ121_cc_b94_2</th>
<th style="text-align: right;">NIOZ136_cc_b15_2</th>
<th style="text-align: right;">GCA_002502135</th>
<th style="text-align: right;">NIOZ134_mb_b293_2</th>
<th style="text-align: right;">NIOZ132_mb_b260_2</th>
<th style="text-align: right;">NIOZ126_mb_b137_2</th>
<th style="text-align: right;">NIOZ129_mb_b254_2</th>
<th style="text-align: right;">NIOZ127_mb_b282_2</th>
<th style="text-align: right;">NIOZ122_mb_b305_2</th>
<th style="text-align: right;">NIOZ136_mb_b335_2</th>
<th style="text-align: right;">NIOZ132_mb_b198_2</th>
<th style="text-align: right;">NIOZ136_mb_b104_2</th>
<th style="text-align: right;">NIOZ125_mb_b178_2</th>
<th style="text-align: right;">NIOZ121_mb_b48_2</th>
<th style="text-align: right;">NIOZ134_cc_b149_2</th>
<th style="text-align: right;">NIOZ134_mb_b361_2</th>
<th style="text-align: right;">NIOZ136_mb_b226_2</th>
<th style="text-align: right;">NIOZ123_bs_b392_2</th>
<th style="text-align: right;">NIOZ124_mb_b130_2</th>
<th style="text-align: right;">NIOZ132_cc_b149_2</th>
<th style="text-align: right;">NIOZ125_cc_b75_2</th>
<th style="text-align: right;">NIOZ132_mx_b314_2</th>
<th style="text-align: right;">NIOZ120_mb_b229_2</th>
<th style="text-align: right;">NIOZ126_mb_b304_2</th>
<th style="text-align: right;">NIOZ119_mb_b5_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">K00362</td>
<td style="text-align: left;">ID14</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N2</td>
<td style="text-align: left;">nirB; nitrite reductase (NADH) large subunit [EC:1.7.1.15]</td>
<td style="text-align: left;">Dissimilatory</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">TIGR02374</td>
<td style="text-align: left;">ID15</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N2</td>
<td style="text-align: left;">“NirB, nitrite reductase [NAD(P)H]”</td>
<td style="text-align: left;">Dissimilatory</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TIGR02378</td>
<td style="text-align: left;">ID17</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N2</td>
<td style="text-align: left;">“NirD, nitrite reductase [NAD(P)H]”</td>
<td style="text-align: left;">Dissimilatory</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">K04748</td>
<td style="text-align: left;">ID31</td>
<td style="text-align: left;">no_pathway</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">norQ; nitric oxide reductase NorQ protein</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">K00376</td>
<td style="text-align: left;">ID36</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N5</td>
<td style="text-align: left;">nosZ; nitrous-oxide reductase [EC:1.7.2.4]</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">TIGR04246</td>
<td style="text-align: left;">ID37</td>
<td style="text-align: left;">Nitrogen</td>
<td style="text-align: left;">N5</td>
<td style="text-align: left;">“nosZ, nitrous-oxide reductase, Sec-dependent”</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb226"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb226-1"><a href="#cb226-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb226-2"><a href="#cb226-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print</span></span>
<span id="cb226-3"><a href="#cb226-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(KEGG_Metabolism, <span class="st">"../output_examples/KEGG_Metabolism.txt"</span>,  <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> F, <span class="at">row.names =</span> T, <span class="at">na =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="for-the-clusters" class="level3" data-number="1.32.2">
<h3 data-number="1.32.2" class="anchored" data-anchor-id="for-the-clusters"><span class="header-section-number">1.32.2</span> For the clusters</h3>
<p>The process works exactly the same as above. So try by yourself if you can merge things ;-).</p>
</section>
</section>
<section id="plot-the-data-for-our-genes-of-interest" class="level2" data-number="1.33">
<h2 data-number="1.33" class="anchored" data-anchor-id="plot-the-data-for-our-genes-of-interest"><span class="header-section-number">1.33</span> Plot the data for our genes of interest</h2>
<p>Here, we are not interested in plotting all genes but just want to plot things that are listed under the lipid pathway.</p>
<p>Since we are only interested in the Lipid pathway for the genes of interest (the table, among others, also lists genes involved in informational processing), we first need to make a gene list of just the genes we are interested in.</p>
<p>New functions:</p>
<ul>
<li><code>subset()</code> = subsets a dataframe. Syntax –&gt; subset(dataframe, column, %in% pattern_we_look_for)</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb227"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb227-1"><a href="#cb227-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb227-2"><a href="#cb227-2" aria-hidden="true" tabindex="-1"></a><span class="co">#subset genes of interest and clean factors</span></span>
<span id="cb227-3"><a href="#cb227-3" aria-hidden="true" tabindex="-1"></a>Genes_Lipids <span class="ot">&lt;-</span> <span class="fu">subset</span>(Genes_of_interest, Pathway_1 <span class="sc">%in%</span> <span class="st">"Lipids"</span>)</span>
<span id="cb227-4"><a href="#cb227-4" aria-hidden="true" tabindex="-1"></a>Genes_Lipids<span class="sc">$</span>Gene <span class="ot">&lt;-</span> <span class="fu">factor</span>(Genes_Lipids<span class="sc">$</span>Gene)</span>
<span id="cb227-5"><a href="#cb227-5" aria-hidden="true" tabindex="-1"></a>Genes_Lipids<span class="sc">$</span>arcog <span class="ot">&lt;-</span> <span class="fu">factor</span>(Genes_Lipids<span class="sc">$</span>arcog)</span>
<span id="cb227-6"><a href="#cb227-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb227-7"><a href="#cb227-7" aria-hidden="true" tabindex="-1"></a><span class="co">#check how many genes we have</span></span>
<span id="cb227-8"><a href="#cb227-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Genes_Lipids)</span>
<span id="cb227-9"><a href="#cb227-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 16 12</span></span>
<span id="cb227-10"><a href="#cb227-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb227-11"><a href="#cb227-11" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Genes_Lipids)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 1%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 0%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 21%">
<col style="width: 51%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Pathway_1</th>
<th style="text-align: left;">pathway_2</th>
<th style="text-align: left;">arcog</th>
<th style="text-align: left;">xx</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">Gene_short</th>
<th style="text-align: left;">Gene_name</th>
<th style="text-align: left;">Comment</th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Order</th>
<th style="text-align: right;">Order2</th>
<th style="text-align: left;">Other</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">262</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Mevalonate</td>
<td style="text-align: left;">arCOG01767</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">HmgB</td>
<td style="text-align: left;">3-hydroxy-3-methylglutaryl CoA synthase</td>
<td style="text-align: left;">Catalyzes the condensation of acetyl-CoA with acetoacetyl-CoA to form 3-hydroxy-3-methylglutaryl-CoA (HMG-CoA). Functions in the mevalonate (MVA) pathway</td>
<td style="text-align: left;">HmgB</td>
<td style="text-align: left;">Lipid_1</td>
<td style="text-align: right;">262</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">263</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Mevalonate</td>
<td style="text-align: left;">arCOG04260</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">HmgA</td>
<td style="text-align: left;">“3-hydroxy-3-methylglutaryl-coenzyme A reductase, HMG-CoA reductase”</td>
<td style="text-align: left;">Catalyzes the NADPH-dependent reductive deacylation of (S)-3-hydroxy-3-methylglutaryl-CoA (HMG-CoA) to (R)-mevalonate. Functions in the mevalonate (MVA) pathway</td>
<td style="text-align: left;">HmgA</td>
<td style="text-align: left;">Lipid_1</td>
<td style="text-align: right;">263</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">264</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Mevalonate</td>
<td style="text-align: left;">arCOG01028</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">MvK</td>
<td style="text-align: left;">Mevalonate kinase</td>
<td style="text-align: left;">Catalyzes the phosphorylation of (R)-mevalonate (MVA) to (R)-mevalonate 5-phosphate (MVAP).</td>
<td style="text-align: left;">MvK</td>
<td style="text-align: left;">Lipid_1</td>
<td style="text-align: right;">264</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">265</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Mevalonate</td>
<td style="text-align: left;">K18689</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">M3k</td>
<td style="text-align: left;">Mevalonate 3-kinase</td>
<td style="text-align: left;">Catalyzes the phosphorylation of mevalonate (MVA) to yield mevalonate-3-phosphate. Functions in an alternative mevalonate pathway</td>
<td style="text-align: left;">M3k</td>
<td style="text-align: left;">Lipid_1</td>
<td style="text-align: right;">265</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">266</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Mevalonate</td>
<td style="text-align: left;">arCOG01967</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">M3P5K</td>
<td style="text-align: left;">Mevalonate-3-phosphate 5-kinase</td>
<td style="text-align: left;">“Phosphorylates mevalonate 3-phosphate to form mevalonate 3,5-bisphosphate. Functions in an alternative mevalonate pathway”</td>
<td style="text-align: left;">M3P5K</td>
<td style="text-align: left;">Lipid_1</td>
<td style="text-align: right;">266</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">267</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Mevalonate</td>
<td style="text-align: left;">arCOG02937</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">I</td>
<td style="text-align: left;">mvaD</td>
<td style="text-align: left;">Diphosphomevalonate decarboxylase</td>
<td style="text-align: left;">Catalyzes the decarboxylation of mevalonate 5-phosphate (MVAP) to isopentenyl phosphate (IP). Functions in the mevalonate (MVA) pathway leading to IPP</td>
<td style="text-align: left;">mvaD</td>
<td style="text-align: left;">Lipid_1</td>
<td style="text-align: right;">267</td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Next, we want to make sure that the order is ok. In this specific example, we manually defined two columns for ordering (Order and Order2). We sort based on these columns and make a vector to order our genes of interest and our pathways of interest. For the lipid genes we look at the mevalonate pathway and general lipid biosynthesis genes</p>
<ul>
<li>``length()``` - lets us check the length of a vector, here it allows us to see that we would expect 16 genes</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb228"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb228-1"><a href="#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="co">#define an order (we arange the dataframe based on two columsn, Order and Order2)</span></span>
<span id="cb228-2"><a href="#cb228-2" aria-hidden="true" tabindex="-1"></a>Genes_Lipids_order_temp <span class="ot">&lt;-</span> Genes_Lipids <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(Order, Order2)</span>
<span id="cb228-3"><a href="#cb228-3" aria-hidden="true" tabindex="-1"></a>Genes_Lipids_order <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(Genes_Lipids_order_temp<span class="sc">$</span>Gene))</span>
<span id="cb228-4"><a href="#cb228-4" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(Genes_Lipids_order)</span>
<span id="cb228-5"><a href="#cb228-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 16</span></span>
<span id="cb228-6"><a href="#cb228-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb228-7"><a href="#cb228-7" aria-hidden="true" tabindex="-1"></a>Genes_Lipids_order</span>
<span id="cb228-8"><a href="#cb228-8" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] "HmgB"          "HmgA"          "MvK"           "M3k"          </span></span>
<span id="cb228-9"><a href="#cb228-9" aria-hidden="true" tabindex="-1"></a><span class="do">##  [5] "M3P5K"         "mvaD"          "Ipk"           "GFPS"         </span></span>
<span id="cb228-10"><a href="#cb228-10" aria-hidden="true" tabindex="-1"></a><span class="do">##  [9] "EgsA"          "GGGP_synthase" "UbiA"          "GGR"          </span></span>
<span id="cb228-11"><a href="#cb228-11" aria-hidden="true" tabindex="-1"></a><span class="do">## [13] "CarS"          "PgsA"          "ASS"           "Asd"</span></span>
<span id="cb228-12"><a href="#cb228-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb228-13"><a href="#cb228-13" aria-hidden="true" tabindex="-1"></a><span class="co">#The lipids belong to two different pathways, these 2 pathways we want to show in two separate heatmaps</span></span>
<span id="cb228-14"><a href="#cb228-14" aria-hidden="true" tabindex="-1"></a>Lipids_Pathway_order <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(Genes_Lipids<span class="sc">$</span>pathway_2))</span>
<span id="cb228-15"><a href="#cb228-15" aria-hidden="true" tabindex="-1"></a>Lipids_Pathway_order</span>
<span id="cb228-16"><a href="#cb228-16" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "Mevalonate" "Lipids"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that we know what genes we are interested in, lets subset our original count table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb229"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb229-1"><a href="#cb229-1" aria-hidden="true" tabindex="-1"></a><span class="co">#subset our original count table for genes of interest and clean factors</span></span>
<span id="cb229-2"><a href="#cb229-2" aria-hidden="true" tabindex="-1"></a>Genes_Lipids_counts <span class="ot">&lt;-</span> <span class="fu">subset</span>(Counts_Table_long_Tax_sum, geneID <span class="sc">%in%</span> <span class="fu">as.character</span>(Genes_Lipids<span class="sc">$</span>arcog))</span>
<span id="cb229-3"><a href="#cb229-3" aria-hidden="true" tabindex="-1"></a>Genes_Lipids_counts<span class="sc">$</span>geneID <span class="ot">&lt;-</span> <span class="fu">factor</span>(Genes_Lipids_counts<span class="sc">$</span>geneID)</span>
<span id="cb229-4"><a href="#cb229-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb229-5"><a href="#cb229-5" aria-hidden="true" tabindex="-1"></a><span class="co">#control that all went fine</span></span>
<span id="cb229-6"><a href="#cb229-6" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(Genes_Lipids_counts<span class="sc">$</span>geneID))</span>
<span id="cb229-7"><a href="#cb229-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 15</span></span>
<span id="cb229-8"><a href="#cb229-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb229-9"><a href="#cb229-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Genes_Lipids_counts)</span>
<span id="cb229-10"><a href="#cb229-10" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 60  5</span></span>
<span id="cb229-11"><a href="#cb229-11" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Genes_Lipids_counts)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ClusterName</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">NrGenomes</th>
<th style="text-align: right;">quantity</th>
<th style="text-align: right;">percentage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1924</td>
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">arCOG00476</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">100</td>
</tr>
<tr class="even">
<td style="text-align: left;">1947</td>
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">100</td>
</tr>
<tr class="odd">
<td style="text-align: left;">1967</td>
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">arCOG00670</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">75</td>
</tr>
<tr class="even">
<td style="text-align: left;">1968</td>
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">arCOG00671</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">100</td>
</tr>
<tr class="odd">
<td style="text-align: left;">1997</td>
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">arCOG00860</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">75</td>
</tr>
<tr class="even">
<td style="text-align: left;">2032</td>
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: left;">arCOG00982</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">75</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>With length we see that now we just have 15 genes. How can we find out what gene is missing?</p>
<ul>
<li><code>setdiff()</code> = we compare two vectors and print the elements that differ.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb230"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb230-1"><a href="#cb230-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb230-2"><a href="#cb230-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setdiff</span>(Genes_Lipids<span class="sc">$</span>arcog, Genes_Lipids_counts<span class="sc">$</span>geneID)</span>
<span id="cb230-3"><a href="#cb230-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "K18689"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that we miss K18689. If we check our original annotations input, we can see that K18689 does not exist in that table. So we can nout pull information because of that. This gives a good example, why it is important to check your data as we do not know whether this is an issue with the code or what the problem could be.</p>
<hr>
<p>Now, since our count data does not know that we categorize our different genes into different pathways, lets add this info in with <code>merge()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb231"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb231-1"><a href="#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="co">#add in metadata (from the pathway info)</span></span>
<span id="cb231-2"><a href="#cb231-2" aria-hidden="true" tabindex="-1"></a>Key_Lipids_genes_cluster <span class="ot">&lt;-</span> <span class="fu">merge</span>(Genes_Lipids_counts, Genes_Lipids, <span class="at">by.x =</span><span class="st">"geneID"</span>, <span class="at">by.y =</span> <span class="st">'arcog'</span> , <span class="at">all.x =</span> T )</span>
<span id="cb231-3"><a href="#cb231-3" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>((<span class="fu">head</span>(Key_Lipids_genes_cluster)), <span class="at">format=</span><span class="st">'markdown'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<colgroup>
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 0%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 21%">
<col style="width: 42%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">geneID</th>
<th style="text-align: left;">ClusterName</th>
<th style="text-align: right;">NrGenomes</th>
<th style="text-align: right;">quantity</th>
<th style="text-align: right;">percentage</th>
<th style="text-align: left;">Pathway_1</th>
<th style="text-align: left;">pathway_2</th>
<th style="text-align: left;">xx</th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: left;">Gene_short</th>
<th style="text-align: left;">Gene_name</th>
<th style="text-align: left;">Comment</th>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Order</th>
<th style="text-align: right;">Order2</th>
<th style="text-align: left;">Other</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">arCOG00476</td>
<td style="text-align: left;">Aquifer (4)</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">100</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">L</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">geranylgeranylglycerol-phosphate geranylgeranyltransferase [EC:2.5.1.42]</td>
<td style="text-align: left;">Notice: only the homologs identified in the aquifer MAGs harbor the characteristic DGGGP synthase domain (IPR023547)</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">Lipid_2</td>
<td style="text-align: right;">272</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00476</td>
<td style="text-align: left;">Basal (2)</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">100</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">L</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">geranylgeranylglycerol-phosphate geranylgeranyltransferase [EC:2.5.1.42]</td>
<td style="text-align: left;">Notice: only the homologs identified in the aquifer MAGs harbor the characteristic DGGGP synthase domain (IPR023547)</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">Lipid_2</td>
<td style="text-align: right;">272</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00476</td>
<td style="text-align: left;">Clade1 (11)</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">100</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">L</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">geranylgeranylglycerol-phosphate geranylgeranyltransferase [EC:2.5.1.42]</td>
<td style="text-align: left;">Notice: only the homologs identified in the aquifer MAGs harbor the characteristic DGGGP synthase domain (IPR023547)</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">Lipid_2</td>
<td style="text-align: right;">272</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00476</td>
<td style="text-align: left;">Marine (6)</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">100</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">L</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">geranylgeranylglycerol-phosphate geranylgeranyltransferase [EC:2.5.1.42]</td>
<td style="text-align: left;">Notice: only the homologs identified in the aquifer MAGs harbor the characteristic DGGGP synthase domain (IPR023547)</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">Lipid_2</td>
<td style="text-align: right;">272</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">arCOG00476</td>
<td style="text-align: left;">Clade2 (23)</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">83</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">L</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">geranylgeranylglycerol-phosphate geranylgeranyltransferase [EC:2.5.1.42]</td>
<td style="text-align: left;">Notice: only the homologs identified in the aquifer MAGs harbor the characteristic DGGGP synthase domain (IPR023547)</td>
<td style="text-align: left;">UbiA</td>
<td style="text-align: left;">Lipid_2</td>
<td style="text-align: right;">272</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">arCOG00570</td>
<td style="text-align: left;">Clade2 (23)</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">22</td>
<td style="text-align: right;">96</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">Lipids</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">L</td>
<td style="text-align: left;">GGR</td>
<td style="text-align: left;">digeranylgeranylglycerophospholipid reductase [EC:1.3.1.101 1.3.7.11]</td>
<td style="text-align: left;">“reduction of 2,3-digeranylgeranylglycerophospholipids (unsaturated archaeols) into 2,3-diphytanylglycerophospholipids (saturated archaeols)”</td>
<td style="text-align: left;">GGR</td>
<td style="text-align: left;">Lipid_2</td>
<td style="text-align: right;">273</td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="categorizing-data" class="level3" data-number="1.33.1">
<h3 data-number="1.33.1" class="anchored" data-anchor-id="categorizing-data"><span class="header-section-number">1.33.1</span> Categorizing data</h3>
<p>There are different ways to color code data. By default the way we do it, we use a gradual color scale from 0-100%. However, we could also define categories with the ifelse statement we learned before. Here, we define 4 categories (100%, 75-100 and, 33-75 and 0-33%)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb232"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb232-1"><a href="#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="co">#define color code (not used for the current figure, but can be changed)</span></span>
<span id="cb232-2"><a href="#cb232-2" aria-hidden="true" tabindex="-1"></a><span class="co">#here , we define 3 color levels, which sometimes is useful to show very clear cutoffs</span></span>
<span id="cb232-3"><a href="#cb232-3" aria-hidden="true" tabindex="-1"></a>Key_Lipids_genes_cluster<span class="sc">$</span>category <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(Key_Lipids_genes_cluster<span class="sc">$</span>percentage <span class="sc">==</span> <span class="dv">100</span>, <span class="st">"1"</span>,</span>
<span id="cb232-4"><a href="#cb232-4" aria-hidden="true" tabindex="-1"></a>                                         <span class="fu">ifelse</span>(Key_Lipids_genes_cluster<span class="sc">$</span>percentage <span class="sc">&gt;=</span> <span class="dv">75</span>, <span class="st">"0.75"</span>,</span>
<span id="cb232-5"><a href="#cb232-5" aria-hidden="true" tabindex="-1"></a>                                                <span class="fu">ifelse</span>(Key_Lipids_genes_cluster<span class="sc">$</span>percentage <span class="sc">&gt;=</span> <span class="dv">33</span>, <span class="st">"0.33"</span>, <span class="st">"0"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="order-our-data" class="level3" data-number="1.33.2">
<h3 data-number="1.33.2" class="anchored" data-anchor-id="order-our-data"><span class="header-section-number">1.33.2</span> Order our data</h3>
<p>Remember the annoying thing that R sorts alphabetically? Let’s make sure we ahve the order we want.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb233"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb233-1"><a href="#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="co">#define order for the plot</span></span>
<span id="cb233-2"><a href="#cb233-2" aria-hidden="true" tabindex="-1"></a>Key_Lipids_genes_cluster<span class="sc">$</span>ClusterName2 <span class="ot">&lt;-</span>  <span class="fu">factor</span>(Key_Lipids_genes_cluster<span class="sc">$</span>ClusterName, <span class="at">levels =</span> <span class="fu">rev</span>(Cluster_order))</span>
<span id="cb233-3"><a href="#cb233-3" aria-hidden="true" tabindex="-1"></a>Key_Lipids_genes_cluster<span class="sc">$</span>Gene2 <span class="ot">&lt;-</span>  <span class="fu">factor</span>(Key_Lipids_genes_cluster<span class="sc">$</span>Gene, <span class="at">levels =</span> Genes_Lipids_order)</span>
<span id="cb233-4"><a href="#cb233-4" aria-hidden="true" tabindex="-1"></a>Key_Lipids_genes_cluster<span class="sc">$</span>pathway_2b <span class="ot">&lt;-</span>  <span class="fu">factor</span>(Key_Lipids_genes_cluster<span class="sc">$</span>pathway_2, <span class="at">levels =</span> Lipids_Pathway_order)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting-1" class="level3" data-number="1.33.3">
<h3 data-number="1.33.3" class="anchored" data-anchor-id="plotting-1"><span class="header-section-number">1.33.3</span> Plotting</h3>
<p>In the example here, we use a gradual scale. If we would want to use our 4 categories we can use this code <code>#scale_fill_manual(values= c("white", "blue", "blue", "dodgerblue"))</code> and replacing the <code>fill = percentage</code> with <code>fill = category</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb234"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb234-1"><a href="#cb234-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-2"><a href="#cb234-2" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb234-3"><a href="#cb234-3" aria-hidden="true" tabindex="-1"></a>p1_Lipids <span class="ot">&lt;-</span> </span>
<span id="cb234-4"><a href="#cb234-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(Key_Lipids_genes_cluster, <span class="fu">aes</span>(<span class="at">x=</span>Gene2, <span class="at">y=</span>ClusterName2)) <span class="sc">+</span> </span>
<span id="cb234-5"><a href="#cb234-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">fill =</span> percentage)) <span class="sc">+</span></span>
<span id="cb234-6"><a href="#cb234-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept=</span><span class="fl">2.5</span>) <span class="sc">+</span></span>
<span id="cb234-7"><a href="#cb234-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>( <span class="sc">~</span> pathway_2b, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">scales=</span><span class="st">'free_x'</span>) <span class="sc">+</span></span>
<span id="cb234-8"><a href="#cb234-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_distiller</span>(<span class="at">palette =</span> <span class="st">"Blues"</span>, <span class="at">direction =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb234-9"><a href="#cb234-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb234-10"><a href="#cb234-10" aria-hidden="true" tabindex="-1"></a>  <span class="co">#scale_fill_manual(values= c("white", "blue", "blue", "dodgerblue")) +</span></span>
<span id="cb234-11"><a href="#cb234-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">""</span>, <span class="at">y=</span><span class="st">""</span>, <span class="at">fill=</span><span class="st">"Percentage"</span>) <span class="sc">+</span> </span>
<span id="cb234-12"><a href="#cb234-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb234-13"><a href="#cb234-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb234-14"><a href="#cb234-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb234-15"><a href="#cb234-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>,</span>
<span id="cb234-16"><a href="#cb234-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>,<span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust=</span><span class="dv">1</span>, <span class="at">size=</span><span class="dv">8</span>),</span>
<span id="cb234-17"><a href="#cb234-17" aria-hidden="true" tabindex="-1"></a>        <span class="co">#axis.text.y=element_blank(),</span></span>
<span id="cb234-18"><a href="#cb234-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks=</span><span class="fu">element_blank</span>(),</span>
<span id="cb234-19"><a href="#cb234-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.line=</span><span class="fu">element_blank</span>(),</span>
<span id="cb234-20"><a href="#cb234-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin=</span><span class="fu">unit</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="st">"mm"</span>))</span>
<span id="cb234-21"><a href="#cb234-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb234-22"><a href="#cb234-22" aria-hidden="true" tabindex="-1"></a>p1_Lipids</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-218-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb235"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb235-1"><a href="#cb235-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb235-2"><a href="#cb235-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(<span class="fu">levels</span>(Key_Lipids_genes_cluster<span class="sc">$</span>ClusterName2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["levels(Key_Lipids_genes_cluster$ClusterName2)"],"name":[1],"type":["chr"],"align":["left"]}],"data":[{"1":"Clade2 (23)"},{"1":"Clade1 (11)"},{"1":"Marine (6)"},{"1":"Aquifer (4)"},{"1":"Basal (2)"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb236"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb236-1"><a href="#cb236-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb236-2"><a href="#cb236-2" aria-hidden="true" tabindex="-1"></a>p1_Lipids2 <span class="ot">&lt;-</span> </span>
<span id="cb236-3"><a href="#cb236-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(Key_Lipids_genes_cluster, <span class="fu">aes</span>(<span class="at">x=</span>Gene2, <span class="at">y=</span>ClusterName2)) <span class="sc">+</span> </span>
<span id="cb236-4"><a href="#cb236-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">fill =</span> percentage)) <span class="sc">+</span></span>
<span id="cb236-5"><a href="#cb236-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept=</span><span class="fl">2.5</span>) <span class="sc">+</span></span>
<span id="cb236-6"><a href="#cb236-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>( <span class="sc">~</span> pathway_2b, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">scales=</span><span class="st">'free_x'</span>) <span class="sc">+</span></span>
<span id="cb236-7"><a href="#cb236-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_distiller</span>(<span class="at">palette =</span> <span class="st">"Blues"</span>, <span class="at">direction =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb236-8"><a href="#cb236-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb236-9"><a href="#cb236-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">#scale_fill_manual(values= c("white", "blue", "blue", "dodgerblue")) +</span></span>
<span id="cb236-10"><a href="#cb236-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">""</span>, <span class="at">y=</span><span class="st">""</span>, <span class="at">fill=</span><span class="st">"Percentage"</span>) <span class="sc">+</span> </span>
<span id="cb236-11"><a href="#cb236-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb236-12"><a href="#cb236-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb236-13"><a href="#cb236-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb236-14"><a href="#cb236-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>,</span>
<span id="cb236-15"><a href="#cb236-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>,<span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust=</span><span class="dv">1</span>, <span class="at">size=</span><span class="dv">8</span>),</span>
<span id="cb236-16"><a href="#cb236-16" aria-hidden="true" tabindex="-1"></a>        <span class="co">#axis.text.y=element_blank(),</span></span>
<span id="cb236-17"><a href="#cb236-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks=</span><span class="fu">element_blank</span>(),</span>
<span id="cb236-18"><a href="#cb236-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.line=</span><span class="fu">element_blank</span>(),</span>
<span id="cb236-19"><a href="#cb236-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin=</span><span class="fu">unit</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="st">"mm"</span>))</span>
<span id="cb236-20"><a href="#cb236-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb236-21"><a href="#cb236-21" aria-hidden="true" tabindex="-1"></a>p1_Lipids2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-218-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>If we plot with facets we can sometimes have the problem that different genes have different widths. We can correct this behaviour with ggplotGrob.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb237"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb237-1"><a href="#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert ggplot object to grob object (used to rescale plot)</span></span>
<span id="cb237-2"><a href="#cb237-2" aria-hidden="true" tabindex="-1"></a>gp_lipid <span class="ot">&lt;-</span> <span class="fu">ggplotGrob</span>(p1_Lipids2)</span>
<span id="cb237-3"><a href="#cb237-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-4"><a href="#cb237-4" aria-hidden="true" tabindex="-1"></a><span class="co"># optional: take a look at the grob object's layout</span></span>
<span id="cb237-5"><a href="#cb237-5" aria-hidden="true" tabindex="-1"></a>gtable<span class="sc">::</span><span class="fu">gtable_show_layout</span>(gp_lipid)</span>
<span id="cb237-6"><a href="#cb237-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-7"><a href="#cb237-7" aria-hidden="true" tabindex="-1"></a><span class="co"># get gtable columns corresponding to the facets (5 &amp; 9, in this case)</span></span>
<span id="cb237-8"><a href="#cb237-8" aria-hidden="true" tabindex="-1"></a>facet.columns <span class="ot">&lt;-</span> gp_lipid<span class="sc">$</span>layout<span class="sc">$</span>l[<span class="fu">grepl</span>(<span class="st">"panel"</span>, gp_lipid<span class="sc">$</span>layout<span class="sc">$</span>name)]</span>
<span id="cb237-9"><a href="#cb237-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-10"><a href="#cb237-10" aria-hidden="true" tabindex="-1"></a><span class="co"># get the number of unique x-axis values per facet (1 &amp; 3, in this case)</span></span>
<span id="cb237-11"><a href="#cb237-11" aria-hidden="true" tabindex="-1"></a>x.var <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">ggplot_build</span>(p1_Lipids)<span class="sc">$</span>layout<span class="sc">$</span>panel_scales_x,</span>
<span id="cb237-12"><a href="#cb237-12" aria-hidden="true" tabindex="-1"></a>                <span class="cf">function</span>(l) <span class="fu">length</span>(l<span class="sc">$</span>range<span class="sc">$</span>range))</span>
<span id="cb237-13"><a href="#cb237-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-14"><a href="#cb237-14" aria-hidden="true" tabindex="-1"></a><span class="co"># change the relative widths of the facet columns based on</span></span>
<span id="cb237-15"><a href="#cb237-15" aria-hidden="true" tabindex="-1"></a><span class="co"># how many unique x-axis values are in each facet</span></span>
<span id="cb237-16"><a href="#cb237-16" aria-hidden="true" tabindex="-1"></a>gp_lipid<span class="sc">$</span>widths[facet.columns] <span class="ot">&lt;-</span> gp_lipid<span class="sc">$</span>widths[facet.columns] <span class="sc">*</span> x.var</span>
<span id="cb237-17"><a href="#cb237-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-18"><a href="#cb237-18" aria-hidden="true" tabindex="-1"></a><span class="co"># plot result</span></span>
<span id="cb237-19"><a href="#cb237-19" aria-hidden="true" tabindex="-1"></a>grid<span class="sc">::</span><span class="fu">grid.draw</span>(gp_lipid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="1_intro_files/figure-html/unnamed-chunk-219-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb238"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb238-1"><a href="#cb238-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb238-2"><a href="#cb238-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print</span></span>
<span id="cb238-3"><a href="#cb238-3" aria-hidden="true" tabindex="-1"></a><span class="co">#pdf("2_Output/Figure_S64.pdf", paper="special", family="sans",width=8, height=7, useDingbats=FALSE)</span></span>
<span id="cb238-4"><a href="#cb238-4" aria-hidden="true" tabindex="-1"></a><span class="co">#grid::grid.draw(gp_lipid)</span></span>
<span id="cb238-5"><a href="#cb238-5" aria-hidden="true" tabindex="-1"></a><span class="co">#dev.off() </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this example we see that only the aquifer and the basal clade have all required genes for the mevalonate pathway and the lipid pathway that is required to make a key archaeal lipid.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../index.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../code/2_misc.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R basics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>